<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #20794d; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007ba5; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #007ba5; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #007ba5; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #20794d; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

  <!--radix_placeholder_meta_tags-->
  <title>C094: Annotating Pellicci gamma-delta T-cell dataset</title>



  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2021-07-01"/>
  <meta property="article:created" itemprop="dateCreated" content="2021-07-01"/>
  <meta name="article:author" content="William Ho"/>

  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="C094: Annotating Pellicci gamma-delta T-cell dataset"/>
  <meta property="og:type" content="article"/>
  <meta property="og:locale" content="en_US"/>
  <meta property="og:site_name" content="C094"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="C094: Annotating Pellicci gamma-delta T-cell dataset"/>

  <!--/radix_placeholder_meta_tags-->
  
  <meta name="citation_reference" content="citation_title=An improved bonferroni procedure for multiple tests of significance;citation_publication_date=1986;citation_volume=73;citation_author=R. J. Simes"/>
  <meta name="citation_reference" content="citation_title=Identification of cell types from single-cell transcriptomes using a novel clustering method;citation_publication_date=2015;citation_volume=31;citation_author=C. Xu;citation_author=Z. Su"/>
  <meta name="citation_reference" content="citation_title=Bias, robustness and scalability in single-cell differential expression analysis;citation_publication_date=2018;citation_volume=15;citation_author=C. Soneson;citation_author=M. D. Robinson"/>
  <meta name="citation_reference" content="citation_title=Reference-based analysis of lung single-cell sequencing reveals a transitional profibrotic macrophage;citation_publication_date=2019;citation_volume=20;citation_author=D. Aran;citation_author=A. P. Looney;citation_author=L. Liu;citation_author=E. Wu;citation_author=V. Fong;citation_author=A. Hsu;citation_author=S. Chak;citation_author=R. P. Naikawadi;citation_author=P. J. Wolters;citation_author=A. R. Abate;citation_author=A. J. Butte;citation_author=M. Bhattacharya"/>
  <meta name="citation_reference" content="citation_title=Testing significance relative to a fold-change threshold is a TREAT;citation_publication_date=2009;citation_volume=25;citation_author=D. J. McCarthy;citation_author=G. K. Smyth"/>
  <meta name="citation_reference" content="citation_title=An expression atlas of human primary cells: Inference of gene function from coexpression networks;citation_publication_date=2013;citation_volume=14;citation_doi=10.1186/1471-2164-14-632;citation_author=Neil A. Mabbott;citation_author=J. K. Baillie;citation_author=Helen Brown;citation_author=Tom C. Freeman;citation_author=David A. Hume"/>
  <meta name="citation_reference" content="citation_title=BLUEPRINT: Mapping human blood cell epigenomes;citation_publication_date=2013;citation_volume=98;citation_doi=10.3324/haematol.2013.094243;citation_author=Joost H A Martens;citation_author=Hendrik G. Stunnenberg"/>
  <meta name="citation_reference" content="citation_title=An integrated encyclopedia of DNA elements in the human genome;citation_publication_date=2012;citation_doi=10.1038/nature11247;citation_issn=14764687;citation_author=An integrated encyclopedia of DNA elements in the human genome"/>
  <meta name="citation_reference" content="citation_title=Impact of genetic polymorphisms on human immune cell gene expression;citation_publication_date=2018;citation_volume=175;citation_doi=10.1016/j.cell.2018.10.022;citation_issn=0092-8674;citation_author=Benjamin J. Schmiedel;citation_author=Divya Singh;citation_author=Ariel Madrigal;citation_author=Alan G. Valdovino-Gonzalez;citation_author=Brandie M. White;citation_author=Jose Zapardiel-Gonzalo;citation_author=Brendan Ha;citation_author=Gokmen Altay;citation_author=Jason A. Greenbaum;citation_author=Graham McVicker;citation_author=Grégory Seumois;citation_author=Anjana Rao;citation_author=Mitchell Kronenberg;citation_author=Bjoern Peters;citation_author=Pandurangan Vijayanand"/>
  <meta name="citation_reference" content="citation_title=Densely interconnected transcriptional circuits control cell states in human hematopoiesis;citation_publication_date=2011;citation_volume=144;citation_doi=10.1016/j.cell.2011.01.004;citation_issn=1097-4172;citation_author=Noa Novershtern;citation_author=Aravind Subramanian;citation_author=Lee N. Lawton;citation_author=Raymond H. Mak;citation_author=W. Nicholas Haining;citation_author=Marie E. McConkey;citation_author=Naomi Habib;citation_author=Nir Yosef;citation_author=Cindy Y. Chang;citation_author=Tal Shay;citation_author=Garrett M. Frampton;citation_author=Adam C. B. Drake;citation_author=Ilya Leskov;citation_author=Bjorn Nilsson;citation_author=Fred Preffer;citation_author=David Dombkowski;citation_author=John W. Evans;citation_author=Ted Liefeld;citation_author=John S. Smutko;citation_author=Jianzhu Chen;citation_author=Nir Friedman;citation_author=Richard A. Young;citation_author=Todd R. Golub;citation_author=Aviv Regev;citation_author=Benjamin L. Ebert"/>
  <meta name="citation_reference" content="citation_title=RNA-seq signatures normalized by mRNA abundance allow absolute deconvolution of human immune cell types;citation_publication_date=2019;citation_volume=26;citation_doi=10.1016/j.celrep.2019.01.041;citation_issn=2211-1247;citation_author=Gianni Monaco;citation_author=Bernett Lee;citation_author=Weili Xu;citation_author=Seri Mustafah;citation_author=You Yi Hwang;citation_author=Christophe Carré;citation_author=Nicolas Burdin;citation_author=Lucian Visan;citation_author=Michele Ceccarelli;citation_author=Michael Poidinger;citation_author=Alfred Zippelius;citation_author=João Pedro de Magalhães;citation_author=Anis Larbi"/>
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","author","date","editor_options","bibliography"]}},"value":[{"type":"character","attributes":{},"value":["Annotating Pellicci gamma-delta T-cell dataset"]},{"type":"NULL"},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","affiliation_url"]}},"value":[{"type":"character","attributes":{},"value":["William Ho"]},{"type":"character","attributes":{},"value":["Single Cell Open Research Endeavour (SCORE), WEHI"]},{"type":"character","attributes":{},"value":["https://www.wehi.edu.au/people/rory-bowden/4536/wehi-advanced-genomics-facility"]}]}]},{"type":"character","attributes":{},"value":["2021-07-01"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["chunk_output_type"]}},"value":[{"type":"character","attributes":{},"value":["console"]}]},{"type":"character","attributes":{},"value":["ref.bib"]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  <!--radix_placeholder_navigation_in_header-->
  <meta name="distill:offset" content=""/>

  <script type="application/javascript">

    window.headroom_prevent_pin = false;

    window.document.addEventListener("DOMContentLoaded", function (event) {

      // initialize headroom for banner
      var header = $('header').get(0);
      var headerHeight = header.offsetHeight;
      var headroom = new Headroom(header, {
        tolerance: 5,
        onPin : function() {
          if (window.headroom_prevent_pin) {
            window.headroom_prevent_pin = false;
            headroom.unpin();
          }
        }
      });
      headroom.init();
      if(window.location.hash)
        headroom.unpin();
      $(header).addClass('headroom--transition');

      // offset scroll location for banner on hash change
      // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
      window.addEventListener("hashchange", function(event) {
        window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
      });

      // responsive menu
      $('.distill-site-header').each(function(i, val) {
        var topnav = $(this);
        var toggle = topnav.find('.nav-toggle');
        toggle.on('click', function() {
          topnav.toggleClass('responsive');
        });
      });

      // nav dropdowns
      $('.nav-dropbtn').click(function(e) {
        $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
        $(this).parent().siblings('.nav-dropdown')
           .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $("body").click(function(e){
        $('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $(".nav-dropdown").click(function(e){
        e.stopPropagation();
      });
    });
  </script>

  <style type="text/css">

  /* Theme (user-documented overrideables for nav appearance) */

  .distill-site-nav {
    color: rgba(255, 255, 255, 0.8);
    background-color: #0F2E3D;
    font-size: 15px;
    font-weight: 300;
  }

  .distill-site-nav a {
    color: inherit;
    text-decoration: none;
  }

  .distill-site-nav a:hover {
    color: white;
  }

  @media print {
    .distill-site-nav {
      display: none;
    }
  }

  .distill-site-header {

  }

  .distill-site-footer {

  }


  /* Site Header */

  .distill-site-header {
    width: 100%;
    box-sizing: border-box;
    z-index: 3;
  }

  .distill-site-header .nav-left {
    display: inline-block;
    margin-left: 8px;
  }

  @media screen and (max-width: 768px) {
    .distill-site-header .nav-left {
      margin-left: 0;
    }
  }


  .distill-site-header .nav-right {
    float: right;
    margin-right: 8px;
  }

  .distill-site-header a,
  .distill-site-header .title {
    display: inline-block;
    text-align: center;
    padding: 14px 10px 14px 10px;
  }

  .distill-site-header .title {
    font-size: 18px;
    min-width: 150px;
  }

  .distill-site-header .logo {
    padding: 0;
  }

  .distill-site-header .logo img {
    display: none;
    max-height: 20px;
    width: auto;
    margin-bottom: -4px;
  }

  .distill-site-header .nav-image img {
    max-height: 18px;
    width: auto;
    display: inline-block;
    margin-bottom: -3px;
  }



  @media screen and (min-width: 1000px) {
    .distill-site-header .logo img {
      display: inline-block;
    }
    .distill-site-header .nav-left {
      margin-left: 20px;
    }
    .distill-site-header .nav-right {
      margin-right: 20px;
    }
    .distill-site-header .title {
      padding-left: 12px;
    }
  }


  .distill-site-header .nav-toggle {
    display: none;
  }

  .nav-dropdown {
    display: inline-block;
    position: relative;
  }

  .nav-dropdown .nav-dropbtn {
    border: none;
    outline: none;
    color: rgba(255, 255, 255, 0.8);
    padding: 16px 10px;
    background-color: transparent;
    font-family: inherit;
    font-size: inherit;
    font-weight: inherit;
    margin: 0;
    margin-top: 1px;
    z-index: 2;
  }

  .nav-dropdown-content {
    display: none;
    position: absolute;
    background-color: white;
    min-width: 200px;
    border: 1px solid rgba(0,0,0,0.15);
    border-radius: 4px;
    box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
    z-index: 1;
    margin-top: 2px;
    white-space: nowrap;
    padding-top: 4px;
    padding-bottom: 4px;
  }

  .nav-dropdown-content hr {
    margin-top: 4px;
    margin-bottom: 4px;
    border: none;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .nav-dropdown-active {
    display: block;
  }

  .nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
    color: black;
    padding: 6px 24px;
    text-decoration: none;
    display: block;
    text-align: left;
  }

  .nav-dropdown-content .nav-dropdown-header {
    display: block;
    padding: 5px 24px;
    padding-bottom: 0;
    text-transform: uppercase;
    font-size: 14px;
    color: #999999;
    white-space: nowrap;
  }

  .nav-dropdown:hover .nav-dropbtn {
    color: white;
  }

  .nav-dropdown-content a:hover {
    background-color: #ddd;
    color: black;
  }

  .nav-right .nav-dropdown-content {
    margin-left: -45%;
    right: 0;
  }

  @media screen and (max-width: 768px) {
    .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
    .distill-site-header a.nav-toggle {
      float: right;
      display: block;
    }
    .distill-site-header .title {
      margin-left: 0;
    }
    .distill-site-header .nav-right {
      margin-right: 0;
    }
    .distill-site-header {
      overflow: hidden;
    }
    .nav-right .nav-dropdown-content {
      margin-left: 0;
    }
  }


  @media screen and (max-width: 768px) {
    .distill-site-header.responsive {position: relative; min-height: 500px; }
    .distill-site-header.responsive a.nav-toggle {
      position: absolute;
      right: 0;
      top: 0;
    }
    .distill-site-header.responsive a,
    .distill-site-header.responsive .nav-dropdown {
      display: block;
      text-align: left;
    }
    .distill-site-header.responsive .nav-left,
    .distill-site-header.responsive .nav-right {
      width: 100%;
    }
    .distill-site-header.responsive .nav-dropdown {float: none;}
    .distill-site-header.responsive .nav-dropdown-content {position: relative;}
    .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
      display: block;
      width: 100%;
      text-align: left;
    }
  }

  /* Site Footer */

  .distill-site-footer {
    width: 100%;
    overflow: hidden;
    box-sizing: border-box;
    z-index: 3;
    margin-top: 30px;
    padding-top: 30px;
    padding-bottom: 30px;
    text-align: center;
  }

  /* Headroom */

  d-title {
    padding-top: 6rem;
  }

  @media print {
    d-title {
      padding-top: 4rem;
    }
  }

  .headroom {
    z-index: 1000;
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
  }

  .headroom--transition {
    transition: all .4s ease-in-out;
  }

  .headroom--unpinned {
    top: -100px;
  }

  .headroom--pinned {
    top: 0;
  }

  /* adjust viewport for navbar height */
  /* helps vertically center bootstrap (non-distill) content */
  .min-vh-100 {
    min-height: calc(100vh - 100px) !important;
  }

  </style>

  <script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
  <link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet"/>
  <link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet"/>
  <script src="site_libs/headroom-0.9.4/headroom.min.js"></script>
  <script src="site_libs/autocomplete-0.37.1/autocomplete.min.js"></script>
  <script src="site_libs/fuse-6.4.1/fuse.min.js"></script>

  <script type="application/javascript">

  function getMeta(metaName) {
    var metas = document.getElementsByTagName('meta');
    for (let i = 0; i < metas.length; i++) {
      if (metas[i].getAttribute('name') === metaName) {
        return metas[i].getAttribute('content');
      }
    }
    return '';
  }

  function offsetURL(url) {
    var offset = getMeta('distill:offset');
    return offset ? offset + '/' + url : url;
  }

  function createFuseIndex() {

    // create fuse index
    var options = {
      keys: [
        { name: 'title', weight: 20 },
        { name: 'categories', weight: 15 },
        { name: 'description', weight: 10 },
        { name: 'contents', weight: 5 },
      ],
      ignoreLocation: true,
      threshold: 0
    };
    var fuse = new window.Fuse([], options);

    // fetch the main search.json
    return fetch(offsetURL('search.json'))
      .then(function(response) {
        if (response.status == 200) {
          return response.json().then(function(json) {
            // index main articles
            json.articles.forEach(function(article) {
              fuse.add(article);
            });
            // download collections and index their articles
            return Promise.all(json.collections.map(function(collection) {
              return fetch(offsetURL(collection)).then(function(response) {
                if (response.status === 200) {
                  return response.json().then(function(articles) {
                    articles.forEach(function(article) {
                      fuse.add(article);
                    });
                  })
                } else {
                  return Promise.reject(
                    new Error('Unexpected status from search index request: ' +
                              response.status)
                  );
                }
              });
            })).then(function() {
              return fuse;
            });
          });

        } else {
          return Promise.reject(
            new Error('Unexpected status from search index request: ' +
                        response.status)
          );
        }
      });
  }

  window.document.addEventListener("DOMContentLoaded", function (event) {

    // get search element (bail if we don't have one)
    var searchEl = window.document.getElementById('distill-search');
    if (!searchEl)
      return;

    createFuseIndex()
      .then(function(fuse) {

        // make search box visible
        searchEl.classList.remove('hidden');

        // initialize autocomplete
        var options = {
          autoselect: true,
          hint: false,
          minLength: 2,
        };
        window.autocomplete(searchEl, options, [{
          source: function(query, callback) {
            const searchOptions = {
              isCaseSensitive: false,
              shouldSort: true,
              minMatchCharLength: 2,
              limit: 10,
            };
            var results = fuse.search(query, searchOptions);
            callback(results
              .map(function(result) { return result.item; })
            );
          },
          templates: {
            suggestion: function(suggestion) {
              var img = suggestion.preview && Object.keys(suggestion.preview).length > 0
                ? `<img src="${offsetURL(suggestion.preview)}"</img>`
                : '';
              var html = `
                <div class="search-item">
                  <h3>${suggestion.title}</h3>
                  <div class="search-item-description">
                    ${suggestion.description || ''}
                  </div>
                  <div class="search-item-preview">
                    ${img}
                  </div>
                </div>
              `;
              return html;
            }
          }
        }]).on('autocomplete:selected', function(event, suggestion) {
          window.location.href = offsetURL(suggestion.path);
        });
        // remove inline display style on autocompleter (we want to
        // manage responsive display via css)
        $('.algolia-autocomplete').css("display", "");
      })
      .catch(function(error) {
        console.log(error);
      });

  });

  </script>

  <style type="text/css">

  .nav-search {
    font-size: x-small;
  }

  /* Algolioa Autocomplete */

  .algolia-autocomplete {
    display: inline-block;
    margin-left: 10px;
    vertical-align: sub;
    background-color: white;
    color: black;
    padding: 6px;
    padding-top: 8px;
    padding-bottom: 0;
    border-radius: 6px;
    border: 1px #0F2E3D solid;
    width: 180px;
  }


  @media screen and (max-width: 768px) {
    .distill-site-nav .algolia-autocomplete {
      display: none;
      visibility: hidden;
    }
    .distill-site-nav.responsive .algolia-autocomplete {
      display: inline-block;
      visibility: visible;
    }
    .distill-site-nav.responsive .algolia-autocomplete .aa-dropdown-menu {
      margin-left: 0;
      width: 400px;
      max-height: 400px;
    }
  }

  .algolia-autocomplete .aa-input, .algolia-autocomplete .aa-hint {
    width: 90%;
    outline: none;
    border: none;
  }

  .algolia-autocomplete .aa-hint {
    color: #999;
  }
  .algolia-autocomplete .aa-dropdown-menu {
    width: 550px;
    max-height: 70vh;
    overflow-x: visible;
    overflow-y: scroll;
    padding: 5px;
    margin-top: 3px;
    margin-left: -150px;
    background-color: #fff;
    border-radius: 5px;
    border: 1px solid #999;
    border-top: none;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion {
    cursor: pointer;
    padding: 5px 4px;
    border-bottom: 1px solid #eee;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion:last-of-type {
    border-bottom: none;
    margin-bottom: 2px;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item {
    overflow: hidden;
    font-size: 0.8em;
    line-height: 1.4em;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item h3 {
    font-size: 1rem;
    margin-block-start: 0;
    margin-block-end: 5px;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-description {
    display: inline-block;
    overflow: hidden;
    height: 2.8em;
    width: 80%;
    margin-right: 4%;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview {
    display: inline-block;
    width: 15%;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview img {
    height: 3em;
    width: auto;
    display: none;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview img[src] {
    display: initial;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion.aa-cursor {
    background-color: #eee;
  }
  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion em {
    font-weight: bold;
    font-style: normal;
  }

  </style>


  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: auto;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  d-article details {
    grid-column: text;
    margin-bottom: 0.8em;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure img {
    width: 100%;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }

  /* Citation hover box */

  .tippy-box[data-theme~=light-border] {
    background-color: rgba(250, 250, 250, 0.95);
  }

  .tippy-content > p {
    margin-bottom: 0;
    padding: 2px;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('details, div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme
    $('code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // hoverable references
      $('span.citation[data-cites]').each(function() {
        var refHtml = $('#ref-' + $(this).attr('data-cites')).html();
        window.tippy(this, {
          allowHTML: true,
          content: refHtml,
          maxWidth: 500,
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        });
      });

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace("index.html", "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <!--/radix_placeholder_distill-->
  <script src="site_libs/header-attrs-2.7/header-attrs.js"></script>
  <link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
  <script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
  <script src="site_libs/popper-2.6.0/popper.min.js"></script>
  <link href="site_libs/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="site_libs/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="site_libs/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="site_libs/anchor-4.2.2/anchor.min.js"></script>
  <script src="site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <!--/radix_placeholder_site_in_header-->


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"Annotating Pellicci gamma-delta T-cell dataset","authors":[{"author":"William Ho","authorURL":"#","affiliation":"Single Cell Open Research Endeavour (SCORE), WEHI","affiliationURL":"https://www.wehi.edu.au/people/rory-bowden/4536/wehi-advanced-genomics-facility","orcidID":""}],"publishedDate":"2021-07-01T00:00:00.000+10:00","citationText":"Ho, 2021"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a href="index.html" class="title">C094</a>
<a href="index.html">Introduction</a>
<a href="C094_Pellicci.preprocess.html">Preprocessing</a>
<a href="C094_Pellicci.cell_selection.html">Selecting biologically relevant cells</a>
<a href="C094_Pellicci.annotate.html">Annotating cells</a>
<a href="C094_Pellicci.multi-sample_comparisons.html">Differential analyses</a>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Supplementary files
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="../output/CellRanger/">CellRanger reports</a>
<a href="../output/DEGs/">Differentially expressed genes</a>
<a href="../output/glimma-plots/">Glimma plots</a>
</div>
</div>
</div>
<div class="nav-right">
<a href="https://github.com/WEHISCORE/C094_Pellicci">
<i class="fa fa-github" aria-hidden="true"></i>
</a>
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>Annotating Pellicci gamma-delta T-cell dataset</h1>
<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->

</div>

<div class="d-byline">
  William Ho  (Single Cell Open Research Endeavour (SCORE), WEHI)<a href="https://www.wehi.edu.au/people/rory-bowden/4536/wehi-advanced-genomics-facility" class="uri">https://www.wehi.edu.au/people/rory-bowden/4536/wehi-advanced-genomics-facility</a>
  
<br/>2021-07-01
</div>

<div class="d-article">
<div class="d-contents d-contents-float">
<nav class="l-text toc figcaption" id="TOC">
<h3>Contents</h3>
<ul>
<li><a href="#preparing-the-data">Preparing the data</a></li>
<li><a href="#clustering-cells">Clustering cells</a>
<ul>
<li><a href="#clusters-vs.-cell-types">Clusters vs. cell types</a></li>
<li><a href="#graph-based-clustering">Graph-based clustering</a>
<ul>
<li><a href="#assessing-cluster-separation">Assessing cluster separation</a></li>
</ul></li>
<li><a href="#visual-inspection-of-clustering-results">Visual inspection of clustering results</a></li>
</ul></li>
<li><a href="#assigning-cell-labels-from-reference-data">Assigning cell labels from reference data</a>
<ul>
<li><a href="#overview">Overview</a></li>
<li><a href="#available-references">Available references</a></li>
<li><a href="#results">Results</a></li>
<li><a href="#monacoimmunedata"><code>MonacoImmuneData</code></a>
<ul>
<li><a href="#fine-label">Fine label</a></li>
</ul></li>
</ul></li>
<li><a href="#marker-gene-detection">Marker gene detection</a>
<ul>
<li><a href="#using-pairwise-t-tests">Using pairwise <em>t</em>-tests</a>
<ul>
<li><a href="#standard-application">Standard application</a></li>
<li><a href="#using-the-log-fold-change">Using the log-fold change</a></li>
<li><a href="#finding-cluster-specific-markers">Finding cluster-specific markers</a></li>
</ul></li>
<li><a href="#summary">Summary</a>
<ul>
<li><a href="#using-marker-genes-to-perform-cell-type-annotation">Using marker genes to perform cell type annotation</a></li>
</ul></li>
</ul></li>
<li><a href="#cell-cycle-assignment">Cell cycle assignment</a>
<ul>
<li><a href="#using-the-cyclins">Using the cyclins</a></li>
</ul></li>
<li><a href="#concluding-remarks">Concluding remarks</a></li>
<li><a href="#additional-information">Additional information</a></li>
<li><a href="#session-info">Session info</a></li>
</ul>
</nav>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>SingleCellExperiment</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://here.r-lib.org/'>here</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://bioconductor.org/packages/scater/'>scater</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>scran</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://ggplot2.tidyverse.org'>ggplot2</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://wilkelab.org/cowplot/'>cowplot</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://bioinf.wehi.edu.au/edgeR'>edgeR</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/hasaru-k/GlimmaV2'>Glimma</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/Bioconductor/BiocParallel'>BiocParallel</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://patchwork.data-imaginist.com'>patchwork</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/sfirke/janitor'>janitor</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>pheatmap</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>batchelor</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/rstudio/rmarkdown'>rmarkdown</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/Bioconductor/BiocStyle'>BiocStyle</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://ggplot2.tidyverse.org'>ggplot2</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://readxl.tidyverse.org'>readxl</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://dplyr.tidyverse.org'>dplyr</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://tidyr.tidyverse.org'>tidyr</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/source.html'>source</a></span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"code"</span>, <span class='st'>"helper_functions.R"</span><span class='op'>)</span><span class='op'>)</span>
<span class='co'># NOTE: Using multiple cores seizes up my laptop. Can use more on unix box.</span>
<span class='fu'><a href='https://rdrr.io/r/base/options.html'>options</a></span><span class='op'>(</span><span class='st'>"mc.cores"</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/Sys.info.html'>Sys.info</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>[[</span><span class='st'>"nodename"</span><span class='op'>]</span><span class='op'>]</span> <span class='op'>==</span> <span class='st'>"PC1331"</span>, <span class='fl'>2L</span>, <span class='fl'>8L</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/register.html'>register</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html'>MulticoreParam</a></span><span class='op'>(</span>workers <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/options.html'>getOption</a></span><span class='op'>(</span><span class='st'>"mc.cores"</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'>knitr</span><span class='fu'>::</span><span class='va'><a href='https://rdrr.io/pkg/knitr/man/opts_chunk.html'>opts_chunk</a></span><span class='op'>$</span><span class='fu'>set</span><span class='op'>(</span>fig.path <span class='op'>=</span> <span class='st'>"C094_Pellicci.annotate_files/"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<h1 id="preparing-the-data">Preparing the data</h1>
<p>We start from the cell selected <em>SingleCellExperiment</em> object created in <a href="C094_Pellicci.cell_selection.html">‘Selection of biologically relevant cells for Pellicci gamma-delta T-cell dataset’</a>.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"data"</span>, <span class='st'>"SCEs"</span>, <span class='st'>"C094_Pellicci.cells_selected.SCE.rds"</span><span class='op'>)</span><span class='op'>)</span>

<span class='fu'><a href='https://rdrr.io/r/base/files2.html'>dir.create</a></span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"data"</span>, <span class='st'>"marker_genes"</span><span class='op'>)</span>, recursive <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/files2.html'>dir.create</a></span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"output"</span>, <span class='st'>"marker_genes"</span><span class='op'>)</span>, recursive <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># Some useful colours</span>
<span class='va'>sample_type_colours</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/setNames.html'>setNames</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>sample_type_colours</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>sample_type_colours</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>sample_name_colours</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/setNames.html'>setNames</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>sample_name_colours</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>sample_name_colours</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># NOTE: need to update samplecolours stored in SCE after introducing "post_hoc_sample_gate"</span>
<span class='va'>sample_gate_colours</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/setNames.html'>setNames</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/grDevices/palette.html'>palette.colors</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/nlevels.html'>nlevels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>post_hoc_sample_gate</span><span class='op'>)</span>, <span class='st'>"Tableau 10"</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>post_hoc_sample_gate</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>sample_gate_colours</span> <span class='op'>&lt;-</span> <span class='va'>sample_gate_colours</span><span class='op'>[</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>post_hoc_sample_gate</span><span class='op'>]</span>

<span class='va'>plate_number_colours</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/setNames.html'>setNames</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>plate_number_colours</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>plate_number_colours</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>tissue_colours</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/setNames.html'>setNames</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>tissue_colours</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>tissue_colours</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>donor_colours</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/setNames.html'>setNames</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>donor_colours</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>donor_colours</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># Some useful gene sets</span>
<span class='va'>mito_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/any.html'>any</a></span><span class='op'>(</span><span class='fu'>rowData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>$</span><span class='va'>ENSEMBL.SEQNAME</span> <span class='op'>==</span> <span class='st'>"MT"</span><span class='op'>)</span><span class='op'>]</span>
<span class='va'>ribo_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/grep.html'>grep</a></span><span class='op'>(</span><span class='st'>"^RP(S|L)"</span>, <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>, value <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='co'># NOTE: A more curated approach for identifying ribosomal protein genes </span>
<span class='co'>#       (https://github.com/Bioconductor/OrchestratingSingleCellAnalysis-base/blob/ae201bf26e3e4fa82d9165d8abf4f4dc4b8e5a68/feature-selection.Rmd#L376-L380)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://igordot.github.io/msigdbr'>msigdbr</a></span><span class='op'>)</span>
<span class='va'>c2_sets</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://igordot.github.io/msigdbr/reference/msigdbr.html'>msigdbr</a></span><span class='op'>(</span>species <span class='op'>=</span> <span class='st'>"Homo sapiens"</span>, category <span class='op'>=</span> <span class='st'>"C2"</span><span class='op'>)</span>
<span class='va'>ribo_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://generics.r-lib.org/reference/setops.html'>union</a></span><span class='op'>(</span>
  <span class='va'>ribo_set</span>,
  <span class='va'>c2_sets</span><span class='op'>[</span><span class='va'>c2_sets</span><span class='op'>$</span><span class='va'>gs_name</span> <span class='op'>==</span> <span class='st'>"KEGG_RIBOSOME"</span>, <span class='op'>]</span><span class='op'>$</span><span class='va'>gene_symbol</span><span class='op'>)</span>
<span class='va'>ribo_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://generics.r-lib.org/reference/setops.html'>intersect</a></span><span class='op'>(</span><span class='va'>ribo_set</span>, <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>sex_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/any.html'>any</a></span><span class='op'>(</span><span class='fu'>rowData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>$</span><span class='va'>ENSEMBL.SEQNAME</span> <span class='op'>%in%</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"X"</span>, <span class='st'>"Y"</span><span class='op'>)</span><span class='op'>)</span><span class='op'>]</span>
<span class='va'>pseudogene_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span>
  <span class='fu'><a href='https://rdrr.io/r/base/any.html'>any</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/grep.html'>grepl</a></span><span class='op'>(</span><span class='st'>"pseudogene"</span>, <span class='fu'>rowData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>$</span><span class='va'>ENSEMBL.GENEBIOTYPE</span><span class='op'>)</span><span class='op'>)</span><span class='op'>]</span>
</code></pre>
</div>
</details>
</div>
<h1 id="clustering-cells">Clustering cells</h1>
<p>Clustering is an unsupervised learning procedure that is used in scRNA-seq data analysis to empirically define groups of cells with similar expression profiles. Its primary purpose is to summarize the data in a digestible format for human interpretation. This allows us to describe population heterogeneity in terms of discrete labels that are easily understood, rather than attempting to comprehend the high-dimensional manifold on which the cells truly reside. After annotation based on marker genes, the clusters can be treated as proxies for more abstract biological concepts such as cell types or states. Clustering is thus a critical step for extracting biological insights from scRNA-seq data.</p>
<h2 id="clusters-vs.-cell-types">Clusters vs. cell types</h2>
<p>It is worth stressing the distinction between clusters and cell types. The former is an empirical construct while the latter is a biological truth (albeit a vaguely defined one). For this reason, questions like “what is the true number of clusters?” are usually meaningless. We can define as many clusters as we like, with whatever algorithm we like - each clustering will represent its own partitioning of the high-dimensional expression space, and is as “real” as any other clustering.</p>
<p>A more relevant question is “how well do the clusters approximate the cell types?” Unfortunately, this is difficult to answer given the context-dependent interpretation of biological truth. Some analysts will be satisfied with resolution of the major cell types; other analysts may want resolution of subtypes; and others still may require resolution of different states (e.g., metabolic activity, stress) within those subtypes. Two clusterings can also be highly inconsistent yet both valid, simply partitioning the cells based on different aspects of biology. Indeed, asking for an unqualified “best” clustering is akin to asking for the best magnification on a microscope without any context.</p>
<p>It is helpful to realize that clustering, like a microscope, is simply a tool to explore the data. We can zoom in and out by changing the resolution of the clustering parameters, and we can experiment with different clustering algorithms to obtain alternative perspectives of the data. This iterative approach is entirely permissible for data exploration, which constitutes the majority of all scRNA-seq data analysis.</p>
<h2 id="graph-based-clustering">Graph-based clustering</h2>
<p>We build a shared nearest neighbour graph <span class="citation" data-cites="xu2015identification">(<a href="#ref-xu2015identification" role="doc-biblioref">Xu and Su 2015</a>)</span> and use the Louvain algorithm to identify clusters. Typically, we would build the graph using the principal components, but here we use the MNN-corrected values so as to incorporate the batch effect correction.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>4759</span><span class='op'>)</span>
<span class='va'>snn_gr</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scran/man/buildSNNGraph.html'>buildSNNGraph</a></span><span class='op'>(</span><span class='va'>sce</span>, use.dimred <span class='op'>=</span> <span class='st'>"corrected"</span><span class='op'>)</span>
<span class='va'>clusters</span> <span class='op'>&lt;-</span> <span class='fu'>igraph</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/igraph/man/cluster_louvain.html'>cluster_louvain</a></span><span class='op'>(</span><span class='va'>snn_gr</span><span class='op'>)</span>
<span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>clusters</span><span class='op'>$</span><span class='va'>membership</span><span class='op'>)</span>
<span class='va'>cluster_colours</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/setNames.html'>setNames</a></span><span class='op'>(</span>
  <span class='fu'>scater</span><span class='fu'>:::</span><span class='fu'>.get_palette</span><span class='op'>(</span><span class='st'>"tableau10medium"</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_len</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/nlevels.html'>nlevels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span><span class='op'>)</span><span class='op'>)</span><span class='op'>]</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>cluster_colours</span> <span class='op'>&lt;-</span> <span class='va'>cluster_colours</span><span class='op'>[</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span><span class='op'>]</span>
</code></pre>
</div>
</details>
</div>
<h3 id="assessing-cluster-separation">Assessing cluster separation</h3>
<p>We look at the ratio of the observed and expected edge weights to confirm that the clusters are modular. Figure <a href="#fig:clustermod">1</a> indicates that most of the clusters are reasonably separated, with few strong off-diagonal entries.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>cluster_mod</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scran/man/clusterModularity.html'>clusterModularity</a></span><span class='op'>(</span>
  graph <span class='op'>=</span> <span class='va'>snn_gr</span>, 
  clusters <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span>, 
  get.weights <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='va'>log_ratio</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log2</a></span><span class='op'>(</span><span class='va'>cluster_mod</span><span class='op'>$</span><span class='va'>observed</span> <span class='op'>/</span> <span class='va'>cluster_mod</span><span class='op'>$</span><span class='va'>expected</span> <span class='op'>+</span> <span class='fl'>1</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>pheatmap</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span>
  mat <span class='op'>=</span> <span class='va'>log_ratio</span>,
  cluster_rows <span class='op'>=</span> <span class='cn'>FALSE</span>, 
  cluster_cols <span class='op'>=</span> <span class='cn'>FALSE</span>, 
  color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/colorRamp.html'>colorRampPalette</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"white"</span>, <span class='st'>"blue"</span><span class='op'>)</span><span class='op'>)</span><span class='op'>(</span><span class='fl'>100</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:clustermod"></span>
<img src="C094_Pellicci.annotate_files/clustermod-1.png" alt="Heatmap of the log~10~-ratio of the total weight between nodes in the same cluster or in different clusters, relative to the total weight expected under a null model of random links." width="624" />
<p class="caption">
Figure 1: Heatmap of the log<sub>10</sub>-ratio of the total weight between nodes in the same cluster or in different clusters, relative to the total weight expected under a null model of random links.
</p>
</div>
</div>
<h2 id="visual-inspection-of-clustering-results">Visual inspection of clustering results</h2>
<p>There are 4 clusters detected, shown on the UMAP plot Figure <a href="#fig:clusterplot-umap">2</a> and broken down by experimental factors in Figure <a href="#fig:cluster-barplot">3</a>. A sensible partitioning was confirmed to be generated.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scater/man/ggsce.html'>ggcells</a></span><span class='op'>(</span><span class='va'>sce</span>, <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>UMAP.1</span>, y <span class='op'>=</span> <span class='va'>UMAP.2</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='va'>cluster</span><span class='op'>)</span>, size <span class='op'>=</span> <span class='fl'>0.25</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>cluster_colours</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>"Dimension 1"</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Dimension 2"</span><span class='op'>)</span>

<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scater/man/ggsce.html'>ggcells</a></span><span class='op'>(</span><span class='va'>sce</span>, <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>UMAP.1</span>, y <span class='op'>=</span> <span class='va'>UMAP.2</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='va'>sample_type</span><span class='op'>)</span>, size <span class='op'>=</span> <span class='fl'>0.25</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>sample_type_colours</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>"Dimension 1"</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Dimension 2"</span><span class='op'>)</span>

<span class='va'>p3</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scater/man/ggsce.html'>ggcells</a></span><span class='op'>(</span><span class='va'>sce</span>, <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>UMAP.1</span>, y <span class='op'>=</span> <span class='va'>UMAP.2</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='va'>sample_name</span><span class='op'>)</span>, size <span class='op'>=</span> <span class='fl'>0.25</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>sample_name_colours</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>"Dimension 1"</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Dimension 2"</span><span class='op'>)</span>

<span class='va'>p4</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scater/man/ggsce.html'>ggcells</a></span><span class='op'>(</span><span class='va'>sce</span>, <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>UMAP.1</span>, y <span class='op'>=</span> <span class='va'>UMAP.2</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='va'>post_hoc_sample_gate</span><span class='op'>)</span>, size <span class='op'>=</span> <span class='fl'>0.25</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>sample_gate_colours</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>"Dimension 1"</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Dimension 2"</span><span class='op'>)</span>

<span class='va'>p5</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scater/man/ggsce.html'>ggcells</a></span><span class='op'>(</span><span class='va'>sce</span>, <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>UMAP.1</span>, y <span class='op'>=</span> <span class='va'>UMAP.2</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='va'>plate_number</span><span class='op'>)</span>, size <span class='op'>=</span> <span class='fl'>0.25</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>plate_number_colours</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>"Dimension 1"</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Dimension 2"</span><span class='op'>)</span>

<span class='va'>p6</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scater/man/ggsce.html'>ggcells</a></span><span class='op'>(</span><span class='va'>sce</span>, <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>UMAP.1</span>, y <span class='op'>=</span> <span class='va'>UMAP.2</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='va'>tissue</span><span class='op'>)</span>, size <span class='op'>=</span> <span class='fl'>0.25</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>tissue_colours</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>"Dimension 1"</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Dimension 2"</span><span class='op'>)</span>

<span class='va'>p7</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scater/man/ggsce.html'>ggcells</a></span><span class='op'>(</span><span class='va'>sce</span>, <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>UMAP.1</span>, y <span class='op'>=</span> <span class='va'>UMAP.2</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='va'>donor</span><span class='op'>)</span>, size <span class='op'>=</span> <span class='fl'>0.25</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>donor_colours</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>"Dimension 1"</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Dimension 2"</span><span class='op'>)</span>

<span class='op'>(</span><span class='va'>p1</span> <span class='op'>|</span> <span class='va'>p2</span><span class='op'>)</span> <span class='op'>/</span> <span class='op'>(</span><span class='va'>p3</span> <span class='op'>|</span> <span class='va'>p4</span><span class='op'>)</span> <span class='op'>/</span> <span class='op'>(</span><span class='va'>p5</span> <span class='op'>|</span> <span class='va'>p6</span><span class='op'>)</span> <span class='op'>/</span> <span class='op'>(</span><span class='va'>p7</span> <span class='op'>|</span> <span class='fu'><a href='https://patchwork.data-imaginist.com/reference/plot_spacer.html'>plot_spacer</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:clusterplot-umap"></span>
<img src="C094_Pellicci.annotate_files/clusterplot-umap-1.png" alt="UMAP plot, where each point represents a cell and is coloured according to the legend." width="624" />
<p class="caption">
Figure 2: UMAP plot, where each point represents a cell and is coloured according to the legend.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scater/man/ggsce.html'>ggcells</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>cluster</span>, fill <span class='op'>=</span> <span class='va'>sample_type</span><span class='op'>)</span>,
    position <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/position_stack.html'>position_fill</a></span><span class='op'>(</span>reverse <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Frequency"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>sample_type_colours</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='co'># NOTE: Lots of levels, so need smaller font size.</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span>

<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scater/man/ggsce.html'>ggcells</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>cluster</span>, fill <span class='op'>=</span> <span class='va'>sample_name</span><span class='op'>)</span>,
    position <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/position_stack.html'>position_fill</a></span><span class='op'>(</span>reverse <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Frequency"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>sample_name_colours</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span>

<span class='va'>p3</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scater/man/ggsce.html'>ggcells</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>cluster</span>, fill <span class='op'>=</span> <span class='va'>post_hoc_sample_gate</span><span class='op'>)</span>,
    position <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/position_stack.html'>position_fill</a></span><span class='op'>(</span>reverse <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Frequency"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>sample_gate_colours</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span>

<span class='va'>p4</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scater/man/ggsce.html'>ggcells</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>cluster</span>, fill <span class='op'>=</span> <span class='va'>plate_number</span><span class='op'>)</span>,
    position <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/position_stack.html'>position_fill</a></span><span class='op'>(</span>reverse <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Frequency"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>plate_number_colours</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span>

<span class='va'>p5</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scater/man/ggsce.html'>ggcells</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>cluster</span>, fill <span class='op'>=</span> <span class='va'>tissue</span><span class='op'>)</span>,
    position <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/position_stack.html'>position_fill</a></span><span class='op'>(</span>reverse <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Frequency"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>tissue_colours</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span>

<span class='va'>p6</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scater/man/ggsce.html'>ggcells</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>cluster</span>, fill <span class='op'>=</span> <span class='va'>donor</span><span class='op'>)</span>,
    position <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/position_stack.html'>position_fill</a></span><span class='op'>(</span>reverse <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Frequency"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>donor_colours</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span>

<span class='op'>(</span><span class='va'>p1</span> <span class='op'>|</span> <span class='va'>p2</span><span class='op'>)</span> <span class='op'>/</span> <span class='op'>(</span><span class='va'>p3</span> <span class='op'>|</span> <span class='va'>p4</span><span class='op'>)</span> <span class='op'>/</span> <span class='op'>(</span><span class='va'>p5</span> <span class='op'>|</span> <span class='va'>p6</span><span class='op'>)</span> 
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:cluster-barplot"></span>
<img src="C094_Pellicci.annotate_files/cluster-barplot-1.png" alt="Breakdown of clusters by experimental factors." width="624" />
<p class="caption">
Figure 3: Breakdown of clusters by experimental factors.
</p>
</div>
</div>
<h1 id="assigning-cell-labels-from-reference-data">Assigning cell labels from reference data</h1>
<h2 id="overview">Overview</h2>
<p>A conceptually straightforward annotation approach is to compare the single-cell expression profiles with previously annotated reference data sets. Labels can then be assigned to each cell in our uncharacterized test data set based on the most similar reference sample(s), for some definition of “similar.” Any published and labelled RNA-seq data set (bulk or single-cell) can be used as a reference, though its reliability depends greatly on the expertise of the original authors who assigned the labels in the first place.</p>
<p><em><a href="https://bioconductor.org/packages/3.12/SingleR">SingleR</a></em> is one such automatic annotation method for scRNA-seq data <span class="citation" data-cites="aran2019reference">(<a href="#ref-aran2019reference" role="doc-biblioref">Aran et al. 2019</a>)</span>. Given a reference data set(s) of samples<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> with known labels, it labels new cells from a test data set based on similarity to the reference set(s). Specifically, for each test cell:</p>
<ol type="1">
<li>We compute the Spearman correlation between its expression profile and that of each reference sample.</li>
<li>We define the per-label score as a fixed quantile (by default, 0.8) of the distribution of correlations.</li>
<li>We repeat this for all labels and we take the label with the highest score as the annotation for this cell.</li>
<li>Finally, we perform a fine-tuning step:</li>
</ol>
<ul>
<li>The reference data set is subsetted to only include labels with scores close to the maximum.</li>
<li>Scores are recomputed using only marker genes for the subset of labels.</li>
<li>This is iterated until one label remains.</li>
</ul>
<p>For visualization purposes we use the normalized scores, which are the scores linearly adjusted for each cell so that the smallest score is 0 and the largest score is 1. This is followed by cubing of the adjusted scores to improve dynamic range near 1.</p>
<p>Automatic annotation provides a convenient way of transferring biological knowledge across data sets. In this manner, the burden of interpreting clusters and defining marker genes only has to be done once (i.e. for the reference set).</p>
<h2 id="available-references">Available references</h2>
<p><em><a href="https://bioconductor.org/packages/3.12/SingleR">SingleR</a></em> includes several in-built reference data sets, some of which are relevant for this study:</p>
<ul>
<li><code>HumanPrimaryCellAtlasData</code> (<code>HPCA</code>): The Human Primary Cell Atlas <span class="citation" data-cites="hpcaRef">(<a href="#ref-hpcaRef" role="doc-biblioref">Mabbott et al. 2013</a>)</span></li>
<li><code>BlueprintEncodeData</code> (<code>BE</code>): Blueprint <span class="citation" data-cites="blueprintRef">(<a href="#ref-blueprintRef" role="doc-biblioref">Martens and Stunnenberg 2013</a>)</span> and Encode <span class="citation" data-cites="encodeRef">(<a href="#ref-encodeRef" role="doc-biblioref">The ENCODE Project Consortium 2012</a>)</span></li>
<li><code>DatabaseImmuneCellExpressionData</code> (<code>DICE</code>): The <a href="https://dice-database.org/">Database of Immune Cell Expression (DICE)</a> <span class="citation" data-cites="diceRef">(<a href="#ref-diceRef" role="doc-biblioref">Schmiedel et al. 2018</a>)</span></li>
<li><code>NovershternHematopoieticData</code> (<code>NH</code>): The <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE24759">Novershtern Hematopoietic Cell Data (GSE24759)</a> <span class="citation" data-cites="nhRef">(<a href="#ref-nhRef" role="doc-biblioref">Novershtern et al. 2011</a>)</span></li>
<li><code>MonacoImmuneData</code> (<code>MI</code>): The <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE107011">Monaco Immune Cell Data (GSE107011)</a> <span class="citation" data-cites="monaco_immuneRef">(<a href="#ref-monaco_immuneRef" role="doc-biblioref">Monaco et al. 2019</a>)</span>.</li>
</ul>
<p>These bulk RNA-seq and microarray data sets were obtained from pre-sorted cell populations, i.e., the cell labels of these samples were mostly derived based on the respective sorting/purification strategy, not via <em>in silico</em> prediction methods.</p>
<p>The characteristics of each data set are summarized below<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>:</p>
<table>
<colgroup>
<col style="width: 21%" />
<col style="width: 11%" />
<col style="width: 15%" />
<col style="width: 25%" />
<col style="width: 14%" />
<col style="width: 11%" />
</colgroup>
<thead>
<tr class="header">
<th>Reference</th>
<th>Samples</th>
<th>Sample types</th>
<th>No. of main labels</th>
<th>No. of fine labels</th>
<th>Cell type focus</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>HPCA</code></td>
<td>713</td>
<td>microarrays of sorted cell populations</td>
<td>37</td>
<td>157</td>
<td>Non-specific</td>
</tr>
<tr class="even">
<td><code>BE</code></td>
<td>259</td>
<td>RNA-seq</td>
<td>24</td>
<td>43</td>
<td>Non-specific</td>
</tr>
<tr class="odd">
<td><code>DICE</code></td>
<td>1561</td>
<td>RNA-seq</td>
<td>5</td>
<td>15</td>
<td>Immune</td>
</tr>
<tr class="even">
<td><code>NH</code></td>
<td>211</td>
<td>microarrays of sorted cell populations</td>
<td>17</td>
<td>38</td>
<td>Hematopoietic &amp; Immune</td>
</tr>
<tr class="odd">
<td><code>MI</code></td>
<td>114</td>
<td>RNA-seq</td>
<td>11</td>
<td>29</td>
<td>Immune</td>
</tr>
</tbody>
</table>
<p>The available sample types in each set can be viewed in the collapsible sections below.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/LTLA/SingleR'>SingleR</a></span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<details>
<summary>
<code>HumanPrimaryCellAtlasData</code> Labels
</summary>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>hpca</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/SingleR/man/datasets.html'>HumanPrimaryCellAtlasData</a></span><span class='op'>(</span><span class='op'>)</span>
<span class='fu'>.adf</span><span class='op'>(</span><span class='fu'>colData</span><span class='op'>(</span><span class='va'>hpca</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/count.html'>count</a></span><span class='op'>(</span><span class='va'>label.main</span>, <span class='va'>label.fine</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/arrange.html'>arrange</a></span><span class='op'>(</span><span class='va'>label.main</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>knitr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/knitr/man/kable.html'>kable</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<table>
<thead>
<tr class="header">
<th style="text-align: left;">label.main</th>
<th style="text-align: left;">label.fine</th>
<th style="text-align: right;">n</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Astrocyte</td>
<td style="text-align: left;">Astrocyte:Embryonic_stem_cell-derived</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">B_cell</td>
<td style="text-align: left;">B_cell</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">B_cell</td>
<td style="text-align: left;">B_cell:CXCR4-_centrocyte</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">B_cell</td>
<td style="text-align: left;">B_cell:CXCR4+_centroblast</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">B_cell</td>
<td style="text-align: left;">B_cell:Germinal_center</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">B_cell</td>
<td style="text-align: left;">B_cell:immature</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">B_cell</td>
<td style="text-align: left;">B_cell:Memory</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">B_cell</td>
<td style="text-align: left;">B_cell:Naive</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">B_cell</td>
<td style="text-align: left;">B_cell:Plasma_cell</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">BM</td>
<td style="text-align: left;">BM</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="odd">
<td style="text-align: left;">BM &amp; Prog.</td>
<td style="text-align: left;">BM</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">Chondrocytes</td>
<td style="text-align: left;">Chondrocytes:MSC-derived</td>
<td style="text-align: right;">8</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CMP</td>
<td style="text-align: left;">CMP</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">DC</td>
<td style="text-align: left;">DC:monocyte-derived</td>
<td style="text-align: right;">19</td>
</tr>
<tr class="odd">
<td style="text-align: left;">DC</td>
<td style="text-align: left;">DC:monocyte-derived:A._fumigatus_germ_tubes_6h</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">DC</td>
<td style="text-align: left;">DC:monocyte-derived:AEC-conditioned</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="odd">
<td style="text-align: left;">DC</td>
<td style="text-align: left;">DC:monocyte-derived:AM580</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">DC</td>
<td style="text-align: left;">DC:monocyte-derived:anti-DC-SIGN_2h</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">DC</td>
<td style="text-align: left;">DC:monocyte-derived:antiCD40/VAF347</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">DC</td>
<td style="text-align: left;">DC:monocyte-derived:CD40L</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">DC</td>
<td style="text-align: left;">DC:monocyte-derived:Galectin-1</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">DC</td>
<td style="text-align: left;">DC:monocyte-derived:immature</td>
<td style="text-align: right;">20</td>
</tr>
<tr class="odd">
<td style="text-align: left;">DC</td>
<td style="text-align: left;">DC:monocyte-derived:LPS</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="even">
<td style="text-align: left;">DC</td>
<td style="text-align: left;">DC:monocyte-derived:mature</td>
<td style="text-align: right;">11</td>
</tr>
<tr class="odd">
<td style="text-align: left;">DC</td>
<td style="text-align: left;">DC:monocyte-derived:Poly(IC)</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">DC</td>
<td style="text-align: left;">DC:monocyte-derived:rosiglitazone</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">DC</td>
<td style="text-align: left;">DC:monocyte-derived:rosiglitazone/AGN193109</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">DC</td>
<td style="text-align: left;">DC:monocyte-derived:Schuler_treatment</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Embryonic_stem_cells</td>
<td style="text-align: left;">Embryonic_stem_cells</td>
<td style="text-align: right;">17</td>
</tr>
<tr class="even">
<td style="text-align: left;">Endothelial_cells</td>
<td style="text-align: left;">Endothelial_cells:blood_vessel</td>
<td style="text-align: right;">8</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Endothelial_cells</td>
<td style="text-align: left;">Endothelial_cells:HUVEC</td>
<td style="text-align: right;">16</td>
</tr>
<tr class="even">
<td style="text-align: left;">Endothelial_cells</td>
<td style="text-align: left;">Endothelial_cells:HUVEC:B._anthracis_LT</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Endothelial_cells</td>
<td style="text-align: left;">Endothelial_cells:HUVEC:Borrelia_burgdorferi</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Endothelial_cells</td>
<td style="text-align: left;">Endothelial_cells:HUVEC:FPV-infected</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Endothelial_cells</td>
<td style="text-align: left;">Endothelial_cells:HUVEC:H5N1-infected</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">Endothelial_cells</td>
<td style="text-align: left;">Endothelial_cells:HUVEC:IFNg</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Endothelial_cells</td>
<td style="text-align: left;">Endothelial_cells:HUVEC:IL-1b</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">Endothelial_cells</td>
<td style="text-align: left;">Endothelial_cells:HUVEC:PR8-infected</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Endothelial_cells</td>
<td style="text-align: left;">Endothelial_cells:HUVEC:Serum_Amyloid_A</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="even">
<td style="text-align: left;">Endothelial_cells</td>
<td style="text-align: left;">Endothelial_cells:HUVEC:VEGF</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Endothelial_cells</td>
<td style="text-align: left;">Endothelial_cells:lymphatic</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="even">
<td style="text-align: left;">Endothelial_cells</td>
<td style="text-align: left;">Endothelial_cells:lymphatic:KSHV</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Endothelial_cells</td>
<td style="text-align: left;">Endothelial_cells:lymphatic:TNFa_48h</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">Epithelial_cells</td>
<td style="text-align: left;">Epithelial_cells:bladder</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Epithelial_cells</td>
<td style="text-align: left;">Epithelial_cells:bronchial</td>
<td style="text-align: right;">10</td>
</tr>
<tr class="even">
<td style="text-align: left;">Erythroblast</td>
<td style="text-align: left;">Erythroblast</td>
<td style="text-align: right;">8</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Fibroblasts</td>
<td style="text-align: left;">Fibroblasts:breast</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="even">
<td style="text-align: left;">Fibroblasts</td>
<td style="text-align: left;">Fibroblasts:foreskin</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Gametocytes</td>
<td style="text-align: left;">Gametocytes:oocyte</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">Gametocytes</td>
<td style="text-align: left;">Gametocytes:spermatocyte</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GMP</td>
<td style="text-align: left;">GMP</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Hepatocytes</td>
<td style="text-align: left;">Hepatocytes</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">HSC_-G-CSF</td>
<td style="text-align: left;">HSC_-G-CSF</td>
<td style="text-align: right;">10</td>
</tr>
<tr class="even">
<td style="text-align: left;">HSC_CD34+</td>
<td style="text-align: left;">HSC_CD34+</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="odd">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:adipose_stem_cell-derived:lentiviral</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:adipose_stem_cell-derived:minicircle-derived</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:adipose_stem_cells</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:CRL2097_foreskin</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:CRL2097_foreskin-derived:d20_hepatic_diff</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:CRL2097_foreskin-derived:undiff.</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:fibroblast-derived:Direct_del._reprog</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:fibroblast-derived:Retroviral_transf</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:fibroblasts</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:foreskin_fibrobasts</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:iPS:minicircle-derived</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="even">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:PDB_1lox-17Puro-10</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:PDB_1lox-17Puro-5</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:PDB_1lox-21Puro-20</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:PDB_1lox-21Puro-26</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:PDB_2lox-17</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:PDB_2lox-21</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:PDB_2lox-22</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:PDB_2lox-5</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:PDB_fibroblasts</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:skin_fibroblast</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">iPS_cells</td>
<td style="text-align: left;">iPS_cells:skin_fibroblast-derived</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Keratinocytes</td>
<td style="text-align: left;">Keratinocytes</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">Keratinocytes</td>
<td style="text-align: left;">Keratinocytes:IFNg</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Keratinocytes</td>
<td style="text-align: left;">Keratinocytes:IL19</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">Keratinocytes</td>
<td style="text-align: left;">Keratinocytes:IL1b</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Keratinocytes</td>
<td style="text-align: left;">Keratinocytes:IL20</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">Keratinocytes</td>
<td style="text-align: left;">Keratinocytes:IL22</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Keratinocytes</td>
<td style="text-align: left;">Keratinocytes:IL24</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">Keratinocytes</td>
<td style="text-align: left;">Keratinocytes:IL26</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Keratinocytes</td>
<td style="text-align: left;">Keratinocytes:KGF</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">Macrophage</td>
<td style="text-align: left;">Macrophage:Alveolar</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Macrophage</td>
<td style="text-align: left;">Macrophage:Alveolar:B._anthacis_spores</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">Macrophage</td>
<td style="text-align: left;">Macrophage:monocyte-derived</td>
<td style="text-align: right;">26</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Macrophage</td>
<td style="text-align: left;">Macrophage:monocyte-derived:IFNa</td>
<td style="text-align: right;">9</td>
</tr>
<tr class="even">
<td style="text-align: left;">Macrophage</td>
<td style="text-align: left;">Macrophage:monocyte-derived:IL-4/cntrl</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Macrophage</td>
<td style="text-align: left;">Macrophage:monocyte-derived:IL-4/Dex/cntrl</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="even">
<td style="text-align: left;">Macrophage</td>
<td style="text-align: left;">Macrophage:monocyte-derived:IL-4/Dex/TGFb</td>
<td style="text-align: right;">10</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Macrophage</td>
<td style="text-align: left;">Macrophage:monocyte-derived:IL-4/TGFb</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="even">
<td style="text-align: left;">Macrophage</td>
<td style="text-align: left;">Macrophage:monocyte-derived:M-CSF</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Macrophage</td>
<td style="text-align: left;">Macrophage:monocyte-derived:M-CSF/IFNg</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Macrophage</td>
<td style="text-align: left;">Macrophage:monocyte-derived:M-CSF/IFNg/Pam3Cys</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Macrophage</td>
<td style="text-align: left;">Macrophage:monocyte-derived:M-CSF/Pam3Cys</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Macrophage</td>
<td style="text-align: left;">Macrophage:monocyte-derived:S._aureus</td>
<td style="text-align: right;">15</td>
</tr>
<tr class="odd">
<td style="text-align: left;">MEP</td>
<td style="text-align: left;">MEP</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Monocyte</td>
<td style="text-align: left;">Monocyte</td>
<td style="text-align: right;">27</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Monocyte</td>
<td style="text-align: left;">Monocyte:anti-FcgRIIB</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Monocyte</td>
<td style="text-align: left;">Monocyte:CD14+</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Monocyte</td>
<td style="text-align: left;">Monocyte:CD16-</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="even">
<td style="text-align: left;">Monocyte</td>
<td style="text-align: left;">Monocyte:CD16+</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Monocyte</td>
<td style="text-align: left;">Monocyte:CXCL4</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Monocyte</td>
<td style="text-align: left;">Monocyte:F._tularensis_novicida</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Monocyte</td>
<td style="text-align: left;">Monocyte:leukotriene_D4</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Monocyte</td>
<td style="text-align: left;">Monocyte:MCSF</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Monocyte</td>
<td style="text-align: left;">Monocyte:S._typhimurium_flagellin</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">MSC</td>
<td style="text-align: left;">MSC</td>
<td style="text-align: right;">9</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Myelocyte</td>
<td style="text-align: left;">Myelocyte</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Neuroepithelial_cell</td>
<td style="text-align: left;">Neuroepithelial_cell:ESC-derived</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Neurons</td>
<td style="text-align: left;">Neurons:adrenal_medulla_cell_line</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="even">
<td style="text-align: left;">Neurons</td>
<td style="text-align: left;">Neurons:ES_cell-derived_neural_precursor</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Neurons</td>
<td style="text-align: left;">Neurons:Schwann_cell</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Neutrophils</td>
<td style="text-align: left;">Neutrophil</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Neutrophils</td>
<td style="text-align: left;">Neutrophil:commensal_E._coli_MG1655</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Neutrophils</td>
<td style="text-align: left;">Neutrophil:GM-CSF_IFNg</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Neutrophils</td>
<td style="text-align: left;">Neutrophil:inflam</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Neutrophils</td>
<td style="text-align: left;">Neutrophil:LPS</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Neutrophils</td>
<td style="text-align: left;">Neutrophil:uropathogenic_E._coli_UTI89</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">NK_cell</td>
<td style="text-align: left;">NK_cell</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NK_cell</td>
<td style="text-align: left;">NK_cell:CD56hiCD62L+</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">NK_cell</td>
<td style="text-align: left;">NK_cell:IL2</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Osteoblasts</td>
<td style="text-align: left;">Osteoblasts</td>
<td style="text-align: right;">9</td>
</tr>
<tr class="even">
<td style="text-align: left;">Osteoblasts</td>
<td style="text-align: left;">Osteoblasts:BMP2</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Platelets</td>
<td style="text-align: left;">Platelets</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="even">
<td style="text-align: left;">Pre-B_cell_CD34-</td>
<td style="text-align: left;">Pre-B_cell_CD34-</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Pro-B_cell_CD34+</td>
<td style="text-align: left;">Pro-B_cell_CD34+</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Pro-Myelocyte</td>
<td style="text-align: left;">Pro-Myelocyte</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Smooth_muscle_cells</td>
<td style="text-align: left;">Smooth_muscle_cells:bronchial</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">Smooth_muscle_cells</td>
<td style="text-align: left;">Smooth_muscle_cells:bronchial:vit_D</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Smooth_muscle_cells</td>
<td style="text-align: left;">Smooth_muscle_cells:umbilical_vein</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Smooth_muscle_cells</td>
<td style="text-align: left;">Smooth_muscle_cells:vascular</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Smooth_muscle_cells</td>
<td style="text-align: left;">Smooth_muscle_cells:vascular:IL-17</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">T_cells</td>
<td style="text-align: left;">T_cell:CCR10-CLA+1,25(OH)2_vit_D3/IL-12</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">T_cells</td>
<td style="text-align: left;">T_cell:CCR10+CLA+1,25(OH)2_vit_D3/IL-12</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">T_cells</td>
<td style="text-align: left;">T_cell:CD4+</td>
<td style="text-align: right;">12</td>
</tr>
<tr class="odd">
<td style="text-align: left;">T_cells</td>
<td style="text-align: left;">T_cell:CD4+_central_memory</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="even">
<td style="text-align: left;">T_cells</td>
<td style="text-align: left;">T_cell:CD4+_effector_memory</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">T_cells</td>
<td style="text-align: left;">T_cell:CD4+_Naive</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="even">
<td style="text-align: left;">T_cells</td>
<td style="text-align: left;">T_cell:CD8+</td>
<td style="text-align: right;">16</td>
</tr>
<tr class="odd">
<td style="text-align: left;">T_cells</td>
<td style="text-align: left;">T_cell:CD8+_Central_memory</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">T_cells</td>
<td style="text-align: left;">T_cell:CD8+_effector_memory</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">T_cells</td>
<td style="text-align: left;">T_cell:CD8+_effector_memory_RA</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">T_cells</td>
<td style="text-align: left;">T_cell:CD8+_naive</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">T_cells</td>
<td style="text-align: left;">T_cell:effector</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">T_cells</td>
<td style="text-align: left;">T_cell:gamma-delta</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">T_cells</td>
<td style="text-align: left;">T_cell:Treg:Naive</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Tissue_stem_cells</td>
<td style="text-align: left;">Tissue_stem_cells:adipose-derived_MSC_AM3</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Tissue_stem_cells</td>
<td style="text-align: left;">Tissue_stem_cells:BM_MSC</td>
<td style="text-align: right;">8</td>
</tr>
<tr class="even">
<td style="text-align: left;">Tissue_stem_cells</td>
<td style="text-align: left;">Tissue_stem_cells:BM_MSC:BMP2</td>
<td style="text-align: right;">12</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Tissue_stem_cells</td>
<td style="text-align: left;">Tissue_stem_cells:BM_MSC:osteogenic</td>
<td style="text-align: right;">8</td>
</tr>
<tr class="even">
<td style="text-align: left;">Tissue_stem_cells</td>
<td style="text-align: left;">Tissue_stem_cells:BM_MSC:TGFb3</td>
<td style="text-align: right;">11</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Tissue_stem_cells</td>
<td style="text-align: left;">Tissue_stem_cells:CD326-CD56+</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">Tissue_stem_cells</td>
<td style="text-align: left;">Tissue_stem_cells:dental_pulp</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Tissue_stem_cells</td>
<td style="text-align: left;">Tissue_stem_cells:iliac_MSC</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">Tissue_stem_cells</td>
<td style="text-align: left;">Tissue_stem_cells:lipoma-derived_MSC</td>
<td style="text-align: right;">2</td>
</tr>
</tbody>
</table>
</div>
</details>
<details>
<summary>
<code>BlueprintEncodeData</code> Labels
</summary>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>be</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/SingleR/man/datasets.html'>BlueprintEncodeData</a></span><span class='op'>(</span><span class='op'>)</span>
<span class='fu'>.adf</span><span class='op'>(</span><span class='fu'>colData</span><span class='op'>(</span><span class='va'>be</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/count.html'>count</a></span><span class='op'>(</span><span class='va'>label.main</span>, <span class='va'>label.fine</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/arrange.html'>arrange</a></span><span class='op'>(</span><span class='va'>label.main</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>knitr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/knitr/man/kable.html'>kable</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<table>
<thead>
<tr class="header">
<th style="text-align: left;">label.main</th>
<th style="text-align: left;">label.fine</th>
<th style="text-align: right;">n</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Adipocytes</td>
<td style="text-align: left;">Adipocytes</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="even">
<td style="text-align: left;">Adipocytes</td>
<td style="text-align: left;">Preadipocytes</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Astrocytes</td>
<td style="text-align: left;">Astrocytes</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">B-cells</td>
<td style="text-align: left;">Class-switched memory B-cells</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">B-cells</td>
<td style="text-align: left;">Memory B-cells</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">B-cells</td>
<td style="text-align: left;">naive B-cells</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">B-cells</td>
<td style="text-align: left;">Plasma cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">CD4+ T-cells</td>
<td style="text-align: left;">CD4+ T-cells</td>
<td style="text-align: right;">11</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CD4+ T-cells</td>
<td style="text-align: left;">CD4+ Tcm</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">CD4+ T-cells</td>
<td style="text-align: left;">CD4+ Tem</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CD4+ T-cells</td>
<td style="text-align: left;">Tregs</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">CD8+ T-cells</td>
<td style="text-align: left;">CD8+ T-cells</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CD8+ T-cells</td>
<td style="text-align: left;">CD8+ Tcm</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">CD8+ T-cells</td>
<td style="text-align: left;">CD8+ Tem</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Chondrocytes</td>
<td style="text-align: left;">Chondrocytes</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">DC</td>
<td style="text-align: left;">DC</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Endothelial cells</td>
<td style="text-align: left;">Endothelial cells</td>
<td style="text-align: right;">18</td>
</tr>
<tr class="even">
<td style="text-align: left;">Endothelial cells</td>
<td style="text-align: left;">mv Endothelial cells</td>
<td style="text-align: right;">8</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Eosinophils</td>
<td style="text-align: left;">Eosinophils</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">Epithelial cells</td>
<td style="text-align: left;">Epithelial cells</td>
<td style="text-align: right;">18</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Erythrocytes</td>
<td style="text-align: left;">Erythrocytes</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="even">
<td style="text-align: left;">Fibroblasts</td>
<td style="text-align: left;">Fibroblasts</td>
<td style="text-align: right;">20</td>
</tr>
<tr class="odd">
<td style="text-align: left;">HSC</td>
<td style="text-align: left;">CLP</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="even">
<td style="text-align: left;">HSC</td>
<td style="text-align: left;">CMP</td>
<td style="text-align: right;">11</td>
</tr>
<tr class="odd">
<td style="text-align: left;">HSC</td>
<td style="text-align: left;">GMP</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">HSC</td>
<td style="text-align: left;">HSC</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="odd">
<td style="text-align: left;">HSC</td>
<td style="text-align: left;">Megakaryocytes</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="even">
<td style="text-align: left;">HSC</td>
<td style="text-align: left;">MEP</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">HSC</td>
<td style="text-align: left;">MPP</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Keratinocytes</td>
<td style="text-align: left;">Keratinocytes</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Macrophages</td>
<td style="text-align: left;">Macrophages</td>
<td style="text-align: right;">18</td>
</tr>
<tr class="even">
<td style="text-align: left;">Macrophages</td>
<td style="text-align: left;">Macrophages M1</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Macrophages</td>
<td style="text-align: left;">Macrophages M2</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Melanocytes</td>
<td style="text-align: left;">Melanocytes</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Mesangial cells</td>
<td style="text-align: left;">Mesangial cells</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Monocytes</td>
<td style="text-align: left;">Monocytes</td>
<td style="text-align: right;">16</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Myocytes</td>
<td style="text-align: left;">Myocytes</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Neurons</td>
<td style="text-align: left;">Neurons</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Neutrophils</td>
<td style="text-align: left;">Neutrophils</td>
<td style="text-align: right;">23</td>
</tr>
<tr class="even">
<td style="text-align: left;">NK cells</td>
<td style="text-align: left;">NK cells</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Pericytes</td>
<td style="text-align: left;">Pericytes</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Skeletal muscle</td>
<td style="text-align: left;">Skeletal muscle</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Smooth muscle</td>
<td style="text-align: left;">Smooth muscle</td>
<td style="text-align: right;">16</td>
</tr>
</tbody>
</table>
</div>
</details>
<details>
<summary>
<code>DatabaseImmuneCellExpressionData</code> Labels
</summary>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>dice</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/SingleR/man/datasets.html'>DatabaseImmuneCellExpressionData</a></span><span class='op'>(</span><span class='op'>)</span>
<span class='fu'>.adf</span><span class='op'>(</span><span class='fu'>colData</span><span class='op'>(</span><span class='va'>dice</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/count.html'>count</a></span><span class='op'>(</span><span class='va'>label.main</span>, <span class='va'>label.fine</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/arrange.html'>arrange</a></span><span class='op'>(</span><span class='va'>label.main</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>knitr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/knitr/man/kable.html'>kable</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<table>
<thead>
<tr class="header">
<th style="text-align: left;">label.main</th>
<th style="text-align: left;">label.fine</th>
<th style="text-align: right;">n</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">B cells</td>
<td style="text-align: left;">B cells, naive</td>
<td style="text-align: right;">106</td>
</tr>
<tr class="even">
<td style="text-align: left;">Monocytes</td>
<td style="text-align: left;">Monocytes, CD14+</td>
<td style="text-align: right;">106</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Monocytes</td>
<td style="text-align: left;">Monocytes, CD16+</td>
<td style="text-align: right;">105</td>
</tr>
<tr class="even">
<td style="text-align: left;">NK cells</td>
<td style="text-align: left;">NK cells</td>
<td style="text-align: right;">105</td>
</tr>
<tr class="odd">
<td style="text-align: left;">T cells, CD4+</td>
<td style="text-align: left;">T cells, CD4+, memory TREG</td>
<td style="text-align: right;">104</td>
</tr>
<tr class="even">
<td style="text-align: left;">T cells, CD4+</td>
<td style="text-align: left;">T cells, CD4+, naive</td>
<td style="text-align: right;">103</td>
</tr>
<tr class="odd">
<td style="text-align: left;">T cells, CD4+</td>
<td style="text-align: left;">T cells, CD4+, naive TREG</td>
<td style="text-align: right;">104</td>
</tr>
<tr class="even">
<td style="text-align: left;">T cells, CD4+</td>
<td style="text-align: left;">T cells, CD4+, naive, stimulated</td>
<td style="text-align: right;">102</td>
</tr>
<tr class="odd">
<td style="text-align: left;">T cells, CD4+</td>
<td style="text-align: left;">T cells, CD4+, TFH</td>
<td style="text-align: right;">104</td>
</tr>
<tr class="even">
<td style="text-align: left;">T cells, CD4+</td>
<td style="text-align: left;">T cells, CD4+, Th1</td>
<td style="text-align: right;">104</td>
</tr>
<tr class="odd">
<td style="text-align: left;">T cells, CD4+</td>
<td style="text-align: left;">T cells, CD4+, Th1_17</td>
<td style="text-align: right;">104</td>
</tr>
<tr class="even">
<td style="text-align: left;">T cells, CD4+</td>
<td style="text-align: left;">T cells, CD4+, Th17</td>
<td style="text-align: right;">104</td>
</tr>
<tr class="odd">
<td style="text-align: left;">T cells, CD4+</td>
<td style="text-align: left;">T cells, CD4+, Th2</td>
<td style="text-align: right;">104</td>
</tr>
<tr class="even">
<td style="text-align: left;">T cells, CD8+</td>
<td style="text-align: left;">T cells, CD8+, naive</td>
<td style="text-align: right;">104</td>
</tr>
<tr class="odd">
<td style="text-align: left;">T cells, CD8+</td>
<td style="text-align: left;">T cells, CD8+, naive, stimulated</td>
<td style="text-align: right;">102</td>
</tr>
</tbody>
</table>
</div>
</details>
<details>
<summary>
<code>NovershternHematopoieticData</code> Labels
</summary>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>nh</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/SingleR/man/datasets.html'>NovershternHematopoieticData</a></span><span class='op'>(</span><span class='op'>)</span>
<span class='fu'>.adf</span><span class='op'>(</span><span class='fu'>colData</span><span class='op'>(</span><span class='va'>nh</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/count.html'>count</a></span><span class='op'>(</span><span class='va'>label.main</span>, <span class='va'>label.fine</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/arrange.html'>arrange</a></span><span class='op'>(</span><span class='va'>label.main</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>knitr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/knitr/man/kable.html'>kable</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<table>
<thead>
<tr class="header">
<th style="text-align: left;">label.main</th>
<th style="text-align: left;">label.fine</th>
<th style="text-align: right;">n</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">B cells</td>
<td style="text-align: left;">Early B cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">B cells</td>
<td style="text-align: left;">Mature B cells</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="odd">
<td style="text-align: left;">B cells</td>
<td style="text-align: left;">Mature B cells class able to switch</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="even">
<td style="text-align: left;">B cells</td>
<td style="text-align: left;">Mature B cells class switched</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="odd">
<td style="text-align: left;">B cells</td>
<td style="text-align: left;">Naive B cells</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="even">
<td style="text-align: left;">B cells</td>
<td style="text-align: left;">Pro B cells</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Basophils</td>
<td style="text-align: left;">Basophils</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="even">
<td style="text-align: left;">CD4+ T cells</td>
<td style="text-align: left;">CD4+ Central Memory</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CD4+ T cells</td>
<td style="text-align: left;">CD4+ Effector Memory</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="even">
<td style="text-align: left;">CD4+ T cells</td>
<td style="text-align: left;">Naive CD4+ T cells</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CD8+ T cells</td>
<td style="text-align: left;">CD8+ Central Memory</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="even">
<td style="text-align: left;">CD8+ T cells</td>
<td style="text-align: left;">CD8+ Effector Memory</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CD8+ T cells</td>
<td style="text-align: left;">CD8+ Effector Memory RA</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">CD8+ T cells</td>
<td style="text-align: left;">Naive CD8+ T cells</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CMPs</td>
<td style="text-align: left;">Common myeloid progenitors</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Dendritic cells</td>
<td style="text-align: left;">Myeloid Dendritic Cells</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Dendritic cells</td>
<td style="text-align: left;">Plasmacytoid Dendritic Cells</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="even">
<td style="text-align: left;">Eosinophils</td>
<td style="text-align: left;">Eosinophils</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Erythroid cells</td>
<td style="text-align: left;">Erythroid_CD34- CD71- GlyA+</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="even">
<td style="text-align: left;">Erythroid cells</td>
<td style="text-align: left;">Erythroid_CD34- CD71+ GlyA-</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Erythroid cells</td>
<td style="text-align: left;">Erythroid_CD34- CD71+ GlyA+</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="even">
<td style="text-align: left;">Erythroid cells</td>
<td style="text-align: left;">Erythroid_CD34- CD71lo GlyA+</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Erythroid cells</td>
<td style="text-align: left;">Erythroid_CD34+ CD71+ GlyA-</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="even">
<td style="text-align: left;">GMPs</td>
<td style="text-align: left;">Granulocyte/monocyte progenitors</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Granulocytes</td>
<td style="text-align: left;">Colony Forming Unit-Granulocytes</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="even">
<td style="text-align: left;">Granulocytes</td>
<td style="text-align: left;">Granulocytes (Neutrophilic Metamyelocytes)</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Granulocytes</td>
<td style="text-align: left;">Granulocytes (Neutrophils)</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">HSCs</td>
<td style="text-align: left;">Hematopoietic stem cells_CD133+ CD34dim</td>
<td style="text-align: right;">10</td>
</tr>
<tr class="odd">
<td style="text-align: left;">HSCs</td>
<td style="text-align: left;">Hematopoietic stem cells_CD38- CD34+</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Megakaryocytes</td>
<td style="text-align: left;">Colony Forming Unit-Megakaryocytic</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Megakaryocytes</td>
<td style="text-align: left;">Megakaryocytes</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="even">
<td style="text-align: left;">MEPs</td>
<td style="text-align: left;">Megakaryocyte/erythroid progenitors</td>
<td style="text-align: right;">9</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Monocytes</td>
<td style="text-align: left;">Colony Forming Unit-Monocytes</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Monocytes</td>
<td style="text-align: left;">Monocytes</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NK cells</td>
<td style="text-align: left;">Mature NK cells_CD56- CD16- CD3-</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="even">
<td style="text-align: left;">NK cells</td>
<td style="text-align: left;">Mature NK cells_CD56- CD16+ CD3-</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NK cells</td>
<td style="text-align: left;">Mature NK cells_CD56+ CD16+ CD3-</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="even">
<td style="text-align: left;">NK T cells</td>
<td style="text-align: left;">NK T cells</td>
<td style="text-align: right;">4</td>
</tr>
</tbody>
</table>
</div>
</details>
<details>
<summary>
<code>MonacoImmuneData</code> Labels
</summary>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>mi</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/SingleR/man/datasets.html'>MonacoImmuneData</a></span><span class='op'>(</span><span class='op'>)</span>
<span class='fu'>.adf</span><span class='op'>(</span><span class='fu'>colData</span><span class='op'>(</span><span class='va'>mi</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/count.html'>count</a></span><span class='op'>(</span><span class='va'>label.main</span>, <span class='va'>label.fine</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/arrange.html'>arrange</a></span><span class='op'>(</span><span class='va'>label.main</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>knitr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/knitr/man/kable.html'>kable</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<table>
<thead>
<tr class="header">
<th style="text-align: left;">label.main</th>
<th style="text-align: left;">label.fine</th>
<th style="text-align: right;">n</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">B cells</td>
<td style="text-align: left;">Exhausted B cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">B cells</td>
<td style="text-align: left;">Naive B cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">B cells</td>
<td style="text-align: left;">Non-switched memory B cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">B cells</td>
<td style="text-align: left;">Plasmablasts</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">B cells</td>
<td style="text-align: left;">Switched memory B cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Basophils</td>
<td style="text-align: left;">Low-density basophils</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CD4+ T cells</td>
<td style="text-align: left;">Follicular helper T cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">CD4+ T cells</td>
<td style="text-align: left;">Naive CD4 T cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CD4+ T cells</td>
<td style="text-align: left;">T regulatory cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">CD4+ T cells</td>
<td style="text-align: left;">Terminal effector CD4 T cells</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CD4+ T cells</td>
<td style="text-align: left;">Th1 cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">CD4+ T cells</td>
<td style="text-align: left;">Th1/Th17 cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CD4+ T cells</td>
<td style="text-align: left;">Th17 cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">CD4+ T cells</td>
<td style="text-align: left;">Th2 cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CD8+ T cells</td>
<td style="text-align: left;">Central memory CD8 T cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">CD8+ T cells</td>
<td style="text-align: left;">Effector memory CD8 T cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CD8+ T cells</td>
<td style="text-align: left;">Naive CD8 T cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">CD8+ T cells</td>
<td style="text-align: left;">Terminal effector CD8 T cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Dendritic cells</td>
<td style="text-align: left;">Myeloid dendritic cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Dendritic cells</td>
<td style="text-align: left;">Plasmacytoid dendritic cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Monocytes</td>
<td style="text-align: left;">Classical monocytes</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Monocytes</td>
<td style="text-align: left;">Intermediate monocytes</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Monocytes</td>
<td style="text-align: left;">Non classical monocytes</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Neutrophils</td>
<td style="text-align: left;">Low-density neutrophils</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NK cells</td>
<td style="text-align: left;">Natural killer cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Progenitors</td>
<td style="text-align: left;">Progenitor cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">T cells</td>
<td style="text-align: left;">MAIT cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">T cells</td>
<td style="text-align: left;">Non-Vd2 gd T cells</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">T cells</td>
<td style="text-align: left;">Vd2 gd T cells</td>
<td style="text-align: right;">4</td>
</tr>
</tbody>
</table>
</div>
</details>
<h2 id="results">Results</h2>
<p><em><a href="https://bioconductor.org/packages/3.12/SingleR">SingleR</a></em> can be used to annotate at the cell-level or the cluster-level. The latter is generally used to speed-up calculations, which is not necessary for this sized data set. Therefore, I have annotated at the cell-level to obtain a higher resolution analysis and choose <code>MonacoImmuneData</code> (<code>MI</code>) as our annotation reference due to its higher relevance to this samples.</p>
<h2 id="monacoimmunedata"><code>MonacoImmuneData</code></h2>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>mi</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/SingleR/man/datasets.html'>MonacoImmuneData</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>Each sample has a ‘main’ and a ‘fine’ cell type label, summarised below.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>.adf</span><span class='op'>(</span><span class='fu'>colData</span><span class='op'>(</span><span class='va'>mi</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='https://rdrr.io/pkg/janitor/man/tabyl.html'>tabyl</a></span><span class='op'>(</span><span class='va'>label.main</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='https://rdrr.io/pkg/janitor/man/adorn_pct_formatting.html'>adorn_pct_formatting</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>knitr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/knitr/man/kable.html'>kable</a></span><span class='op'>(</span>
    caption <span class='op'>=</span> <span class='st'>"Summary of main cell types in the `MonacoImmuneData` reference set."</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<table>
<caption><span id="tab:unnamed-chunk-11">Table 1: </span>Summary of main cell types in the <code>MonacoImmuneData</code> reference set.</caption>
<thead>
<tr class="header">
<th style="text-align: left;">label.main</th>
<th style="text-align: right;">n</th>
<th style="text-align: left;">percent</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">B cells</td>
<td style="text-align: right;">20</td>
<td style="text-align: left;">17.5%</td>
</tr>
<tr class="even">
<td style="text-align: left;">Basophils</td>
<td style="text-align: right;">4</td>
<td style="text-align: left;">3.5%</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CD4+ T cells</td>
<td style="text-align: right;">30</td>
<td style="text-align: left;">26.3%</td>
</tr>
<tr class="even">
<td style="text-align: left;">CD8+ T cells</td>
<td style="text-align: right;">16</td>
<td style="text-align: left;">14.0%</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Dendritic cells</td>
<td style="text-align: right;">8</td>
<td style="text-align: left;">7.0%</td>
</tr>
<tr class="even">
<td style="text-align: left;">Monocytes</td>
<td style="text-align: right;">12</td>
<td style="text-align: left;">10.5%</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Neutrophils</td>
<td style="text-align: right;">4</td>
<td style="text-align: left;">3.5%</td>
</tr>
<tr class="even">
<td style="text-align: left;">NK cells</td>
<td style="text-align: right;">4</td>
<td style="text-align: left;">3.5%</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Progenitors</td>
<td style="text-align: right;">4</td>
<td style="text-align: left;">3.5%</td>
</tr>
<tr class="even">
<td style="text-align: left;">T cells</td>
<td style="text-align: right;">12</td>
<td style="text-align: left;">10.5%</td>
</tr>
</tbody>
</table>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://pkgs.rstudio.com/rmarkdown/reference/paged_table.html'>paged_table</a></span><span class='op'>(</span>
  <span class='fu'>.adf</span><span class='op'>(</span><span class='fu'>colData</span><span class='op'>(</span><span class='va'>mi</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
    <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/count.html'>count</a></span><span class='op'>(</span><span class='va'>label.main</span>, <span class='va'>label.fine</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
    <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/arrange.html'>arrange</a></span><span class='op'>(</span><span class='va'>label.main</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["label.main"],"name":[1],"type":["chr"],"align":["left"]},{"label":["label.fine"],"name":[2],"type":["chr"],"align":["left"]},{"label":["n"],"name":[3],"type":["int"],"align":["right"]}],"data":[{"1":"B cells","2":"Exhausted B cells","3":"4"},{"1":"B cells","2":"Naive B cells","3":"4"},{"1":"B cells","2":"Non-switched memory B cells","3":"4"},{"1":"B cells","2":"Plasmablasts","3":"4"},{"1":"B cells","2":"Switched memory B cells","3":"4"},{"1":"Basophils","2":"Low-density basophils","3":"4"},{"1":"CD4+ T cells","2":"Follicular helper T cells","3":"4"},{"1":"CD4+ T cells","2":"Naive CD4 T cells","3":"4"},{"1":"CD4+ T cells","2":"T regulatory cells","3":"4"},{"1":"CD4+ T cells","2":"Terminal effector CD4 T cells","3":"2"},{"1":"CD4+ T cells","2":"Th1 cells","3":"4"},{"1":"CD4+ T cells","2":"Th1/Th17 cells","3":"4"},{"1":"CD4+ T cells","2":"Th17 cells","3":"4"},{"1":"CD4+ T cells","2":"Th2 cells","3":"4"},{"1":"CD8+ T cells","2":"Central memory CD8 T cells","3":"4"},{"1":"CD8+ T cells","2":"Effector memory CD8 T cells","3":"4"},{"1":"CD8+ T cells","2":"Naive CD8 T cells","3":"4"},{"1":"CD8+ T cells","2":"Terminal effector CD8 T cells","3":"4"},{"1":"Dendritic cells","2":"Myeloid dendritic cells","3":"4"},{"1":"Dendritic cells","2":"Plasmacytoid dendritic cells","3":"4"},{"1":"Monocytes","2":"Classical monocytes","3":"4"},{"1":"Monocytes","2":"Intermediate monocytes","3":"4"},{"1":"Monocytes","2":"Non classical monocytes","3":"4"},{"1":"Neutrophils","2":"Low-density neutrophils","3":"4"},{"1":"NK cells","2":"Natural killer cells","3":"4"},{"1":"Progenitors","2":"Progenitor cells","3":"4"},{"1":"T cells","2":"MAIT cells","3":"4"},{"1":"T cells","2":"Non-Vd2 gd T cells","3":"4"},{"1":"T cells","2":"Vd2 gd T cells","3":"4"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<p><strong>NOTE</strong>: Stemmed from the fact that only the <code>Fine label</code> of <code>MonacoImmuneData</code> is relevant to this dataset, we will only focus on exploration with the <code>Fine label</code></p>
<h3 id="fine-label">Fine label</h3>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>pred_mi_cell_fine</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/SingleR/man/SingleR.html'>SingleR</a></span><span class='op'>(</span>
  test <span class='op'>=</span> <span class='va'>sce</span>,
  ref <span class='op'>=</span> <span class='va'>mi</span><span class='op'>[</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/grep.html'>grepl</a></span><span class='op'>(</span><span class='st'>"^mt|^Rps|^Rpl"</span>, <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>mi</span><span class='op'>)</span><span class='op'>)</span>, <span class='op'>]</span>, 
  labels <span class='op'>=</span> <span class='va'>mi</span><span class='op'>$</span><span class='va'>label.fine</span>,
  BPPARAM <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html'>SerialParam</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>Annotating against the fine labels, we obtain the following cell type labels.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/janitor/man/tabyl.html'>tabyl</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>label.fine <span class='op'>=</span> <span class='va'>pred_mi_cell_fine</span><span class='op'>$</span><span class='va'>labels</span><span class='op'>)</span>, <span class='va'>label.fine</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='https://rdrr.io/pkg/janitor/man/adorn_pct_formatting.html'>adorn_pct_formatting</a></span><span class='op'>(</span>digits <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>%&gt;%</span>  
  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/arrange.html'>arrange</a></span><span class='op'>(</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/desc.html'>desc</a></span><span class='op'>(</span><span class='va'>n</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>knitr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/knitr/man/kable.html'>kable</a></span><span class='op'>(</span>
    caption <span class='op'>=</span> <span class='st'>"Cell label assignments using the fine labels of the `MonacoImmuneData` reference data."</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<table>
<caption><span id="tab:unnamed-chunk-14">Table 2: </span>Cell label assignments using the fine labels of the <code>MonacoImmuneData</code> reference data.</caption>
<thead>
<tr class="header">
<th style="text-align: left;">label.fine</th>
<th style="text-align: right;">n</th>
<th style="text-align: left;">percent</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Vd2 gd T cells</td>
<td style="text-align: right;">867</td>
<td style="text-align: left;">43.1%</td>
</tr>
<tr class="even">
<td style="text-align: left;">Th1 cells</td>
<td style="text-align: right;">297</td>
<td style="text-align: left;">14.8%</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Non-Vd2 gd T cells</td>
<td style="text-align: right;">139</td>
<td style="text-align: left;">6.9%</td>
</tr>
<tr class="even">
<td style="text-align: left;">MAIT cells</td>
<td style="text-align: right;">136</td>
<td style="text-align: left;">6.8%</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Naive CD8 T cells</td>
<td style="text-align: right;">102</td>
<td style="text-align: left;">5.1%</td>
</tr>
<tr class="even">
<td style="text-align: left;">Naive CD4 T cells</td>
<td style="text-align: right;">99</td>
<td style="text-align: left;">4.9%</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Natural killer cells</td>
<td style="text-align: right;">80</td>
<td style="text-align: left;">4.0%</td>
</tr>
<tr class="even">
<td style="text-align: left;">Terminal effector CD4 T cells</td>
<td style="text-align: right;">70</td>
<td style="text-align: left;">3.5%</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Central memory CD8 T cells</td>
<td style="text-align: right;">53</td>
<td style="text-align: left;">2.6%</td>
</tr>
<tr class="even">
<td style="text-align: left;">Follicular helper T cells</td>
<td style="text-align: right;">49</td>
<td style="text-align: left;">2.4%</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Th1/Th17 cells</td>
<td style="text-align: right;">48</td>
<td style="text-align: left;">2.4%</td>
</tr>
<tr class="even">
<td style="text-align: left;">Th17 cells</td>
<td style="text-align: right;">34</td>
<td style="text-align: left;">1.7%</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Effector memory CD8 T cells</td>
<td style="text-align: right;">20</td>
<td style="text-align: left;">1.0%</td>
</tr>
<tr class="even">
<td style="text-align: left;">Th2 cells</td>
<td style="text-align: right;">11</td>
<td style="text-align: left;">0.5%</td>
</tr>
<tr class="odd">
<td style="text-align: left;">T regulatory cells</td>
<td style="text-align: right;">3</td>
<td style="text-align: left;">0.1%</td>
</tr>
<tr class="even">
<td style="text-align: left;">Intermediate monocytes</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">0.1%</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Terminal effector CD8 T cells</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">0.1%</td>
</tr>
<tr class="even">
<td style="text-align: left;">Naive B cells</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">0.0%</td>
</tr>
</tbody>
</table>
</div>
<p>Heatmaps of the annotation scores are shown, with and without clustering by these normalized scores, in Figures <a href="#fig:heat-mi-fine-clustered">4</a> and <a href="#fig:heat-mi-fine">5</a>, and clustering by FACS gates in in Figures <a href="#fig:heat-mi-fine-facs-gating"><strong>??</strong></a>.</p>
<div class="layout-chunk" data-layout="l-body-outset">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/stopifnot.html'>stopifnot</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/identical.html'>identical</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>pred_mi_cell_fine</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/SingleR/man/plotScoreHeatmap.html'>plotScoreHeatmap</a></span><span class='op'>(</span>
  <span class='va'>pred_mi_cell_fine</span>,
  annotation_col <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
    cluster <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span>,
    sample_name <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>sample_name</span>,
    post_hoc_sample_gate <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>post_hoc_sample_gate</span>,
    plate_number <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span>,
    tissue <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>tissue</span>,
    donor <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>donor</span>,
    row.names <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>pred_mi_cell_fine</span><span class='op'>)</span><span class='op'>)</span>,
  annotation_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
    cluster <span class='op'>=</span> <span class='va'>cluster_colours</span>,
    sample_name <span class='op'>=</span> <span class='va'>sample_name_colours</span>,
    post_hoc_sample_gate <span class='op'>=</span> <span class='va'>sample_gate_colours</span>,
    plate_number <span class='op'>=</span> <span class='va'>plate_number_colours</span>,    
    tissue <span class='op'>=</span> <span class='va'>tissue_colours</span>,
    donor <span class='op'>=</span> <span class='va'>donor_colours</span><span class='op'>)</span>,
  labels.use <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='va'>pred_mi_cell_fine</span><span class='op'>$</span><span class='va'>labels</span><span class='op'>)</span>,
  max.labels <span class='op'>=</span> <span class='cn'>Inf</span>,
  normalize <span class='op'>=</span> <span class='cn'>TRUE</span>,
  show.labels <span class='op'>=</span> <span class='cn'>FALSE</span>,
  fontsize <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:heat-mi-fine-clustered"></span>
<img src="C094_Pellicci.annotate_files/heat-mi-fine-clustered-1.png" alt="Heatmap of normalized scores using the fine labels from the `MonacoImmuneData` reference set. Cells are columns, rows are labels. Columns are ordered by clustering the normalized scores." width="624" />
<p class="caption">
Figure 4: Heatmap of normalized scores using the fine labels from the <code>MonacoImmuneData</code> reference set. Cells are columns, rows are labels. Columns are ordered by clustering the normalized scores.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body-outset">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/stopifnot.html'>stopifnot</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/identical.html'>identical</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>pred_mi_cell_fine</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/SingleR/man/plotScoreHeatmap.html'>plotScoreHeatmap</a></span><span class='op'>(</span>
  <span class='va'>pred_mi_cell_fine</span>,
  annotation_col <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
    cluster <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span>,
    sample_name <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>sample_name</span>,
    post_hoc_sample_gate <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>post_hoc_sample_gate</span>,
    plate_number <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span>,
    tissue <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>tissue</span>,
    donor <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>donor</span>,
    row.names <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>pred_mi_cell_fine</span><span class='op'>)</span><span class='op'>)</span>,
  annotation_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
    cluster <span class='op'>=</span> <span class='va'>cluster_colours</span>,
    sample_name <span class='op'>=</span> <span class='va'>sample_name_colours</span>,
    post_hoc_sample_gate <span class='op'>=</span> <span class='va'>sample_gate_colours</span>,
    plate_number <span class='op'>=</span> <span class='va'>plate_number_colours</span>,    
    tissue <span class='op'>=</span> <span class='va'>tissue_colours</span>,
    donor <span class='op'>=</span> <span class='va'>donor_colours</span><span class='op'>)</span>,
  cells.order <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span>,
  labels.use <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='va'>pred_mi_cell_fine</span><span class='op'>$</span><span class='va'>labels</span><span class='op'>)</span>,
  max.labels <span class='op'>=</span> <span class='cn'>Inf</span>,
  normalize <span class='op'>=</span> <span class='cn'>TRUE</span>,
  show.labels <span class='op'>=</span> <span class='cn'>FALSE</span>,
  fontsize <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:heat-mi-fine"></span>
<img src="C094_Pellicci.annotate_files/heat-mi-fine-1.png" alt="Heatmap of normalized scores using the fine labels from the `MonacoImmuneData` reference set. Cells are columns, rows are labels. Columns are ordered by and annotated by the column labels." width="624" />
<p class="caption">
Figure 5: Heatmap of normalized scores using the fine labels from the <code>MonacoImmuneData</code> reference set. Cells are columns, rows are labels. Columns are ordered by and annotated by the column labels.
</p>
</div>
</div>
<p>Figure <a href="#fig:umap-mi-fine">6</a> overlays the normalized scores for the 10 labels with the largest maximum scores on the UMAP plot.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>plotScoreReducedDim</span><span class='op'>(</span>
  <span class='va'>pred_mi_cell_fine</span>, <span class='va'>sce</span>,
  max.labels <span class='op'>=</span> <span class='fl'>10</span>,
  ncol <span class='op'>=</span> <span class='fl'>2</span>,
  dimred <span class='op'>=</span> <span class='st'>"UMAP_corrected"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:umap-mi-fine"></span>
<img src="C094_Pellicci.annotate_files/umap-mi-fine-1.png" alt="UMAP overlaid with normalized scores using the fine labels from the `MonacoImmuneData` reference set. Only the 10 labels with the largest maximum scores are shown." width="624" />
<p class="caption">
Figure 6: UMAP overlaid with normalized scores using the fine labels from the <code>MonacoImmuneData</code> reference set. Only the 10 labels with the largest maximum scores are shown.
</p>
</div>
</div>
<p>Figure <a href="#fig:barplot-mi-fine">7</a> breaks down the cell labels for each sample.</p>
<div class="layout-chunk" data-layout="l-body-outset">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>sce</span><span class='op'>$</span><span class='va'>mi_label_fine</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>pred_mi_cell_fine</span><span class='op'>$</span><span class='va'>pruned.labels</span><span class='op'>)</span>
<span class='va'>sce</span><span class='op'>$</span><span class='va'>mi_label_fine_collapsed</span> <span class='op'>&lt;-</span> <span class='fu'>.collapseLabel</span><span class='op'>(</span>
  <span class='va'>pred_mi_cell_fine</span><span class='op'>$</span><span class='va'>pruned.labels</span>,
  <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span><span class='op'>)</span>
<span class='va'>mi_label_fine_colours</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/setNames.html'>setNames</a></span><span class='op'>(</span>
  <span class='fu'>Polychrome</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/Polychrome/man/palettes.html'>glasbey.colors</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/nlevels.html'>nlevels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>mi_label_fine_collapsed</span><span class='op'>)</span> <span class='op'>+</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>[</span><span class='op'>-</span><span class='fl'>1</span><span class='op'>]</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>mi_label_fine_collapsed</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://ggplot2.tidyverse.org/reference/ggplot.html'>ggplot</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/as.data.frame.html'>as.data.frame</a></span><span class='op'>(</span><span class='fu'>colData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>plate_number</span>, fill <span class='op'>=</span> <span class='va'>mi_label_fine_collapsed</span><span class='op'>)</span>,
    position <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/position_stack.html'>position_fill</a></span><span class='op'>(</span>reverse <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Frequency"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>mi_label_fine_colours</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span> 
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:barplot-mi-fine"></span>
<img src="C094_Pellicci.annotate_files/barplot-mi-fine-1.png" alt="Barplot showing cell label assignments for cells from each donor. Cell labels with fewer than 1% of cells in all donor are grouped together as 'other'. Cell labels are ordered by their overall frequency in the data set." width="624" />
<p class="caption">
Figure 7: Barplot showing cell label assignments for cells from each donor. Cell labels with fewer than 1% of cells in all donor are grouped together as ‘other.’ Cell labels are ordered by their overall frequency in the data set.
</p>
</div>
</div>
<p>Finally, we examine the expression of the marker genes for each label in the test dataset. We extract the identity of the markers in the reference data sets and use them to create a heatmap of their expression in the test data set. If a cell in the test dataset is confidently assigned to a particular label, we would expect it to have strong expression of that label’s markers.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>all_markers</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://pkgs.rstudio.com/rmarkdown/reference/metadata.html'>metadata</a></span><span class='op'>(</span><span class='va'>pred_mi_cell_fine</span><span class='op'>)</span><span class='op'>$</span><span class='va'>de.genes</span>
<span class='co'># Get top-10 marker genes compared to each other cell type</span>
<span class='va'>example_label</span> <span class='op'>&lt;-</span> <span class='st'>"Vd2 gd T cells"</span>
<span class='va'>top_markers</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/funprog.html'>Reduce</a></span><span class='op'>(</span><span class='va'>union</span>, <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span><span class='va'>all_markers</span><span class='op'>[[</span><span class='va'>example_label</span><span class='op'>]</span><span class='op'>]</span>, <span class='va'>head</span>, <span class='fl'>10</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>Figures <a href="#fig:heat-mi-fine-markers-row-normalized">8</a> demonstrates this for cells labelled as Vd2 gd T cells. At the very least, it should exhibit up-regulation of those markers relative to cells assigned to other labels. This can be visualized by row-normalizing the data, as done in Figure <a href="#fig:heat-mi-fine-markers-row-normalized">8</a>.</p>
<div class="layout-chunk" data-layout="l-page">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotHeatmap.html'>plotHeatmap</a></span><span class='op'>(</span>
  <span class='va'>sce</span>,
  features <span class='op'>=</span> <span class='va'>top_markers</span>,
  columns <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span>
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>mi_label_fine_collapsed</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>sample_name</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>post_hoc_sample_gate</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>tissue</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>donor</span><span class='op'>)</span>,
  order_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>
    <span class='st'>"mi_label_fine_collapsed"</span>,
    <span class='st'>"cluster"</span>,
    <span class='st'>"sample_name"</span>,
    <span class='st'>"post_hoc_sample_gate"</span>,
    <span class='st'>"plate_number"</span>,
    <span class='st'>"tissue"</span>,
    <span class='st'>"donor"</span><span class='op'>)</span>,
  cluster_cols <span class='op'>=</span> <span class='cn'>FALSE</span>, 
  center <span class='op'>=</span> <span class='cn'>TRUE</span>,
  symmetric <span class='op'>=</span> <span class='cn'>TRUE</span>,
  zlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>3</span>, <span class='fl'>3</span><span class='op'>)</span>,
  show_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
  main <span class='op'>=</span> <span class='va'>example_label</span>,
  column_annotation_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
     mi_label_fine_collapsed <span class='op'>=</span> <span class='va'>mi_label_fine_colours</span>,
     cluster <span class='op'>=</span> <span class='va'>cluster_colours</span>,
     sample_name <span class='op'>=</span> <span class='va'>sample_name_colours</span>,
     post_hoc_sample_gate <span class='op'>=</span> <span class='va'>sample_gate_colours</span>,
     plate_number <span class='op'>=</span> <span class='va'>plate_number_colours</span>,    
     tissue <span class='op'>=</span> <span class='va'>tissue_colours</span>,
     donor <span class='op'>=</span> <span class='va'>donor_colours</span><span class='op'>)</span>,
  color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>hcl.colors</a></span><span class='op'>(</span><span class='fl'>101</span>, <span class='st'>"Blue-Red 3"</span><span class='op'>)</span>,
  fontsize <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:heat-mi-fine-markers-row-normalized"></span>
<img src="C094_Pellicci.annotate_files/heat-mi-fine-markers-row-normalized-1.png" alt="Heatmap of row-normalized log-expression values for the top-10 marker genes of `r example_cluster`. Each column is a sample, each row a gene. Colours are capped at -3 and 3 to preserve dynamic range." width="624" />
<p class="caption">
Figure 8: Heatmap of row-normalized log-expression values for the top-10 marker genes of <code>r example_cluster</code>. Each column is a sample, each row a gene. Colours are capped at -3 and 3 to preserve dynamic range.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-page">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotHeatmap.html'>plotHeatmap</a></span><span class='op'>(</span>
  <span class='va'>sce</span>,
  features <span class='op'>=</span> <span class='va'>top_markers</span>,
  columns <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span>
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>mi_label_fine_collapsed</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>sample_name</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>post_hoc_sample_gate</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>tissue</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>donor</span><span class='op'>)</span>,
  order_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>
    <span class='st'>"cluster"</span>,
    <span class='st'>"mi_label_fine_collapsed"</span>,
    <span class='st'>"sample_name"</span>,
    <span class='st'>"post_hoc_sample_gate"</span>,
    <span class='st'>"plate_number"</span>,
    <span class='st'>"tissue"</span>,
    <span class='st'>"donor"</span><span class='op'>)</span>,
  cluster_cols <span class='op'>=</span> <span class='cn'>FALSE</span>, 
  center <span class='op'>=</span> <span class='cn'>TRUE</span>,
  symmetric <span class='op'>=</span> <span class='cn'>TRUE</span>,
  zlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>3</span>, <span class='fl'>3</span><span class='op'>)</span>,
  show_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
  main <span class='op'>=</span> <span class='va'>example_label</span>,
  column_annotation_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
     cluster <span class='op'>=</span> <span class='va'>cluster_colours</span>,
     mi_label_fine_collapsed <span class='op'>=</span> <span class='va'>mi_label_fine_colours</span>,
     sample_name <span class='op'>=</span> <span class='va'>sample_name_colours</span>,
     post_hoc_sample_gate <span class='op'>=</span> <span class='va'>sample_gate_colours</span>,
     plate_number <span class='op'>=</span> <span class='va'>plate_number_colours</span>,    
     tissue <span class='op'>=</span> <span class='va'>tissue_colours</span>,
     donor <span class='op'>=</span> <span class='va'>donor_colours</span><span class='op'>)</span>,
  color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>hcl.colors</a></span><span class='op'>(</span><span class='fl'>101</span>, <span class='st'>"Blue-Red 3"</span><span class='op'>)</span>,
  fontsize <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:heat-mi-fine-markers-row-normalized-ordered-by-cluster"></span>
<img src="C094_Pellicci.annotate_files/heat-mi-fine-markers-row-normalized-ordered-by-cluster-1.png" alt="Heatmap of row-normalized log-expression values for the top-10 marker genes of `r example_cluster`. Each column is a sample, each row a gene. Colours are capped at -3 and 3 to preserve dynamic range." width="624" />
<p class="caption">
Figure 9: Heatmap of row-normalized log-expression values for the top-10 marker genes of <code>r example_cluster</code>. Each column is a sample, each row a gene. Colours are capped at -3 and 3 to preserve dynamic range.
</p>
</div>
</div>
<p>Similar figures showing the top-10 marker genes for each label are available in <a href="../output/marker_genes/SingleR/"><code>output/marker_genes/SingleR/</code></a>.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># prevent "/" contained in annotation "Th1/Th17 cells" from leading to a new path </span>
<span class='va'>sce</span><span class='op'>$</span><span class='va'>mi_label_fine_collapsed</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/grep.html'>gsub</a></span><span class='op'>(</span><span class='st'>"Th1/Th17 cells"</span>, <span class='st'>"Th1_Th17 cells"</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>mi_label_fine_collapsed</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>mi_label_fine_colours</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/grep.html'>gsub</a></span><span class='op'>(</span><span class='st'>"Th1/Th17 cells"</span>, <span class='st'>"Th1_Th17 cells"</span>, <span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>mi_label_fine_colours</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>all_markers</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/grep.html'>gsub</a></span><span class='op'>(</span><span class='st'>"Th1/Th17 cells"</span>, <span class='st'>"Th1_Th17 cells"</span>, <span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>all_markers</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/files2.html'>dir.create</a></span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"output"</span>, <span class='st'>"marker_genes"</span>, <span class='st'>"SingleR"</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/grDevices/dev.html'>dev.off</a></span><span class='op'>(</span><span class='op'>)</span> <span class='co'># Needed to ensure next figures save properly.</span>

<span class='kw'>for</span> <span class='op'>(</span><span class='va'>label</span> <span class='kw'>in</span> <span class='fu'><a href='https://generics.r-lib.org/reference/setops.html'>setdiff</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>mi_label_fine_collapsed</span><span class='op'>)</span>, <span class='st'>"other"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='fu'><a href='https://rdrr.io/r/grDevices/png.html'>png</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span>
      <span class='st'>"output"</span>,
      <span class='st'>"marker_genes"</span>,
      <span class='st'>"SingleR"</span>,
      <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"mi."</span>, <span class='st'>"heatmaps."</span>, <span class='va'>label</span>, <span class='st'>".png"</span><span class='op'>)</span><span class='op'>)</span>,
    width <span class='op'>=</span> <span class='fl'>10</span>,
    height <span class='op'>=</span> <span class='fl'>12</span>,
    units <span class='op'>=</span> <span class='st'>"in"</span>,
    res <span class='op'>=</span> <span class='fl'>250</span><span class='op'>)</span>
  
  <span class='va'>top_markers</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/funprog.html'>Reduce</a></span><span class='op'>(</span><span class='va'>union</span>, <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span><span class='va'>all_markers</span><span class='op'>[[</span><span class='va'>label</span><span class='op'>]</span><span class='op'>]</span>, <span class='va'>head</span>, <span class='fl'>10</span><span class='op'>)</span><span class='op'>)</span>
  
<span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotHeatmap.html'>plotHeatmap</a></span><span class='op'>(</span>
    <span class='va'>sce</span>,
    features <span class='op'>=</span> <span class='va'>top_markers</span>,
    columns <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span>
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>mi_label_fine_collapsed</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>sample_name</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>post_hoc_sample_gate</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>tissue</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>donor</span><span class='op'>)</span>,
    order_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>
      <span class='st'>"mi_label_fine_collapsed"</span>,
      <span class='st'>"cluster"</span>,
      <span class='st'>"sample_name"</span>,
      <span class='st'>"post_hoc_sample_gate"</span>,
      <span class='st'>"plate_number"</span>,
      <span class='st'>"tissue"</span>,
      <span class='st'>"donor"</span><span class='op'>)</span>,
    cluster_cols <span class='op'>=</span> <span class='cn'>FALSE</span>, 
    center <span class='op'>=</span> <span class='cn'>TRUE</span>,
    symmetric <span class='op'>=</span> <span class='cn'>TRUE</span>,
    zlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>3</span>, <span class='fl'>3</span><span class='op'>)</span>,
    show_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
    column_annotation_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
       mi_label_fine_collapsed <span class='op'>=</span> <span class='va'>mi_label_fine_colours</span>,
       cluster <span class='op'>=</span> <span class='va'>cluster_colours</span>,
       sample_name <span class='op'>=</span> <span class='va'>sample_name_colours</span>,
       post_hoc_sample_gate <span class='op'>=</span> <span class='va'>sample_gate_colours</span>,
       plate_number <span class='op'>=</span> <span class='va'>plate_number_colours</span>,    
       tissue <span class='op'>=</span> <span class='va'>tissue_colours</span>,
       donor <span class='op'>=</span> <span class='va'>donor_colours</span><span class='op'>)</span>,
    color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>hcl.colors</a></span><span class='op'>(</span><span class='fl'>101</span>, <span class='st'>"Blue-Red 3"</span><span class='op'>)</span>,
    main <span class='op'>=</span> <span class='va'>label</span>,
    fontsize <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span>
  
  <span class='fu'><a href='https://rdrr.io/r/grDevices/dev.html'>dev.off</a></span><span class='op'>(</span><span class='op'>)</span>
<span class='op'>}</span>
</code></pre>
</div>
</details>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/files2.html'>dir.create</a></span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"output"</span>, <span class='st'>"marker_genes"</span>, <span class='st'>"SingleR"</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/grDevices/dev.html'>dev.off</a></span><span class='op'>(</span><span class='op'>)</span> <span class='co'># Needed to ensure next figures save properly.</span>
<span class='kw'>for</span> <span class='op'>(</span><span class='va'>label</span> <span class='kw'>in</span> <span class='fu'><a href='https://generics.r-lib.org/reference/setops.html'>setdiff</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>mi_label_fine_collapsed</span><span class='op'>)</span>, <span class='st'>"other"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='fu'><a href='https://rdrr.io/r/grDevices/png.html'>png</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span>
      <span class='st'>"output"</span>,
      <span class='st'>"marker_genes"</span>,
      <span class='st'>"SingleR"</span>,
      <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"mi."</span>, <span class='st'>"heatmaps."</span>, <span class='va'>label</span>, <span class='st'>"ordered.by.cluster"</span>, <span class='st'>".png"</span><span class='op'>)</span><span class='op'>)</span>,
    width <span class='op'>=</span> <span class='fl'>10</span>,
    height <span class='op'>=</span> <span class='fl'>12</span>,
    units <span class='op'>=</span> <span class='st'>"in"</span>,
    res <span class='op'>=</span> <span class='fl'>250</span><span class='op'>)</span>
  
  <span class='va'>top_markers</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/funprog.html'>Reduce</a></span><span class='op'>(</span><span class='va'>union</span>, <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span><span class='va'>all_markers</span><span class='op'>[[</span><span class='va'>label</span><span class='op'>]</span><span class='op'>]</span>, <span class='va'>head</span>, <span class='fl'>10</span><span class='op'>)</span><span class='op'>)</span>
  
<span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotHeatmap.html'>plotHeatmap</a></span><span class='op'>(</span>
    <span class='va'>sce</span>,
    features <span class='op'>=</span> <span class='va'>top_markers</span>,
    columns <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span>
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>mi_label_fine_collapsed</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>sample_name</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>post_hoc_sample_gate</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>tissue</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>donor</span><span class='op'>)</span>,
    colour_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>
      <span class='st'>"cluster"</span>,
      <span class='st'>"mi_label_fine_collapsed"</span>,
      <span class='st'>"sample_name"</span>,
      <span class='st'>"post_hoc_sample_gate"</span>,
      <span class='st'>"plate_number"</span>,
      <span class='st'>"tissue"</span>,
      <span class='st'>"donor"</span><span class='op'>)</span>,
    cluster_cols <span class='op'>=</span> <span class='cn'>FALSE</span>, 
    center <span class='op'>=</span> <span class='cn'>TRUE</span>,
    symmetric <span class='op'>=</span> <span class='cn'>TRUE</span>,
    zlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>3</span>, <span class='fl'>3</span><span class='op'>)</span>,
    show_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
    column_annotation_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
       cluster <span class='op'>=</span> <span class='va'>cluster_colours</span>,
       mi_label_fine_collapsed <span class='op'>=</span> <span class='va'>mi_label_fine_colours</span>,
       sample_name <span class='op'>=</span> <span class='va'>sample_name_colours</span>,
       post_hoc_sample_gate <span class='op'>=</span> <span class='va'>sample_gate_colours</span>,
       plate_number <span class='op'>=</span> <span class='va'>plate_number_colours</span>,    
       tissue <span class='op'>=</span> <span class='va'>tissue_colours</span>,
       donor <span class='op'>=</span> <span class='va'>donor_colours</span><span class='op'>)</span>,
    color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>hcl.colors</a></span><span class='op'>(</span><span class='fl'>101</span>, <span class='st'>"Blue-Red 3"</span><span class='op'>)</span>,
    main <span class='op'>=</span> <span class='va'>label</span>,
    fontsize <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span>
  
  <span class='fu'><a href='https://rdrr.io/r/grDevices/dev.html'>dev.off</a></span><span class='op'>(</span><span class='op'>)</span>
<span class='op'>}</span>
</code></pre>
</div>
</details>
</div>
<h1 id="marker-gene-detection">Marker gene detection</h1>
<p>To interpret our clustering results, we identify the genes that drive separation between clusters. These marker genes allow us to assign biological meaning to each cluster based on their functional annotation. In the most obvious case, the marker genes for each cluster are <em>a priori</em> associated with particular cell types, allowing us to treat the clustering as a proxy for cell type identity. The same principle can be applied to more subtle differences in activation status or differentiation state.</p>
<p>Identification of marker genes is usually based around the retrospective detection of differential expression between clusters<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a>. Genes that are more strongly DE are more likely to have driven cluster separation in the first place. The top DE genes are likely to be good candidate markers as they can effectively distinguish between cells in different clusters.</p>
<h2 id="using-pairwise-t-tests">Using pairwise <em>t</em>-tests</h2>
<h3 id="standard-application">Standard application</h3>
<p>The Welch <em>t</em>-test is an obvious choice of statistical method to test for differences in expression between clusters. It is quickly computed and has good statistical properties for large numbers of cells <span class="citation" data-cites="soneson2018bias">(<a href="#ref-soneson2018bias" role="doc-biblioref">Soneson and Robinson 2018</a>)</span>. We perform intra-<code>donor</code> comparisons by blocking on the <code>plate_number</code> of origin<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a> to avoid confounding effects from library-to-library variability or differential expression between mice<a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a>.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>sce</span><span class='op'>$</span><span class='va'>block</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>For each cluster, the DE results of the relevant comparisons are consolidated into a single output table. This allows a set of marker genes to be easily defined by taking the top DE genes from each pairwise comparison between clusters. Other statistics are also reported for each gene, including the adjusted p-values<a href="#fn6" class="footnote-ref" id="fnref6" role="doc-noteref"><sup>6</sup></a> and the log-fold changes relative to every other cluster.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>markers_union_any</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scran/man/findMarkers.html'>findMarkers</a></span><span class='op'>(</span>
  <span class='va'>sce</span>, 
  groups <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span>,
  block <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>block</span>,
  pval.type <span class='op'>=</span> <span class='st'>"any"</span>,
  direction <span class='op'>=</span> <span class='st'>"any"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span>
  <span class='va'>markers_union_any</span>, 
  <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"data"</span>, <span class='st'>"marker_genes"</span>, <span class='st'>"C094_Pellicci.markers_union_any.rds"</span><span class='op'>)</span>,
  compress <span class='op'>=</span> <span class='st'>"xz"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/files2.html'>dir.create</a></span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"output"</span>, <span class='st'>"marker_genes"</span>, <span class='st'>"union_any"</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/message.html'>message</a></span><span class='op'>(</span><span class='st'>"Writing 'union_any' marker genes to file."</span><span class='op'>)</span>
<span class='kw'>for</span> <span class='op'>(</span><span class='va'>n</span> <span class='kw'>in</span> <span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>markers_union_any</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='fu'><a href='https://rdrr.io/r/base/message.html'>message</a></span><span class='op'>(</span><span class='va'>n</span><span class='op'>)</span>
  <span class='va'>gzout</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/connections.html'>gzfile</a></span><span class='op'>(</span>
    description <span class='op'>=</span> <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span>
      <span class='st'>"output"</span>, 
      <span class='st'>"marker_genes"</span>, 
      <span class='st'>"union_any"</span>,
      <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"cluster_"</span>, <span class='va'>n</span>, <span class='st'>".union_any.csv.gz"</span><span class='op'>)</span><span class='op'>)</span>,
    open <span class='op'>=</span> <span class='st'>"wb"</span><span class='op'>)</span>
  
  <span class='fu'><a href='https://rdrr.io/r/utils/write.table.html'>write.table</a></span><span class='op'>(</span>
    x <span class='op'>=</span> <span class='va'>markers_union_any</span><span class='op'>[[</span><span class='va'>n</span><span class='op'>]</span><span class='op'>]</span> <span class='op'>%&gt;%</span>
      <span class='fu'><a href='https://rdrr.io/r/base/as.data.frame.html'>as.data.frame</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
      <span class='fu'>tibble</span><span class='fu'>::</span><span class='fu'><a href='https://tibble.tidyverse.org/reference/rownames.html'>rownames_to_column</a></span><span class='op'>(</span><span class='st'>"gene_ID"</span><span class='op'>)</span>, 
    file <span class='op'>=</span> <span class='va'>gzout</span>, 
    sep <span class='op'>=</span> <span class='st'>","</span>, 
    quote <span class='op'>=</span> <span class='cn'>FALSE</span>, 
    row.names <span class='op'>=</span> <span class='cn'>FALSE</span>,
    col.names <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/base/connections.html'>close</a></span><span class='op'>(</span><span class='va'>gzout</span><span class='op'>)</span>
<span class='op'>}</span>
</code></pre>
</div>
</details>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>chosen</span> <span class='op'>&lt;-</span> <span class='st'>"4"</span>
<span class='va'>interesting</span> <span class='op'>&lt;-</span> <span class='va'>markers_union_any</span><span class='op'>[[</span><span class='va'>chosen</span><span class='op'>]</span><span class='op'>]</span>
</code></pre>
</div>
</details>
</div>
<p>To demonstrate, we use cluster 4 as our cluster of interest for this section. The relevant table contains log<sub>2</sub>-fold changes of expression in cluster 4 over each other cluster, along with several statistics obtained by combining <span class="math inline">\(p\)</span>-values <span class="citation" data-cites="simes1986improved">(<a href="#ref-simes1986improved" role="doc-biblioref">Simes 1986</a>)</span> across the pairwise comparisons involving 4.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>.adf</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>interesting</span>, <span class='fl'>10</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>tibble</span><span class='fu'>::</span><span class='fu'><a href='https://tibble.tidyverse.org/reference/rownames.html'>rownames_to_column</a></span><span class='op'>(</span><span class='st'>"Gene"</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>knitr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/knitr/man/kable.html'>kable</a></span><span class='op'>(</span>
    caption <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span>
      <span class='st'>"First 10 marker genes for cluster "</span>,
      <span class='va'>chosen</span>,
      <span class='st'>" across comparisons."</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<table>
<caption><span id="tab:unnamed-chunk-22">Table 3: </span>First 10 marker genes for cluster 4 across comparisons.</caption>
<thead>
<tr class="header">
<th style="text-align: left;">Gene</th>
<th style="text-align: right;">Top</th>
<th style="text-align: right;">p.value</th>
<th style="text-align: right;">FDR</th>
<th style="text-align: right;">summary.logFC</th>
<th style="text-align: right;">logFC.1</th>
<th style="text-align: right;">logFC.2</th>
<th style="text-align: right;">logFC.3</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">HLA-B</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">-1.0959431</td>
<td style="text-align: right;">-1.0959431</td>
<td style="text-align: right;">-0.7598154</td>
<td style="text-align: right;">0.0455160</td>
</tr>
<tr class="even">
<td style="text-align: left;">AC016700.2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1.1785304</td>
<td style="text-align: right;">0.1750436</td>
<td style="text-align: right;">0.1954106</td>
<td style="text-align: right;">1.1785304</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CCL5</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">-1.4682427</td>
<td style="text-align: right;">-1.4682427</td>
<td style="text-align: right;">-0.5837482</td>
<td style="text-align: right;">-0.7541207</td>
</tr>
<tr class="even">
<td style="text-align: left;">TMSB4XP6</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.8845827</td>
<td style="text-align: right;">0.0355193</td>
<td style="text-align: right;">0.2750976</td>
<td style="text-align: right;">0.8845827</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SOX4</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.6792162</td>
<td style="text-align: right;">0.6792162</td>
<td style="text-align: right;">0.6453342</td>
<td style="text-align: right;">0.5623182</td>
</tr>
<tr class="even">
<td style="text-align: left;">AC078817.1</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.8507711</td>
<td style="text-align: right;">0.3191471</td>
<td style="text-align: right;">0.3675160</td>
<td style="text-align: right;">0.8507711</td>
</tr>
<tr class="odd">
<td style="text-align: left;">HLA-C</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">-0.9518689</td>
<td style="text-align: right;">-0.9518689</td>
<td style="text-align: right;">-0.6541720</td>
<td style="text-align: right;">0.1090136</td>
</tr>
<tr class="even">
<td style="text-align: left;">BCL11B</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.7641963</td>
<td style="text-align: right;">0.7641963</td>
<td style="text-align: right;">0.7144840</td>
<td style="text-align: right;">0.8102414</td>
</tr>
<tr class="odd">
<td style="text-align: left;">MIR1244-2</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.6615599</td>
<td style="text-align: right;">0.2911494</td>
<td style="text-align: right;">0.3135224</td>
<td style="text-align: right;">0.6615599</td>
</tr>
<tr class="even">
<td style="text-align: left;">TOX2</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.4584782</td>
<td style="text-align: right;">0.4584782</td>
<td style="text-align: right;">0.4438104</td>
<td style="text-align: right;">0.4484043</td>
</tr>
</tbody>
</table>
</div>
<p>Of particular interest is the <code>Top</code> field, which contains the highest rank for each gene across all pairwise comparisons involving cluster 4. The set of genes with <code>Top</code> values of 1 contains the gene with the lowest <span class="math inline">\(p\)</span>-value from each comparison. Similarly, the set of genes with <code>Top</code> values less than or equal to 10 contains the top 10 genes from each comparison.</p>
<p>We use the <code>Top</code> value to identify a set of genes that is guaranteed to distinguish cluster 4 from any other cluster. Figure <a href="#fig:heat-logFC">10</a> plots the log-fold changes for the top 10 genes from each pairwise comparison.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>best_set</span> <span class='op'>&lt;-</span> <span class='va'>interesting</span><span class='op'>[</span><span class='va'>interesting</span><span class='op'>$</span><span class='va'>Top</span> <span class='op'>&lt;=</span> <span class='fl'>10</span>, <span class='op'>]</span>
<span class='va'>logFCs</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span><span class='op'>(</span><span class='va'>best_set</span><span class='op'>[</span>, <span class='op'>-</span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>4</span><span class='op'>)</span><span class='op'>]</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>logFCs</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/grep.html'>sub</a></span><span class='op'>(</span><span class='st'>"logFC."</span>, <span class='st'>""</span>, <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>logFCs</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>logFCs</span>, breaks <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>5</span>, <span class='fl'>5</span>, length.out <span class='op'>=</span> <span class='fl'>101</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:heat-logFC"></span>
<img src="C094_Pellicci.annotate_files/heat-logFC-1.png" alt="Heatmap of log-fold changes for the top 10 marker genes for cluster 4 over all other clusters. Each column is a cluster, each row a gene. Colours are capped at -5 and 5 to preserve dynamic range." width="624" />
<p class="caption">
Figure 10: Heatmap of log-fold changes for the top 10 marker genes for cluster 4 over all other clusters. Each column is a cluster, each row a gene. Colours are capped at -5 and 5 to preserve dynamic range.
</p>
</div>
</div>
<p>Complementing Figure <a href="#fig:heat-logFC">10</a>, Figure <a href="#fig:heat-logExp">11</a> plots the log-expression values of each cells in each cluster for the top 5 genes</p>
<div class="layout-chunk" data-layout="l-page">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotHeatmap.html'>plotHeatmap</a></span><span class='op'>(</span>
  <span class='va'>sce</span>,
  features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>best_set</span><span class='op'>)</span>,
  columns <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span>
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>sample_name</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>post_hoc_sample_gate</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>tissue</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>donor</span><span class='op'>)</span>,
  order_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>
    <span class='st'>"cluster"</span>,
    <span class='st'>"sample_name"</span>,
    <span class='st'>"post_hoc_sample_gate"</span>,
    <span class='st'>"plate_number"</span>,
    <span class='st'>"tissue"</span>,
    <span class='st'>"donor"</span><span class='op'>)</span>,
  cluster_cols <span class='op'>=</span> <span class='cn'>FALSE</span>, 
  center <span class='op'>=</span> <span class='cn'>TRUE</span>,
  symmetric <span class='op'>=</span> <span class='cn'>TRUE</span>,
  zlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>5</span>, <span class='fl'>5</span><span class='op'>)</span>,
  show_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
  main <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"Cluster: "</span>, <span class='va'>chosen</span><span class='op'>)</span>,
  column_annotation_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
     cluster <span class='op'>=</span> <span class='va'>cluster_colours</span>,
     sample_name <span class='op'>=</span> <span class='va'>sample_name_colours</span>,
     post_hoc_sample_gate <span class='op'>=</span> <span class='va'>sample_gate_colours</span>,
     plate_number <span class='op'>=</span> <span class='va'>plate_number_colours</span>,    
     tissue <span class='op'>=</span> <span class='va'>tissue_colours</span>,
     donor <span class='op'>=</span> <span class='va'>donor_colours</span><span class='op'>)</span>,
  color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>hcl.colors</a></span><span class='op'>(</span><span class='fl'>101</span>, <span class='st'>"Blue-Red 3"</span><span class='op'>)</span>,
  fontsize <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:heat-logExp"></span>
<img src="C094_Pellicci.annotate_files/heat-logExp-1.png" alt="Heatmap of log-expression values in each sample for the top 10 marker genes in cluster 4. Each column is a sample, each row a gene. Colours are capped at -5 and 5 to preserve dynamic range." width="624" />
<p class="caption">
Figure 11: Heatmap of log-expression values in each sample for the top 10 marker genes in cluster 4. Each column is a sample, each row a gene. Colours are capped at -5 and 5 to preserve dynamic range.
</p>
</div>
</div>
<p>Similar figures showing the top 10 marker genes for each cluster are available in <a href="../output/marker_genes/union_any/heatmaps.union_any.pdf"><code>output/marker_genes/union_any/heatmaps.union_any.pdf</code></a>.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/grDevices/dev.html'>dev.off</a></span><span class='op'>(</span><span class='op'>)</span> <span class='co'># Needed to ensure next figure saves properly.</span>
<span class='fu'><a href='https://rdrr.io/r/grDevices/pdf.html'>pdf</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span>
    <span class='st'>"output"</span>,
    <span class='st'>"marker_genes"</span>,
    <span class='st'>"union_any"</span>,
    <span class='st'>"heatmaps.union_any.pdf"</span><span class='op'>)</span>,
  width <span class='op'>=</span> <span class='fl'>10</span>,
  height <span class='op'>=</span> <span class='fl'>12</span><span class='op'>)</span>
<span class='kw'>for</span> <span class='op'>(</span><span class='va'>cluster</span> <span class='kw'>in</span> <span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='va'>marker_set</span> <span class='op'>&lt;-</span> <span class='va'>markers_union_any</span><span class='op'>[[</span><span class='va'>cluster</span><span class='op'>]</span><span class='op'>]</span>
  <span class='va'>best_set</span> <span class='op'>&lt;-</span> <span class='va'>marker_set</span><span class='op'>[</span><span class='va'>marker_set</span><span class='op'>$</span><span class='va'>Top</span> <span class='op'>&lt;=</span> <span class='fl'>10</span>, <span class='op'>]</span>
  
  <span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotHeatmap.html'>plotHeatmap</a></span><span class='op'>(</span>
    <span class='va'>sce</span>,
    features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>best_set</span><span class='op'>)</span>,
    columns <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span>
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>sample_name</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>post_hoc_sample_gate</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>tissue</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>donor</span><span class='op'>)</span>,
    colour_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>
        <span class='st'>"cluster"</span>,
        <span class='st'>"sample_name"</span>,
        <span class='st'>"post_hoc_sample_gate"</span>,
        <span class='st'>"plate_number"</span>,
        <span class='st'>"tissue"</span>,
        <span class='st'>"donor"</span><span class='op'>)</span>,
    cluster_cols <span class='op'>=</span> <span class='cn'>FALSE</span>, 
    center <span class='op'>=</span> <span class='cn'>TRUE</span>,
    symmetric <span class='op'>=</span> <span class='cn'>TRUE</span>,
    zlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>5</span>, <span class='fl'>5</span><span class='op'>)</span>,
    show_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
    main <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"Cluster: "</span>, <span class='va'>cluster</span><span class='op'>)</span>,
    column_annotation_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
         cluster <span class='op'>=</span> <span class='va'>cluster_colours</span>,
         sample_name <span class='op'>=</span> <span class='va'>sample_name_colours</span>,
         post_hoc_sample_gate <span class='op'>=</span> <span class='va'>sample_gate_colours</span>,
         plate_number <span class='op'>=</span> <span class='va'>plate_number_colours</span>,    
         tissue <span class='op'>=</span> <span class='va'>tissue_colours</span>,
         donor <span class='op'>=</span> <span class='va'>donor_colours</span><span class='op'>)</span>,
    color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>hcl.colors</a></span><span class='op'>(</span><span class='fl'>101</span>, <span class='st'>"Blue-Red 3"</span><span class='op'>)</span>,
    fontsize <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span>
<span class='op'>}</span>
<span class='fu'><a href='https://rdrr.io/r/grDevices/dev.html'>dev.off</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>We intentionally use pairwise comparisons between clusters rather than comparing each cluster to the average of all other cells. The latter approach is sensitive to the population composition, potentially resulting in wildly different sets of markers when cell type abundances change in different contexts. In the worst case, the presence of a single dominant subpopulation will drive the selection of top markers for every other cluster, pushing out useful genes that can resolve the various minor subpopulations. Moreover, pairwise comparisons naturally provide more information to interpret of the utility of a marker, e.g., by providing log-fold changes to indicate which clusters are distinguished by this gene.</p>
<h3 id="using-the-log-fold-change">Using the log-fold change</h3>
<p>The differential expression analysis in <a href="#standard-application">Standard application</a> considers both up- and down-regulated genes to be potential markers. However, down-regulated genes are less appealing as markers as it is more difficult to interpret and experimentally validate an absence of expression. To focus on upregulated markers, we can instead perform a one-sided <span class="math inline">\(t\)</span>-test to identify genes that are upregulated in each cluster compared to the others.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>markers_union_up</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scran/man/findMarkers.html'>findMarkers</a></span><span class='op'>(</span>
  <span class='va'>sce</span>,
  groups <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span>,
  block <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>block</span>,
  pval.type <span class='op'>=</span> <span class='st'>"any"</span>,
  direction <span class='op'>=</span> <span class='st'>"up"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span>
  <span class='va'>markers_union_up</span>, 
  <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"data"</span>, <span class='st'>"marker_genes"</span>, <span class='st'>"C094_Pellicci.markers_union_up.rds"</span><span class='op'>)</span>,
  compress <span class='op'>=</span> <span class='st'>"xz"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/files2.html'>dir.create</a></span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"output"</span>, <span class='st'>"marker_genes"</span>, <span class='st'>"union_up"</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/message.html'>message</a></span><span class='op'>(</span><span class='st'>"Writing 'union_up' marker genes to file."</span><span class='op'>)</span>
<span class='kw'>for</span> <span class='op'>(</span><span class='va'>n</span> <span class='kw'>in</span> <span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>markers_union_up</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='fu'><a href='https://rdrr.io/r/base/message.html'>message</a></span><span class='op'>(</span><span class='va'>n</span><span class='op'>)</span>
  <span class='va'>gzout</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/connections.html'>gzfile</a></span><span class='op'>(</span>
    description <span class='op'>=</span> <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span>
      <span class='st'>"output"</span>, 
      <span class='st'>"marker_genes"</span>, 
      <span class='st'>"union_up"</span>,
      <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"cluster_"</span>, <span class='va'>n</span>, <span class='st'>".union_up.csv.gz"</span><span class='op'>)</span><span class='op'>)</span>,
    open <span class='op'>=</span> <span class='st'>"wb"</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/utils/write.table.html'>write.table</a></span><span class='op'>(</span>
    x <span class='op'>=</span> <span class='va'>markers_union_up</span><span class='op'>[[</span><span class='va'>n</span><span class='op'>]</span><span class='op'>]</span> <span class='op'>%&gt;%</span>
      <span class='fu'><a href='https://rdrr.io/r/base/as.data.frame.html'>as.data.frame</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
      <span class='fu'>tibble</span><span class='fu'>::</span><span class='fu'><a href='https://tibble.tidyverse.org/reference/rownames.html'>rownames_to_column</a></span><span class='op'>(</span><span class='st'>"gene_ID"</span><span class='op'>)</span>, 
    file <span class='op'>=</span> <span class='va'>gzout</span>, 
    sep <span class='op'>=</span> <span class='st'>","</span>, 
    quote <span class='op'>=</span> <span class='cn'>FALSE</span>, 
    row.names <span class='op'>=</span> <span class='cn'>FALSE</span>,
    col.names <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/base/connections.html'>close</a></span><span class='op'>(</span><span class='va'>gzout</span><span class='op'>)</span>
<span class='op'>}</span>
</code></pre>
</div>
</details>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>interesting_up</span> <span class='op'>&lt;-</span> <span class='va'>markers_union_up</span><span class='op'>[[</span><span class='va'>chosen</span><span class='op'>]</span><span class='op'>]</span>
</code></pre>
</div>
</details>
</div>
<p>The <span class="math inline">\(t\)</span>-test also allows us to specify a non-zero log-fold change as the null hypothesis. This allows us to consider the magnitude of the log-fold change in our <span class="math inline">\(p\)</span>-value calculations, in a manner that is more rigorous than simply filtering directly on the log-fold changes<a href="#fn7" class="footnote-ref" id="fnref7" role="doc-noteref"><sup>7</sup></a>. For example, we can test for genes with log-fold changes that are significantly greater than 1.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>markers_union_up_lfc1</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scran/man/findMarkers.html'>findMarkers</a></span><span class='op'>(</span>
  <span class='va'>sce</span>,
  groups <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span>,
  block <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>block</span>,
  pval.type <span class='op'>=</span> <span class='st'>"any"</span>,
  direction <span class='op'>=</span> <span class='st'>"up"</span>,
  lfc <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span>
  <span class='va'>markers_union_up_lfc1</span>, 
  <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"data"</span>, <span class='st'>"marker_genes"</span>, <span class='st'>"C094_Pellicci.markers_union_up_lfc1.rds"</span><span class='op'>)</span>,
  compress <span class='op'>=</span> <span class='st'>"xz"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/files2.html'>dir.create</a></span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"output"</span>, <span class='st'>"marker_genes"</span>, <span class='st'>"union_up_lfc1"</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/message.html'>message</a></span><span class='op'>(</span><span class='st'>"Writing 'union_up_lfc1' marker genes to file."</span><span class='op'>)</span>
<span class='kw'>for</span> <span class='op'>(</span><span class='va'>n</span> <span class='kw'>in</span> <span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>markers_union_up_lfc1</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='fu'><a href='https://rdrr.io/r/base/message.html'>message</a></span><span class='op'>(</span><span class='va'>n</span><span class='op'>)</span>
  <span class='va'>gzout</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/connections.html'>gzfile</a></span><span class='op'>(</span>
    description <span class='op'>=</span> <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span>
      <span class='st'>"output"</span>, 
      <span class='st'>"marker_genes"</span>, 
      <span class='st'>"union_up_lfc1"</span>,
      <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"cluster_"</span>, <span class='va'>n</span>, <span class='st'>".union_up_lfc1.csv.gz"</span><span class='op'>)</span><span class='op'>)</span>,
    open <span class='op'>=</span> <span class='st'>"wb"</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/utils/write.table.html'>write.table</a></span><span class='op'>(</span>
    x <span class='op'>=</span> <span class='va'>markers_union_up_lfc1</span><span class='op'>[[</span><span class='va'>n</span><span class='op'>]</span><span class='op'>]</span> <span class='op'>%&gt;%</span>
      <span class='fu'><a href='https://rdrr.io/r/base/as.data.frame.html'>as.data.frame</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
      <span class='fu'>tibble</span><span class='fu'>::</span><span class='fu'><a href='https://tibble.tidyverse.org/reference/rownames.html'>rownames_to_column</a></span><span class='op'>(</span><span class='st'>"gene_ID"</span><span class='op'>)</span>, 
    file <span class='op'>=</span> <span class='va'>gzout</span>, 
    sep <span class='op'>=</span> <span class='st'>","</span>, 
    quote <span class='op'>=</span> <span class='cn'>FALSE</span>, 
    row.names <span class='op'>=</span> <span class='cn'>FALSE</span>,
    col.names <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/base/connections.html'>close</a></span><span class='op'>(</span><span class='va'>gzout</span><span class='op'>)</span>
<span class='op'>}</span>
</code></pre>
</div>
</details>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>interesting_up_lfc1</span> <span class='op'>&lt;-</span> <span class='va'>markers_union_up_lfc1</span><span class='op'>[[</span><span class='va'>chosen</span><span class='op'>]</span><span class='op'>]</span>
</code></pre>
</div>
</details>
</div>
<p>Combining these two settings yield a more focused set of candidate marker genes that are ‘upregulated with a log-fold change &gt; 1’ in cluster 4 (Figures <a href="#fig:heat-focused-logFC">12</a> and <a href="#fig:heat-focused-logExp">13</a>).</p>
<div class="layout-chunk" data-layout="l-body-outset">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>best_set</span> <span class='op'>&lt;-</span> <span class='va'>interesting_up_lfc1</span><span class='op'>[</span><span class='va'>interesting_up_lfc1</span><span class='op'>$</span><span class='va'>Top</span> <span class='op'>&lt;=</span> <span class='fl'>10</span>,<span class='op'>]</span>
<span class='va'>logFCs</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span><span class='op'>(</span><span class='va'>best_set</span><span class='op'>[</span>,<span class='op'>-</span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>4</span><span class='op'>)</span><span class='op'>]</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>logFCs</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/grep.html'>sub</a></span><span class='op'>(</span><span class='st'>"logFC."</span>, <span class='st'>""</span>, <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>logFCs</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>logFCs</span>, breaks <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>5</span>, <span class='fl'>5</span>, length.out <span class='op'>=</span> <span class='fl'>101</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:heat-focused-logFC"></span>
<img src="C094_Pellicci.annotate_files/heat-focused-logFC-1.png" alt="Heatmap of log-fold changes for cluster 4 over all other clusters. Colours are capped at -5 and 5 to preserve dynamic range." width="624" />
<p class="caption">
Figure 12: Heatmap of log-fold changes for cluster 4 over all other clusters. Colours are capped at -5 and 5 to preserve dynamic range.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-page">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotHeatmap.html'>plotHeatmap</a></span><span class='op'>(</span>
  <span class='va'>sce</span>,
  features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>best_set</span><span class='op'>)</span>,
  columns <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span>
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>sample_name</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>post_hoc_sample_gate</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>tissue</span>,
    <span class='va'>sce</span><span class='op'>$</span><span class='va'>donor</span><span class='op'>)</span>,
  order_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>
    <span class='st'>"cluster"</span>,
    <span class='st'>"sample_name"</span>,
    <span class='st'>"post_hoc_sample_gate"</span>,
    <span class='st'>"plate_number"</span>,
    <span class='st'>"tissue"</span>,
    <span class='st'>"donor"</span><span class='op'>)</span>,
  cluster_cols <span class='op'>=</span> <span class='cn'>FALSE</span>, 
  center <span class='op'>=</span> <span class='cn'>TRUE</span>,
  symmetric <span class='op'>=</span> <span class='cn'>TRUE</span>,
  zlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>5</span>, <span class='fl'>5</span><span class='op'>)</span>,
  show_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
  main <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"Cluster: "</span>, <span class='va'>chosen</span><span class='op'>)</span>,
  column_annotation_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
     mi_label_fine_collapsed <span class='op'>=</span> <span class='va'>mi_label_fine_colours</span>,
     cluster <span class='op'>=</span> <span class='va'>cluster_colours</span>,
     sample_name <span class='op'>=</span> <span class='va'>sample_name_colours</span>,
     post_hoc_sample_gate <span class='op'>=</span> <span class='va'>sample_gate_colours</span>,
     plate_number <span class='op'>=</span> <span class='va'>plate_number_colours</span>,    
     tissue <span class='op'>=</span> <span class='va'>tissue_colours</span>,
     donor <span class='op'>=</span> <span class='va'>donor_colours</span><span class='op'>)</span>,
  color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>hcl.colors</a></span><span class='op'>(</span><span class='fl'>101</span>, <span class='st'>"Blue-Red 3"</span><span class='op'>)</span>,
  fontsize <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:heat-focused-logExp"></span>
<img src="C094_Pellicci.annotate_files/heat-focused-logExp-1.png" alt="Heatmap of log-expression values in each sample for the top 10 'upregulated with a log-fold change &gt; 1' marker genes in cluster 4. Each column is a sample, each row a gene. Colours are capped at -5 and 5 to preserve dynamic range." width="624" />
<p class="caption">
Figure 13: Heatmap of log-expression values in each sample for the top 10 ‘upregulated with a log-fold change &gt; 1’ marker genes in cluster 4. Each column is a sample, each row a gene. Colours are capped at -5 and 5 to preserve dynamic range.
</p>
</div>
</div>
<p>Similar figures showing the top 10 upregulated and ‘upregulated with a log-fold change &gt; 1’ marker genes for each cluster are available in <a href="../output/marker_genes/union_up/heatmaps.union_up.pdf"><code>output/marker_genes/union_up/heatmaps.union_up.pdf</code></a>. and <a href="../output/marker_genes/union_up_lfc1/heatmaps.union_up_lfc1.pdf"><code>output/marker_genes/union_up_lfc1/heatmaps.union_up_lfc1.pdf</code></a>.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/grDevices/dev.html'>dev.off</a></span><span class='op'>(</span><span class='op'>)</span> <span class='co'># Needed to ensure next figure saves properly.</span>
<span class='fu'><a href='https://rdrr.io/r/grDevices/pdf.html'>pdf</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span>
    <span class='st'>"output"</span>,
    <span class='st'>"marker_genes"</span>,
    <span class='st'>"union_up"</span>,
    <span class='st'>"heatmaps.union_up.pdf"</span><span class='op'>)</span>,
  width <span class='op'>=</span> <span class='fl'>10</span>,
  height <span class='op'>=</span> <span class='fl'>12</span><span class='op'>)</span>
<span class='kw'>for</span> <span class='op'>(</span><span class='va'>cluster</span> <span class='kw'>in</span> <span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='va'>marker_set</span> <span class='op'>&lt;-</span> <span class='va'>markers_union_up</span><span class='op'>[[</span><span class='va'>cluster</span><span class='op'>]</span><span class='op'>]</span>
  <span class='va'>best_set</span> <span class='op'>&lt;-</span> <span class='va'>marker_set</span><span class='op'>[</span><span class='va'>marker_set</span><span class='op'>$</span><span class='va'>Top</span> <span class='op'>&lt;=</span> <span class='fl'>10</span>, <span class='op'>]</span>
  
  <span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotHeatmap.html'>plotHeatmap</a></span><span class='op'>(</span>
    <span class='va'>sce</span>,
    features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>best_set</span><span class='op'>)</span>,
    columns <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span>
        <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span>,
        <span class='va'>sce</span><span class='op'>$</span><span class='va'>sample_name</span>,
        <span class='va'>sce</span><span class='op'>$</span><span class='va'>post_hoc_sample_gate</span>,
        <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span>,
        <span class='va'>sce</span><span class='op'>$</span><span class='va'>tissue</span>,
        <span class='va'>sce</span><span class='op'>$</span><span class='va'>donor</span><span class='op'>)</span>,
    colour_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>
        <span class='st'>"cluster"</span>,
        <span class='st'>"sample_name"</span>,
        <span class='st'>"post_hoc_sample_gate"</span>,
        <span class='st'>"plate_number"</span>,
        <span class='st'>"tissue"</span>,
        <span class='st'>"donor"</span><span class='op'>)</span>,
    cluster_cols <span class='op'>=</span> <span class='cn'>FALSE</span>, 
    center <span class='op'>=</span> <span class='cn'>TRUE</span>,
    symmetric <span class='op'>=</span> <span class='cn'>TRUE</span>,
    zlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>5</span>, <span class='fl'>5</span><span class='op'>)</span>,
    show_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
    main <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"Cluster: "</span>, <span class='va'>cluster</span><span class='op'>)</span>,
    column_annotation_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
       cluster <span class='op'>=</span> <span class='va'>cluster_colours</span>,
       sample_name <span class='op'>=</span> <span class='va'>sample_name_colours</span>,
       post_hoc_sample_gate <span class='op'>=</span> <span class='va'>sample_gate_colours</span>,
       plate_number <span class='op'>=</span> <span class='va'>plate_number_colours</span>,    
       tissue <span class='op'>=</span> <span class='va'>tissue_colours</span>,
       donor <span class='op'>=</span> <span class='va'>donor_colours</span><span class='op'>)</span>,
    color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>hcl.colors</a></span><span class='op'>(</span><span class='fl'>101</span>, <span class='st'>"Blue-Red 3"</span><span class='op'>)</span>,
    fontsize <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span>
<span class='op'>}</span>
<span class='fu'><a href='https://rdrr.io/r/grDevices/dev.html'>dev.off</a></span><span class='op'>(</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/grDevices/pdf.html'>pdf</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span>
    <span class='st'>"output"</span>,
    <span class='st'>"marker_genes"</span>,
    <span class='st'>"union_up_lfc1"</span>,
    <span class='st'>"heatmaps.union_up_lfc1.pdf"</span><span class='op'>)</span>,
  width <span class='op'>=</span> <span class='fl'>10</span>,
  height <span class='op'>=</span> <span class='fl'>12</span><span class='op'>)</span>
<span class='kw'>for</span> <span class='op'>(</span><span class='va'>cluster</span> <span class='kw'>in</span> <span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='va'>marker_set</span> <span class='op'>&lt;-</span> <span class='va'>markers_union_up_lfc1</span><span class='op'>[[</span><span class='va'>cluster</span><span class='op'>]</span><span class='op'>]</span>
  <span class='va'>best_set</span> <span class='op'>&lt;-</span> <span class='va'>marker_set</span><span class='op'>[</span><span class='va'>marker_set</span><span class='op'>$</span><span class='va'>Top</span> <span class='op'>&lt;=</span> <span class='fl'>10</span>, <span class='op'>]</span>
  
  <span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotHeatmap.html'>plotHeatmap</a></span><span class='op'>(</span>
    <span class='va'>sce</span>,
    features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>best_set</span><span class='op'>)</span>,
    columns <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span>
        <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span>,
        <span class='va'>sce</span><span class='op'>$</span><span class='va'>sample_name</span>,
        <span class='va'>sce</span><span class='op'>$</span><span class='va'>post_hoc_sample_gate</span>,
        <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span>,
        <span class='va'>sce</span><span class='op'>$</span><span class='va'>tissue</span>,
        <span class='va'>sce</span><span class='op'>$</span><span class='va'>donor</span><span class='op'>)</span>,
    colour_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>
        <span class='st'>"cluster"</span>,
        <span class='st'>"sample_name"</span>,
        <span class='st'>"post_hoc_sample_gate"</span>,
        <span class='st'>"plate_number"</span>,
        <span class='st'>"tissue"</span>,
        <span class='st'>"donor"</span><span class='op'>)</span>,
    cluster_cols <span class='op'>=</span> <span class='cn'>FALSE</span>, 
    center <span class='op'>=</span> <span class='cn'>TRUE</span>,
    symmetric <span class='op'>=</span> <span class='cn'>TRUE</span>,
    zlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>5</span>, <span class='fl'>5</span><span class='op'>)</span>,
    show_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
    main <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"Cluster: "</span>, <span class='va'>cluster</span><span class='op'>)</span>,
    column_annotation_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
       cluster <span class='op'>=</span> <span class='va'>cluster_colours</span>,
       sample_name <span class='op'>=</span> <span class='va'>sample_name_colours</span>,
       post_hoc_sample_gate <span class='op'>=</span> <span class='va'>sample_gate_colours</span>,
       plate_number <span class='op'>=</span> <span class='va'>plate_number_colours</span>,    
       tissue <span class='op'>=</span> <span class='va'>tissue_colours</span>,
       donor <span class='op'>=</span> <span class='va'>donor_colours</span><span class='op'>)</span>,
    color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>hcl.colors</a></span><span class='op'>(</span><span class='fl'>101</span>, <span class='st'>"Blue-Red 3"</span><span class='op'>)</span>,
    fontsize <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span>
<span class='op'>}</span>
<span class='fu'><a href='https://rdrr.io/r/grDevices/dev.html'>dev.off</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>Of course, this increased stringency is not without cost. If only upregulated genes are requested, any cluster defined by down-regulation of a marker gene will not contain that gene among the top set of features in its table. This is occasionally relevant for subtypes or other states that are distinguished by high versus low expression of particular genes. Similarly, setting an excessively high log-fold change threshold may discard otherwise useful genes. For example, a gene upregulated in a small proportion of cells of a cluster will have a small log-fold change but can still be an effective marker if the focus is on specificity rather than sensitivity.</p>
<h3 id="finding-cluster-specific-markers">Finding cluster-specific markers</h3>
<p>The genes detected in <a href="#standard-application">Standard application</a> and <a href="#using-the-log-fold-change">Using the log-fold change</a> will give a high ranking to genes that are differentially expressed in any pairwise comparison. This is because a gene only needs a very low <span class="math inline">\(p\)</span>-value in a single pairwise comparison to achieve a low <code>Top</code> value. A more stringent approach would only consider genes that are differentially expressed in all pairwise comparisons involving the cluster of interest. To achieve this, we use an intersection-union test <span class="citation" data-cites="berger1996bioequivalence">(<a href="#ref-berger1996bioequivalence" role="doc-biblioref">Berger and Hsu 1996</a>)</span> where the combined <span class="math inline">\(p\)</span>-value for each gene is the maximum of the <span class="math inline">\(p\)</span>-values from all pairwise comparisons. A gene will only achieve a low combined <span class="math inline">\(p\)</span>-value if it is strongly DE in all comparisons to other clusters.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>markers_uniquely_up</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scran/man/findMarkers.html'>findMarkers</a></span><span class='op'>(</span>
  <span class='va'>sce</span>,
  groups <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span>,
  block <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>block</span>,
  pval.type <span class='op'>=</span> <span class='st'>"all"</span>,
  direction <span class='op'>=</span> <span class='st'>"up"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span>
  <span class='va'>markers_uniquely_up</span>, 
  <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"data"</span>, <span class='st'>"marker_genes"</span>, <span class='st'>"C094_Pellicci.markers_uniquely_up.rds"</span><span class='op'>)</span>,
  compress <span class='op'>=</span> <span class='st'>"xz"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/files2.html'>dir.create</a></span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"output"</span>, <span class='st'>"marker_genes"</span>, <span class='st'>"uniquely_up"</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/message.html'>message</a></span><span class='op'>(</span><span class='st'>"Writing 'uniquely_up' marker genes to file."</span><span class='op'>)</span>
<span class='kw'>for</span> <span class='op'>(</span><span class='va'>n</span> <span class='kw'>in</span> <span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>markers_uniquely_up</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='fu'><a href='https://rdrr.io/r/base/message.html'>message</a></span><span class='op'>(</span><span class='va'>n</span><span class='op'>)</span>
  <span class='va'>gzout</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/connections.html'>gzfile</a></span><span class='op'>(</span>
    description <span class='op'>=</span> <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span>
      <span class='st'>"output"</span>, 
      <span class='st'>"marker_genes"</span>, 
      <span class='st'>"uniquely_up"</span>,
      <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"cluster_"</span>, <span class='va'>n</span>, <span class='st'>".uniquely_up.csv.gz"</span><span class='op'>)</span><span class='op'>)</span>,
    open <span class='op'>=</span> <span class='st'>"wb"</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/utils/write.table.html'>write.table</a></span><span class='op'>(</span>
    x <span class='op'>=</span> <span class='va'>markers_uniquely_up</span><span class='op'>[[</span><span class='va'>n</span><span class='op'>]</span><span class='op'>]</span> <span class='op'>%&gt;%</span>
      <span class='fu'><a href='https://rdrr.io/r/base/as.data.frame.html'>as.data.frame</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
      <span class='fu'>tibble</span><span class='fu'>::</span><span class='fu'><a href='https://tibble.tidyverse.org/reference/rownames.html'>rownames_to_column</a></span><span class='op'>(</span><span class='st'>"gene_ID"</span><span class='op'>)</span>, 
    file <span class='op'>=</span> <span class='va'>gzout</span>, 
    sep <span class='op'>=</span> <span class='st'>","</span>, 
    quote <span class='op'>=</span> <span class='cn'>FALSE</span>, 
    row.names <span class='op'>=</span> <span class='cn'>FALSE</span>,
    col.names <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/base/connections.html'>close</a></span><span class='op'>(</span><span class='va'>gzout</span><span class='op'>)</span>
<span class='op'>}</span>
</code></pre>
</div>
</details>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>interesting_uniquely_up</span> <span class='op'>&lt;-</span> <span class='va'>markers_uniquely_up</span><span class='op'>[[</span><span class='va'>chosen</span><span class='op'>]</span><span class='op'>]</span>
</code></pre>
</div>
</details>
</div>
<p>This strategy will only report genes that are highly specific to the cluster of interest. When it works, it can be highly effective as it generates a small focused set of candidate markers. However, any gene that is expressed at the same level in two or more clusters will simply not be detected. This is likely to discard many interesting genes, especially if the clusters are finely resolved with weak separation. To give a concrete example, consider a mixed population of CD4<sup>+</sup>-only, CD8<sup>+</sup>-only, double-positive and double-negative T cells. Using this strategy, neither <em>Cd4</em> or <em>Cd8</em> would be detected as subpopulation-specific markers because each gene is expressed in two subpopulations. In comparison, the strategy in <a href="#standard-application">Standard application</a> will detect both of these genes as they will be DE between at least one pair of subpopulations.</p>
<p>Figures <a href="#fig:heat-uniquely-up-logFC">14</a> and <a href="#fig:heat-uniquely-up-logExp">15</a> show examples of this strategy for cluster 4.</p>
<div class="layout-chunk" data-layout="l-body-outset">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>best_set</span> <span class='op'>&lt;-</span> <span class='va'>interesting_uniquely_up</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>20</span>, <span class='op'>]</span>
<span class='va'>logFCs</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span><span class='op'>(</span><span class='va'>best_set</span><span class='op'>[</span>, <span class='op'>-</span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>3</span><span class='op'>)</span><span class='op'>]</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>logFCs</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/grep.html'>sub</a></span><span class='op'>(</span><span class='st'>"logFC."</span>, <span class='st'>""</span>, <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>logFCs</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>logFCs</span>, breaks <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>5</span>, <span class='fl'>5</span>, length.out <span class='op'>=</span> <span class='fl'>101</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:heat-uniquely-up-logFC"></span>
<img src="C094_Pellicci.annotate_files/heat-uniquely-up-logFC-1.png" alt="Heatmap of log-fold changes for cluster 4 over all other clusters. Colours are capped at -5 and 5 to preserve dynamic range." width="624" />
<p class="caption">
Figure 14: Heatmap of log-fold changes for cluster 4 over all other clusters. Colours are capped at -5 and 5 to preserve dynamic range.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body-outset">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotHeatmap.html'>plotHeatmap</a></span><span class='op'>(</span>
  <span class='va'>sce</span>,
  features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>best_set</span><span class='op'>)</span>,
  columns <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span>
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>sample_name</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>post_hoc_sample_gate</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>tissue</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>donor</span><span class='op'>)</span>,
  colour_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>
      <span class='st'>"cluster"</span>,
      <span class='st'>"sample_name"</span>,
      <span class='st'>"post_hoc_sample_gate"</span>,
      <span class='st'>"plate_number"</span>,
      <span class='st'>"tissue"</span>,
      <span class='st'>"donor"</span><span class='op'>)</span>,
  cluster_cols <span class='op'>=</span> <span class='cn'>FALSE</span>, 
  center <span class='op'>=</span> <span class='cn'>TRUE</span>,
  symmetric <span class='op'>=</span> <span class='cn'>TRUE</span>,
  zlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>5</span>, <span class='fl'>5</span><span class='op'>)</span>,
  show_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
  annotation_row <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
    Sig <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span>
      <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span><span class='va'>best_set</span><span class='op'>[</span>, <span class='st'>"FDR"</span><span class='op'>]</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span>, <span class='st'>"Yes"</span>, <span class='st'>"No"</span><span class='op'>)</span>,
      levels <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"No"</span>, <span class='st'>"Yes"</span><span class='op'>)</span><span class='op'>)</span>,
    row.names <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>best_set</span><span class='op'>)</span><span class='op'>)</span>,
  main <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"Cluster: "</span>, <span class='va'>chosen</span><span class='op'>)</span>,
  column_annotation_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
     mi_label_fine_collapsed <span class='op'>=</span> <span class='va'>mi_label_fine_colours</span>,
     cluster <span class='op'>=</span> <span class='va'>cluster_colours</span>,
     sample_name <span class='op'>=</span> <span class='va'>sample_name_colours</span>,
     post_hoc_sample_gate <span class='op'>=</span> <span class='va'>sample_gate_colours</span>,
     plate_number <span class='op'>=</span> <span class='va'>plate_number_colours</span>,    
     tissue <span class='op'>=</span> <span class='va'>tissue_colours</span>,
     donor <span class='op'>=</span> <span class='va'>donor_colours</span>,
     Sig <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Yes"</span> <span class='op'>=</span> <span class='st'>"red"</span>, <span class='st'>"No"</span> <span class='op'>=</span> <span class='st'>"lightgrey"</span><span class='op'>)</span><span class='op'>)</span>,
  color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>hcl.colors</a></span><span class='op'>(</span><span class='fl'>101</span>, <span class='st'>"Blue-Red 3"</span><span class='op'>)</span>,
  fontsize <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:heat-uniquely-up-logExp"></span>
<img src="C094_Pellicci.annotate_files/heat-uniquely-up-logExp-1.png" alt="Heatmap of log-expression values in each sample for the top 10 uniquely upregulated marker genes in cluster 4. Each column is a sample, each row a gene. Colours are capped at -5 and 5 to preserve dynamic range." width="624" />
<p class="caption">
Figure 15: Heatmap of log-expression values in each sample for the top 10 uniquely upregulated marker genes in cluster 4. Each column is a sample, each row a gene. Colours are capped at -5 and 5 to preserve dynamic range.
</p>
</div>
</div>
<p>Similar figures showing the top 100 uniquely upregulated marker genes for each cluster are available in <a href="../output/marker_genes/uniquely_up/heatmaps.uniquely_up.pdf"><code>output/marker_genes/uniquely_up/heatmaps.uniquely_up.pdf</code></a>. Recall that the stringency of the ‘cluster-specific marker’ criteria is such that a cluster may only have a small number of statistically significant marker genes or indeed none at all. Therefore, these plots include a row annotation highlighting if the gene is a statistically significant cluster-specific marker (i.e. <span class="math inline">\(FDR &lt; 0.05\)</span>).</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/grDevices/dev.html'>dev.off</a></span><span class='op'>(</span><span class='op'>)</span> <span class='co'># Needed to ensure next figure saves properly.</span>
<span class='fu'><a href='https://rdrr.io/r/grDevices/pdf.html'>pdf</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span>
    <span class='st'>"output"</span>,
    <span class='st'>"marker_genes"</span>,
    <span class='st'>"uniquely_up"</span>,
    <span class='st'>"heatmaps.uniquely_up.pdf"</span><span class='op'>)</span>,
  width <span class='op'>=</span> <span class='fl'>10</span>,
  height <span class='op'>=</span> <span class='fl'>12</span><span class='op'>)</span>
<span class='kw'>for</span> <span class='op'>(</span><span class='va'>cluster</span> <span class='kw'>in</span> <span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='va'>marker_set</span> <span class='op'>&lt;-</span> <span class='va'>markers_uniquely_up</span><span class='op'>[[</span><span class='va'>cluster</span><span class='op'>]</span><span class='op'>]</span>
  <span class='va'>best_set</span> <span class='op'>&lt;-</span> <span class='va'>marker_set</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>100</span>, <span class='op'>]</span>
  
<span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotHeatmap.html'>plotHeatmap</a></span><span class='op'>(</span>
  <span class='va'>sce</span>,
  features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>best_set</span><span class='op'>)</span>,
  columns <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span>
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>sample_name</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>post_hoc_sample_gate</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>tissue</span>,
      <span class='va'>sce</span><span class='op'>$</span><span class='va'>donor</span><span class='op'>)</span>,
  colour_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>
      <span class='st'>"cluster"</span>,
      <span class='st'>"sample_name"</span>,
      <span class='st'>"post_hoc_sample_gate"</span>,
      <span class='st'>"plate_number"</span>,
      <span class='st'>"tissue"</span>,
      <span class='st'>"donor"</span><span class='op'>)</span>,
  cluster_cols <span class='op'>=</span> <span class='cn'>FALSE</span>, 
  center <span class='op'>=</span> <span class='cn'>TRUE</span>,
  symmetric <span class='op'>=</span> <span class='cn'>TRUE</span>,
  zlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>5</span>, <span class='fl'>5</span><span class='op'>)</span>,
  show_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
  annotation_row <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
    Sig <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span>
      <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span><span class='va'>best_set</span><span class='op'>[</span>, <span class='st'>"FDR"</span><span class='op'>]</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span>, <span class='st'>"Yes"</span>, <span class='st'>"No"</span><span class='op'>)</span>,
      levels <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"No"</span>, <span class='st'>"Yes"</span><span class='op'>)</span><span class='op'>)</span>,
    row.names <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>best_set</span><span class='op'>)</span><span class='op'>)</span>,
  main <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"Cluster: "</span>, <span class='va'>cluster</span><span class='op'>)</span>,
  column_annotation_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
     mi_label_fine_collapsed <span class='op'>=</span> <span class='va'>mi_label_fine_colours</span>,
     cluster <span class='op'>=</span> <span class='va'>cluster_colours</span>,
     sample_name <span class='op'>=</span> <span class='va'>sample_name_colours</span>,
     post_hoc_sample_gate <span class='op'>=</span> <span class='va'>sample_gate_colours</span>,
     plate_number <span class='op'>=</span> <span class='va'>plate_number_colours</span>,    
     tissue <span class='op'>=</span> <span class='va'>tissue_colours</span>,
     donor <span class='op'>=</span> <span class='va'>donor_colours</span>,
     Sig <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Yes"</span> <span class='op'>=</span> <span class='st'>"red"</span>, <span class='st'>"No"</span> <span class='op'>=</span> <span class='st'>"lightgrey"</span><span class='op'>)</span><span class='op'>)</span>,
  color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>hcl.colors</a></span><span class='op'>(</span><span class='fl'>101</span>, <span class='st'>"Blue-Red 3"</span><span class='op'>)</span>,
  fontsize <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span>
<span class='op'>}</span>
</code></pre>
</div>
</details>
</div>
<p>Finally, a common visualisation of cluster-specific marker genes is in a so-called ‘dot plot’ or ‘bubble plot.’ In these plots, the size of the dot corresponds to the percentage of cells expressing the feature in each cluster while the colour represents the average expression level.</p>
<p>Figure <a href="#fig:dot-plot">16</a> is a dot plot of the top-3 ‘uniquely up-regulated’ marker genes for each cluster.</p>
<div class="layout-chunk" data-layout="l-body-outset">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>markers</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/unlist.html'>unlist</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>lapply</a></span><span class='op'>(</span><span class='va'>markers_uniquely_up</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span> <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>, <span class='fl'>3</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>x</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/SingleCellExperiment/man/assays.html'>logcounts</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span><span class='va'>markers</span>, <span class='op'>]</span>
<span class='va'>y</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/do.call.html'>do.call</a></span><span class='op'>(</span>
  <span class='va'>rbind</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>lapply</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span><span class='op'>)</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>cluster</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
      exp <span class='op'>=</span> <span class='fu'>Matrix</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/Matrix/man/colSums.html'>rowMeans</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span> <span class='op'>==</span> <span class='va'>cluster</span><span class='op'>]</span><span class='op'>)</span>,
      frac_exp <span class='op'>=</span> <span class='fl'>100</span> <span class='op'>*</span> <span class='fu'>Matrix</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/Matrix/man/colSums.html'>rowMeans</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span> <span class='op'>==</span> <span class='va'>cluster</span><span class='op'>]</span> <span class='op'>&gt;</span> <span class='fl'>0</span><span class='op'>)</span>,
      cluster <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>cluster</span>, levels <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span><span class='op'>)</span><span class='op'>)</span>,
      gene <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>,
      stringsAsFactors <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
  <span class='op'>}</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/group_by.html'>group_by</a></span><span class='op'>(</span><span class='va'>gene</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/mutate.html'>mutate</a></span><span class='op'>(</span>
    centered_exp <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/scale.html'>scale</a></span><span class='op'>(</span><span class='va'>exp</span>, scale <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>,
    z_scaled_exp <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/scale.html'>scale</a></span><span class='op'>(</span><span class='va'>exp</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/group_by.html'>ungroup</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/mutate.html'>mutate</a></span><span class='op'>(</span>gene <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>gene</span>, levels <span class='op'>=</span> <span class='va'>markers</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://ggplot2.tidyverse.org/reference/ggplot.html'>ggplot</a></span><span class='op'>(</span>data <span class='op'>=</span> <span class='va'>y</span>, <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>gene</span>, y <span class='op'>=</span> <span class='va'>cluster</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='va'>frac_exp</span>, colour <span class='op'>=</span> <span class='va'>exp</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_viridis.html'>scale_colour_viridis_c</a></span><span class='op'>(</span>
    option <span class='op'>=</span> <span class='st'>"plasma"</span>, 
    <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guide_legend.html'>guide_legend</a></span><span class='op'>(</span>title <span class='op'>=</span> <span class='st'>"Average Expression"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_size.html'>scale_size</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guide_legend.html'>guide_legend</a></span><span class='op'>(</span>title <span class='op'>=</span> <span class='st'>"Percent Expressed"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>10</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>angle <span class='op'>=</span> <span class='fl'>45</span>, hjust <span class='op'>=</span> <span class='fl'>1</span>, size <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:dot-plot"></span>
<img src="C094_Pellicci.annotate_files/dot-plot-1.png" alt="Dot plot of the top-3 'uniquely up-regulated' marker genes for each cluster. The size of the dot corresponds to the percentage of cells expressing the feature in each cluster while the colour represents the average expression level" width="624" />
<p class="caption">
Figure 16: Dot plot of the top-3 ‘uniquely up-regulated’ marker genes for each cluster. The size of the dot corresponds to the percentage of cells expressing the feature in each cluster while the colour represents the average expression level
</p>
</div>
</div>
<p>As can be seen in Figure <a href="#fig:dot-plot">16</a>, several of the marker genes are relatively highly expressed across all clusters. To compensate for this, Figure <a href="#fig:dot-plot-centered">17</a> first centres the average expression level of each gene before visualization.</p>
<div class="layout-chunk" data-layout="l-body-outset">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>ylim</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/MathFun.html'>abs</a></span><span class='op'>(</span><span class='va'>y</span><span class='op'>$</span><span class='va'>z_scaled_exp</span><span class='op'>)</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/MathFun.html'>abs</a></span><span class='op'>(</span><span class='va'>y</span><span class='op'>$</span><span class='va'>z_scaled_exp</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://ggplot2.tidyverse.org/reference/ggplot.html'>ggplot</a></span><span class='op'>(</span>data <span class='op'>=</span> <span class='va'>y</span>, <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>gene</span>, y <span class='op'>=</span> <span class='va'>cluster</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='va'>frac_exp</span>, colour <span class='op'>=</span> <span class='va'>z_scaled_exp</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_size.html'>scale_size</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guide_legend.html'>guide_legend</a></span><span class='op'>(</span>title <span class='op'>=</span> <span class='st'>"Percent Expressed"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_gradient.html'>scale_colour_gradient2</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guide_legend.html'>guide_legend</a></span><span class='op'>(</span>title <span class='op'>=</span> <span class='st'>"Centered Average Expression"</span><span class='op'>)</span>,
    low <span class='op'>=</span> <span class='st'>"blue3"</span>,
    mid <span class='op'>=</span> <span class='st'>"white"</span>,
    high <span class='op'>=</span> <span class='st'>"red3"</span>,
    limits <span class='op'>=</span> <span class='va'>ylim</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>cowplot</span><span class='fu'>::</span><span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>angle <span class='op'>=</span> <span class='fl'>45</span>, hjust <span class='op'>=</span> <span class='fl'>1</span>, size <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:dot-plot-centered"></span>
<img src="C094_Pellicci.annotate_files/dot-plot-centered-1.png" alt="Dot plot of the top-3 'uniquely up-regulated' marker genes for each cluster. The size of the dot corresponds to the percentage of cells expressing the feature in each cluster while the colour represents the average (centered) expression level. Each gene (column) is centred by the average expression over all clusters." width="624" />
<p class="caption">
Figure 17: Dot plot of the top-3 ‘uniquely up-regulated’ marker genes for each cluster. The size of the dot corresponds to the percentage of cells expressing the feature in each cluster while the colour represents the average (centered) expression level. Each gene (column) is centred by the average expression over all clusters.
</p>
</div>
</div>
<h2 id="summary">Summary</h2>
<p>The tables of marker genes are available as CSV files in <a href="../output/marker_genes/"><code>output/marker_genes/</code></a> and are named as follows:</p>
<ul>
<li><code>union_any</code>: Marker genes from <a href="#standard-application">Standard application</a> whereby marker genes are formed from the <strong>union</strong> of genes differentially expressed in <strong>any</strong> direction.</li>
<li><code>union_up</code>: Marker genes from <a href="#using-the-log-fold-change">Using the log-fold change</a> whereby marker genes are formed from the <strong>union</strong> of genes differentially expressed that are <strong>up</strong>regulated.</li>
<li><code>union_up_lfc1</code>: Marker genes from <a href="#using-the-log-fold-change">Using the log-fold change</a> whereby marker genes are formed from the <strong>union</strong> of genes differentially expressed that are <strong>up</strong>regulated with a log-fold change of at least 1 (<strong>lfc1</strong>).</li>
<li><code>uniquely_up</code>: Marker genes from <a href="#finding-cluster-specific-markers">Finding cluster-specific markers</a> whereby marker genes are cluster-specific (<strong>uniquely</strong>) differentially expressed genes that are <strong>up</strong>regulated.</li>
</ul>
<h3 id="using-marker-genes-to-perform-cell-type-annotation">Using marker genes to perform cell type annotation</h3>
<p>Regardless of the strategy employed, <a href="#standard-application">Standard application</a>, <a href="#using-the-log-fold-change">Using the log-fold change</a> or <a href="#finding-cluster-specific-markers">Finding cluster-specific markers</a>, the result is a gene list that needs to be examined to try to identify the cell type. As such, they can be analysed like any other gene list, just as you might a list of differentially expressed genes from a bulk RNA-seq experiment.</p>
<p>This includes:</p>
<ul>
<li>Staring at heatmaps and spreadsheets, such as those included with this report, and googling some of the more informative genes, such as the cluster-specific marker genes.</li>
<li>Gene set enrichment analysis, e.g., using GO or KEGG terms. This can require a fair amount of manual work and knowledge of the biological system but is an essential complement to the automated approach to cell type labelling.</li>
</ul>
<h1 id="cell-cycle-assignment">Cell cycle assignment</h1>
<p>On occasion, it can be desirable to determine cell cycle activity from scRNA-seq data. In and of itself, the distribution of cells across phases of the cell cycle is not usually informative, but we can use this to determine if there are differences in cell cycle progression between subpopulations or across treatment conditions. Many of the key events in the cell cycle (e.g., passage through checkpoints) are post-translational and thus not directly visible in transcriptomic data; nonetheless, there are enough changes in expression that can be exploited to determine cell cycle phase.</p>
<h2 id="using-the-cyclins">Using the cyclins</h2>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>cyclin_genes</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/grep.html'>grep</a></span><span class='op'>(</span><span class='st'>"^CCN[ABDE][0-9]$"</span>, <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span><span class='op'>]</span>
<span class='va'>cyclin_genes</span>
</code></pre>
</div>
</details>
</div>
<p>The cyclins control progression through the cell cycle and have well-characterized patterns of expression across cell cycle phases. The cyclins available in this dataset are as follows: CCND3, CCNB1, CCNB2, CCNE2, CCNE1, CCNA2, CCND2, CCND1, CCNB3, and CCNA1. Cyclin D is expressed throughout but peaks at G1; cyclin E is expressed highest in the G1/S transition; cyclin A is expressed across S and G2; and cyclin B is expressed highest in late G2 and mitosis. Inspection of the relative expression of cyclins across the population can often be sufficient to determine the relative cell cycle activity in each cluster.</p>
<p>Figure <a href="#fig:cyclin-heatmap">18</a> shows the expression of the cyclin genes in the dataset. Unfortunately, no obviously different expression of cyclins were observed in a particular cluster of this dataset, suggesting there could have no cell cycle-related difference between the clusters.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotHeatmap.html'>plotHeatmap</a></span><span class='op'>(</span>
  <span class='va'>sce</span>,
  colour_columns_by <span class='op'>=</span> <span class='st'>"cluster"</span>,
  cluster_cols <span class='op'>=</span> <span class='cn'>FALSE</span>, 
  columns <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span><span class='op'>)</span>,
  cluster_rows <span class='op'>=</span> <span class='cn'>FALSE</span>, 
  features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/sort.html'>sort</a></span><span class='op'>(</span><span class='va'>cyclin_genes</span><span class='op'>)</span>,
  column_annotation_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>cluster <span class='op'>=</span> <span class='va'>cluster_colours</span><span class='op'>)</span>,
  show_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
  zlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>4</span><span class='op'>)</span>,
  color <span class='op'>=</span> <span class='fu'>viridisLite</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/viridisLite/man/viridis.html'>viridis</a></span><span class='op'>(</span><span class='fl'>101</span>, option <span class='op'>=</span> <span class='st'>"inferno"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:cyclin-heatmap"></span>
<img src="C094_Pellicci.annotate_files/cyclin-heatmap-1.png" alt="Heatmap of the log-normalized expression values of the cyclin genes in the dataset." width="624" />
<p class="caption">
Figure 18: Heatmap of the log-normalized expression values of the cyclin genes in the dataset.
</p>
</div>
</div>
<p>We can also use standard differential expression methods to look for upregulation of each cyclin, allowing us to determine if there are more cells in the corresponding phase of the cell cycle between subpopulations.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>markers</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scran/man/findMarkers.html'>findMarkers</a></span><span class='op'>(</span>
  <span class='va'>sce</span>,
  subset.row <span class='op'>=</span> <span class='va'>cyclin_genes</span>, 
  groups <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span>,
  test.type <span class='op'>=</span> <span class='st'>"wilcox"</span>, 
  direction <span class='op'>=</span> <span class='st'>"up"</span>, 
  block <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>Here we use pairwise Wilcoxon rank sum tests to compute differential expression of cyclin genes between each pair of clusters. The Wilcoxon test estimates effect sizes as the area under the curve (<code>AUC</code> in the tables below). Consider the distribution of expression values for gene <span class="math inline">\(X\)</span> within each of two groups <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span>. The <code>AUC</code> is the probability that a randomly selected cell in <span class="math inline">\(A\)</span> has a greater expression of <span class="math inline">\(X\)</span> than a randomly selected cell in <span class="math inline">\(B\)</span><a href="#fn8" class="footnote-ref" id="fnref8" role="doc-noteref"><sup>8</sup></a>. Concordance probabilities near 0 indicate that most observations in <span class="math inline">\(A\)</span> are lower than most observations in <span class="math inline">\(B\)</span>; conversely, probabilities near 1 indicate that most observations in <span class="math inline">\(A\)</span> are higher than those in <span class="math inline">\(B\)</span>. The Wilcoxon rank sum test effectively tests for significant deviations from an <code>AUC</code> of 0.5.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>knitr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/knitr/man/kable.html'>kable</a></span><span class='op'>(</span>
  <span class='fu'>.adf</span><span class='op'>(</span><span class='va'>markers</span><span class='op'>[[</span><span class='fl'>4</span><span class='op'>]</span><span class='op'>]</span><span class='op'>)</span>,
  caption <span class='op'>=</span> <span class='st'>"Cyclin DE analysis comparing cluster 4 to all other clusters."</span>,
  digits <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<table>
<caption><span id="tab:unnamed-chunk-34">Table 4: </span>Cyclin DE analysis comparing cluster 4 to all other clusters.</caption>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">Top</th>
<th style="text-align: right;">p.value</th>
<th style="text-align: right;">FDR</th>
<th style="text-align: right;">summary.AUC</th>
<th style="text-align: right;">AUC.1</th>
<th style="text-align: right;">AUC.2</th>
<th style="text-align: right;">AUC.3</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">CCND3</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0.001</td>
<td style="text-align: right;">0.007</td>
<td style="text-align: right;">0.543</td>
<td style="text-align: right;">0.506</td>
<td style="text-align: right;">0.543</td>
<td style="text-align: right;">0.504</td>
</tr>
<tr class="even">
<td style="text-align: left;">CCNE2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0.007</td>
<td style="text-align: right;">0.033</td>
<td style="text-align: right;">0.508</td>
<td style="text-align: right;">0.508</td>
<td style="text-align: right;">0.510</td>
<td style="text-align: right;">0.504</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CCND2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0.157</td>
<td style="text-align: right;">0.393</td>
<td style="text-align: right;">0.529</td>
<td style="text-align: right;">0.499</td>
<td style="text-align: right;">0.500</td>
<td style="text-align: right;">0.529</td>
</tr>
<tr class="even">
<td style="text-align: left;">CCNB2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0.033</td>
<td style="text-align: right;">0.109</td>
<td style="text-align: right;">0.512</td>
<td style="text-align: right;">0.512</td>
<td style="text-align: right;">0.508</td>
<td style="text-align: right;">0.502</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CCNA2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0.388</td>
<td style="text-align: right;">0.776</td>
<td style="text-align: right;">0.508</td>
<td style="text-align: right;">0.503</td>
<td style="text-align: right;">0.502</td>
<td style="text-align: right;">0.508</td>
</tr>
<tr class="even">
<td style="text-align: left;">CCNB1</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0.543</td>
<td style="text-align: right;">0.905</td>
<td style="text-align: right;">0.501</td>
<td style="text-align: right;">0.501</td>
<td style="text-align: right;">0.503</td>
<td style="text-align: right;">0.504</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CCNE1</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">1.000</td>
<td style="text-align: right;">1.000</td>
<td style="text-align: right;">0.506</td>
<td style="text-align: right;">0.506</td>
<td style="text-align: right;">0.505</td>
<td style="text-align: right;">0.506</td>
</tr>
<tr class="even">
<td style="text-align: left;">CCND1</td>
<td style="text-align: right;">8</td>
<td style="text-align: right;">1.000</td>
<td style="text-align: right;">1.000</td>
<td style="text-align: right;">0.497</td>
<td style="text-align: right;">0.497</td>
<td style="text-align: right;">0.499</td>
<td style="text-align: right;">0.500</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CCNB3</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">1.000</td>
<td style="text-align: right;">1.000</td>
<td style="text-align: right;">0.501</td>
<td style="text-align: right;">0.501</td>
<td style="text-align: right;">0.499</td>
<td style="text-align: right;">0.500</td>
</tr>
<tr class="even">
<td style="text-align: left;">CCNA1</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">1.000</td>
<td style="text-align: right;">1.000</td>
<td style="text-align: right;">0.500</td>
<td style="text-align: right;">0.500</td>
<td style="text-align: right;">0.500</td>
<td style="text-align: right;">0.500</td>
</tr>
</tbody>
</table>
</div>
<p>Direct examination of cyclin expression is easily understood, interpreted and validated with other technologies. However, it is best suited for statements about relative cell cycle activity; for example, we would find it difficult to assign cell cycle phase in Figure <a href="#fig:cyclin-heatmap">18</a> without the presence of clusters spanning all phases to provide benchmarks for “high” and “low” expression of each cyclin. We also assume that cyclin expression is not affected by biological processes other than the cell cycle, which may be a strong assumption in some cases, e.g., malignant cells.</p>
<h1 id="concluding-remarks">Concluding remarks</h1>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span>
  <span class='va'>sce</span>,
  <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"data"</span>, <span class='st'>"SCEs"</span>, <span class='st'>"C094_Pellicci_annotate.SCE.rds"</span><span class='op'>)</span>,
  compress <span class='op'>=</span> <span class='st'>"xz"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>The processed <em>SingleCellExperiment</em> object is available (see <a href="../data/SCEs/C094_Pellicci_annotate.SCE.rds"><code>data/SCEs/C094_Pellicci_annotate.SCE.rds</code></a>). This will be used in downstream analyses, e.g., differential expression analysis between conditions within each cluster and differential abundance analyses between conditions for each cluster.</p>
<h1 class="appendix" id="additional-information">Additional information</h1>
<p>The following are available on request:</p>
<ul>
<li>Full CSV tables of any data presented.</li>
<li>PDF/PNG files of any static plots.</li>
</ul>
<h1 class="appendix" id="session-info">Session info</h1>
<p>The analysis and this document were prepared using the following software:</p>
<summary>
The analysis and this document were prepared using the following software (click triangle to expand)
</summary>
<details>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>sessioninfo</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/sessioninfo/man/session_info.html'>session_info</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<pre><code>─ Session info ─────────────────────────────────────────────────────
 setting  value                       
 version  R version 4.0.3 (2020-10-10)
 os       CentOS Linux 7 (Core)       
 system   x86_64, linux-gnu           
 ui       X11                         
 language (EN)                        
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 
 tz       Australia/Melbourne         
 date     2021-07-01                  

─ Packages ─────────────────────────────────────────────────────────
 ! package                * version  date       lib source        
 P annotate                 1.68.0   2020-10-27 [?] Bioconductor  
 P AnnotationDbi            1.52.0   2020-10-27 [?] Bioconductor  
 P AnnotationHub            2.22.0   2020-10-27 [?] Bioconductor  
 P assertthat               0.2.1    2019-03-21 [?] CRAN (R 4.0.0)
   batchelor              * 1.6.3    2021-04-16 [1] Bioconductor  
 P beachmat                 2.6.4    2020-12-20 [?] Bioconductor  
 P beeswarm                 0.3.1    2021-03-07 [?] CRAN (R 4.0.2)
 P Biobase                * 2.50.0   2020-10-27 [?] Bioconductor  
 P BiocFileCache            1.14.0   2020-10-27 [?] Bioconductor  
 P BiocGenerics           * 0.36.0   2020-10-27 [?] Bioconductor  
 P BiocManager              1.30.12  2021-03-28 [?] CRAN (R 4.0.3)
 P BiocNeighbors            1.8.2    2020-12-07 [?] Bioconductor  
 P BiocParallel           * 1.24.1   2020-11-06 [?] Bioconductor  
 P BiocSingular             1.6.0    2020-10-27 [?] Bioconductor  
 P BiocStyle              * 2.18.1   2020-11-24 [?] Bioconductor  
 P BiocVersion              3.12.0   2020-04-27 [?] Bioconductor  
 P bit                      4.0.4    2020-08-04 [?] CRAN (R 4.0.0)
 P bit64                    4.0.5    2020-08-30 [?] CRAN (R 4.0.0)
 P bitops                   1.0-6    2013-08-17 [?] CRAN (R 4.0.0)
 P blob                     1.2.1    2020-01-20 [?] CRAN (R 4.0.0)
 P bluster                  1.0.0    2020-10-27 [?] Bioconductor  
 P bslib                    0.2.4    2021-01-25 [?] CRAN (R 4.0.2)
 P cachem                   1.0.4    2021-02-13 [?] CRAN (R 4.0.2)
   celldex                * 1.0.0    2020-10-29 [1] Bioconductor  
 P cellranger               1.1.0    2016-07-27 [?] CRAN (R 4.0.0)
 P cli                      2.4.0    2021-04-05 [?] CRAN (R 4.0.3)
 P colorspace               2.0-0    2020-11-11 [?] CRAN (R 4.0.2)
 P cowplot                * 1.1.1    2020-12-30 [?] CRAN (R 4.0.0)
 P crayon                   1.4.1    2021-02-08 [?] CRAN (R 4.0.3)
 P curl                     4.3      2019-12-02 [?] CRAN (R 4.0.0)
 P DBI                      1.1.1    2021-01-15 [?] CRAN (R 4.0.2)
 P dbplyr                   2.1.0    2021-02-03 [?] CRAN (R 4.0.2)
 P DelayedArray             0.16.3   2021-03-24 [?] Bioconductor  
 P DelayedMatrixStats       1.12.3   2021-02-03 [?] Bioconductor  
 P DESeq2                   1.30.1   2021-02-19 [?] Bioconductor  
 P digest                   0.6.27   2020-10-24 [?] CRAN (R 4.0.0)
 P distill                  1.2      2021-01-13 [?] CRAN (R 4.0.0)
 P downlit                  0.2.1    2020-11-04 [?] CRAN (R 4.0.0)
 P dplyr                  * 1.0.5    2021-03-05 [?] CRAN (R 4.0.2)
 P dqrng                    0.2.1    2019-05-17 [?] CRAN (R 4.0.0)
 P edgeR                  * 3.32.1   2021-01-14 [?] Bioconductor  
 P ellipsis                 0.3.1    2020-05-15 [?] CRAN (R 4.0.0)
 P evaluate                 0.14     2019-05-28 [?] CRAN (R 4.0.0)
   ExperimentHub            1.16.1   2021-04-16 [1] Bioconductor  
 P fansi                    0.4.2    2021-01-15 [?] CRAN (R 4.0.2)
 P farver                   2.1.0    2021-02-28 [?] CRAN (R 4.0.2)
 P fastmap                  1.1.0    2021-01-25 [?] CRAN (R 4.0.2)
 P genefilter               1.72.1   2021-01-21 [?] Bioconductor  
 P geneplotter              1.68.0   2020-10-27 [?] Bioconductor  
 P generics                 0.1.0    2020-10-31 [?] CRAN (R 4.0.0)
 P GenomeInfoDb           * 1.26.4   2021-03-10 [?] Bioconductor  
 P GenomeInfoDbData         1.2.4    2020-12-07 [?] Bioconductor  
 P GenomicRanges          * 1.42.0   2020-10-27 [?] Bioconductor  
 P ggbeeswarm               0.6.0    2017-08-07 [?] CRAN (R 4.0.0)
 P ggplot2                * 3.3.3    2020-12-30 [?] CRAN (R 4.0.2)
 P Glimma                 * 2.0.0    2020-10-27 [?] Bioconductor  
 P glue                     1.4.2    2020-08-27 [?] CRAN (R 4.0.0)
 P gridExtra                2.3      2017-09-09 [?] CRAN (R 4.0.0)
 P gtable                   0.3.0    2019-03-25 [?] CRAN (R 4.0.0)
 P here                   * 1.0.1    2020-12-13 [?] CRAN (R 4.0.0)
 P highr                    0.8      2019-03-20 [?] CRAN (R 4.0.0)
 P htmltools                0.5.1.1  2021-01-22 [?] CRAN (R 4.0.2)
 P htmlwidgets              1.5.3    2020-12-10 [?] CRAN (R 4.0.2)
 P httpuv                   1.5.5    2021-01-13 [?] CRAN (R 4.0.0)
 P httr                     1.4.2    2020-07-20 [?] CRAN (R 4.0.0)
 P igraph                   1.2.6    2020-10-06 [?] CRAN (R 4.0.0)
 P interactiveDisplayBase   1.28.0   2020-10-27 [?] Bioconductor  
 P IRanges                * 2.24.1   2020-12-12 [?] Bioconductor  
 P irlba                    2.3.3    2019-02-05 [?] CRAN (R 4.0.0)
 P janitor                * 2.1.0    2021-01-05 [?] CRAN (R 4.0.3)
 P jquerylib                0.1.3    2020-12-17 [?] CRAN (R 4.0.2)
 P jsonlite                 1.7.2    2020-12-09 [?] CRAN (R 4.0.2)
 P knitr                    1.31     2021-01-27 [?] CRAN (R 4.0.2)
 P labeling                 0.4.2    2020-10-20 [?] CRAN (R 4.0.0)
 P later                    1.1.0.1  2020-06-05 [?] CRAN (R 4.0.0)
 P lattice                  0.20-41  2020-04-02 [3] CRAN (R 4.0.3)
 P lifecycle                1.0.0    2021-02-15 [?] CRAN (R 4.0.2)
 P limma                  * 3.46.0   2020-10-27 [?] Bioconductor  
 P locfit                   1.5-9.4  2020-03-25 [?] CRAN (R 4.0.0)
 P lubridate                1.7.10   2021-02-26 [?] CRAN (R 4.0.3)
 P magrittr                 2.0.1    2020-11-17 [?] CRAN (R 4.0.2)
 P Matrix                   1.2-18   2019-11-27 [3] CRAN (R 4.0.3)
 P MatrixGenerics         * 1.2.1    2021-01-30 [?] Bioconductor  
 P matrixStats            * 0.58.0   2021-01-29 [?] CRAN (R 4.0.2)
 P memoise                  2.0.0    2021-01-26 [?] CRAN (R 4.0.2)
 P mime                     0.10     2021-02-13 [?] CRAN (R 4.0.2)
 P msigdbr                * 7.2.1    2020-10-02 [?] CRAN (R 4.0.0)
 P munsell                  0.5.0    2018-06-12 [?] CRAN (R 4.0.0)
 P patchwork              * 1.1.1    2020-12-17 [?] CRAN (R 4.0.0)
 P pheatmap               * 1.0.12   2019-01-04 [?] CRAN (R 4.0.0)
 P pillar                   1.5.1    2021-03-05 [?] CRAN (R 4.0.2)
 P pkgconfig                2.0.3    2019-09-22 [?] CRAN (R 4.0.0)
 P Polychrome               1.2.6    2020-11-11 [?] CRAN (R 4.0.0)
 P promises                 1.2.0.1  2021-02-11 [?] CRAN (R 4.0.2)
 P purrr                    0.3.4    2020-04-17 [?] CRAN (R 4.0.0)
 P R6                       2.5.0    2020-10-28 [?] CRAN (R 4.0.0)
 P rappdirs                 0.3.3    2021-01-31 [?] CRAN (R 4.0.2)
 P RColorBrewer             1.1-2    2014-12-07 [?] CRAN (R 4.0.0)
 P Rcpp                     1.0.6    2021-01-15 [?] CRAN (R 4.0.2)
 P RCurl                    1.98-1.3 2021-03-16 [?] CRAN (R 4.0.5)
 P readxl                 * 1.3.1    2019-03-13 [?] CRAN (R 4.0.0)
   ResidualMatrix           1.0.0    2020-10-27 [1] Bioconductor  
 P rlang                    0.4.10   2020-12-30 [?] CRAN (R 4.0.2)
 P rmarkdown              * 2.7      2021-02-19 [?] CRAN (R 4.0.2)
 P rprojroot                2.0.2    2020-11-15 [?] CRAN (R 4.0.2)
 P RSQLite                  2.2.5    2021-03-27 [?] CRAN (R 4.0.3)
 P rsvd                     1.0.3    2020-02-17 [?] CRAN (R 4.0.0)
 P S4Vectors              * 0.28.1   2020-12-09 [?] Bioconductor  
 P sass                     0.3.1    2021-01-24 [?] CRAN (R 4.0.2)
 P scales                   1.1.1    2020-05-11 [?] CRAN (R 4.0.0)
 P scater                 * 1.18.6   2021-02-26 [?] Bioconductor  
 P scatterplot3d            0.3-41   2018-03-14 [?] CRAN (R 4.0.0)
 P scran                  * 1.18.5   2021-02-04 [?] Bioconductor  
 P scuttle                  1.0.4    2020-12-17 [?] Bioconductor  
 P sessioninfo              1.1.1    2018-11-05 [?] CRAN (R 4.0.0)
 P shiny                    1.6.0    2021-01-25 [?] CRAN (R 4.0.2)
 P SingleCellExperiment   * 1.12.0   2020-10-27 [?] Bioconductor  
   SingleR                * 1.4.1    2021-02-02 [1] Bioconductor  
 P snakecase                0.11.0   2019-05-25 [?] CRAN (R 4.0.0)
 P sparseMatrixStats        1.2.1    2021-02-02 [?] Bioconductor  
 P statmod                  1.4.35   2020-10-19 [?] CRAN (R 4.0.0)
 P stringi                  1.5.3    2020-09-09 [?] CRAN (R 4.0.0)
 P stringr                  1.4.0    2019-02-10 [?] CRAN (R 4.0.0)
 P SummarizedExperiment   * 1.20.0   2020-10-27 [?] Bioconductor  
 P survival                 3.2-7    2020-09-28 [3] CRAN (R 4.0.3)
 P tibble                   3.1.0    2021-02-25 [?] CRAN (R 4.0.2)
 P tidyr                  * 1.1.3    2021-03-03 [?] CRAN (R 4.0.3)
 P tidyselect               1.1.0    2020-05-11 [?] CRAN (R 4.0.0)
 P utf8                     1.2.1    2021-03-12 [?] CRAN (R 4.0.2)
 P vctrs                    0.3.7    2021-03-29 [?] CRAN (R 4.0.3)
 P vipor                    0.4.5    2017-03-22 [?] CRAN (R 4.0.0)
 P viridis                  0.5.1    2018-03-29 [?] CRAN (R 4.0.0)
 P viridisLite              0.3.0    2018-02-01 [?] CRAN (R 4.0.0)
 P withr                    2.4.1    2021-01-26 [?] CRAN (R 4.0.2)
 P xfun                     0.22     2021-03-11 [?] CRAN (R 4.0.2)
 P XML                      3.99-0.6 2021-03-16 [?] CRAN (R 4.0.5)
 P xtable                   1.8-4    2019-04-21 [?] CRAN (R 4.0.0)
 P XVector                  0.30.0   2020-10-27 [?] Bioconductor  
 P yaml                     2.2.1    2020-02-01 [?] CRAN (R 4.0.0)
 P zlibbioc                 1.36.0   2020-10-27 [?] Bioconductor  

[1] /stornext/Home/data/allstaff/h/ho.w/C094_Pellicci/renv/library/R-4.0/x86_64-pc-linux-gnu
[2] /tmp/RtmpRtKfI1/renv-system-library
[3] /stornext/System/data/apps/R/R-4.0.3/lib64/R/library

 P ── Loaded and on-disk path mismatch.</code></pre>
</div>
</details>
<div class="sourceCode" id="cb2"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography">
<div id="ref-aran2019reference" class="csl-entry" role="doc-biblioentry">
Aran, D., A. P. Looney, L. Liu, E. Wu, V. Fong, A. Hsu, S. Chak, et al. 2019. <span>“Reference-Based Analysis of Lung Single-Cell Sequencing Reveals a Transitional Profibrotic Macrophage.”</span> <em>Nat. Immunol.</em> 20 (2): 163–72.
</div>
<div id="ref-berger1996bioequivalence" class="csl-entry" role="doc-biblioentry">
Berger, R. L., and J. C. Hsu. 1996. <span>“Bioequivalence Trials, Intersection-Union Tests and Equivalence Confidence Sets.”</span> <em>Statist. Sci.</em> 11 (4): 283–319. <a href="https://doi.org/10.1214/ss/1032280304">https://doi.org/10.1214/ss/1032280304</a>.
</div>
<div id="ref-hpcaRef" class="csl-entry" role="doc-biblioentry">
Mabbott, Neil A., J. K. Baillie, Helen Brown, Tom C. Freeman, and David A. Hume. 2013. <span>“An Expression Atlas of Human Primary Cells: Inference of Gene Function from Coexpression Networks.”</span> <em>BMC Genomics</em> 14. <a href="https://doi.org/10.1186/1471-2164-14-632">https://doi.org/10.1186/1471-2164-14-632</a>.
</div>
<div id="ref-blueprintRef" class="csl-entry" role="doc-biblioentry">
Martens, Joost H A, and Hendrik G. Stunnenberg. 2013. <span>“BLUEPRINT: Mapping Human Blood Cell Epigenomes.”</span> <em>Haematologica</em> 98: 1487–89. <a href="https://doi.org/10.3324/haematol.2013.094243">https://doi.org/10.3324/haematol.2013.094243</a>.
</div>
<div id="ref-mccarthy2009treat" class="csl-entry" role="doc-biblioentry">
McCarthy, D. J., and G. K. Smyth. 2009. <span>“Testing Significance Relative to a Fold-Change Threshold Is a TREAT.”</span> <em>Bioinformatics</em> 25 (6): 765–71.
</div>
<div id="ref-monaco_immuneRef" class="csl-entry" role="doc-biblioentry">
Monaco, Gianni, Bernett Lee, Weili Xu, Seri Mustafah, You Yi Hwang, Christophe Carré, Nicolas Burdin, et al. 2019. <span>“RNA-Seq Signatures Normalized by mRNA Abundance Allow Absolute Deconvolution of Human Immune Cell Types.”</span> <em>Cell Reports</em> 26 (6): 1627–1640.e7. <a href="https://doi.org/10.1016/j.celrep.2019.01.041">https://doi.org/10.1016/j.celrep.2019.01.041</a>.
</div>
<div id="ref-nhRef" class="csl-entry" role="doc-biblioentry">
Novershtern, Noa, Aravind Subramanian, Lee N. Lawton, Raymond H. Mak, W. Nicholas Haining, Marie E. McConkey, Naomi Habib, et al. 2011. <span>“Densely Interconnected Transcriptional Circuits Control Cell States in Human Hematopoiesis.”</span> <em>Cell</em> 144 (2): 296–309. <a href="https://doi.org/10.1016/j.cell.2011.01.004">https://doi.org/10.1016/j.cell.2011.01.004</a>.
</div>
<div id="ref-diceRef" class="csl-entry" role="doc-biblioentry">
Schmiedel, Benjamin J., Divya Singh, Ariel Madrigal, Alan G. Valdovino-Gonzalez, Brandie M. White, Jose Zapardiel-Gonzalo, Brendan Ha, et al. 2018. <span>“Impact of Genetic Polymorphisms on Human Immune Cell Gene Expression.”</span> <em>Cell</em> 175 (6): 1701–1715.e16. <a href="https://doi.org/10.1016/j.cell.2018.10.022">https://doi.org/10.1016/j.cell.2018.10.022</a>.
</div>
<div id="ref-simes1986improved" class="csl-entry" role="doc-biblioentry">
Simes, R. J. 1986. <span>“An Improved Bonferroni Procedure for Multiple Tests of Significance.”</span> <em>Biometrika</em> 73 (3): 751–54.
</div>
<div id="ref-soneson2018bias" class="csl-entry" role="doc-biblioentry">
Soneson, C., and M. D. Robinson. 2018. <span>“Bias, Robustness and Scalability in Single-Cell Differential Expression Analysis.”</span> <em>Nat. Methods</em> 15 (4): 255–61. <a href="https://www.ncbi.nlm.nih.gov/pubmed/29481549">https://www.ncbi.nlm.nih.gov/pubmed/29481549</a>.
</div>
<div id="ref-encodeRef" class="csl-entry" role="doc-biblioentry">
The ENCODE Project Consortium. 2012. <span>“An Integrated Encyclopedia of DNA Elements in the Human Genome.”</span> <em>Nature</em>. <a href="https://doi.org/10.1038/nature11247">https://doi.org/10.1038/nature11247</a>.
</div>
<div id="ref-xu2015identification" class="csl-entry" role="doc-biblioentry">
Xu, C., and Z. Su. 2015. <span>“Identification of Cell Types from Single-Cell Transcriptomes Using a Novel Clustering Method.”</span> <em>Bioinformatics</em> 31 (12): 1974–80.
</div>
</div>
<section class="footnotes" role="doc-endnotes">
<hr />
<ol>
<li id="fn1" role="doc-endnote"><p>This can be individual cells or bulk samples.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2" role="doc-endnote"><p>Table adapted from <a href="https://bioconductor.org/packages/release/bioc/vignettes/SingleR/inst/doc/SingleR.html#5_available_references">https://bioconductor.org/packages/release/bioc/vignettes/SingleR/inst/doc/SingleR.html#5_available_references</a><a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3" role="doc-endnote"><p>Differential expression analyses is performed on the original log-expression values. We do not use the MNN-corrected values here except to obtain the clusters to be characterized.<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4" role="doc-endnote"><p>Equivalent to blocking on the plate of origin in this data set.<a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn5" role="doc-endnote"><p>Intra-<code>donor</code> comparisons are robust to differences between libraries but assume that each pair of clusters is present in at least one library. When this is not the case, we cannot commute a logFC for that pair of clusters.<a href="#fnref5" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn6" role="doc-endnote"><p>It must be stressed that the (adjusted) p-values computed here cannot be properly interpreted as measures of significance because the clusters have been empirically identified from the data. Nonetheless, they can still be used as way to rank ‘interesting’ genes.<a href="#fnref6" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn7" role="doc-endnote"><p>Specifically, a simple threshold does not consider the variance and can enrich for genes that have both large log-fold changes and large variances <span class="citation" data-cites="mccarthy2009treat">(<a href="#ref-mccarthy2009treat" role="doc-biblioref">McCarthy and Smyth 2009</a>)</span><a href="#fnref7" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn8" role="doc-endnote"><p>Ties are assumed to be randomly broken<a href="#fnref8" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
<h3 id="references">References</h3>
<div id="references-listing"></div>
</div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
