<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #20794d; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007ba5; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #007ba5; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #007ba5; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #20794d; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

  <!--radix_placeholder_meta_tags-->
  <title>C094: Preprocessing the Pellicci gamma-delta T-cell mini-bulk data set</title>



  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2021-08-23"/>
  <meta property="article:created" itemprop="dateCreated" content="2021-08-23"/>
  <meta name="article:author" content="Peter Hickey"/>

  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="C094: Preprocessing the Pellicci gamma-delta T-cell mini-bulk data set"/>
  <meta property="og:type" content="article"/>
  <meta property="og:locale" content="en_US"/>
  <meta property="og:site_name" content="C094"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="C094: Preprocessing the Pellicci gamma-delta T-cell mini-bulk data set"/>

  <!--/radix_placeholder_meta_tags-->
  
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","author","date","editor_options","bibliography"]}},"value":[{"type":"character","attributes":{},"value":["Preprocessing the Pellicci gamma-delta T-cell mini-bulk data set"]},{"type":"NULL"},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","url","affiliation","affiliation_url"]}},"value":[{"type":"character","attributes":{},"value":["Peter Hickey"]},{"type":"character","attributes":{},"value":["https://peterhickey.org"]},{"type":"character","attributes":{},"value":["Single Cell Open Research Endeavour (SCORE), WEHI"]},{"type":"character","attributes":{},"value":["https://www.wehi.edu.au/people/rory-bowden/4536/wehi-advanced-genomics-facility"]}]}]},{"type":"character","attributes":{},"value":["2021-08-23"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["chunk_output_type"]}},"value":[{"type":"character","attributes":{},"value":["console"]}]},{"type":"character","attributes":{},"value":["ref.bib"]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  <!--radix_placeholder_navigation_in_header-->
  <meta name="distill:offset" content=""/>

  <script type="application/javascript">

    window.headroom_prevent_pin = false;

    window.document.addEventListener("DOMContentLoaded", function (event) {

      // initialize headroom for banner
      var header = $('header').get(0);
      var headerHeight = header.offsetHeight;
      var headroom = new Headroom(header, {
        tolerance: 5,
        onPin : function() {
          if (window.headroom_prevent_pin) {
            window.headroom_prevent_pin = false;
            headroom.unpin();
          }
        }
      });
      headroom.init();
      if(window.location.hash)
        headroom.unpin();
      $(header).addClass('headroom--transition');

      // offset scroll location for banner on hash change
      // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
      window.addEventListener("hashchange", function(event) {
        window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
      });

      // responsive menu
      $('.distill-site-header').each(function(i, val) {
        var topnav = $(this);
        var toggle = topnav.find('.nav-toggle');
        toggle.on('click', function() {
          topnav.toggleClass('responsive');
        });
      });

      // nav dropdowns
      $('.nav-dropbtn').click(function(e) {
        $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
        $(this).parent().siblings('.nav-dropdown')
           .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $("body").click(function(e){
        $('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $(".nav-dropdown").click(function(e){
        e.stopPropagation();
      });
    });
  </script>

  <style type="text/css">

  /* Theme (user-documented overrideables for nav appearance) */

  .distill-site-nav {
    color: rgba(255, 255, 255, 0.8);
    background-color: #0F2E3D;
    font-size: 15px;
    font-weight: 300;
  }

  .distill-site-nav a {
    color: inherit;
    text-decoration: none;
  }

  .distill-site-nav a:hover {
    color: white;
  }

  @media print {
    .distill-site-nav {
      display: none;
    }
  }

  .distill-site-header {

  }

  .distill-site-footer {

  }


  /* Site Header */

  .distill-site-header {
    width: 100%;
    box-sizing: border-box;
    z-index: 3;
  }

  .distill-site-header .nav-left {
    display: inline-block;
    margin-left: 8px;
  }

  @media screen and (max-width: 768px) {
    .distill-site-header .nav-left {
      margin-left: 0;
    }
  }


  .distill-site-header .nav-right {
    float: right;
    margin-right: 8px;
  }

  .distill-site-header a,
  .distill-site-header .title {
    display: inline-block;
    text-align: center;
    padding: 14px 10px 14px 10px;
  }

  .distill-site-header .title {
    font-size: 18px;
    min-width: 150px;
  }

  .distill-site-header .logo {
    padding: 0;
  }

  .distill-site-header .logo img {
    display: none;
    max-height: 20px;
    width: auto;
    margin-bottom: -4px;
  }

  .distill-site-header .nav-image img {
    max-height: 18px;
    width: auto;
    display: inline-block;
    margin-bottom: -3px;
  }



  @media screen and (min-width: 1000px) {
    .distill-site-header .logo img {
      display: inline-block;
    }
    .distill-site-header .nav-left {
      margin-left: 20px;
    }
    .distill-site-header .nav-right {
      margin-right: 20px;
    }
    .distill-site-header .title {
      padding-left: 12px;
    }
  }


  .distill-site-header .nav-toggle {
    display: none;
  }

  .nav-dropdown {
    display: inline-block;
    position: relative;
  }

  .nav-dropdown .nav-dropbtn {
    border: none;
    outline: none;
    color: rgba(255, 255, 255, 0.8);
    padding: 16px 10px;
    background-color: transparent;
    font-family: inherit;
    font-size: inherit;
    font-weight: inherit;
    margin: 0;
    margin-top: 1px;
    z-index: 2;
  }

  .nav-dropdown-content {
    display: none;
    position: absolute;
    background-color: white;
    min-width: 200px;
    border: 1px solid rgba(0,0,0,0.15);
    border-radius: 4px;
    box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
    z-index: 1;
    margin-top: 2px;
    white-space: nowrap;
    padding-top: 4px;
    padding-bottom: 4px;
  }

  .nav-dropdown-content hr {
    margin-top: 4px;
    margin-bottom: 4px;
    border: none;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .nav-dropdown-active {
    display: block;
  }

  .nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
    color: black;
    padding: 6px 24px;
    text-decoration: none;
    display: block;
    text-align: left;
  }

  .nav-dropdown-content .nav-dropdown-header {
    display: block;
    padding: 5px 24px;
    padding-bottom: 0;
    text-transform: uppercase;
    font-size: 14px;
    color: #999999;
    white-space: nowrap;
  }

  .nav-dropdown:hover .nav-dropbtn {
    color: white;
  }

  .nav-dropdown-content a:hover {
    background-color: #ddd;
    color: black;
  }

  .nav-right .nav-dropdown-content {
    margin-left: -45%;
    right: 0;
  }

  @media screen and (max-width: 768px) {
    .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
    .distill-site-header a.nav-toggle {
      float: right;
      display: block;
    }
    .distill-site-header .title {
      margin-left: 0;
    }
    .distill-site-header .nav-right {
      margin-right: 0;
    }
    .distill-site-header {
      overflow: hidden;
    }
    .nav-right .nav-dropdown-content {
      margin-left: 0;
    }
  }


  @media screen and (max-width: 768px) {
    .distill-site-header.responsive {position: relative; min-height: 500px; }
    .distill-site-header.responsive a.nav-toggle {
      position: absolute;
      right: 0;
      top: 0;
    }
    .distill-site-header.responsive a,
    .distill-site-header.responsive .nav-dropdown {
      display: block;
      text-align: left;
    }
    .distill-site-header.responsive .nav-left,
    .distill-site-header.responsive .nav-right {
      width: 100%;
    }
    .distill-site-header.responsive .nav-dropdown {float: none;}
    .distill-site-header.responsive .nav-dropdown-content {position: relative;}
    .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
      display: block;
      width: 100%;
      text-align: left;
    }
  }

  /* Site Footer */

  .distill-site-footer {
    width: 100%;
    overflow: hidden;
    box-sizing: border-box;
    z-index: 3;
    margin-top: 30px;
    padding-top: 30px;
    padding-bottom: 30px;
    text-align: center;
  }

  /* Headroom */

  d-title {
    padding-top: 6rem;
  }

  @media print {
    d-title {
      padding-top: 4rem;
    }
  }

  .headroom {
    z-index: 1000;
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
  }

  .headroom--transition {
    transition: all .4s ease-in-out;
  }

  .headroom--unpinned {
    top: -100px;
  }

  .headroom--pinned {
    top: 0;
  }

  /* adjust viewport for navbar height */
  /* helps vertically center bootstrap (non-distill) content */
  .min-vh-100 {
    min-height: calc(100vh - 100px) !important;
  }

  </style>

  <script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
  <link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet"/>
  <link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet"/>
  <script src="site_libs/headroom-0.9.4/headroom.min.js"></script>
  <script src="site_libs/autocomplete-0.37.1/autocomplete.min.js"></script>
  <script src="site_libs/fuse-6.4.1/fuse.min.js"></script>

  <script type="application/javascript">

  function getMeta(metaName) {
    var metas = document.getElementsByTagName('meta');
    for (let i = 0; i < metas.length; i++) {
      if (metas[i].getAttribute('name') === metaName) {
        return metas[i].getAttribute('content');
      }
    }
    return '';
  }

  function offsetURL(url) {
    var offset = getMeta('distill:offset');
    return offset ? offset + '/' + url : url;
  }

  function createFuseIndex() {

    // create fuse index
    var options = {
      keys: [
        { name: 'title', weight: 20 },
        { name: 'categories', weight: 15 },
        { name: 'description', weight: 10 },
        { name: 'contents', weight: 5 },
      ],
      ignoreLocation: true,
      threshold: 0
    };
    var fuse = new window.Fuse([], options);

    // fetch the main search.json
    return fetch(offsetURL('search.json'))
      .then(function(response) {
        if (response.status == 200) {
          return response.json().then(function(json) {
            // index main articles
            json.articles.forEach(function(article) {
              fuse.add(article);
            });
            // download collections and index their articles
            return Promise.all(json.collections.map(function(collection) {
              return fetch(offsetURL(collection)).then(function(response) {
                if (response.status === 200) {
                  return response.json().then(function(articles) {
                    articles.forEach(function(article) {
                      fuse.add(article);
                    });
                  })
                } else {
                  return Promise.reject(
                    new Error('Unexpected status from search index request: ' +
                              response.status)
                  );
                }
              });
            })).then(function() {
              return fuse;
            });
          });

        } else {
          return Promise.reject(
            new Error('Unexpected status from search index request: ' +
                        response.status)
          );
        }
      });
  }

  window.document.addEventListener("DOMContentLoaded", function (event) {

    // get search element (bail if we don't have one)
    var searchEl = window.document.getElementById('distill-search');
    if (!searchEl)
      return;

    createFuseIndex()
      .then(function(fuse) {

        // make search box visible
        searchEl.classList.remove('hidden');

        // initialize autocomplete
        var options = {
          autoselect: true,
          hint: false,
          minLength: 2,
        };
        window.autocomplete(searchEl, options, [{
          source: function(query, callback) {
            const searchOptions = {
              isCaseSensitive: false,
              shouldSort: true,
              minMatchCharLength: 2,
              limit: 10,
            };
            var results = fuse.search(query, searchOptions);
            callback(results
              .map(function(result) { return result.item; })
            );
          },
          templates: {
            suggestion: function(suggestion) {
              var img = suggestion.preview && Object.keys(suggestion.preview).length > 0
                ? `<img src="${offsetURL(suggestion.preview)}"</img>`
                : '';
              var html = `
                <div class="search-item">
                  <h3>${suggestion.title}</h3>
                  <div class="search-item-description">
                    ${suggestion.description || ''}
                  </div>
                  <div class="search-item-preview">
                    ${img}
                  </div>
                </div>
              `;
              return html;
            }
          }
        }]).on('autocomplete:selected', function(event, suggestion) {
          window.location.href = offsetURL(suggestion.path);
        });
        // remove inline display style on autocompleter (we want to
        // manage responsive display via css)
        $('.algolia-autocomplete').css("display", "");
      })
      .catch(function(error) {
        console.log(error);
      });

  });

  </script>

  <style type="text/css">

  .nav-search {
    font-size: x-small;
  }

  /* Algolioa Autocomplete */

  .algolia-autocomplete {
    display: inline-block;
    margin-left: 10px;
    vertical-align: sub;
    background-color: white;
    color: black;
    padding: 6px;
    padding-top: 8px;
    padding-bottom: 0;
    border-radius: 6px;
    border: 1px #0F2E3D solid;
    width: 180px;
  }


  @media screen and (max-width: 768px) {
    .distill-site-nav .algolia-autocomplete {
      display: none;
      visibility: hidden;
    }
    .distill-site-nav.responsive .algolia-autocomplete {
      display: inline-block;
      visibility: visible;
    }
    .distill-site-nav.responsive .algolia-autocomplete .aa-dropdown-menu {
      margin-left: 0;
      width: 400px;
      max-height: 400px;
    }
  }

  .algolia-autocomplete .aa-input, .algolia-autocomplete .aa-hint {
    width: 90%;
    outline: none;
    border: none;
  }

  .algolia-autocomplete .aa-hint {
    color: #999;
  }
  .algolia-autocomplete .aa-dropdown-menu {
    width: 550px;
    max-height: 70vh;
    overflow-x: visible;
    overflow-y: scroll;
    padding: 5px;
    margin-top: 3px;
    margin-left: -150px;
    background-color: #fff;
    border-radius: 5px;
    border: 1px solid #999;
    border-top: none;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion {
    cursor: pointer;
    padding: 5px 4px;
    border-bottom: 1px solid #eee;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion:last-of-type {
    border-bottom: none;
    margin-bottom: 2px;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item {
    overflow: hidden;
    font-size: 0.8em;
    line-height: 1.4em;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item h3 {
    font-size: 1rem;
    margin-block-start: 0;
    margin-block-end: 5px;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-description {
    display: inline-block;
    overflow: hidden;
    height: 2.8em;
    width: 80%;
    margin-right: 4%;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview {
    display: inline-block;
    width: 15%;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview img {
    height: 3em;
    width: auto;
    display: none;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview img[src] {
    display: initial;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion.aa-cursor {
    background-color: #eee;
  }
  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion em {
    font-weight: bold;
    font-style: normal;
  }

  </style>


  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: auto;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  d-article details {
    grid-column: text;
    margin-bottom: 0.8em;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure img {
    width: 100%;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }

  /* Citation hover box */

  .tippy-box[data-theme~=light-border] {
    background-color: rgba(250, 250, 250, 0.95);
  }

  .tippy-content > p {
    margin-bottom: 0;
    padding: 2px;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('details, div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme
    $('code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // hoverable references
      $('span.citation[data-cites]').each(function() {
        var refHtml = $('#ref-' + $(this).attr('data-cites')).html();
        window.tippy(this, {
          allowHTML: true,
          content: refHtml,
          maxWidth: 500,
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        });
      });

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace("index.html", "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <!--/radix_placeholder_distill-->
  <script src="site_libs/header-attrs-2.7/header-attrs.js"></script>
  <script src="site_libs/htmlwidgets-1.5.3/htmlwidgets.js"></script>
  <script src="site_libs/jquery-3.5.1/jquery.min.js"></script>
  <link href="site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
  <script src="site_libs/datatables-binding-0.17/datatables.js"></script>
  <link href="site_libs/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet" />
  <link href="site_libs/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet" />
  <script src="site_libs/dt-core-1.10.20/js/jquery.dataTables.min.js"></script>
  <link href="site_libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet" />
  <script src="site_libs/crosstalk-1.1.1/js/crosstalk.min.js"></script>
  <script src="site_libs/popper-2.6.0/popper.min.js"></script>
  <link href="site_libs/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="site_libs/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="site_libs/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="site_libs/anchor-4.2.2/anchor.min.js"></script>
  <script src="site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <!--/radix_placeholder_site_in_header-->


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"Preprocessing the Pellicci gamma-delta T-cell mini-bulk data set","authors":[{"author":"Peter Hickey","authorURL":"https://peterhickey.org","affiliation":"Single Cell Open Research Endeavour (SCORE), WEHI","affiliationURL":"https://www.wehi.edu.au/people/rory-bowden/4536/wehi-advanced-genomics-facility","orcidID":""}],"publishedDate":"2021-08-23T00:00:00.000+10:00","citationText":"Hickey, 2021"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a href="index.html" class="title">C094</a>
<a href="index.html">Introduction</a>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Preprocessing
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="C094_Pellicci.single-cell.preprocess.html">Single cell</a>
<a href="C094_Pellicci.mini-bulk.preprocess.html">Mini bulk</a>
</div>
</div>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Selecting biologically relevant cells
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="C094_Pellicci.single-cell.cell_selection.html">Single cell</a>
</div>
</div>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Merging cells
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="C094_Pellicci.single-cell.merge.whole_cell.html">Single cell (whole cell)</a>
<a href="C094_Pellicci.single-cell.merge.S1_S2_only.html">Single cell (S1 and S2 only)</a>
<a href="C094_Pellicci.single-cell.merge.S3_only.html">Single cell (S3 only)</a>
<a href="C094_Pellicci.single-cell.merge.thymus_only.html">Single cell (thymus only)</a>
</div>
</div>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Annotating cells
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="C094_Pellicci.single-cell.annotate.whole_cell.html">Single cell (whole cell)</a>
<a href="C094_Pellicci.single-cell.annotate.S1_S2_only.html">Single cell (S1 and S2 only)</a>
<a href="C094_Pellicci.single-cell.annotate.S3_only.html">Single cell (S3 only)</a>
<a href="C094_Pellicci.single-cell.annotate.thymus_only.html">Single cell (thymus only)</a>
</div>
</div>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Mini-bulk differential analyses
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="C094_Pellicci.mini-bulk.multi-sample_comparisons.all_samples.html">All samples</a>
<a href="C094_Pellicci.mini-bulk.multi-sample_comparisons.excluding_blood_1-3.html">Excluding Blood.1 - Blood.3</a>
<a href="C094_Pellicci.mini-bulk.multi-sample_comparisons.excluding_donors_1-3.html">Excluding donors 1-3</a>
</div>
</div>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Supplementary files
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="../output/scPipe/">scPipe reports</a>
<a href="../output/DEGs/">Differentially expressed genes</a>
<a href="../output/glimma-plots/">Glimma plots</a>
</div>
</div>
</div>
<div class="nav-right">
<a href="https://github.com/WEHISCORE/C094_Pellicci">
<i class="fa fa-github" aria-hidden="true"></i>
</a>
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>Preprocessing the Pellicci gamma-delta T-cell mini-bulk data set</h1>
<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->

</div>

<div class="d-byline">
  Peter Hickey <a href="https://peterhickey.org" class="uri">https://peterhickey.org</a> (Single Cell Open Research Endeavour (SCORE), WEHI)<a href="https://www.wehi.edu.au/people/rory-bowden/4536/wehi-advanced-genomics-facility" class="uri">https://www.wehi.edu.au/people/rory-bowden/4536/wehi-advanced-genomics-facility</a>
  
<br/>2021-08-23
</div>

<div class="d-article">
<div class="d-contents d-contents-float">
<nav class="l-text toc figcaption" id="TOC">
<h3>Contents</h3>
<ul>
<li><a href="#setting-up-the-data">Setting up the data</a>
<ul>
<li><a href="#incorporating-cell-based-annotations">Incorporating cell-based annotations</a></li>
<li><a href="#incorporating-gene-based-annotation">Incorporating gene-based annotation</a></li>
</ul></li>
<li><a href="#read-mapping-metrics">Read mapping metrics</a>
<ul>
<li><a href="#motivation">Motivation</a></li>
<li><a href="#analysis">Analysis</a></li>
<li><a href="#summary">Summary</a></li>
</ul></li>
<li><a href="#have-we-sequenced-enough">Have we sequenced enough?</a>
<ul>
<li><a href="#motivation-1">Motivation</a></li>
<li><a href="#analysis-1">Analysis</a></li>
<li><a href="#summary-1">Summary</a></li>
</ul></li>
<li><a href="#umi-counts-vs.-read-counts">UMI counts vs.read counts</a>
<ul>
<li><a href="#motivation-2">Motivation</a></li>
<li><a href="#analysis-2">Analysis</a></li>
<li><a href="#summary-2">Summary</a></li>
</ul></li>
<li><a href="#quality-control-of-cells">Quality control of cells</a>
<ul>
<li><a href="#defining-the-quality-control-metrics">Defining the quality control metrics</a></li>
<li><a href="#visualizing-the-qc-metrics">Visualizing the QC metrics</a></li>
<li><a href="#identifying-outliers-by-each-metric">Identifying outliers by each metric</a></li>
<li><a href="#summary-3">Summary</a></li>
</ul></li>
<li><a href="#examining-gene-level-metrics">Examining gene-level metrics</a>
<ul>
<li><a href="#inspecting-the-most-highly-expressed-genes">Inspecting the most highly expressed genes</a></li>
<li><a href="#sex-check">Sex check</a>
<ul>
<li><a href="#using-xist-expression">Using <em>XIST</em> expression</a></li>
<li><a href="#using-y-chromosome-gene-expression">Using Y chromosome gene expression</a></li>
</ul></li>
</ul></li>
<li><a href="#concluding-remarks">Concluding remarks</a></li>
<li><a href="#additional-information">Additional information</a></li>
<li><a href="#session-info">Session info</a></li>
</ul>
</nav>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>SingleCellExperiment</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://here.r-lib.org/'>here</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/Bioconductor/BiocStyle'>BiocStyle</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://ggplot2.tidyverse.org'>ggplot2</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://wilkelab.org/cowplot/'>cowplot</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://patchwork.data-imaginist.com'>patchwork</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://scales.r-lib.org'>scales</a></span><span class='op'>)</span>

<span class='co'># NOTE: Using &gt;= 4 cores seizes up my laptop. Can use more on unix box.</span>
<span class='fu'><a href='https://rdrr.io/r/base/options.html'>options</a></span><span class='op'>(</span><span class='st'>"mc.cores"</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/Sys.info.html'>Sys.info</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>[[</span><span class='st'>"nodename"</span><span class='op'>]</span><span class='op'>]</span> <span class='op'>==</span> <span class='st'>"PC1331"</span>, <span class='fl'>2L</span>, <span class='fl'>3L</span><span class='op'>)</span><span class='op'>)</span>

<span class='fu'>knitr</span><span class='fu'>::</span><span class='va'><a href='https://rdrr.io/pkg/knitr/man/opts_chunk.html'>opts_chunk</a></span><span class='op'>$</span><span class='fu'>set</span><span class='op'>(</span>
  fig.path <span class='op'>=</span> <span class='st'>"C094_Pellicci.mini-bulk.preprocess_files/"</span><span class='op'>)</span>

<span class='kw'><a href='https://rdrr.io/r/base/source.html'>source</a></span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"code"</span>, <span class='st'>"helper_functions.R"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<h1 id="setting-up-the-data">Setting up the data</h1>
<p>The count data were processed using <em><a href="https://bioconductor.org/packages/3.12/scPipe">scPipe</a></em> and available in a <em>SingleCellExperiment</em> object, along with their metadata.</p>
<aside>
See <a href="../data/SCEs/C094_Pellicci.scPipe.SCE.rds"><code>data/SCEs/C094_Pellicci.scPipe.SCE.rds</code></a>.
</aside>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>sce_deduped</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"data"</span>, <span class='st'>"SCEs"</span>, <span class='st'>"C094_Pellicci.scPipe.SCE.rds"</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>sce_not_deduped</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"data"</span>, <span class='st'>"SCEs"</span>, <span class='st'>"C094_Pellicci.not_UMI_deduped.scPipe.SCE.rds"</span><span class='op'>)</span><span class='op'>)</span>

<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>sce_not_deduped</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/grep.html'>sub</a></span><span class='op'>(</span>
  <span class='st'>"\\.not_UMI_deduped"</span>,
  <span class='st'>""</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>sce_not_deduped</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/stopifnot.html'>stopifnot</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/base/identical.html'>identical</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce_deduped</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce_not_deduped</span><span class='op'>)</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/identical.html'>identical</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>sce_deduped</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>sce_not_deduped</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># Combine UMI and read counts a single SCE.</span>
<span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='va'>sce_deduped</span>
<span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"UMI_counts"</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce_deduped</span>, <span class='st'>"counts"</span><span class='op'>)</span>
<span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"read_counts"</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce_not_deduped</span>, <span class='st'>"counts"</span><span class='op'>)</span>
<span class='co'># NOTE: Have to reorder ERCCs</span>
<span class='fu'>assay</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html'>altExp</a></span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"ERCC"</span><span class='op'>)</span>, <span class='st'>"UMI_counts"</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'>assay</span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html'>altExp</a></span><span class='op'>(</span><span class='va'>sce_deduped</span>, <span class='st'>"ERCC"</span><span class='op'>)</span>, <span class='st'>"counts"</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html'>altExp</a></span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"ERCC"</span><span class='op'>)</span><span class='op'>)</span>, <span class='op'>]</span>
<span class='fu'>assay</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html'>altExp</a></span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"ERCC"</span><span class='op'>)</span>, <span class='st'>"read_counts"</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'>assay</span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html'>altExp</a></span><span class='op'>(</span><span class='va'>sce_not_deduped</span>, <span class='st'>"ERCC"</span><span class='op'>)</span>, <span class='st'>"counts"</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html'>altExp</a></span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"ERCC"</span><span class='op'>)</span><span class='op'>)</span>, <span class='op'>]</span>

<span class='co'># Nullify some now-unrequired data.</span>
<span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"counts"</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='cn'>NULL</span>
<span class='fu'>assay</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html'>altExp</a></span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"ERCC"</span><span class='op'>)</span>, <span class='st'>"counts"</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='cn'>NULL</span>
<span class='va'>sce</span><span class='op'>$</span><span class='va'>UMI_deduped</span> <span class='op'>&lt;-</span> <span class='cn'>NULL</span>
</code></pre>
</div>
</details>
</div>
<p>This data was processed to clean up the sample metadata, tidy the FACS index sorting data, and attach the gene metadata.</p>
<aside>
See <a href="../analysis/C094_Pellicci.preprocess.R"><code>analysis/C094_Pellicci.preprocess.R</code></a>.
</aside>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/source.html'>source</a></span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"analysis"</span>, <span class='st'>"C094_Pellicci.preprocess.R"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p><strong>We then filter the data to only include the mini-bulk samples</strong>.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>sample_type</span> <span class='op'>!=</span> <span class='st'>"Single cell"</span><span class='op'>]</span>
<span class='fu'>colData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/droplevels.html'>droplevels</a></span><span class='op'>(</span><span class='fu'>colData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># Only retain genes with non-zero counts in at least 1 cell.</span>
<span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/colSums.html'>rowSums</a></span><span class='op'>(</span><span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"read_counts"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>&gt;</span> <span class='fl'>0</span>, <span class='op'>]</span>
</code></pre>
</div>
</details>
</div>
<h2 id="incorporating-cell-based-annotations">Incorporating cell-based annotations</h2>
<p>Cell-based annotations are included in the <em>colData</em> of the <em>SingleCellExperiment</em>. Figure <a href="#fig:coldata">1</a> is a visual summary of the cell-based metadata.</p>
<aside>
The SKW-3 sample is a control cell line, originally thought to originate from the peripheral blood of a 61 year old male patient with T cell chronic lymphocytic leukemia. (although <a href="https://web.expasy.org/cellosaurus/CVCL_2197">this is known to be problematic</a>).
</aside>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='fu'>ggcells</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>plate_number</span>, fill <span class='op'>=</span> <span class='va'>donor</span><span class='op'>)</span>,
    position <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/position_stack.html'>position_fill</a></span><span class='op'>(</span>reverse <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Frequency"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>donor_colours</span><span class='op'>)</span>
<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='fu'>ggcells</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>plate_number</span>, fill <span class='op'>=</span> <span class='va'>sample</span><span class='op'>)</span>,
    position <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/position_stack.html'>position_fill</a></span><span class='op'>(</span>reverse <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Frequency"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>sample_colours</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>fill <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guide_legend.html'>guide_legend</a></span><span class='op'>(</span>ncol <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>p3</span> <span class='op'>&lt;-</span> <span class='fu'>ggcells</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>plate_number</span>, fill <span class='op'>=</span> <span class='va'>sample_gate</span><span class='op'>)</span>,
    position <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/position_stack.html'>position_fill</a></span><span class='op'>(</span>reverse <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Frequency"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>sample_gate_colours</span><span class='op'>)</span>
<span class='va'>p4</span> <span class='op'>&lt;-</span>  <span class='fu'>ggcells</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>plate_number</span>, fill <span class='op'>=</span> <span class='va'>sample_type</span><span class='op'>)</span>,
    position <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/position_stack.html'>position_fill</a></span><span class='op'>(</span>reverse <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Frequency"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>sample_type_colours</span><span class='op'>)</span>
<span class='va'>p5</span> <span class='op'>&lt;-</span>  <span class='fu'>ggcells</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>plate_number</span>, fill <span class='op'>=</span> <span class='va'>sequencing_run</span><span class='op'>)</span>,
    position <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/position_stack.html'>position_fill</a></span><span class='op'>(</span>reverse <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Frequency"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>sequencing_run_colours</span><span class='op'>)</span>
<span class='va'>p6</span> <span class='op'>&lt;-</span>  <span class='fu'>ggcells</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>plate_number</span>, fill <span class='op'>=</span> <span class='va'>stage</span><span class='op'>)</span>,
    position <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/position_stack.html'>position_fill</a></span><span class='op'>(</span>reverse <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Frequency"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>stage_colours</span><span class='op'>)</span>
<span class='va'>p7</span> <span class='op'>&lt;-</span>  <span class='fu'>ggcells</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>plate_number</span>, fill <span class='op'>=</span> <span class='va'>tissue</span><span class='op'>)</span>,
    position <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/position_stack.html'>position_fill</a></span><span class='op'>(</span>reverse <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Frequency"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>tissue_colours</span><span class='op'>)</span>
<span class='va'>p8</span> <span class='op'>&lt;-</span> <span class='fu'>ggcells</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>plate_number</span>, fill <span class='op'>=</span> <span class='va'>plate_number</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Number of mini-bulk samples"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>plate_number_colours</span><span class='op'>)</span>

<span class='va'>p1</span> <span class='op'>+</span> <span class='va'>p2</span> <span class='op'>+</span> <span class='va'>p3</span> <span class='op'>+</span> <span class='va'>p4</span> <span class='op'>+</span> <span class='va'>p5</span> <span class='op'>+</span> <span class='va'>p6</span> <span class='op'>+</span> <span class='va'>p7</span> <span class='op'>+</span> <span class='va'>p8</span> <span class='op'>+</span> <span class='fu'><a href='https://patchwork.data-imaginist.com/reference/plot_layout.html'>plot_layout</a></span><span class='op'>(</span>ncol <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:coldata"></span>
<img src="C094_Pellicci.mini-bulk.preprocess_files/coldata-1.png" alt="Breakdown of experiment by plate." width="624" />
<p class="caption">
Figure 1: Breakdown of experiment by plate.
</p>
</div>
</div>
<p>Figure <a href="#fig:blood-stage">2</a> highlights that the <code>Blood.1</code>, <code>Blood.2</code>, and <code>Blood.3</code> samples collected on plate <code>LCE515</code> were from FACS gate <code>P5</code> (corresponding to a mixture of stages 1-3) whereas all subsequent <code>Blood</code> samples were from FACS gate <code>P8</code> (corresponding to stage 3). However, Dan tells me that ~90% of <code>Blood.P5</code> cells are essentially <code>Blood.P8</code> (corresponding to stage 3).</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='fu'>ggcells</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>donor</span>, fill <span class='op'>=</span> <span class='va'>sample_gate</span><span class='op'>)</span>, 
    position <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/position_stack.html'>position_fill</a></span><span class='op'>(</span>reverse <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Frequency"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_grid.html'>facet_grid</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>tissue</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>sample_gate_colours</span><span class='op'>)</span>
<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='fu'>ggcells</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>donor</span>, fill <span class='op'>=</span> <span class='va'>stage</span><span class='op'>)</span>, 
    position <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/position_stack.html'>position_fill</a></span><span class='op'>(</span>reverse <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Frequency"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_grid.html'>facet_grid</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>tissue</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>stage_colours</span><span class='op'>)</span>
<span class='va'>p3</span> <span class='op'>&lt;-</span> <span class='fu'>ggcells</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>donor</span>, fill <span class='op'>=</span> <span class='va'>plate_number</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Number of mini-bulk samples"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_grid.html'>facet_grid</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>tissue</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>plate_number_colours</span><span class='op'>)</span>

<span class='va'>p1</span> <span class='op'>+</span> <span class='va'>p2</span> <span class='op'>+</span> <span class='va'>p3</span> <span class='op'>+</span> <span class='fu'><a href='https://patchwork.data-imaginist.com/reference/plot_layout.html'>plot_layout</a></span><span class='op'>(</span>ncol <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:blood-stage"></span>
<img src="C094_Pellicci.mini-bulk.preprocess_files/blood-stage-1.png" alt="Breakdown of experiment by donor." width="624" />
<p class="caption">
Figure 2: Breakdown of experiment by donor.
</p>
</div>
</div>
<p>Figure <a href="#fig:plate-map">3</a> shows the design of each plate.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># TODO: Fix legend</span>
<span class='va'>p</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>lapply</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span><span class='op'>)</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>p</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='va'>z</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span> <span class='op'>==</span> <span class='va'>p</span><span class='op'>]</span>
  <span class='fu'>plotPlatePosition</span><span class='op'>(</span>
    <span class='va'>z</span>,
    <span class='fu'><a href='https://rdrr.io/r/base/character.html'>as.character</a></span><span class='op'>(</span><span class='va'>z</span><span class='op'>$</span><span class='va'>well_position</span><span class='op'>)</span>,
    point_size <span class='op'>=</span> <span class='fl'>2</span>,
    point_alpha <span class='op'>=</span> <span class='fl'>1</span>,
    theme_size <span class='op'>=</span> <span class='fl'>8</span>,
    colour_by <span class='op'>=</span> <span class='st'>"donor"</span>,
    shape_by <span class='op'>=</span> <span class='st'>"tissue"</span><span class='op'>)</span> <span class='op'>+</span>
    <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>donor_colours</span>, name <span class='op'>=</span> <span class='st'>"donor"</span><span class='op'>)</span> <span class='op'>+</span>
    <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ggtitle</a></span><span class='op'>(</span><span class='va'>p</span><span class='op'>)</span> <span class='op'>+</span>
    <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>
      legend.text <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span>,
      legend.title <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
    <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>
      colour <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guide_legend.html'>guide_legend</a></span><span class='op'>(</span>override.aes <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>2</span>, alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>,
      shape <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guide_legend.html'>guide_legend</a></span><span class='op'>(</span>override.aes <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>2</span>, alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='op'>}</span><span class='op'>)</span>

<span class='fu'><a href='https://patchwork.data-imaginist.com/reference/wrap_plots.html'>wrap_plots</a></span><span class='op'>(</span><span class='va'>p</span>, ncol <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:plate-map"></span>
<img src="C094_Pellicci.mini-bulk.preprocess_files/plate-map-1.png" alt="Plate maps." width="624" />
<p class="caption">
Figure 3: Plate maps.
</p>
</div>
</div>
<h2 id="incorporating-gene-based-annotation">Incorporating gene-based annotation</h2>
<p>I used the <em><a href="https://bioconductor.org/packages/3.12/Homo.sapiens">Homo.sapiens</a></em> package and the <code>EnsDb.Hsapiens.v98</code> resource, which respectively cover the NCBI/RefSeq and Ensembl databases, to obtain gene-based annotations, such as the chromosome and gene symbol.</p>
<p>Having quantified gene expression against the GENCODE gene annotation, we have Ensembl-style identifiers for the genes. These identifiers are used as they are unambiguous and highly stable. However, they are difficult to interpret compared to the gene symbols which are more commonly used in the literature. Henceforth, we will use gene symbols (where available) to refer to genes in our analysis and otherwise use the Ensembl-style gene identifiers<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>.</p>
<p>We also construct some useful gene sets: mitochondrial genes, ribosomal protein genes, sex chromosome genes, and pseudogenes.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># Some useful gene sets</span>
<span class='va'>mito_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/any.html'>any</a></span><span class='op'>(</span><span class='fu'>rowData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>$</span><span class='va'>ENSEMBL.SEQNAME</span> <span class='op'>==</span> <span class='st'>"MT"</span><span class='op'>)</span><span class='op'>]</span>
<span class='va'>ribo_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/grep.html'>grep</a></span><span class='op'>(</span><span class='st'>"^RP(S|L)"</span>, <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>, value <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='co'># NOTE: A more curated approach for identifying ribosomal protein genes </span>
<span class='co'>#       (https://github.com/Bioconductor/OrchestratingSingleCellAnalysis-base/blob/ae201bf26e3e4fa82d9165d8abf4f4dc4b8e5a68/feature-selection.Rmd#L376-L380)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://igordot.github.io/msigdbr'>msigdbr</a></span><span class='op'>)</span>
<span class='va'>c2_sets</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://igordot.github.io/msigdbr/reference/msigdbr.html'>msigdbr</a></span><span class='op'>(</span>species <span class='op'>=</span> <span class='st'>"Homo sapiens"</span>, category <span class='op'>=</span> <span class='st'>"C2"</span><span class='op'>)</span>
<span class='va'>ribo_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/sets.html'>union</a></span><span class='op'>(</span>
  <span class='va'>ribo_set</span>,
  <span class='va'>c2_sets</span><span class='op'>[</span><span class='va'>c2_sets</span><span class='op'>$</span><span class='va'>gs_name</span> <span class='op'>==</span> <span class='st'>"KEGG_RIBOSOME"</span>, <span class='op'>]</span><span class='op'>$</span><span class='va'>gene_symbol</span><span class='op'>)</span>
<span class='va'>ribo_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/sets.html'>intersect</a></span><span class='op'>(</span><span class='va'>ribo_set</span>, <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>sex_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/any.html'>any</a></span><span class='op'>(</span><span class='fu'>rowData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>$</span><span class='va'>ENSEMBL.SEQNAME</span> <span class='op'>%in%</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"X"</span>, <span class='st'>"Y"</span><span class='op'>)</span><span class='op'>)</span><span class='op'>]</span>
<span class='va'>pseudogene_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span>
  <span class='fu'><a href='https://rdrr.io/r/base/any.html'>any</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/grep.html'>grepl</a></span><span class='op'>(</span><span class='st'>"pseudogene"</span>, <span class='fu'>rowData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>$</span><span class='va'>ENSEMBL.GENEBIOTYPE</span><span class='op'>)</span><span class='op'>)</span><span class='op'>]</span>
</code></pre>
</div>
</details>
</div>
<h1 id="read-mapping-metrics">Read mapping metrics</h1>
<h2 id="motivation">Motivation</h2>
<p>We aimed to sequence roughly 2 million reads/sample and for the majority of these reads would be exonic.</p>
<h2 id="analysis">Analysis</h2>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/LuyiTian/scPipe'>scPipe</a></span><span class='op'>)</span>
<span class='va'>sce</span><span class='op'>$</span><span class='va'>nreads</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/apply.html'>apply</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/scPipe/man/QC_metrics.html'>QC_metrics</a></span><span class='op'>(</span><span class='va'>sce_deduped</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>, <span class='op'>]</span>, <span class='fl'>1</span>, <span class='va'>sum</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>Figure <a href="#fig:nreads">4</a> shows that we generated a median of 2.8 million reads per sample with a median 43% of these exonic.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce</span>, <span class='st'>"nreads"</span>, 
  x <span class='op'>=</span> <span class='st'>"sample"</span>, 
  shape_by <span class='op'>=</span> <span class='st'>"plate_number"</span>, 
  colour_by <span class='op'>=</span> <span class='st'>"sample"</span>,
  point_alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_continuous.html'>scale_y_log10</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>sample_colours</span>, name <span class='op'>=</span> <span class='st'>"sample"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>""</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_abline.html'>geom_hline</a></span><span class='op'>(</span>yintercept <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/median.html'>median</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>nreads</span><span class='op'>)</span>, lty <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/annotation_logticks.html'>annotation_logticks</a></span><span class='op'>(</span>
    sides <span class='op'>=</span> <span class='st'>"l"</span>,
    short <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grid/unit.html'>unit</a></span><span class='op'>(</span><span class='fl'>0.06</span>, <span class='st'>"cm"</span><span class='op'>)</span>,
    mid <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grid/unit.html'>unit</a></span><span class='op'>(</span><span class='fl'>0.12</span>, <span class='st'>"cm"</span><span class='op'>)</span>,
    long <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grid/unit.html'>unit</a></span><span class='op'>(</span><span class='fl'>0.18</span>, <span class='st'>"cm"</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce</span>, 
  <span class='fu'><a href='https://rdrr.io/r/base/AsIs.html'>I</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>mapped_to_exon</span> <span class='op'>/</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>nreads</span><span class='op'>)</span>, 
  x <span class='op'>=</span> <span class='st'>"sample"</span>, 
  shape_by <span class='op'>=</span> <span class='st'>"plate_number"</span>, 
  colour_by <span class='op'>=</span> <span class='st'>"sample"</span>,
  point_alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/lims.html'>ylim</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>1</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>sample_colours</span>, name <span class='op'>=</span> <span class='st'>"sample"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>""</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Prop exonic"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_abline.html'>geom_hline</a></span><span class='op'>(</span>yintercept <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/median.html'>median</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>mapped_to_exon</span> <span class='op'>/</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>nreads</span><span class='op'>)</span>, lty <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>
<span class='va'>p3</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce</span>, 
  <span class='fu'><a href='https://rdrr.io/r/base/AsIs.html'>I</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>mapped_to_intron</span> <span class='op'>/</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>nreads</span><span class='op'>)</span>, 
  x <span class='op'>=</span> <span class='st'>"sample"</span>, 
  shape_by <span class='op'>=</span> <span class='st'>"plate_number"</span>, 
  colour_by <span class='op'>=</span> <span class='st'>"sample"</span>,
  point_alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/lims.html'>ylim</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>1</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>sample_colours</span>, name <span class='op'>=</span> <span class='st'>"sample"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>""</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Prop intronic"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_abline.html'>geom_hline</a></span><span class='op'>(</span>yintercept <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/median.html'>median</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>mapped_to_intron</span> <span class='op'>/</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>nreads</span><span class='op'>)</span>, lty <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>

<span class='op'>(</span><span class='va'>p1</span> <span class='op'>|</span> <span class='op'>(</span><span class='va'>p2</span> <span class='op'>/</span> <span class='va'>p3</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> <span class='fu'><a href='https://patchwork.data-imaginist.com/reference/plot_layout.html'>plot_layout</a></span><span class='op'>(</span>guides <span class='op'>=</span> <span class='st'>"collect"</span>, widths <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>2</span>, <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>  
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:nreads"></span>
<img src="C094_Pellicci.mini-bulk.preprocess_files/nreads-1.png" alt="Number of reads generated per sample and the proportion that are exonic and intronic. The dashed horizional line denotes the median across all samples." width="624" />
<p class="caption">
Figure 4: Number of reads generated per sample and the proportion that are exonic and intronic. The dashed horizional line denotes the median across all samples.
</p>
</div>
</div>
<p>We observe that the <code>Blood.1</code>, <code>Blood.2</code>, and <code>Blood.3</code> have notably fewer reads than other samples</p>
<h2 id="summary">Summary</h2>
<p>We roughly achieved our aim of sequencing 2 million reads/sample, although less than 50% of reads are exonic. There is notable plate-specific and sample-specific variation, however, in the number of reads generated.</p>
<p><strong>Henceforth, we only use the exonic reads.</strong> Although the intronic reads are potentially useful, to do so would require substantial changes to the data analysis workflow and additional analysis.</p>
<p>We now remove the control samples.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>tissue</span> <span class='op'>!=</span> <span class='st'>"SKW3"</span><span class='op'>]</span>
<span class='fu'>colData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/droplevels.html'>droplevels</a></span><span class='op'>(</span><span class='fu'>colData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<h1 id="have-we-sequenced-enough">Have we sequenced enough?</h1>
<h2 id="motivation-1">Motivation</h2>
<p>SCOREs mini-bulk protocol is still being finalised and some technical questions remain. In particular, we want to know if we have sequenced enough to answer the biological questions of interest.</p>
<p>One way of estimating if we have sequenced enough uses a metric called sequencing saturation. 10x Genomics have some useful information about sequencing saturation available <a href="https://kb.10xgenomics.com/hc/en-us/articles/115005062366-What-is-sequencing-saturation-">here</a>. In particular,</p>
<blockquote>
<p>Sequencing saturation is a measure of the fraction of library complexity that was sequenced in a given experiment. The inverse of the sequencing saturation can be interpreted as the number of additional reads it would take to detect a new transcript.  Sequencing saturation is dependent on the library complexity and sequencing depth. Different cell types will have different amounts of RNA and thus will differ in the total number of different transcripts in the final library (also known as library complexity).  As sequencing depth increases, more genes are detected, but this reaches saturation at different sequencing depths depending on cell type.  Sequencing depth also affects sequencing saturation; generally, the more sequencing reads, the more additional unique transcripts you can detect. However, this is limited by the library complexity.</p>
</blockquote>
<blockquote>
<p>This naturally leads to the question of <a href="https://kb.10xgenomics.com/hc/en-us/articles/115002474263-How-much-sequencing-saturation-should-I-aim-for-">how much sequencing saturation should I aim for?</a>, which very much depends on the goals of the experiment.</p>
</blockquote>
<p>For this experiment, where we wish to assess differential expression between experimental groups, we will likely want quite high sequencing saturation.</p>
<h2 id="analysis-1">Analysis</h2>
<p>To begin, we simply examine the number of reads sequenced/sample.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>reads_ls</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log10</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colSums.html'>colSums</a></span><span class='op'>(</span><span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"read_counts"</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>Figure <a href="#fig:read-library-size">5</a> plots the distribution of the number of reads mapped to exons in each sample<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/stats/density.html'>density</a></span><span class='op'>(</span><span class='va'>reads_ls</span><span class='op'>)</span>,
  xlab <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/expression.html'>expression</a></span><span class='op'>(</span><span class='va'>log</span><span class='op'>[</span><span class='fl'>10</span><span class='op'>]</span><span class='op'>(</span><span class='va'>n</span><span class='op'>[</span><span class='va'>reads</span><span class='op'>]</span><span class='op'>)</span><span class='op'>)</span>,
  ylab <span class='op'>=</span> <span class='st'>"density"</span>,
  xlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/range.html'>range</a></span><span class='op'>(</span><span class='va'>reads_ls</span><span class='op'>)</span>,
  main <span class='op'>=</span> <span class='st'>""</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:read-library-size"></span>
<img src="C094_Pellicci.mini-bulk.preprocess_files/read-library-size-1.png" alt="Distribution of the number of reads mapped to exons in each sample." width="624" />
<p class="caption">
Figure 5: Distribution of the number of reads mapped to exons in each sample.
</p>
</div>
</div>
<p>We estimate the sequencing saturation using <a href="https://kb.10xgenomics.com/hc/en-us/articles/115003646912-How-is-sequencing-saturation-calculated-">10x Genomics formula</a>,</p>
<p><span class="math display">\[
1 - (n_{deduped reads} / n_{reads})
\]</span> where:</p>
<ul>
<li><span class="math inline">\(n_{deduped reads}\)</span> = Number of unique (valid cell-barcode, valid UMI, gene)-combinations among confidently mapped reads</li>
<li><span class="math inline">\(n_{reads}\)</span> = Total number of confidently mapped, valid cell-barcode, valid UMI reads</li>
</ul>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>saturation</span> <span class='op'>&lt;-</span> <span class='fl'>1</span> <span class='op'>-</span> 
  <span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colSums.html'>colSums</a></span><span class='op'>(</span><span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"UMI_counts"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>/</span> <span class='fu'><a href='https://rdrr.io/r/base/colSums.html'>colSums</a></span><span class='op'>(</span><span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"read_counts"</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>Figure <a href="#fig:saturation">6</a> shows that almost all samples are sequenced to near-total saturation.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/graphics/par.html'>par</a></span><span class='op'>(</span>mfrow <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/stats/density.html'>density</a></span><span class='op'>(</span>
    <span class='va'>saturation</span>,
    from <span class='op'>=</span> <span class='fl'>0</span>,
    to <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span>,
  xlab <span class='op'>=</span> <span class='st'>"saturation"</span>,
  ylab <span class='op'>=</span> <span class='st'>"density"</span>,
  xlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>1</span><span class='op'>)</span>,
  main <span class='op'>=</span> <span class='st'>"Sequencing saturation"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:saturation"></span>
<img src="C094_Pellicci.mini-bulk.preprocess_files/saturation-1.png" alt="Distribution of the sequencing saturation." width="624" />
<p class="caption">
Figure 6: Distribution of the sequencing saturation.
</p>
</div>
</div>
<h2 id="summary-1">Summary</h2>
<p>These results show that the libraries are sequenced to near-saturation.</p>
<p>One way to think of sequencing saturation is <a href="https://kb.10xgenomics.com/hc/en-us/articles/115002474263-How-much-sequencing-saturation-should-I-aim-for-">provided by 10x Genomics</a>:</p>
<blockquote>
<p>The inverse of the sequencing saturation can be interpreted as roughly the number of new transcripts you expect to find with one new read. If sequencing saturation is at 50%, it means that every 2 new reads will result in 1 new UMI count (unique transcript) detected. In contrast, 90% sequencing saturation means that 10 new reads are necessary to obtain one new UMI count. If the sequencing saturation is high, additional sequencing would not recover much new information for the library.</p>
</blockquote>
<p>At this stage, it is not necessary to perform additional sequencing of the library. We considered additional sequencing of the <code>Blood.1</code>, <code>Blood.2</code>, and <code>Blood.3</code> samples. However, these samples are already pooled with the corresponding <code>Thymus.1</code>, <code>Thymus.2, and</code>Thymus.3<code>samples, which means we would also have to sequence these samples. Furthermore, since the</code>Blood.1<code>,</code>Blood.2<code>, and</code>Blood.3<code>used the</code>P5<code>gate rather than</code>P8<code>gate of the other</code>Blood` samples, and we were getting reasonable DE results with the existing data, it was decided not to do any additional sequencing.</p>
<h1 id="umi-counts-vs.-read-counts">UMI counts vs.read counts</h1>
<h2 id="motivation-2">Motivation</h2>
<p>A unique feature of SCOREs mini-bulk RNA-sequencing protocol is the incorporation of unique molecular identifiers (UMIs) into the library preparation; an ordinary bulk RNA-seq protocol does not incorporate UMIs.</p>
<p>UMIs are commonly used in single-cell RNA-sequencing protocols, such as the CEL-Seq2 protocol from which the mini-bulk protocol is adapted. UMIs are molecular tags that are used to detect and quantify unique mRNA transcripts. In this method, mRNA libraries are generated by fragmentation and reverse-transcribed to cDNA. Oligo(dT) primers with specific sequencing linkers are added to the cDNA. Another sequencing linker with a 6 bp random label and an index sequence is added to the 5 end of the template, which is amplified and sequenced.</p>
<aside>
This explanation is adapted from <a href="https://sapac.illumina.com/science/sequencing-method-explorer/kits-and-arrays/umi.html">https://sapac.illumina.com/science/sequencing-method-explorer/kits-and-arrays/umi.html</a>.
</aside>
<p>The UMI count is always less than or equal to the read count and a UMI count of zero also implies a read count of zero.</p>
<p>Pros of using UMIs:</p>
<ul>
<li>Can sequence unique mRNA transcripts</li>
<li>Can detect transcripts occurring at low frequencies</li>
<li>Transcripts can be quantified based on sequencing reads specific to each barcode</li>
</ul>
<p>Cons of using UMIs:</p>
<ul>
<li>SCOREs mini-bulk protocol uses a <span class="math inline">\(k=6\)</span> bp UMI<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a> and so (theoretically) the maximum UMI count for each gene is <span class="math inline">\(4^6 = 4096\)</span>.</li>
<li>There is therefore a risk that highly expressed genes may have their expression measurements artificially truncated, particularly if the sample is deeply sequenced.</li>
</ul>
<p>To investigate whether expression measurements are being truncated in this dataset, and decide whether to use the UMI counts or the read counts, we generated and now analyse two count matrices:</p>
<ol type="1">
<li>UMI counts (i.e.deduplicating UMI counts)</li>
<li>Read counts (i.e.not deduplicating UMI counts, equivalent to ignoring the UMIs)</li>
</ol>
<h2 id="analysis-2">Analysis</h2>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>n</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/apply.html'>apply</a></span><span class='op'>(</span><span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"read_counts"</span><span class='op'>)</span>, <span class='fl'>1</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span> <span class='fu'><a href='https://rdrr.io/r/base/sum.html'>sum</a></span><span class='op'>(</span><span class='va'>x</span> <span class='op'>&gt;</span> <span class='fl'>4</span> <span class='op'>^</span> <span class='fl'>6</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>g</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>n</span><span class='op'>[</span><span class='va'>n</span> <span class='op'>&gt;</span> <span class='fl'>0</span><span class='op'>]</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>Figure <a href="#fig:read-distribution">7</a> plots the distribution of the logarithm of the read counts. From this we see that the vast majority of read counts are small. In fact:</p>
<ul>
<li>67.9% of read counts are 0</li>
<li>95% of the read counts are no greater than 137</li>
<li>99% of the read counts are no greater than 535</li>
<li>99.9% of the read counts are no greater than 3,175</li>
</ul>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>x</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/vector.html'>as.vector</a></span><span class='op'>(</span><span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"read_counts"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> <span class='fl'>1L</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/par.html'>par</a></span><span class='op'>(</span>mfrow <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>2</span><span class='op'>)</span>, cex <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/stats/density.html'>density</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log2</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>, from <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>min</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log2</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span><span class='op'>)</span>, to <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log2</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>,
  xlab <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/expression.html'>expression</a></span><span class='op'>(</span><span class='va'>log</span><span class='op'>[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>(</span><span class='va'>reads</span> <span class='op'>+</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>,
  main <span class='op'>=</span> <span class='st'>""</span>,
  ylab <span class='op'>=</span> <span class='st'>"density"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/abline.html'>abline</a></span><span class='op'>(</span>v <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log2</a></span><span class='op'>(</span><span class='fl'>4</span> <span class='op'>^</span> <span class='fl'>6</span><span class='op'>)</span>, col <span class='op'>=</span> <span class='st'>"orange"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/stats/ecdf.html'>ecdf</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log2</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span><span class='op'>)</span>, 
  xlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/range.html'>range</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log2</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span><span class='op'>)</span>,
  main <span class='op'>=</span> <span class='st'>""</span>,
  xlab <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/expression.html'>expression</a></span><span class='op'>(</span><span class='va'>log</span><span class='op'>[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>(</span><span class='va'>reads</span> <span class='op'>+</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>,
  ylab <span class='op'>=</span> <span class='st'>"cumulative density"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/abline.html'>abline</a></span><span class='op'>(</span>v <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log2</a></span><span class='op'>(</span><span class='fl'>4</span> <span class='op'>^</span> <span class='fl'>6</span><span class='op'>)</span>, col <span class='op'>=</span> <span class='st'>"orange"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:read-distribution"></span>
<img src="C094_Pellicci.mini-bulk.preprocess_files/read-distribution-1.png" alt="Distribution of the logarithm of the read counts as density plot (left) and empirical cumulative distribution function (right). A count of 1 is added to each UMI count to avoid taking the logarithm of zero. The orange vertical line denotes $4^6 = 4096$." width="624" />
<p class="caption">
Figure 7: Distribution of the logarithm of the read counts as density plot (left) and empirical cumulative distribution function (right). A count of 1 is added to each UMI count to avoid taking the logarithm of zero. The orange vertical line denotes <span class="math inline">\(4^6 = 4096\)</span>.
</p>
</div>
</div>
<p>Only 0.072% of read counts exceed the <span class="math inline">\(4^6 = 4096\)</span> threshold. However there are 138 genes for which at least one sample exceeds the <span class="math inline">\(4^6=4096\)</span> threshold; these genes are shown in the table below. If we were to use the UMI counts, the genes in this list may have their expression measurements artificially truncated in those samples where the number of copies of the transcript exceed the <span class="math inline">\(4^6=4096\)</span> threshold.</p>
<p>This table shows that many of these genes are mitochondrial and ribosomal protein genes.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>read_e</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span><span class='op'>(</span><span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"read_counts"</span><span class='op'>)</span><span class='op'>[</span><span class='va'>g</span>, <span class='op'>]</span><span class='op'>)</span>
<span class='va'>umi_e</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span><span class='op'>(</span><span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"UMI_counts"</span><span class='op'>)</span><span class='op'>[</span><span class='va'>g</span>, <span class='op'>]</span><span class='op'>)</span>
<span class='va'>high_count_tbl</span> <span class='op'>&lt;-</span> <span class='fu'>tibble</span><span class='fu'>::</span><span class='fu'><a href='https://tibble.tidyverse.org/reference/tibble.html'>tibble</a></span><span class='op'>(</span>
  gene <span class='op'>=</span> <span class='va'>g</span>,
  read_median_count <span class='op'>=</span> <span class='fu'>matrixStats</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/matrixStats/man/rowMedians.html'>rowMedians</a></span><span class='op'>(</span><span class='va'>read_e</span><span class='op'>)</span>,
  read_exceed_sum <span class='op'>=</span> <span class='fu'>matrixStats</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/matrixStats/man/rowSums2.html'>rowSums2</a></span><span class='op'>(</span><span class='va'>read_e</span> <span class='op'>&gt;</span> <span class='fl'>4</span> <span class='op'>^</span> <span class='fl'>6</span><span class='op'>)</span>,
  read_exceed_percent <span class='op'>=</span> <span class='va'>read_exceed_sum</span> <span class='op'>/</span> <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>ncol</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>,
  umi_median_count <span class='op'>=</span> <span class='fu'>matrixStats</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/matrixStats/man/rowMedians.html'>rowMedians</a></span><span class='op'>(</span><span class='va'>umi_e</span><span class='op'>)</span>,
  umi_exceed_sum <span class='op'>=</span> <span class='fu'>matrixStats</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/matrixStats/man/rowSums2.html'>rowSums2</a></span><span class='op'>(</span><span class='va'>umi_e</span> <span class='op'>&gt;</span> <span class='fl'>4</span> <span class='op'>^</span> <span class='fl'>6</span><span class='op'>)</span>,
  umi_exceed_percent <span class='op'>=</span> <span class='va'>umi_exceed_sum</span> <span class='op'>/</span> <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>ncol</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/arrange.html'>arrange</a></span><span class='op'>(</span><span class='fu'>desc</span><span class='op'>(</span><span class='va'>read_exceed_sum</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>sketch</span> <span class='op'>&lt;-</span> <span class='fu'>htmltools</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/htmltools/man/withTags.html'>withTags</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/table.html'>table</a></span><span class='op'>(</span>
  class <span class='op'>=</span> <span class='st'>"display"</span>,
  <span class='fu'>thead</span><span class='op'>(</span>
    <span class='fu'>tr</span><span class='op'>(</span>
      <span class='fu'>th</span><span class='op'>(</span>rowspan <span class='op'>=</span> <span class='fl'>2</span>, <span class='st'>"gene"</span><span class='op'>)</span>,
      <span class='fu'>th</span><span class='op'>(</span>colspan <span class='op'>=</span> <span class='fl'>3</span>, <span class='st'>"reads"</span><span class='op'>)</span>,
      <span class='fu'>th</span><span class='op'>(</span>colspan <span class='op'>=</span> <span class='fl'>3</span>, <span class='st'>"UMIs"</span><span class='op'>)</span>
    <span class='op'>)</span>,
    <span class='fu'>tr</span><span class='op'>(</span>
      <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>lapply</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"median count"</span>, <span class='st'>"N &gt; 4096"</span>, <span class='st'>"% &gt; 4096"</span><span class='op'>)</span>, <span class='fl'>2</span><span class='op'>)</span>, <span class='va'>th</span><span class='op'>)</span>
    <span class='op'>)</span>
  <span class='op'>)</span>
<span class='op'>)</span><span class='op'>)</span>
<span class='fu'>DT</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/DT/man/datatable.html'>datatable</a></span><span class='op'>(</span>
  <span class='va'>high_count_tbl</span>, 
  caption <span class='op'>=</span> <span class='st'>"Genes for which any sample has a read count exceeding the 4096 threshold and summaries of the corresponding read and UMI counts, namely: the median count ('median count'); the number of samples for which the counts exceed the threshold ('N &gt; 4096'); and the percentage of samples for which the counts exceed the threshold ('% &gt; 4096')."</span>,
  container <span class='op'>=</span> <span class='va'>sketch</span>,
  rownames <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>DT</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/DT/man/formatCurrency.html'>formatPercentage</a></span><span class='op'>(</span>
    columns <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"read_exceed_percent"</span>, <span class='st'>"umi_exceed_percent"</span><span class='op'>)</span>,
    digits <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>DT</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/DT/man/formatCurrency.html'>formatRound</a></span><span class='op'>(</span>
    columns <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"read_median_count"</span>, <span class='st'>"umi_median_count"</span><span class='op'>)</span>,
    digits <span class='op'>=</span> <span class='fl'>0</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div id="htmlwidget-8201f8d795640477ac25" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-8201f8d795640477ac25">{"x":{"filter":"none","class":"display","caption":"<caption>Genes for which any sample has a read count exceeding the 4096 threshold and summaries of the corresponding read and UMI counts, namely: the median count ('median count'); the number of samples for which the counts exceed the threshold ('N &gt; 4096'); and the percentage of samples for which the counts exceed the threshold ('% &gt; 4096').<\/caption>","data":[["MT-RNR2","MALAT1","MT-CO3","MT-CYB","MT-ND4","MT-ATP6","MTRNR2L12","MT-CO1","MT-ND3","MTATP6P1","TMSB4X","B2M","MTRNR2L8","ACTB","MT-CO2","MT-ND4L","MT-ND5","HLA-B","RPS6","MT-ND1","MIR1244-3","RPLP1","MT-RNR1","ACTG1","RPL9","PFN1","TPT1","AC099336.2","RPS14","RPL4","MTCO1P12","RPL15","RPS18","RPS15A","RPL34","RPS20","RPS21","RPS12","HLA-A","PTPRC","RPS2","RPL31","RPLP0","RPL13AP25","LEF1","RPL14","GNAS","RPS8","HLA-C","VIM","PTMA","ARHGDIB","RPS13P2","RPS11","RPS3AP26","RPL28","PABPC1","XIST","RPL15P2","AC079250.1","FYB1","SH3BGRL3","PNISR","RPS16","GAPDH","RPL10P16","RPL35A","RPL11","RPLP2","RPL30","RPL13","EEF1A1","FTH1","SLFN5","RPL23","TRGC1","RAC2","RPL18","RPS29","CFL1","HIST1H4C","BCL11B","SEPTIN6","IL7R","RPS29P16","UBB","GZMK","EEF1A1P6","UBA52","RPL38","TCF7","DDX17","RPL19","RPS27AP16","TUBA1B","MT-ND2","RPL37","EVL","AC090589.1","TOX2","EEF2","TRBC2","MT-ATP8","RPL13A","AC005912.1","ARPC2","HNRNPA2B1","RPL10A","YBX1","RPS25","CCL5","CELF2","IKZF2","STMN1","CCNI","RPS5","GNLY","MTND4P12","SATB1","EEF1B2P6","RPL23AP42","RPS23","RPL21P16","RACK1","GIMAP7","ITGA4","TMSB10","SOX4","H3F3B","SON","AC243919.1","ETS1","CD99","RPL7A","TXNIP","PCBP1","CLDND1","RPS13"],[96690,19977,11597,9569,8076,9142,7908,7943,6528,6033,4854,4752,4554,4541,4667,4344,4287,4153,3954,3556,3262,3773,2991,2960,2941,2424,2512,2630,2160,2250,2728,1955,1820,1924,1990,1844,1935,1738,2061,1585,1965,1797,1732,2006,251,1646,1722,1724,1521,795,1505,1489,1237,1533,1206,1518,1644,589,1129,1649,686,1292,1044,1448,1308,1456,1235,1356,1456,1442,1524,1388,1398,876,1337,1323,1075,1298,1637,1199,548,566,771,1353,179,1173,610,763,1171,1170,446,1000,1021,1103,1077,1574,1060,849,1052,89,838,514,1185,899,905,810,701,960,801,1017,569,558,137,289,493,1030,17,467,365,1045,1028,970,646,898,635,264,736,127,771,692,864,974,608,945,690,856,372,773],[120,109,99,96,95,92,90,90,82,80,71,71,66,66,66,64,63,62,58,55,55,54,45,44,43,42,39,37,33,32,30,29,27,26,26,24,24,22,22,21,21,21,21,18,18,17,16,16,16,15,14,14,14,13,12,12,12,12,11,10,10,10,9,9,9,9,9,9,9,8,8,8,8,8,7,7,6,6,6,6,6,6,6,5,5,5,5,5,5,5,5,4,4,4,4,3,3,3,3,3,3,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1],[0.991735537190083,0.900826446280992,0.818181818181818,0.793388429752066,0.785123966942149,0.760330578512397,0.743801652892562,0.743801652892562,0.677685950413223,0.661157024793388,0.586776859504132,0.586776859504132,0.545454545454545,0.545454545454545,0.545454545454545,0.528925619834711,0.520661157024793,0.512396694214876,0.479338842975207,0.454545454545455,0.454545454545455,0.446280991735537,0.371900826446281,0.363636363636364,0.355371900826446,0.347107438016529,0.322314049586777,0.305785123966942,0.272727272727273,0.264462809917355,0.247933884297521,0.239669421487603,0.223140495867769,0.214876033057851,0.214876033057851,0.198347107438017,0.198347107438017,0.181818181818182,0.181818181818182,0.173553719008264,0.173553719008264,0.173553719008264,0.173553719008264,0.148760330578512,0.148760330578512,0.140495867768595,0.132231404958678,0.132231404958678,0.132231404958678,0.12396694214876,0.115702479338843,0.115702479338843,0.115702479338843,0.107438016528926,0.0991735537190083,0.0991735537190083,0.0991735537190083,0.0991735537190083,0.0909090909090909,0.0826446280991736,0.0826446280991736,0.0826446280991736,0.0743801652892562,0.0743801652892562,0.0743801652892562,0.0743801652892562,0.0743801652892562,0.0743801652892562,0.0743801652892562,0.0661157024793388,0.0661157024793388,0.0661157024793388,0.0661157024793388,0.0661157024793388,0.0578512396694215,0.0578512396694215,0.0495867768595041,0.0495867768595041,0.0495867768595041,0.0495867768595041,0.0495867768595041,0.0495867768595041,0.0495867768595041,0.0413223140495868,0.0413223140495868,0.0413223140495868,0.0413223140495868,0.0413223140495868,0.0413223140495868,0.0413223140495868,0.0413223140495868,0.0330578512396694,0.0330578512396694,0.0330578512396694,0.0330578512396694,0.0247933884297521,0.0247933884297521,0.0247933884297521,0.0247933884297521,0.0247933884297521,0.0247933884297521,0.0165289256198347,0.0165289256198347,0.0165289256198347,0.0165289256198347,0.0165289256198347,0.0165289256198347,0.0165289256198347,0.0165289256198347,0.0165289256198347,0.0165289256198347,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736,0.00826446280991736],[335,320,238,209,255,247,278,233,210,231,181,161,228,167,218,142,161,109,145,197,110,144,153,93,101,102,130,147,93,88,218,87,95,89,95,107,99,111,81,78,103,103,96,124,27,76,75,107,73,37,109,62,61,95,86,107,98,27,61,96,48,41,46,90,90,94,70,96,112,94,123,154,111,44,112,50,38,72,124,88,65,40,40,51,64,59,20,104,66,75,42,46,91,80,85,146,100,38,91,9,58,33,107,84,91,41,54,62,48,80,41,27,12,31,35,72,5,75,18,58,89,90,72,79,25,21,53,22,40,34,75,48,24,73,43,28,13,67],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th rowspan=\"2\">gene<\/th>\n      <th colspan=\"3\">reads<\/th>\n      <th colspan=\"3\">UMIs<\/th>\n    <\/tr>\n    <tr>\n      <th>median count<\/th>\n      <th>N &gt; 4096<\/th>\n      <th>% &gt; 4096<\/th>\n      <th>median count<\/th>\n      <th>N &gt; 4096<\/th>\n      <th>% &gt; 4096<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"targets":1,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 0, 3, \",\", \".\");\n  }"},{"targets":4,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 0, 3, \",\", \".\");\n  }"},{"targets":3,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatPercentage(data, 1, 3, \",\", \".\");\n  }"},{"targets":6,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatPercentage(data, 1, 3, \",\", \".\");\n  }"},{"className":"dt-right","targets":[1,2,3,4,5,6]}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render","options.columnDefs.1.render","options.columnDefs.2.render","options.columnDefs.3.render"],"jsHooks":[]}</script>
</div>
<p>We can investigate this list more formally using a statistical test for over-representation of gene ontology (GO) terms in this gene set. This is done using the <code>goana()</code> function from the <em><a href="https://bioconductor.org/packages/3.12/limma">limma</a></em> package. These results are shown in the table below and confirm that these genes are enriched for GO terms related to ribosomal and mitochondrial processes. This is something we frequently see with the mini-bulk protocol and the CEL-Seq protocol from which it is derived.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://bioinf.wehi.edu.au/limma'>limma</a></span><span class='op'>)</span>
<span class='va'>go</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/limma/man/goana.html'>goana</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/unlist.html'>unlist</a></span><span class='op'>(</span><span class='fu'>rowData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>[</span><span class='va'>g</span>, <span class='op'>]</span><span class='op'>)</span><span class='op'>$</span><span class='va'>NCBI.ENTREZID</span><span class='op'>)</span>, species <span class='op'>=</span> <span class='st'>"Hs"</span><span class='op'>)</span>
<span class='va'>tg</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/limma/man/topGO.html'>topGO</a></span><span class='op'>(</span><span class='va'>go</span>, number <span class='op'>=</span> <span class='cn'>Inf</span><span class='op'>)</span>
<span class='fu'>DT</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/DT/man/datatable.html'>datatable</a></span><span class='op'>(</span>
  <span class='va'>tg</span><span class='op'>[</span><span class='va'>tg</span><span class='op'>$</span><span class='va'>P.DE</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span>, <span class='op'>]</span>,
  caption <span class='op'>=</span> <span class='st'>"Most significant GO terms for genes with read counts that exceed the 4096 threshold. Report columns are the ontology that the GO term belongs to ('Ont'); the number of genes in the GO term ('N'); the number of genes in the DE set ('DE'); and the p-value for over-representation of the GO term in the set ('P.DE'). The p-values returned are unadjusted for multiple testing because GO terms are often overlapping, so standard methods of p-value adjustment may be very conservative. This means that only very small p-values should be used for published results."</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>DT</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/DT/man/formatCurrency.html'>formatRound</a></span><span class='op'>(</span>columns <span class='op'>=</span> <span class='st'>"P.DE"</span>, digits <span class='op'>=</span> <span class='fl'>9</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div id="htmlwidget-1d6f47f806c6d49792fc" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-1d6f47f806c6d49792fc">{"x":{"filter":"none","caption":"<caption>Most significant GO terms for genes with read counts that exceed the 4096 threshold. Report columns are the ontology that the GO term belongs to ('Ont'); the number of genes in the GO term ('N'); the number of genes in the DE set ('DE'); and the p-value for over-representation of the GO term in the set ('P.DE'). The p-values returned are unadjusted for multiple testing because GO terms are often overlapping, so standard methods of p-value adjustment may be very conservative. This means that only very small p-values should be used for published results.<\/caption>","data":[["GO:0022626","GO:0006614","GO:0000184","GO:0006613","GO:0045047","GO:0072599","GO:0019083","GO:0070972","GO:0006413","GO:0019080","GO:0044391","GO:0000956","GO:0003735","GO:0006612","GO:0005840","GO:0006402","GO:0006401","GO:0016032","GO:0044403","GO:0090150","GO:1990904","GO:0006605","GO:0006412","GO:0043043","GO:0022625","GO:0034655","GO:0046700","GO:0044270","GO:0019439","GO:0072594","GO:0016071","GO:1901361","GO:0043604","GO:0072657","GO:0006518","GO:0005198","GO:0005925","GO:0030055","GO:0044419","GO:0015934","GO:0022627","GO:0043603","GO:0033365","GO:0015935","GO:0044265","GO:0003723","GO:1901566","GO:0006886","GO:0070062","GO:1903561","GO:0009057","GO:0043230","GO:0070161","GO:0034613","GO:0032991","GO:0070727","GO:0002181","GO:0051649","GO:0046907","GO:0042788","GO:0010629","GO:0051641","GO:0015031","GO:0008104","GO:0051179","GO:0015833","GO:0042886","GO:0045184","GO:0071705","GO:0005844","GO:0006810","GO:0010468","GO:0048519","GO:0030054","GO:0051234","GO:0033036","GO:1901575","GO:0005615","GO:0044271","GO:0044248","GO:0003676","GO:0010605","GO:0031982","GO:0043232","GO:0043228","GO:0005829","GO:0071702","GO:0005576","GO:0019843","GO:0009892","GO:0009059","GO:0034645","GO:0009056","GO:0060255","GO:0005737","GO:1901363","GO:0010467","GO:0097159","GO:0006139","GO:0019222","GO:0098803","GO:0006119","GO:1901576","GO:0034641","GO:0046483","GO:0044249","GO:0016070","GO:0006725","GO:0009058","GO:0090304","GO:0042775","GO:0042773","GO:0042254","GO:0070469","GO:0043229","GO:1901360","GO:0022904","GO:0005622","GO:0022613","GO:0046034","GO:0044267","GO:0015990","GO:0015988","GO:0098800","GO:0006364","GO:0022900","GO:0016072","GO:0043226","GO:0034097","GO:0071345","GO:0043227","GO:0005746","GO:0016020","GO:0019538","GO:0044085","GO:0034622","GO:0071346","GO:0045335","GO:0045333","GO:0002480","GO:0042612","GO:0045586","GO:1901564","GO:0050789","GO:0042273","GO:1990204","GO:0042274","GO:0050136","GO:0008137","GO:0003954","GO:0034341","GO:1902600","GO:0046643","GO:0006091","GO:0005747","GO:0030964","GO:0045271","GO:0043933","GO:0034470","GO:0042492","GO:0015980","GO:0044260","GO:0006120","GO:0044237","GO:0065003","GO:0016655","GO:0001914","GO:0006417","GO:0030139","GO:0002568","GO:0002681","GO:0033153","GO:0005515","GO:0005488","GO:0002483","GO:0019885","GO:0098798","GO:0008152","GO:0043231","GO:0014069","GO:0046629","GO:0032279","GO:0005634","GO:0001911","GO:0034660","GO:0001913","GO:0010033","GO:0065007","GO:0045588","GO:0001906","GO:0034248","GO:0099572","GO:0045202","GO:0098984","GO:0019883","GO:0031342","GO:0002484","GO:0002486","GO:0046977","GO:0042611","GO:0031341","GO:0046645","GO:0001916","GO:0071840","GO:0005743","GO:0031343","GO:0005740","GO:0048019","GO:0032091","GO:0032981","GO:0010257","GO:0016651","GO:0032956","GO:0042255","GO:0070069","GO:0030666","GO:0002376","GO:0071310","GO:0051272","GO:0098796","GO:0034250","GO:0051100","GO:0002684","GO:0019221","GO:0031975","GO:0031967","GO:1903039","GO:0006807","GO:0002709","GO:0000462","GO:0031966","GO:0019866","GO:0030097","GO:0001910","GO:0006955","GO:0030670","GO:0042981","GO:0033151","GO:0001775","GO:0030547","GO:0009205","GO:0002707","GO:0042110","GO:1990428","GO:0043067","GO:0030217","GO:0032970","GO:0010941","GO:0048534","GO:0045296","GO:0005791","GO:0006414","GO:0009199","GO:1903706","GO:0002682","GO:0045727","GO:0034109","GO:0010608","GO:0009060","GO:0009144","GO:2000147","GO:0043170","GO:1903311","GO:0070887","GO:0040017","GO:0002711","GO:0046631","GO:0110053","GO:0045277","GO:0097452","GO:0022409","GO:0030490","GO:0043393","GO:0002520","GO:0002704","GO:0001912","GO:0019646","GO:0006123","GO:0000028","GO:0007159","GO:0022607","GO:0060548","GO:0002521","GO:0030098","GO:0006754","GO:1901798","GO:0009987","GO:0098794","GO:0033108","GO:0005200","GO:0002456","GO:0070180","GO:1902255","GO:0098973","GO:1990948","GO:1900118","GO:0016491","GO:0036464","GO:0022618","GO:0001909","GO:0030335","GO:0031981","GO:0003823","GO:0046649","GO:0021872","GO:0009141","GO:0012507","GO:0009055","GO:0044238","GO:0003729","GO:0071826","GO:0045321","GO:0005654","GO:0035770","GO:0097433","GO:0042270","GO:0055105","GO:0009206","GO:0003674","GO:0003785","GO:0004129","GO:0015002","GO:0016676","GO:0009145","GO:0016444","GO:0002562","GO:0003924","GO:1903037","GO:0071556","GO:0098553","GO:0016675","GO:0042221","GO:0050863","GO:0006996","GO:0000027","GO:0006396","GO:0001915","GO:0050776","GO:0043066","GO:0009201","GO:0051270","GO:0042605","GO:0045785","GO:0000281","GO:0008064","GO:0009607","GO:0030832","GO:0070013","GO:0031974","GO:0043233","GO:0071353","GO:0015078","GO:0007015","GO:0098552","GO:0035198","GO:0043069","GO:0002479","GO:0035701","GO:0042824","GO:0033077","GO:2000145","GO:0030155","GO:0002200","GO:0030027","GO:0070670","GO:0043254","GO:0010628","GO:0050839","GO:0032272","GO:0010942","GO:1903708","GO:0006950","GO:0098576","GO:0042590","GO:1904667","GO:0046651","GO:0051098","GO:0051493","GO:0009142","GO:0032943","GO:0002703","GO:1902903","GO:0031333","GO:0042769","GO:1902105","GO:0040012","GO:0050870","GO:0055038","GO:0042989","GO:0070181","GO:0042542","GO:0030334","GO:0051707","GO:0098609","GO:0043207","GO:0008150","GO:0031625","GO:1900117","GO:0008154","GO:0045580","GO:0008097","GO:0005522","GO:0006915","GO:0043065","GO:0055114","GO:0060333","GO:0051492","GO:0002822","GO:0002706","GO:0045582","GO:0070482","GO:0051338","GO:0043068","GO:0051702","GO:0043484","GO:0009605","GO:0099186","GO:0032271","GO:0070661","GO:0030029","GO:0030134","GO:0061980","GO:1902107","GO:0044389","GO:0071704","GO:0048598","GO:0030036","GO:0098974","GO:0150105","GO:0031386","GO:0035722","GO:0002474","GO:0051817","GO:0061640","GO:0097435","GO:0051249","GO:0002252","GO:0002819","GO:0071349","GO:0110020","GO:0031252","GO:0005750","GO:0045263","GO:0045275","GO:0070671","GO:0032231","GO:0002824","GO:0045621","GO:0021879","GO:0044092","GO:0046632","GO:0030038","GO:0002708","GO:0043149","GO:0030833","GO:2001171","GO:0099188","GO:0098847","GO:0048524","GO:0016363","GO:0098542","GO:0022407","GO:0002821","GO:0051438","GO:0110165","GO:0045953","GO:1900121","GO:0051444","GO:0045619","GO:0009615","GO:0030660","GO:0045637","GO:0030662","GO:0035821","GO:0048024","GO:0012501","GO:0019887","GO:0048522","GO:0033631","GO:0016045","GO:0002716","GO:0043488","GO:0008219","GO:0051251","GO:0050821","GO:0015672","GO:2000272","GO:0001666","GO:0006979","GO:0098918","GO:0055106","GO:0048523","GO:0030041","GO:0016310","GO:0043487","GO:0042098","GO:0098543","GO:0070233","GO:0002710","GO:0046933","GO:0003746","GO:0002460","GO:0045471","GO:0036293","GO:0051851","GO:0070527","GO:2001244","GO:0071824","GO:0002698","GO:0005525","GO:0002726","GO:2001169","GO:0034463","GO:0002485","GO:0002481","GO:0072749","GO:0046817","GO:1990771","GO:0000461","GO:0033291","GO:1904689","GO:1901194","GO:0090074","GO:0071678","GO:0015453","GO:0035605","GO:0052501","GO:0044533","GO:2000761","GO:0051623","GO:1904571","GO:0040032","GO:1900022","GO:1901193","GO:2000759","GO:0043496","GO:1901328","GO:0032173","GO:0070992","GO:0065008","GO:0032550","GO:0031640","GO:0030837","GO:0006909","GO:0001883","GO:0032535","GO:0032549","GO:0042276","GO:0021854","GO:0046641","GO:0002693","GO:0010310","GO:0007166","GO:0072678","GO:0008380","GO:0048513","GO:0031328","GO:0034399","GO:0061013","GO:0048870","GO:0051674","GO:0001882","GO:0005783","GO:0000398","GO:0000377","GO:0070987","GO:0005753","GO:0000717","GO:0055093","GO:0006812","GO:0016043","GO:0031396","GO:0000375","GO:0030234","GO:0051258","GO:0051301","GO:0019001","GO:0032561","GO:0000715","GO:0006297","GO:1904996","GO:0045259","GO:0002694","GO:0019730","GO:0019207","GO:0042119","GO:0005938","GO:0002705","GO:0009891","GO:0006338","GO:0038096","GO:0002433","GO:0045182","GO:0044305","GO:0098581","GO:0042776","GO:0033177","GO:1904666","GO:0019899","GO:0002696","GO:0036230","GO:0038094","GO:0002687","GO:0010592","GO:0002699","GO:0061844","GO:0006928","GO:0002431","GO:0070911","GO:0048027","GO:0033119","GO:0030810","GO:1900373","GO:0015252","GO:0002347","GO:0008284","GO:0007568","GO:0005739","GO:0001895","GO:0050867","GO:0043903","GO:0015986","GO:0036037","GO:0015985","GO:1900120","GO:0043005","GO:0031397","GO:0005575","GO:0008283","GO:0050896","GO:0000302","GO:0035255","GO:0000470","GO:0050684","GO:0048518","GO:0031729","GO:0140039","GO:0003366","GO:0071283","GO:0071228","GO:1904155","GO:0070288","GO:0043891","GO:0004365","GO:0034669","GO:0004917","GO:0008043","GO:1990630","GO:0097534","GO:0097535","GO:1990145","GO:1902340","GO:0035606","GO:0044855","GO:0044856","GO:1904434","GO:2000473","GO:2000568","GO:1905663","GO:1904437","GO:1904432","GO:2000471","GO:2000567","GO:0051621","GO:2000828","GO:1904569","GO:1904435","GO:0062061","GO:0061153","GO:0061152","GO:0043299","GO:0030099","GO:0050865","GO:0044877","GO:0061099","GO:0006294","GO:0036296","GO:1902904","GO:0003779","GO:0051017","GO:1903320","GO:0050778","GO:0065009","GO:0002443","GO:0007010","GO:1904837","GO:2000765","GO:1900027","GO:0035666","GO:0030176","GO:0120025","GO:0050777","GO:0098662","GO:0002275","GO:0002274","GO:0048013","GO:0070229","GO:0051497","GO:1902624","GO:2000406","GO:1902253","GO:0061572","GO:0031901","GO:0002444","GO:1903312","GO:0007005","GO:0051494","GO:0021915","GO:0042592","GO:0002724","GO:0005730","GO:0009893","GO:0043044","GO:0031227","GO:2001242","GO:1902745","GO:0046640","GO:0002691","GO:0042129","GO:1903321","GO:0010035","GO:0035633","GO:0002756","GO:0032232","GO:0070232","GO:0061097","GO:0099080","GO:1901700","GO:0001894","GO:0000910","GO:0042407","GO:0071357","GO:0097194","GO:0060337","GO:0009152","GO:0002477","GO:0044336","GO:0072717","GO:0071461","GO:0031905","GO:0030538","GO:0048619","GO:0034668","GO:0036194","GO:0036195","GO:0071899","GO:2000978","GO:2000435","GO:0070495","GO:0035898","GO:0052330","GO:0052151","GO:0052042","GO:1903615","GO:1905273","GO:0030298","GO:0071898","GO:1905271","GO:1905661","GO:0070494","GO:0005094","GO:0031509","GO:0140461","GO:0002419","GO:0060249","GO:0046633","GO:0009595","GO:0002755","GO:0043368","GO:0042100","GO:0002449","GO:0061024","GO:0002250","GO:0040011","GO:0002478","GO:0014070","GO:0070423","GO:0006296","GO:0045070","GO:1904994","GO:0034340","GO:0007155","GO:0002702","GO:0035872","GO:2000403","GO:0010591","GO:0005798","GO:0001667","GO:0002429","GO:0002757","GO:0030838","GO:0045639","GO:0042102","GO:0042995","GO:0022610","GO:2001235","GO:0120032","GO:0006959","GO:0002366","GO:0033683","GO:0043312","GO:0042127","GO:0019884","GO:0060491","GO:0002263","GO:0120034","GO:0009260","GO:0002283","GO:0000788","GO:0031492","GO:0000183","GO:0019068","GO:0004857","GO:0098562","GO:0031490","GO:0030863","GO:0001841","GO:0030043","GO:0035910","GO:0042610","GO:0051430","GO:0005853","GO:0021943","GO:0021888","GO:0021886","GO:0002378","GO:0001907","GO:0035709","GO:0031852","GO:0018076","GO:0006933","GO:2000623","GO:0031508","GO:0140462","GO:0097013","GO:0110024","GO:2000566","GO:0042998","GO:0030292","GO:2000622","GO:2000434","GO:0072716","GO:0033590","GO:0006610","GO:0021773","GO:0050832","GO:0042269","GO:0009628","GO:0055037","GO:0008092","GO:0044087","GO:0016477","GO:0098761","GO:1902622","GO:1900371","GO:0098760","GO:0019985","GO:0021953","GO:0002446","GO:0021761","GO:0090079","GO:0048002","GO:0046390","GO:0043028","GO:0002823","GO:0030808","GO:0046688","GO:0002369","GO:0010604","GO:0015629","GO:0051129","GO:0098660","GO:0042493","GO:0045087","GO:2000404","GO:0097178","GO:0006164","GO:0031329","GO:0009150","GO:0031647","GO:0002715","GO:0035094","GO:0002768","GO:0003697","GO:0031032","GO:0000904","GO:0030135","GO:0009888","GO:0050790","GO:0001228","GO:0006310","GO:0002764","GO:0006952","GO:0035254","GO:0010856","GO:1902262","GO:0035905","GO:0003289","GO:0031698","GO:0019960","GO:0031726","GO:0070934","GO:0044194","GO:0098574","GO:0098556","GO:0030284","GO:0071593","GO:0002249","GO:1905098","GO:1903208","GO:0036476","GO:0051343","GO:1905451","GO:0002925","GO:0010729","GO:1902527","GO:0045901","GO:0098871","GO:0060623","GO:1904688","GO:2000977","GO:1903207","GO:0002911","GO:0002667","GO:0072344","GO:0010041","GO:0008186","GO:0050733","GO:0035368","GO:0002870","GO:0001216","GO:0048871","GO:0016055","GO:0032392","GO:0090066","GO:0007420","GO:0198738","GO:0072676","GO:0031410","GO:0031055","GO:0005884","GO:0098655","GO:0097708","GO:0010557","GO:0043009","GO:0002820","GO:1902905","GO:0072522","GO:0009259","GO:0040029","GO:0030658","GO:0030218","GO:0010171","GO:2000107","GO:2000059","GO:0070231","GO:0045058","GO:0070555","GO:1990266","GO:0002685","GO:0014009","GO:1903580","GO:0045981","GO:1900544","GO:0045646","GO:1902743","GO:0051246","GO:0101002","GO:1904813","GO:0031584","GO:0060385","GO:0001740","GO:0051434","GO:0051429","GO:0070937","GO:0050904","GO:0021873","GO:0071204","GO:0006880","GO:1990254","GO:0052433","GO:0052150","GO:0044532","GO:0044531","GO:0052248","GO:0071866","GO:0032696","GO:2000110","GO:1903026","GO:1900028","GO:0033634","GO:2000503","GO:0071896","GO:1902396","GO:0090230","GO:1902525","GO:1903613","GO:0003284","GO:0097227","GO:0002424","GO:0000902","GO:0048255","GO:0030659","GO:0007417","GO:0009986","GO:0001838","GO:0032268","GO:0051651","GO:0006887","GO:0019693","GO:0009792","GO:0043549","GO:2000112","GO:0050792","GO:0007007","GO:0006301","GO:0016469","GO:0071622","GO:0019079","GO:0048812","GO:0019901","GO:0050670","GO:0000122","GO:0045010","GO:0006336","GO:0000731","GO:0061756","GO:0043392","GO:0045744","GO:2000772","GO:0007163","GO:0017148","GO:2000113","GO:0034101","GO:0032944","GO:0045055","GO:0051253","GO:0034724","GO:0009620","GO:0043679","GO:0032989","GO:0006397","GO:0017111","GO:0060322","GO:0042113","GO:0036297","GO:0002720","GO:1904724","GO:0070358","GO:0002428","GO:0019828","GO:0048318","GO:0110021","GO:0031730","GO:0060710","GO:0021902","GO:0060613","GO:0021861","GO:0044806","GO:0007442","GO:0031580","GO:1990247","GO:0070100","GO:0032714","GO:0001842","GO:0003357","GO:0048341","GO:0030578","GO:0010536","GO:0060369","GO:0010726","GO:2000391","GO:0031077","GO:1990405","GO:1903608","GO:0071865","GO:0110022","GO:1905449","GO:0060753","GO:0097577","GO:0032202","GO:0048039","GO:0009790","GO:0045859","GO:1903050","GO:0012506","GO:0120039","GO:0006163","GO:0072175","GO:0043021","GO:0032147","GO:0051171","GO:0034508","GO:0050732","GO:0070228","GO:0031529","GO:0048858","GO:0042743","GO:0075733","GO:0043022","GO:0051495","GO:0050671","GO:0006259","GO:0007566","GO:1902373","GO:0002763","GO:0043489","GO:0032946","GO:0032273","GO:0019882","GO:0002253","GO:0098797","GO:0097305","GO:0043486","GO:0051289","GO:0071839","GO:0060020","GO:0030881","GO:0035740","GO:0007253","GO:0007144","GO:1990712","GO:0061525","GO:0021979","GO:0051665","GO:2000766","GO:0032713","GO:1903979","GO:0051620","GO:2001187","GO:0071864","GO:0030854","GO:1903238","GO:0099571","GO:2000564","GO:1904714","GO:1905097","GO:0034756","GO:2000389","GO:0019058","GO:0007266","GO:0032355","GO:1902494","GO:0034249","GO:0042267","GO:1903902","GO:0090342","GO:0032990","GO:0045595","GO:0010556","GO:0010558","GO:0031326","GO:0070663","GO:0009719","GO:0005102","GO:0038093","GO:0034113","GO:0045824","GO:0032233","GO:0140238","GO:0048488","GO:0050921","GO:0002700","GO:0001738","GO:0065004","GO:0046635","GO:0032059","GO:0048539","GO:0071225","GO:0071316","GO:0031748","GO:0004322","GO:0052040","GO:0090647","GO:1901164","GO:0030223","GO:0031628","GO:0016681","GO:0016724","GO:0034112","GO:1900122","GO:0010533","GO:0042996","GO:0010728","GO:2000501","GO:0051593","GO:0001514","GO:0042608","GO:0006451","GO:0008121"],["cytosolic ribosome","SRP-dependent cotranslational protein targeting to membrane","nuclear-transcribed mRNA catabolic process, nonsense-mediated decay","cotranslational protein targeting to membrane","protein targeting to ER","establishment of protein localization to endoplasmic reticulum","viral transcription","protein localization to endoplasmic reticulum","translational initiation","viral gene expression","ribosomal subunit","nuclear-transcribed mRNA catabolic process","structural constituent of ribosome","protein targeting to membrane","ribosome","mRNA catabolic process","RNA catabolic process","viral process","symbiotic process","establishment of protein localization to membrane","ribonucleoprotein complex","protein targeting","translation","peptide biosynthetic process","cytosolic large ribosomal subunit","nucleobase-containing compound catabolic process","heterocycle catabolic process","cellular nitrogen compound catabolic process","aromatic compound catabolic process","establishment of protein localization to organelle","mRNA metabolic process","organic cyclic compound catabolic process","amide biosynthetic process","protein localization to membrane","peptide metabolic process","structural molecule activity","focal adhesion","cell-substrate junction","interspecies interaction between organisms","large ribosomal subunit","cytosolic small ribosomal subunit","cellular amide metabolic process","protein localization to organelle","small ribosomal subunit","cellular macromolecule catabolic process","RNA binding","organonitrogen compound biosynthetic process","intracellular protein transport","extracellular exosome","extracellular vesicle","macromolecule catabolic process","extracellular organelle","anchoring junction","cellular protein localization","protein-containing complex","cellular macromolecule localization","cytoplasmic translation","establishment of localization in cell","intracellular transport","polysomal ribosome","negative regulation of gene expression","cellular localization","protein transport","protein localization","localization","peptide transport","amide transport","establishment of protein localization","nitrogen compound transport","polysome","transport","regulation of gene expression","negative regulation of biological process","cell junction","establishment of localization","macromolecule localization","organic substance catabolic process","extracellular space","cellular nitrogen compound biosynthetic process","cellular catabolic process","nucleic acid binding","negative regulation of macromolecule metabolic process","vesicle","intracellular non-membrane-bounded organelle","non-membrane-bounded organelle","cytosol","organic substance transport","extracellular region","rRNA binding","negative regulation of metabolic process","macromolecule biosynthetic process","cellular macromolecule biosynthetic process","catabolic process","regulation of macromolecule metabolic process","cytoplasm","heterocyclic compound binding","gene expression","organic cyclic compound binding","nucleobase-containing compound metabolic process","regulation of metabolic process","respiratory chain complex","oxidative phosphorylation","organic substance biosynthetic process","cellular nitrogen compound metabolic process","heterocycle metabolic process","cellular biosynthetic process","RNA metabolic process","cellular aromatic compound metabolic process","biosynthetic process","nucleic acid metabolic process","mitochondrial ATP synthesis coupled electron transport","ATP synthesis coupled electron transport","ribosome biogenesis","respirasome","intracellular organelle","organic cyclic compound metabolic process","respiratory electron transport chain","intracellular","ribonucleoprotein complex biogenesis","ATP metabolic process","cellular protein metabolic process","electron transport coupled proton transport","energy coupled proton transmembrane transport, against electrochemical gradient","inner mitochondrial membrane protein complex","rRNA processing","electron transport chain","rRNA metabolic process","organelle","response to cytokine","cellular response to cytokine stimulus","membrane-bounded organelle","mitochondrial respirasome","membrane","protein metabolic process","cellular component biogenesis","cellular protein-containing complex assembly","cellular response to interferon-gamma","phagocytic vesicle","cellular respiration","antigen processing and presentation of exogenous peptide antigen via MHC class I, TAP-independent","MHC class I protein complex","regulation of gamma-delta T cell differentiation","organonitrogen compound metabolic process","regulation of biological process","ribosomal large subunit biogenesis","oxidoreductase complex","ribosomal small subunit biogenesis","NADH dehydrogenase (quinone) activity","NADH dehydrogenase (ubiquinone) activity","NADH dehydrogenase activity","response to interferon-gamma","proton transmembrane transport","regulation of gamma-delta T cell activation","generation of precursor metabolites and energy","mitochondrial respiratory chain complex I","NADH dehydrogenase complex","respiratory chain complex I","protein-containing complex subunit organization","ncRNA processing","gamma-delta T cell differentiation","energy derivation by oxidation of organic compounds","cellular macromolecule metabolic process","mitochondrial electron transport, NADH to ubiquinone","cellular metabolic process","protein-containing complex assembly","oxidoreductase activity, acting on NAD(P)H, quinone or similar compound as acceptor","regulation of T cell mediated cytotoxicity","regulation of translation","endocytic vesicle","somatic diversification of T cell receptor genes","somatic recombination of T cell receptor gene segments","T cell receptor V(D)J recombination","protein binding","binding","antigen processing and presentation of endogenous peptide antigen","antigen processing and presentation of endogenous peptide antigen via MHC class I","mitochondrial protein complex","metabolic process","intracellular membrane-bounded organelle","postsynaptic density","gamma-delta T cell activation","asymmetric synapse","nucleus","negative regulation of leukocyte mediated cytotoxicity","ncRNA metabolic process","T cell mediated cytotoxicity","response to organic substance","biological regulation","positive regulation of gamma-delta T cell differentiation","cell killing","regulation of cellular amide metabolic process","postsynaptic specialization","synapse","neuron to neuron synapse","antigen processing and presentation of endogenous antigen","negative regulation of cell killing","antigen processing and presentation of endogenous peptide antigen via MHC class I via ER pathway","antigen processing and presentation of endogenous peptide antigen via MHC class I via ER pathway, TAP-independent","TAP binding","MHC protein complex","regulation of cell killing","positive regulation of gamma-delta T cell activation","positive regulation of T cell mediated cytotoxicity","cellular component organization or biogenesis","mitochondrial inner membrane","positive regulation of cell killing","mitochondrial envelope","receptor antagonist activity","negative regulation of protein binding","mitochondrial respiratory chain complex I assembly","NADH dehydrogenase complex assembly","oxidoreductase activity, acting on NAD(P)H","regulation of actin cytoskeleton organization","ribosome assembly","cytochrome complex","endocytic vesicle membrane","immune system process","cellular response to organic substance","positive regulation of cellular component movement","membrane protein complex","positive regulation of cellular amide metabolic process","negative regulation of binding","positive regulation of immune system process","cytokine-mediated signaling pathway","envelope","organelle envelope","positive regulation of leukocyte cell-cell adhesion","nitrogen compound metabolic process","regulation of T cell mediated immunity","maturation of SSU-rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)","mitochondrial membrane","organelle inner membrane","hemopoiesis","regulation of leukocyte mediated cytotoxicity","immune response","phagocytic vesicle membrane","regulation of apoptotic process","V(D)J recombination","cell activation","receptor inhibitor activity","purine ribonucleoside triphosphate metabolic process","negative regulation of lymphocyte mediated immunity","T cell activation","miRNA transport","regulation of programmed cell death","T cell differentiation","regulation of actin filament-based process","regulation of cell death","hematopoietic or lymphoid organ development","cadherin binding","rough endoplasmic reticulum","translational elongation","ribonucleoside triphosphate metabolic process","regulation of hemopoiesis","regulation of immune system process","positive regulation of translation","homotypic cell-cell adhesion","posttranscriptional regulation of gene expression","aerobic respiration","purine nucleoside triphosphate metabolic process","positive regulation of cell motility","macromolecule metabolic process","regulation of mRNA metabolic process","cellular response to chemical stimulus","positive regulation of locomotion","positive regulation of T cell mediated immunity","alpha-beta T cell activation","regulation of actin filament organization","respiratory chain complex IV","GAIT complex","positive regulation of cell-cell adhesion","maturation of SSU-rRNA","regulation of protein binding","immune system development","negative regulation of leukocyte mediated immunity","positive regulation of leukocyte mediated cytotoxicity","aerobic electron transport chain","mitochondrial electron transport, cytochrome c to oxygen","ribosomal small subunit assembly","leukocyte cell-cell adhesion","cellular component assembly","negative regulation of cell death","leukocyte differentiation","lymphocyte differentiation","ATP biosynthetic process","positive regulation of signal transduction by p53 class mediator","cellular process","postsynapse","mitochondrial respiratory chain complex assembly","structural constituent of cytoskeleton","T cell mediated immunity","large ribosomal subunit rRNA binding","positive regulation of intrinsic apoptotic signaling pathway by p53 class mediator","structural constituent of postsynaptic actin cytoskeleton","ubiquitin ligase inhibitor activity","negative regulation of execution phase of apoptosis","oxidoreductase activity","cytoplasmic ribonucleoprotein granule","ribonucleoprotein complex assembly","leukocyte mediated cytotoxicity","positive regulation of cell migration","nuclear lumen","antigen binding","lymphocyte activation","forebrain generation of neurons","nucleoside triphosphate metabolic process","ER to Golgi transport vesicle membrane","electron transfer activity","primary metabolic process","mRNA binding","ribonucleoprotein complex subunit organization","leukocyte activation","nucleoplasm","ribonucleoprotein granule","dense body","protection from natural killer cell mediated cytotoxicity","ubiquitin-protein transferase inhibitor activity","purine ribonucleoside triphosphate biosynthetic process","molecular_function","actin monomer binding","cytochrome-c oxidase activity","heme-copper terminal oxidase activity","oxidoreductase activity, acting on a heme group of donors, oxygen as acceptor","purine nucleoside triphosphate biosynthetic process","somatic cell DNA recombination","somatic diversification of immune receptors via germline recombination within a single locus","GTPase activity","regulation of leukocyte cell-cell adhesion","integral component of lumenal side of endoplasmic reticulum membrane","lumenal side of endoplasmic reticulum membrane","oxidoreductase activity, acting on a heme group of donors","response to chemical","regulation of T cell activation","organelle organization","ribosomal large subunit assembly","RNA processing","negative regulation of T cell mediated cytotoxicity","regulation of immune response","negative regulation of apoptotic process","ribonucleoside triphosphate biosynthetic process","regulation of cellular component movement","peptide antigen binding","positive regulation of cell adhesion","mitotic cytokinesis","regulation of actin polymerization or depolymerization","response to biotic stimulus","regulation of actin filament length","intracellular organelle lumen","membrane-enclosed lumen","organelle lumen","cellular response to interleukin-4","proton transmembrane transporter activity","actin filament organization","side of membrane","miRNA binding","negative regulation of programmed cell death","antigen processing and presentation of exogenous peptide antigen via MHC class I, TAP-dependent","hematopoietic stem cell migration","MHC class I peptide loading complex","T cell differentiation in thymus","regulation of cell motility","regulation of cell adhesion","somatic diversification of immune receptors","lamellipodium","response to interleukin-4","regulation of protein-containing complex assembly","positive regulation of gene expression","cell adhesion molecule binding","negative regulation of protein polymerization","positive regulation of cell death","positive regulation of hemopoiesis","response to stress","lumenal side of membrane","antigen processing and presentation of exogenous peptide antigen via MHC class I","negative regulation of ubiquitin protein ligase activity","lymphocyte proliferation","regulation of binding","regulation of cytoskeleton organization","nucleoside triphosphate biosynthetic process","mononuclear cell proliferation","regulation of leukocyte mediated immunity","regulation of supramolecular fiber organization","negative regulation of protein-containing complex assembly","DNA damage response, detection of DNA damage","regulation of leukocyte differentiation","regulation of locomotion","positive regulation of T cell activation","recycling endosome membrane","sequestering of actin monomers","small ribosomal subunit rRNA binding","response to hydrogen peroxide","regulation of cell migration","response to other organism","cell-cell adhesion","response to external biotic stimulus","biological_process","ubiquitin protein ligase binding","regulation of execution phase of apoptosis","actin polymerization or depolymerization","regulation of T cell differentiation","5S rRNA binding","profilin binding","apoptotic process","positive regulation of apoptotic process","oxidation-reduction process","interferon-gamma-mediated signaling pathway","regulation of stress fiber assembly","regulation of adaptive immune response based on somatic recombination of immune receptors built from immunoglobulin superfamily domains","regulation of lymphocyte mediated immunity","positive regulation of T cell differentiation","response to oxygen levels","regulation of transferase activity","positive regulation of programmed cell death","interaction with symbiont","regulation of RNA splicing","response to external stimulus","structural constituent of postsynapse","regulation of protein polymerization","leukocyte proliferation","actin filament-based process","COPII-coated ER to Golgi transport vesicle","regulatory RNA binding","positive regulation of leukocyte differentiation","ubiquitin-like protein ligase binding","organic substance metabolic process","embryonic morphogenesis","actin cytoskeleton organization","postsynaptic actin cytoskeleton organization","protein localization to cell-cell junction","protein tag","interleukin-12-mediated signaling pathway","antigen processing and presentation of peptide antigen via MHC class I","modulation of process of other organism involved in symbiotic interaction","cytoskeleton-dependent cytokinesis","supramolecular fiber organization","regulation of lymphocyte activation","immune effector process","regulation of adaptive immune response","cellular response to interleukin-12","regulation of actomyosin structure organization","cell leading edge","mitochondrial respiratory chain complex III","proton-transporting ATP synthase complex, coupling factor F(o)","respiratory chain complex III","response to interleukin-12","regulation of actin filament bundle assembly","positive regulation of adaptive immune response based on somatic recombination of immune receptors built from immunoglobulin superfamily domains","positive regulation of lymphocyte differentiation","forebrain neuron differentiation","negative regulation of molecular function","alpha-beta T cell differentiation","contractile actin filament bundle assembly","positive regulation of lymphocyte mediated immunity","stress fiber assembly","regulation of actin filament polymerization","positive regulation of ATP biosynthetic process","postsynaptic cytoskeleton organization","sequence-specific single stranded DNA binding","positive regulation of viral process","nuclear matrix","defense response to other organism","regulation of cell-cell adhesion","positive regulation of adaptive immune response","regulation of ubiquitin-protein transferase activity","cellular anatomical entity","negative regulation of natural killer cell mediated cytotoxicity","negative regulation of receptor binding","negative regulation of ubiquitin-protein transferase activity","regulation of lymphocyte differentiation","response to virus","Golgi-associated vesicle membrane","regulation of myeloid cell differentiation","coated vesicle membrane","modulation of process of other organism","regulation of mRNA splicing, via spliceosome","programmed cell death","protein kinase regulator activity","positive regulation of cellular process","cell-cell adhesion mediated by integrin","detection of bacterium","negative regulation of natural killer cell mediated immunity","regulation of mRNA stability","cell death","positive regulation of lymphocyte activation","protein stabilization","monovalent inorganic cation transport","negative regulation of signaling receptor activity","response to hypoxia","response to oxidative stress","structural constituent of synapse","ubiquitin-protein transferase regulator activity","negative regulation of cellular process","actin filament polymerization","phosphorylation","regulation of RNA stability","T cell proliferation","detection of other organism","negative regulation of T cell apoptotic process","negative regulation of T cell mediated immunity","proton-transporting ATP synthase activity, rotational mechanism","translation elongation factor activity","adaptive immune response based on somatic recombination of immune receptors built from immunoglobulin superfamily domains","response to ethanol","response to decreased oxygen levels","modulation by host of symbiont process","platelet aggregation","positive regulation of intrinsic apoptotic signaling pathway","protein-DNA complex subunit organization","negative regulation of immune effector process","GTP binding","positive regulation of T cell cytokine production","regulation of ATP biosynthetic process","90S preribosome assembly","antigen processing and presentation of endogenous peptide antigen via MHC class I via ER pathway, TAP-dependent","antigen processing and presentation of exogenous protein antigen via MHC class Ib, TAP-dependent","cellular response to cytochalasin B","chemokine receptor antagonist activity","clathrin-dependent extracellular exosome endocytosis","endonucleolytic cleavage to generate mature 3'-end of SSU-rRNA from (SSU-rRNA, 5.8S rRNA, LSU-rRNA)","eukaryotic 80S initiation complex","negative regulation of cytoplasmic translational initiation","negative regulation of formation of translation preinitiation complex","negative regulation of protein homodimerization activity","olfactory bulb axon guidance","oxidoreduction-driven active transmembrane transporter activity","peptidyl-cysteine S-nitrosylase activity","positive regulation by organism of apoptotic process in other organism involved in symbiotic interaction","positive regulation of apoptotic process in other organism","positive regulation of N-terminal peptidyl-lysine acetylation","positive regulation of norepinephrine uptake","positive regulation of selenocysteine incorporation","post-embryonic body morphogenesis","regulation of D-erythro-sphingosine kinase activity","regulation of formation of translation preinitiation complex","regulation of N-terminal peptidyl-lysine acetylation","regulation of protein homodimerization activity","response to cytochalasin B","septin collar","translation initiation complex","regulation of biological quality","purine ribonucleoside binding","killing of cells of other organism","negative regulation of actin filament polymerization","phagocytosis","purine nucleoside binding","regulation of cellular component size","ribonucleoside binding","error-prone translesion synthesis","hypothalamus development","positive regulation of alpha-beta T cell proliferation","positive regulation of cellular extravasation","regulation of hydrogen peroxide metabolic process","cell surface receptor signaling pathway","T cell migration","RNA splicing","animal organ development","positive regulation of cellular biosynthetic process","nuclear periphery","regulation of mRNA catabolic process","cell motility","localization of cell","nucleoside binding","endoplasmic reticulum","mRNA splicing, via spliceosome","RNA splicing, via transesterification reactions with bulged adenosine as nucleophile","error-free translesion synthesis","mitochondrial proton-transporting ATP synthase complex","nucleotide-excision repair, DNA duplex unwinding","response to hyperoxia","cation transport","cellular component organization","regulation of protein ubiquitination","RNA splicing, via transesterification reactions","enzyme regulator activity","protein polymerization","cell division","guanyl nucleotide binding","guanyl ribonucleotide binding","nucleotide-excision repair, DNA damage recognition","nucleotide-excision repair, DNA gap filling","positive regulation of leukocyte adhesion to vascular endothelial cell","proton-transporting ATP synthase complex","regulation of leukocyte activation","antimicrobial humoral response","kinase regulator activity","neutrophil activation","cell cortex","positive regulation of leukocyte mediated immunity","positive regulation of biosynthetic process","chromatin remodeling","Fc-gamma receptor signaling pathway involved in phagocytosis","immune response-regulating cell surface receptor signaling pathway involved in phagocytosis","translation regulator activity","calyx of Held","detection of external biotic stimulus","mitochondrial ATP synthesis coupled proton transport","proton-transporting two-sector ATPase complex, proton-transporting domain","regulation of ubiquitin protein ligase activity","enzyme binding","positive regulation of leukocyte activation","granulocyte activation","Fc-gamma receptor signaling pathway","positive regulation of leukocyte migration","positive regulation of lamellipodium assembly","positive regulation of immune effector process","antimicrobial humoral immune response mediated by antimicrobial peptide","movement of cell or subcellular component","Fc receptor mediated stimulatory signaling pathway","global genome nucleotide-excision repair","mRNA 5'-UTR binding","negative regulation of RNA splicing","positive regulation of nucleotide biosynthetic process","positive regulation of purine nucleotide biosynthetic process","proton channel activity","response to tumor cell","positive regulation of cell population proliferation","aging","mitochondrion","retina homeostasis","positive regulation of cell activation","regulation of symbiotic process","ATP synthesis coupled proton transport","CD8-positive, alpha-beta T cell activation","energy coupled proton transport, down electrochemical gradient","regulation of receptor binding","neuron projection","negative regulation of protein ubiquitination","cellular_component","cell population proliferation","response to stimulus","response to reactive oxygen species","ionotropic glutamate receptor binding","maturation of LSU-rRNA","regulation of mRNA processing","positive regulation of biological process","CCR4 chemokine receptor binding","cell-cell adhesion in response to extracellular stimulus","cell-matrix adhesion involved in ameboidal cell migration","cellular response to iron(III) ion","cellular response to tumor cell","DN2 thymocyte differentiation","ferritin complex","glyceraldehyde-3-phosphate dehydrogenase (NAD(P)+) (phosphorylating) activity","glyceraldehyde-3-phosphate dehydrogenase (NAD+) (phosphorylating) activity","integrin alpha4-beta7 complex","interleukin-7 receptor activity","intracellular ferritin complex","IRE1-RACK1-PP2A complex","lymphoid lineage cell migration","lymphoid lineage cell migration into thymus","maintenance of translational fidelity","negative regulation of chromosome condensation","peptidyl-cysteine S-trans-nitrosylation","plasma membrane raft distribution","plasma membrane raft localization","positive regulation of ferrous iron binding","positive regulation of hematopoietic stem cell migration","positive regulation of memory T cell activation","positive regulation of telomerase RNA reverse transcriptase activity","positive regulation of transferrin receptor binding","regulation of ferrous iron binding","regulation of hematopoietic stem cell migration","regulation of memory T cell activation","regulation of norepinephrine uptake","regulation of parathyroid hormone secretion","regulation of selenocysteine incorporation","regulation of transferrin receptor binding","TAP complex binding","trachea gland development","trachea submucosa development","leukocyte degranulation","myeloid cell differentiation","regulation of cell activation","protein-containing complex binding","negative regulation of protein tyrosine kinase activity","nucleotide-excision repair, preincision complex assembly","response to increased oxygen levels","negative regulation of supramolecular fiber organization","actin binding","actin filament bundle assembly","regulation of protein modification by small protein conjugation or removal","positive regulation of immune response","regulation of molecular function","leukocyte mediated immunity","cytoskeleton organization","beta-catenin-TCF complex assembly","regulation of cytoplasmic translation","regulation of ruffle assembly","TRIF-dependent toll-like receptor signaling pathway","integral component of endoplasmic reticulum membrane","plasma membrane bounded cell projection","negative regulation of immune response","inorganic cation transmembrane transport","myeloid cell activation involved in immune response","myeloid leukocyte activation","ephrin receptor signaling pathway","negative regulation of lymphocyte apoptotic process","negative regulation of stress fiber assembly","positive regulation of neutrophil migration","positive regulation of T cell migration","regulation of intrinsic apoptotic signaling pathway by p53 class mediator","actin filament bundle organization","early endosome membrane","myeloid leukocyte mediated immunity","negative regulation of mRNA metabolic process","mitochondrion organization","negative regulation of cytoskeleton organization","neural tube development","homeostatic process","regulation of T cell cytokine production","nucleolus","positive regulation of metabolic process","ATP-dependent chromatin remodeling","intrinsic component of endoplasmic reticulum membrane","regulation of intrinsic apoptotic signaling pathway","positive regulation of lamellipodium organization","regulation of alpha-beta T cell proliferation","regulation of cellular extravasation","regulation of T cell proliferation","negative regulation of protein modification by small protein conjugation or removal","response to inorganic substance","maintenance of blood-brain barrier","MyD88-independent toll-like receptor signaling pathway","negative regulation of actin filament bundle assembly","regulation of T cell apoptotic process","regulation of protein tyrosine kinase activity","supramolecular complex","response to oxygen-containing compound","tissue homeostasis","cytokinesis","cristae formation","cellular response to type I interferon","execution phase of apoptosis","type I interferon signaling pathway","purine ribonucleotide biosynthetic process","antigen processing and presentation of exogenous peptide antigen via MHC class Ib","canonical Wnt signaling pathway involved in negative regulation of apoptotic process","cellular response to actinomycin D","cellular response to redox state","early endosome lumen","embryonic genitalia morphogenesis","embryonic hindgut morphogenesis","integrin alpha4-beta1 complex","muscle cell projection","muscle cell projection membrane","negative regulation of estrogen receptor binding","negative regulation of forebrain neuron differentiation","negative regulation of protein neddylation","negative regulation of thrombin-activated receptor signaling pathway","parathyroid hormone secretion","positive regulation by organism of programmed cell death in other organism involved in symbiotic interaction","positive regulation by symbiont of host apoptotic process","positive regulation by symbiont of host programmed cell death","positive regulation of protein tyrosine phosphatase activity","positive regulation of proton-transporting ATP synthase activity, rotational mechanism","receptor signaling protein tyrosine kinase activator activity","regulation of estrogen receptor binding","regulation of proton-transporting ATP synthase activity, rotational mechanism","regulation of telomerase RNA reverse transcriptase activity","regulation of thrombin-activated receptor signaling pathway","Rho GDP-dissociation inhibitor activity","subtelomeric heterochromatin assembly","subtelomeric heterochromatin organization","T cell mediated cytotoxicity directed against tumor cell target","anatomical structure homeostasis","alpha-beta T cell proliferation","detection of biotic stimulus","MyD88-dependent toll-like receptor signaling pathway","positive T cell selection","B cell proliferation","lymphocyte mediated immunity","membrane organization","adaptive immune response","locomotion","antigen processing and presentation of exogenous peptide antigen","response to organic cyclic compound","nucleotide-binding oligomerization domain containing signaling pathway","nucleotide-excision repair, DNA incision, 5'-to lesion","positive regulation of viral genome replication","regulation of leukocyte adhesion to vascular endothelial cell","response to type I interferon","cell adhesion","positive regulation of production of molecular mediator of immune response","nucleotide-binding domain, leucine rich repeat containing receptor signaling pathway","positive regulation of lymphocyte migration","regulation of lamellipodium assembly","Golgi-associated vesicle","ameboidal-type cell migration","immune response-activating cell surface receptor signaling pathway","immune response-activating signal transduction","positive regulation of actin filament polymerization","positive regulation of myeloid cell differentiation","positive regulation of T cell proliferation","cell projection","biological adhesion","positive regulation of apoptotic signaling pathway","regulation of plasma membrane bounded cell projection assembly","humoral immune response","leukocyte activation involved in immune response","nucleotide-excision repair, DNA incision","neutrophil degranulation","regulation of cell population proliferation","antigen processing and presentation of exogenous antigen","regulation of cell projection assembly","cell activation involved in immune response","positive regulation of plasma membrane bounded cell projection assembly","ribonucleotide biosynthetic process","neutrophil activation involved in immune response","nuclear nucleosome","nucleosomal DNA binding","rDNA heterochromatin assembly","virion assembly","enzyme inhibitor activity","cytoplasmic side of membrane","chromatin DNA binding","cortical cytoskeleton","neural tube formation","actin filament fragmentation","ascending aorta morphogenesis","CD8 receptor binding","corticotropin-releasing hormone receptor 1 binding","eukaryotic translation elongation factor 1 complex","formation of radial glial scaffolds","hypothalamus gonadotrophin-releasing hormone neuron development","hypothalamus gonadotrophin-releasing hormone neuron differentiation","immunoglobulin biosynthetic process","killing by symbiont of host cells","memory T cell activation","mu-type opioid receptor binding","N-terminal peptidyl-lysine acetylation","negative regulation of cell adhesion involved in substrate-bound cell migration","negative regulation of nuclear-transcribed mRNA catabolic process, nonsense-mediated decay","pericentric heterochromatin assembly","pericentric heterochromatin organization","phagocytic vesicle lumen","positive regulation of cardiac muscle myoblast proliferation","positive regulation of CD8-positive, alpha-beta T cell proliferation","positive regulation of Golgi to plasma membrane protein transport","protein tyrosine kinase inhibitor activity","regulation of nuclear-transcribed mRNA catabolic process, nonsense-mediated decay","regulation of protein neddylation","response to actinomycin D","response to cobalamin","ribosomal protein import into nucleus","striatal medium spiny neuron differentiation","defense response to fungus","regulation of natural killer cell mediated cytotoxicity","response to abiotic stimulus","recycling endosome","cytoskeletal protein binding","regulation of cellular component biogenesis","cell migration","cellular response to interleukin-7","regulation of neutrophil migration","regulation of purine nucleotide biosynthetic process","response to interleukin-7","translesion synthesis","central nervous system neuron differentiation","neutrophil mediated immunity","limbic system development","translation regulator activity, nucleic acid binding","antigen processing and presentation of peptide antigen","ribose phosphate biosynthetic process","cysteine-type endopeptidase regulator activity involved in apoptotic process","negative regulation of adaptive immune response based on somatic recombination of immune receptors built from immunoglobulin superfamily domains","regulation of nucleotide biosynthetic process","response to copper ion","T cell cytokine production","positive regulation of macromolecule metabolic process","actin cytoskeleton","negative regulation of cellular component organization","inorganic ion transmembrane transport","response to drug","innate immune response","regulation of T cell migration","ruffle assembly","purine nucleotide biosynthetic process","regulation of cellular catabolic process","purine ribonucleotide metabolic process","regulation of protein stability","regulation of natural killer cell mediated immunity","response to nicotine","immune response-regulating cell surface receptor signaling pathway","single-stranded DNA binding","actomyosin structure organization","cell morphogenesis involved in differentiation","coated vesicle","tissue development","regulation of catalytic activity","DNA-binding transcription activator activity, RNA polymerase II-specific","DNA recombination","immune response-regulating signaling pathway","defense response","glutamate receptor binding","adenylate cyclase activator activity","apoptotic process involved in blood vessel morphogenesis","ascending aorta development","atrial septum primum morphogenesis","beta-2 adrenergic receptor binding","C-X3-C chemokine binding","CCR1 chemokine receptor binding","CRD-mediated mRNA stabilization","cytolytic granule","cytoplasmic side of lysosomal membrane","cytoplasmic side of rough endoplasmic reticulum membrane","estrogen receptor activity","lymphocyte aggregation","lymphocyte anergy","negative regulation of guanyl-nucleotide exchange factor activity","negative regulation of hydrogen peroxide-induced neuron death","neuron death in response to hydrogen peroxide","positive regulation of cyclic-nucleotide phosphodiesterase activity","positive regulation of Fc-gamma receptor signaling pathway involved in phagocytosis","positive regulation of humoral immune response mediated by circulating immunoglobulin","positive regulation of hydrogen peroxide biosynthetic process","positive regulation of protein monoubiquitination","positive regulation of translational elongation","postsynaptic actin cytoskeleton","regulation of chromosome condensation","regulation of cytoplasmic translational initiation","regulation of forebrain neuron differentiation","regulation of hydrogen peroxide-induced neuron death","regulation of lymphocyte anergy","regulation of T cell anergy","rescue of stalled ribosome","response to iron(III) ion","RNA-dependent ATPase activity","RS domain binding","selenocysteine insertion sequence binding","T cell anergy","DNA-binding transcription activator activity","multicellular organismal homeostasis","Wnt signaling pathway","DNA geometric change","regulation of anatomical structure size","brain development","cell-cell signaling by wnt","lymphocyte migration","cytoplasmic vesicle","chromatin remodeling at centromere","actin filament","cation transmembrane transport","intracellular vesicle","positive regulation of macromolecule biosynthetic process","chordate embryonic development","negative regulation of adaptive immune response","positive regulation of supramolecular fiber organization","purine-containing compound biosynthetic process","ribonucleotide metabolic process","regulation of gene expression, epigenetic","transport vesicle membrane","erythrocyte differentiation","body morphogenesis","negative regulation of leukocyte apoptotic process","negative regulation of ubiquitin-dependent protein catabolic process","T cell apoptotic process","T cell selection","response to interleukin-1","neutrophil migration","regulation of leukocyte migration","glial cell proliferation","positive regulation of ATP metabolic process","positive regulation of nucleotide metabolic process","positive regulation of purine nucleotide metabolic process","regulation of erythrocyte differentiation","regulation of lamellipodium organization","regulation of protein metabolic process","ficolin-1-rich granule","ficolin-1-rich granule lumen","activation of phospholipase D activity","axonogenesis involved in innervation","Barr body","BH3 domain binding","corticotropin-releasing hormone receptor binding","CRD-mediated mRNA stability complex","diapedesis","forebrain neuroblast division","histone pre-mRNA 3'end processing complex","intracellular sequestering of iron ion","keratin filament binding","modulation by organism of apoptotic process in other organism involved in symbiotic interaction","modulation by symbiont of host apoptotic process","modulation of apoptotic process in other organism","modulation of programmed cell death in other organism","modulation of programmed cell death in other organism involved in symbiotic interaction","negative regulation of apoptotic process in bone marrow cell","negative regulation of interleukin-13 production","negative regulation of macrophage apoptotic process","negative regulation of RNA polymerase II regulatory region sequence-specific DNA binding","negative regulation of ruffle assembly","positive regulation of cell-cell adhesion mediated by integrin","positive regulation of natural killer cell chemotaxis","protein localization to adherens junction","protein localization to bicellular tight junction","regulation of centromere complex assembly","regulation of protein monoubiquitination","regulation of protein tyrosine phosphatase activity","septum primum development","sperm annulus","T cell mediated immune response to tumor cell","cell morphogenesis","mRNA stabilization","cytoplasmic vesicle membrane","central nervous system development","cell surface","embryonic epithelial tube formation","regulation of cellular protein metabolic process","maintenance of location in cell","exocytosis","ribose phosphate metabolic process","embryo development ending in birth or egg hatching","regulation of kinase activity","regulation of cellular macromolecule biosynthetic process","regulation of viral process","inner mitochondrial membrane organization","postreplication repair","proton-transporting two-sector ATPase complex","regulation of granulocyte chemotaxis","viral genome replication","neuron projection morphogenesis","protein kinase binding","regulation of lymphocyte proliferation","negative regulation of transcription by RNA polymerase II","actin nucleation","DNA replication-independent nucleosome assembly","DNA synthesis involved in DNA repair","leukocyte adhesion to vascular endothelial cell","negative regulation of DNA binding","negative regulation of G protein-coupled receptor signaling pathway","regulation of cellular senescence","establishment or maintenance of cell polarity","negative regulation of translation","negative regulation of cellular macromolecule biosynthetic process","erythrocyte homeostasis","regulation of mononuclear cell proliferation","regulated exocytosis","negative regulation of RNA metabolic process","DNA replication-independent nucleosome organization","response to fungus","axon terminus","cellular component morphogenesis","mRNA processing","nucleoside-triphosphatase activity","head development","B cell activation","interstrand cross-link repair","positive regulation of cytokine production involved in immune response","tertiary granule lumen","actin polymerization-dependent cell motility","antigen processing and presentation of peptide antigen via MHC class Ib","aspartic-type endopeptidase inhibitor activity","axial mesoderm development","cardiac muscle myoblast proliferation","CCR5 chemokine receptor binding","chorio-allantoic fusion","commitment of neuronal cell to specific neuron type in forebrain","fat pad development","forebrain radial glial cell differentiation","G-quadruplex DNA unwinding","hindgut morphogenesis","membrane raft distribution","N6-methyladenosine-containing RNA binding","negative regulation of chemokine-mediated signaling pathway","negative regulation of interleukin-5 production","neural fold formation","noradrenergic neuron differentiation","paraxial mesoderm formation","PML body organization","positive regulation of activation of Janus kinase activity","positive regulation of Fc receptor mediated stimulatory signaling pathway","positive regulation of hydrogen peroxide metabolic process","positive regulation of neutrophil extravasation","post-embryonic camera-type eye development","protein antigen binding","protein localization to cytoplasmic stress granule","regulation of apoptotic process in bone marrow cell","regulation of cardiac muscle myoblast proliferation","regulation of Fc-gamma receptor signaling pathway involved in phagocytosis","regulation of mast cell chemotaxis","sequestering of iron ion","telomere assembly","ubiquinone binding","embryo development","regulation of protein kinase activity","regulation of proteolysis involved in cellular protein catabolic process","vesicle membrane","plasma membrane bounded cell projection morphogenesis","purine nucleotide metabolic process","epithelial tube formation","ribonucleoprotein complex binding","activation of protein kinase activity","regulation of nitrogen compound metabolic process","centromere complex assembly","negative regulation of peptidyl-tyrosine phosphorylation","regulation of lymphocyte apoptotic process","ruffle organization","cell projection morphogenesis","hydrogen peroxide metabolic process","intracellular transport of virus","ribosome binding","positive regulation of cytoskeleton organization","positive regulation of lymphocyte proliferation","DNA metabolic process","embryo implantation","negative regulation of mRNA catabolic process","positive regulation of myeloid leukocyte differentiation","RNA stabilization","positive regulation of mononuclear cell proliferation","positive regulation of protein polymerization","antigen processing and presentation","activation of immune response","plasma membrane protein complex","response to alcohol","histone exchange","protein homotetramerization","apoptotic process in bone marrow cell","Bergmann glial cell differentiation","beta-2-microglobulin binding","CD8-positive, alpha-beta T cell proliferation","cytoplasmic sequestering of NF-kappaB","female meiosis I","HFE-transferrin receptor complex","hindgut development","hypothalamus cell differentiation","membrane raft localization","negative regulation of cytoplasmic translation","negative regulation of interleukin-4 production","negative regulation of microglial cell activation","norepinephrine uptake","positive regulation of CD8-positive, alpha-beta T cell activation","positive regulation of cell proliferation in bone marrow","positive regulation of granulocyte differentiation","positive regulation of leukocyte tethering or rolling","postsynaptic cytoskeleton","regulation of CD8-positive, alpha-beta T cell proliferation","regulation of chaperone-mediated autophagy","regulation of guanyl-nucleotide exchange factor activity","regulation of iron ion transport","regulation of neutrophil extravasation","viral life cycle","Rho protein signal transduction","response to estradiol","catalytic complex","negative regulation of cellular amide metabolic process","natural killer cell mediated cytotoxicity","positive regulation of viral life cycle","regulation of cell aging","cell part morphogenesis","regulation of cell differentiation","regulation of macromolecule biosynthetic process","negative regulation of macromolecule biosynthetic process","regulation of cellular biosynthetic process","regulation of leukocyte proliferation","response to endogenous stimulus","signaling receptor binding","Fc receptor signaling pathway","heterotypic cell-cell adhesion","negative regulation of innate immune response","positive regulation of actin filament bundle assembly","presynaptic endocytosis","synaptic vesicle endocytosis","positive regulation of chemotaxis","regulation of production of molecular mediator of immune response","morphogenesis of a polarized epithelium","protein-DNA complex assembly","positive regulation of alpha-beta T cell activation","bleb","bone marrow development","cellular response to muramyl dipeptide","cellular response to nicotine","D1 dopamine receptor binding","ferroxidase activity","modulation by symbiont of host programmed cell death","modulation of age-related behavioral decline","negative regulation of trophoblast cell migration","neutrophil differentiation","opioid receptor binding","oxidoreductase activity, acting on diphenols and related substances as donors, cytochrome as acceptor","oxidoreductase activity, oxidizing metal ions, oxygen as acceptor","positive regulation of homotypic cell-cell adhesion","positive regulation of receptor binding","regulation of activation of Janus kinase activity","regulation of Golgi to plasma membrane protein transport","regulation of hydrogen peroxide biosynthetic process","regulation of natural killer cell chemotaxis","response to folic acid","selenocysteine incorporation","T cell receptor binding","translational readthrough","ubiquinol-cytochrome-c reductase activity"],["CC","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","BP","MF","BP","CC","BP","BP","BP","BP","BP","CC","BP","BP","BP","CC","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","MF","CC","CC","BP","CC","CC","BP","BP","CC","BP","MF","BP","BP","CC","CC","BP","CC","CC","BP","CC","BP","BP","BP","BP","CC","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","BP","BP","BP","CC","BP","BP","BP","CC","BP","BP","MF","BP","CC","CC","CC","CC","BP","CC","MF","BP","BP","BP","BP","BP","CC","MF","BP","MF","BP","BP","CC","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","CC","BP","BP","CC","BP","BP","BP","BP","BP","CC","BP","BP","BP","CC","BP","BP","CC","CC","CC","BP","BP","BP","BP","CC","BP","BP","CC","BP","BP","BP","BP","CC","BP","MF","MF","MF","BP","BP","BP","BP","CC","CC","CC","BP","BP","BP","BP","BP","BP","BP","BP","MF","BP","BP","CC","BP","BP","BP","MF","MF","BP","BP","CC","BP","CC","CC","BP","CC","CC","BP","BP","BP","BP","BP","BP","BP","BP","CC","CC","CC","BP","BP","BP","BP","MF","CC","BP","BP","BP","BP","CC","BP","CC","MF","BP","BP","BP","MF","BP","BP","CC","CC","BP","BP","BP","CC","BP","BP","BP","BP","CC","CC","BP","BP","BP","BP","CC","CC","BP","BP","BP","CC","BP","BP","BP","MF","BP","BP","BP","BP","BP","BP","BP","BP","BP","MF","CC","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","CC","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","BP","MF","BP","MF","BP","MF","MF","BP","MF","CC","BP","BP","BP","CC","MF","BP","BP","BP","CC","MF","BP","MF","BP","BP","CC","CC","CC","BP","MF","BP","MF","MF","MF","MF","MF","BP","BP","BP","MF","BP","CC","CC","MF","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","MF","BP","BP","BP","BP","BP","CC","CC","CC","BP","MF","BP","CC","MF","BP","BP","BP","CC","BP","BP","BP","BP","CC","BP","BP","BP","MF","BP","BP","BP","BP","CC","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","BP","MF","BP","BP","BP","BP","BP","BP","MF","BP","BP","BP","MF","MF","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","MF","BP","BP","BP","CC","MF","BP","MF","BP","BP","BP","BP","BP","MF","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","CC","CC","CC","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","MF","BP","CC","BP","BP","BP","BP","CC","BP","BP","BP","BP","BP","CC","BP","CC","BP","BP","BP","MF","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","MF","MF","BP","BP","BP","BP","BP","BP","BP","BP","MF","MF","BP","BP","BP","BP","BP","BP","BP","BP","MF","BP","BP","BP","BP","BP","BP","MF","BP","BP","CC","BP","BP","BP","BP","MF","MF","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","CC","BP","MF","BP","BP","BP","MF","BP","MF","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","BP","BP","BP","MF","CC","BP","BP","BP","CC","BP","BP","BP","BP","BP","BP","MF","BP","BP","MF","MF","BP","BP","BP","CC","BP","BP","MF","BP","CC","BP","BP","BP","BP","BP","MF","CC","BP","BP","CC","BP","MF","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","MF","BP","BP","BP","MF","BP","BP","BP","CC","BP","BP","BP","BP","BP","BP","BP","CC","BP","CC","BP","BP","BP","MF","BP","BP","BP","MF","BP","BP","BP","BP","BP","CC","MF","MF","CC","MF","CC","CC","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","MF","BP","BP","BP","BP","BP","MF","BP","BP","BP","BP","MF","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","CC","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","BP","BP","BP","BP","BP","BP","BP","CC","BP","BP","CC","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","BP","BP","CC","CC","CC","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","MF","BP","BP","BP","BP","MF","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","BP","BP","BP","BP","BP","BP","CC","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","MF","BP","BP","MF","CC","MF","CC","BP","BP","BP","MF","MF","CC","BP","BP","BP","BP","BP","BP","MF","BP","BP","BP","BP","BP","CC","BP","BP","BP","MF","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","MF","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","MF","BP","BP","MF","BP","BP","BP","BP","BP","CC","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","MF","BP","BP","CC","BP","BP","MF","BP","BP","BP","MF","MF","BP","BP","BP","MF","MF","MF","BP","CC","CC","CC","MF","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","BP","BP","BP","BP","BP","BP","BP","BP","MF","MF","MF","BP","MF","BP","BP","BP","BP","BP","BP","BP","CC","BP","CC","BP","CC","BP","BP","BP","BP","BP","BP","BP","CC","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","CC","BP","BP","CC","MF","MF","CC","BP","BP","CC","BP","MF","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","BP","BP","BP","CC","BP","CC","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","BP","BP","BP","MF","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","BP","BP","MF","BP","BP","BP","BP","CC","BP","BP","MF","BP","BP","MF","BP","BP","BP","BP","BP","BP","BP","MF","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","MF","BP","BP","BP","BP","BP","BP","BP","MF","BP","BP","BP","CC","BP","BP","BP","MF","BP","BP","BP","BP","BP","BP","BP","BP","BP","MF","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","BP","BP","BP","BP","BP","MF","BP","BP","BP","CC","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","BP","BP","BP","BP","BP","BP","BP","BP","CC","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","MF","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","CC","BP","BP","BP","MF","MF","BP","BP","BP","BP","MF","MF","MF","BP","BP","BP","BP","BP","BP","BP","BP","MF","BP","MF"],[110,105,120,109,120,124,178,152,192,195,187,210,202,210,270,376,415,924,983,342,744,444,738,763,58,555,602,603,620,585,890,653,906,657,909,730,415,423,2250,116,47,1194,982,74,1203,1988,1829,1183,2167,2289,1433,2291,831,1984,5501,1993,103,2795,1762,32,2766,3493,1985,2788,6865,2026,2061,2080,2370,69,5224,5425,6078,2093,5352,3173,2222,3616,4960,2316,4303,3578,4073,5775,5784,5213,2800,4665,64,3818,5091,4963,2683,7012,11620,6213,6921,6299,6225,7553,85,149,6204,7016,6399,6104,5213,6444,6290,5709,97,98,310,101,13340,6678,116,15166,482,311,5350,5,5,138,227,184,238,14482,1227,1136,13330,89,9562,5894,3257,1134,182,140,187,9,9,9,6846,12364,74,110,75,46,46,46,202,159,11,534,50,50,50,1984,400,12,278,8408,55,11384,1706,61,35,450,313,5,5,5,13697,16428,17,17,265,12706,11502,337,19,343,8076,20,493,44,3385,13030,7,172,509,361,1357,368,24,24,8,8,8,25,92,9,28,6717,489,61,770,32,106,65,65,109,360,67,34,164,3287,2756,611,1310,167,169,1110,808,1219,1219,235,10609,73,38,727,549,926,75,2306,77,1592,15,1482,41,79,42,483,3,1617,253,405,1748,967,332,84,135,85,498,1646,136,86,1199,87,87,593,10650,344,3371,607,50,146,278,20,4,279,51,211,1024,52,52,21,21,21,364,3012,1035,538,368,54,22,17917,640,102,104,104,5,5,5,5,24,751,233,233,108,567,5457,168,765,61,110,62,111,11109,577,240,1321,3978,243,6,6,6,65,18352,28,28,28,28,66,66,66,326,329,29,29,29,4785,332,4048,30,1484,7,1027,919,71,1143,31,428,72,190,1625,191,6574,6574,6574,32,127,434,625,33,940,75,8,8,76,1058,736,77,201,35,449,2313,544,79,746,204,4162,36,80,9,283,367,554,82,286,211,373,142,38,290,1097,214,85,10,10,146,994,1591,884,1593,18866,297,41,221,150,11,11,1996,684,1014,91,91,153,154,92,396,1022,693,93,156,2976,12,231,313,807,95,45,159,316,12140,602,707,13,13,13,47,99,99,100,718,515,1299,168,49,102,421,14,14,14,50,104,105,105,51,1197,106,106,106,106,174,15,15,15,107,109,1213,439,110,54,19352,16,16,16,181,349,112,263,183,113,113,2139,185,5722,17,17,17,186,2281,357,187,554,58,359,458,18,18,5131,193,2455,195,195,19,19,19,19,19,370,122,371,62,62,62,282,124,376,20,20,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,4102,380,65,65,382,383,383,383,21,21,21,21,21,3219,67,487,3674,1955,131,208,1823,1823,389,1430,390,390,22,22,22,22,1181,6526,211,393,1063,300,603,397,397,23,23,23,23,605,137,216,502,305,138,1991,218,139,139,139,24,24,24,24,24,2277,406,509,142,142,25,223,75,2298,145,26,26,26,26,26,26,26,979,319,1622,78,421,231,27,27,27,27,1367,80,19559,2047,9398,235,28,28,152,6346,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,537,431,650,1258,29,29,29,156,437,157,242,772,3084,892,1399,30,30,30,30,158,2221,159,777,549,662,87,31,31,31,31,31,161,161,555,88,556,163,163,1963,32,1423,3880,90,166,166,33,33,33,167,92,569,34,34,34,34,93,1312,1715,261,172,35,95,95,95,173,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,469,36,36,36,36,97,366,944,701,2016,177,946,37,37,37,37,99,1474,100,38,38,38,181,481,481,481,101,101,101,2321,1481,183,183,377,716,39,487,1766,185,185,720,104,186,490,40,40,40,40,383,188,106,106,106,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,41,41,1239,190,980,982,1650,42,42,42,42,42,192,501,109,109,194,194,43,43,43,43,43,3574,506,744,868,397,996,44,44,197,874,401,296,45,45,514,114,200,753,298,2107,2398,516,299,517,1824,46,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,519,519,519,116,520,760,521,117,2413,47,118,890,2417,1841,645,48,208,208,416,209,209,121,49,49,49,49,49,211,122,212,50,50,50,50,50,50,2891,124,124,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,1041,51,784,1042,912,125,2754,216,914,427,663,915,3991,217,52,52,52,52,127,667,667,219,922,53,53,53,53,53,53,53,220,220,1602,129,221,797,1465,54,54,130,800,554,802,803,328,55,55,55,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,1066,805,225,806,681,442,133,133,331,5980,56,56,56,56,685,57,57,57,230,136,948,58,58,58,58,137,137,232,570,694,234,59,59,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,341,140,141,1372,238,61,61,61,705,1947,4094,1659,4255,240,1662,1662,241,62,62,62,62,62,144,144,145,243,63,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9],[40,39,40,39,39,39,41,39,41,41,40,40,39,39,41,43,43,55,56,40,49,41,48,48,23,43,43,43,43,42,48,43,48,43,48,44,36,36,63,23,18,48,43,18,44,54,52,43,55,55,45,55,36,51,81,51,17,57,46,12,56,62,47,55,85,47,47,47,49,13,72,73,77,45,72,55,46,58,68,46,63,57,60,72,72,68,49,64,11,57,66,65,46,77,101,71,75,71,70,78,10,12,69,74,70,68,62,70,69,65,10,10,15,10,105,70,10,111,17,14,60,4,4,10,12,11,12,107,25,24,102,8,83,61,42,23,10,9,10,4,4,4,66,96,7,8,7,6,6,6,10,9,4,15,6,6,6,30,13,4,11,74,6,90,27,6,5,13,11,3,3,3,100,111,4,4,10,95,89,11,4,11,70,4,13,5,39,96,3,8,13,11,22,11,4,4,3,3,3,4,6,3,4,60,12,5,15,4,6,5,5,6,10,5,4,7,36,32,13,20,7,7,18,15,19,19,8,81,5,4,14,12,16,5,28,5,22,3,21,4,5,4,11,2,22,8,10,23,16,9,5,6,5,11,22,6,5,18,5,5,12,80,9,35,12,4,6,8,3,2,8,4,7,16,4,4,3,3,3,9,32,16,11,9,4,3,113,12,5,5,5,2,2,2,2,3,13,7,7,5,11,48,6,13,4,5,4,5,81,11,7,18,38,7,2,2,2,4,114,3,3,3,3,4,4,4,8,8,3,3,3,43,8,38,3,19,2,15,14,4,16,3,9,4,6,20,6,54,54,54,3,5,9,11,3,14,4,2,2,4,15,12,4,6,3,9,25,10,4,12,6,38,3,4,2,7,8,10,4,7,6,8,5,3,7,15,6,4,2,2,5,14,19,13,19,115,7,3,6,5,2,2,22,11,14,4,4,5,5,4,8,14,11,4,5,29,2,6,7,12,4,3,5,7,84,10,11,2,2,2,3,4,4,4,11,9,16,5,3,4,8,2,2,2,3,4,4,4,3,15,4,4,4,4,5,2,2,2,4,4,15,8,4,3,116,2,2,2,5,7,4,6,5,4,4,22,5,46,2,2,2,5,23,7,5,9,3,7,8,2,2,42,5,24,5,5,2,2,2,2,2,7,4,7,3,3,3,6,4,7,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,35,7,3,3,7,7,7,7,2,2,2,2,2,29,3,8,32,20,4,5,19,19,7,16,7,7,2,2,2,2,14,50,5,7,13,6,9,7,7,2,2,2,2,9,4,5,8,6,4,20,5,4,4,4,2,2,2,2,2,22,7,8,4,4,2,5,3,22,4,2,2,2,2,2,2,2,12,6,17,3,7,5,2,2,2,2,15,3,116,20,66,5,2,2,4,48,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,8,7,9,14,2,2,2,4,7,4,5,10,27,11,15,2,2,2,2,4,21,4,10,8,9,3,2,2,2,2,2,4,4,8,3,8,4,4,19,2,15,32,3,4,4,2,2,2,4,3,8,2,2,2,2,3,14,17,5,4,2,3,3,3,4,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,7,2,2,2,2,3,6,11,9,19,4,11,2,2,2,2,3,15,3,2,2,2,4,7,7,7,3,3,3,21,15,4,4,6,9,2,7,17,4,4,9,3,4,7,2,2,2,2,6,4,3,3,3,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,13,4,11,11,16,2,2,2,2,2,4,7,3,3,4,4,2,2,2,2,2,29,7,9,10,6,11,2,2,4,10,6,5,2,2,7,3,4,9,5,19,21,7,5,7,17,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,7,7,7,3,7,9,7,3,21,2,3,10,21,17,8,2,4,4,6,4,4,3,2,2,2,2,2,4,3,4,2,2,2,2,2,2,24,3,3,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,11,2,9,11,10,3,23,4,10,6,8,10,31,4,2,2,2,2,3,8,8,4,10,2,2,2,2,2,2,2,4,4,15,3,4,9,14,2,2,3,9,7,9,9,5,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,11,9,4,9,8,6,3,3,5,43,2,2,2,2,8,2,2,2,4,3,10,2,2,2,2,3,3,4,7,8,4,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,5,3,3,13,4,2,2,2,8,17,31,15,32,4,15,15,4,2,2,2,2,2,3,3,3,4,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1],[8.33797442522071e-64,1.33865377633415e-62,5.66675360301809e-62,8.00293386219689e-62,7.35575373575501e-60,3.35796821501479e-59,5.60025199597941e-56,3.32759812273066e-55,1.7610754823685e-54,3.55135615939244e-54,3.9377712895357e-53,6.35662702003484e-51,7.51276563303277e-50,3.89259499176044e-49,6.36807261244907e-48,7.49341163071413e-45,5.79372320239897e-43,8.33022199505419e-43,1.20583349860326e-42,5.73222074483667e-42,9.63464399336405e-40,8.86344091106651e-39,1.33729317672019e-38,6.40642374218872e-38,1.52149365666171e-37,1.68822453205497e-37,5.37172892854627e-36,5.76341567101431e-36,1.87065300966478e-35,3.37292984513706e-35,8.39194797564481e-35,1.66804275435416e-34,1.91274726169118e-34,2.15677542275589e-34,2.22828670348814e-34,1.02838863035828e-33,7.0086749382107e-33,1.39273591124837e-32,2.0168398299596e-30,1.52189862604556e-29,3.54848032969929e-29,5.49888573325676e-29,3.39589870354277e-27,4.9911891985149e-25,1.10588081333677e-24,1.18091169719017e-24,1.61022501161991e-24,5.71212121166739e-24,9.51140252903296e-24,1.34429114068419e-22,1.39164051769679e-22,1.40197561916222e-22,2.17483098288559e-22,5.29784950962174e-22,5.89167022618548e-22,6.4885368214228e-22,1.41955502991185e-20,4.54102814090033e-20,7.34081874774627e-20,1.22872118354696e-19,1.69823179444821e-19,4.27273604018884e-19,1.29712932939442e-18,1.46675556866091e-18,2.42545447839701e-18,2.95205054881513e-18,5.86290303764291e-18,8.4586814486177e-18,4.67323828960751e-17,9.02490426969031e-17,1.01478644466543e-16,1.89032247111123e-16,3.37350746496213e-16,3.80924750560336e-16,4.06911388941174e-16,5.16299027733237e-16,6.52092848859898e-16,1.6793533737932e-15,2.65392840488814e-15,3.14250057073651e-15,3.27833130507283e-15,4.86228689016522e-15,2.0553067748804e-14,2.94451002556207e-14,3.20982590945525e-14,3.67886060182376e-14,3.69862001582075e-14,4.08233960723493e-14,6.75656086449359e-14,8.85849199080492e-14,1.85195999613333e-13,2.10464849260271e-13,7.21164789197883e-13,1.64011607921021e-12,2.02522293284931e-12,5.90238335126393e-12,1.08484532809994e-11,1.20576409398402e-11,2.34103442101371e-11,3.08835215031715e-11,4.95310090523204e-11,4.97868780346932e-11,6.86382058869944e-11,7.93693822702504e-11,9.36506899932511e-11,1.0545676727173e-10,1.08082867785958e-10,1.32748456849133e-10,1.35058491192564e-10,1.70159800823279e-10,1.88118533201176e-10,2.0851002535429e-10,2.45089847276202e-10,2.82010603076184e-10,3.39660818348562e-10,7.65957229494073e-10,1.11558614708327e-09,1.28901719520671e-09,1.85595016281771e-09,2.56950099637963e-09,3.4532899130208e-09,4.83315494990284e-09,4.83315494990284e-09,6.09458687021137e-09,6.31964685451349e-09,7.80116507980793e-09,1.07356221330922e-08,1.62842299457964e-08,1.7592494372184e-08,1.80146505826886e-08,2.42394960340557e-08,3.95666380974856e-08,4.95937991641585e-08,5.87317756734093e-08,7.53097390716371e-08,7.90153684630164e-08,8.5591930334422e-08,1.02023718322377e-07,1.10374305094523e-07,1.19681336935433e-07,1.19681336935433e-07,1.19681336935433e-07,1.7158478314249e-07,1.80789789797856e-07,2.02906731877898e-07,2.08012448890603e-07,2.22755198915509e-07,2.23685537968242e-07,2.23685537968242e-07,2.23685537968242e-07,2.26603825363454e-07,3.02806697505267e-07,3.1071988821017e-07,3.46537993431369e-07,3.72562467605309e-07,3.72562467605309e-07,3.72562467605309e-07,4.01359318325101e-07,4.36005928360271e-07,4.64045253464232e-07,5.20436580573303e-07,6.27858753286737e-07,6.642245546125e-07,6.73339852828499e-07,7.37290633866621e-07,1.23761501631642e-06,1.5034028321286e-06,1.62815769988536e-06,1.66326832472444e-06,1.74743360737965e-06,1.74743360737965e-06,1.74743360737965e-06,1.97645945508448e-06,2.13685400298235e-06,2.1829125563871e-06,2.1829125563871e-06,2.68441547092596e-06,3.04058662500734e-06,3.25021839822111e-06,3.39046158643896e-06,3.52408830446647e-06,4.01357056148124e-06,4.08847717839255e-06,4.38590304580575e-06,4.4211098948908e-06,4.82944285872201e-06,4.9363774606142e-06,5.25002524049905e-06,6.06560158112072e-06,6.11604904238031e-06,6.24104606423843e-06,6.5240307590175e-06,6.54145709330782e-06,7.81979105194497e-06,9.45250174814865e-06,9.45250174814865e-06,9.66489376311811e-06,9.66489376311811e-06,9.66489376311811e-06,1.12039485301993e-05,1.3785139386731e-05,1.44374999762094e-05,1.78985168351828e-05,1.83871857579486e-05,2.15049020895592e-05,2.45062754103008e-05,2.94016362503074e-05,3.08911084586426e-05,3.09284470081657e-05,3.34182077406807e-05,3.34182077406807e-05,3.62128813656894e-05,3.84424357639617e-05,3.87235087629004e-05,3.94929093192619e-05,4.12733593163547e-05,4.13224712254021e-05,4.20071732667536e-05,4.25173230064501e-05,4.31225085814968e-05,4.63194512281145e-05,4.99564763618868e-05,5.05705088955342e-05,5.09532649148213e-05,5.21114253343939e-05,5.21114253343939e-05,5.78488078297936e-05,5.81086991395485e-05,5.86233339769452e-05,6.17728916522556e-05,6.25593411169096e-05,6.5933026594597e-05,6.59711561137833e-05,6.67596066548125e-05,6.7379215472824e-05,7.57400204212543e-05,7.57587648381596e-05,7.62878033868646e-05,7.9464352994184e-05,8.36484135518124e-05,8.56232057860932e-05,9.20517052318344e-05,9.38089797668375e-05,9.47477467964435e-05,9.52237061554831e-05,9.68195954944974e-05,0.000102423000950512,0.000106568898681861,0.000109363660564904,0.000114372975179784,0.000114683578312762,0.000119257710984247,0.000121305619076379,0.000122775712314675,0.000123312335320184,0.000124192579488009,0.000128216905762061,0.000134810947699522,0.000135425930958918,0.000135425930958918,0.000136310804236262,0.000147852634944778,0.000149389373471279,0.00016547161455721,0.000169253650392972,0.000182918280217599,0.000183017195741282,0.00018496377248129,0.000187234708692807,0.000188793620256965,0.000189537596772015,0.000197624351337112,0.000199472366107809,0.00021041868650345,0.000213164011101062,0.000213164011101062,0.00021754145006996,0.00021754145006996,0.00021754145006996,0.000227506886201437,0.000230117262667767,0.000237314909968162,0.000239787323616366,0.000246625992508885,0.000246854439057942,0.000250853640125931,0.000253616085836374,0.000274901942195593,0.000285418093111668,0.000312319891300469,0.000312319891300469,0.000313491078023355,0.000313491078023355,0.000313491078023355,0.000313491078023355,0.000326986730620269,0.000328419954132913,0.000364231916015456,0.000364231916015456,0.000371846502672881,0.000374548319777048,0.000387479204232924,0.000389564703802761,0.000391555095948242,0.000395139439999779,0.000404642531531636,0.000420560345837988,0.000421836922422546,0.000425449715201064,0.000433788339276508,0.000434991289534878,0.000439450128619103,0.000444412803587846,0.000468494622209091,0.000468496401485564,0.000468496401485564,0.000468496401485564,0.000503817919290552,0.000517731607576407,0.000520605398170865,0.000520605398170865,0.000520605398170865,0.000520605398170865,0.000534000031626041,0.000534000031626041,0.000534000031626041,0.00053683099840061,0.00057008677299984,0.000578290035954794,0.000578290035954794,0.000578290035954794,0.000601206631236682,0.000604992260073521,0.000631907978749863,0.00063990578325196,0.000646227221383911,0.000653468721447342,0.000679792390807144,0.000689465677388796,0.000704465042855541,0.000705029045199209,0.000705558415523531,0.000735050049112179,0.000742677458243158,0.00074518606576883,0.00075370364068298,0.000765857239612202,0.00077446423736693,0.00077446423736693,0.00077446423736693,0.0007753515852199,0.000778985031680236,0.000811034501811273,0.000839293806144117,0.000849386846379322,0.000858426113185936,0.000866092016301398,0.000868069997342082,0.000868069997342082,0.000910255979501654,0.000919853734429445,0.000947251855057815,0.000955982881369072,0.000997874874229139,0.00101057951317372,0.00102936384980587,0.00103839096884223,0.00104773290877081,0.00105224567542048,0.00106367223086032,0.00107705024847167,0.00109174598584364,0.00109792975302416,0.00110284164737355,0.00111196499658947,0.00114603682774313,0.00115551996317083,0.00120115991787483,0.0012091128503917,0.00121785028975425,0.00128054954178958,0.00128098074887597,0.00128398951413113,0.00128660507827417,0.00131904297622685,0.00132104034307385,0.00137625562726553,0.00138166743086257,0.00138482127409203,0.00138482127409203,0.0014520837450511,0.00146096989541535,0.00147592585611676,0.0014813968485598,0.0014975741043004,0.00150207692326865,0.00151187373395043,0.00160589733007925,0.00162080167611434,0.0016358694517309,0.00168630915187062,0.00168630915187062,0.00170595079516695,0.00172736952689806,0.00175971221349799,0.00177706422700545,0.00177706422700545,0.00178452416015823,0.00183620808565237,0.00184982884666892,0.00186607668380481,0.00189271248680839,0.00191369298235599,0.00192463360624141,0.00194285621509172,0.00196831281011306,0.00201610169883808,0.00202588273417232,0.002035320693077,0.00206169307131411,0.00208048072086984,0.00210289910647534,0.00211123431777776,0.00214742339911794,0.00218151363392115,0.00221501472507022,0.00223594498447301,0.00237387471071002,0.00237387471071002,0.00237387471071002,0.00238338080086115,0.00241794017366181,0.00241794017366181,0.00250784133037806,0.00251903019251732,0.00262475751990178,0.00264293504126571,0.00268035762448522,0.00268597930518282,0.00269448765477599,0.00272309228778812,0.00275930669005197,0.00275930669005197,0.00275930669005197,0.00284574915784018,0.0028904483967626,0.00299199208394236,0.00299199208394236,0.00301125695824662,0.00306956182441035,0.00309594859389888,0.00309594859389888,0.00309594859389888,0.00309594859389888,0.00311683023960024,0.00317207882646199,0.00317207882646199,0.00317207882646199,0.00320234570825599,0.00342257214147309,0.00347669385550101,0.00351123554669052,0.0035364562776973,0.00354286045537598,0.00354464551629495,0.00361187497688773,0.00361187497688773,0.00361187497688773,0.00368836574761787,0.00372201801386819,0.00377190226168045,0.00384706247720938,0.00386461179092905,0.00389351797587632,0.00389351797587632,0.00401590496548932,0.00404681623535353,0.00406764963727696,0.00407838164607732,0.00407838164607732,0.00407838164607732,0.00414018732443359,0.00414061198596826,0.00421038470156807,0.00423508934139083,0.00424123714072874,0.00433581792378205,0.00433973740644307,0.0045246752703216,0.00457128796716299,0.00457128796716299,0.00476870571230532,0.00483741895619067,0.00499125001520879,0.00505106333201497,0.00505106333201497,0.00509028568237679,0.00509028568237679,0.00509028568237679,0.00509028568237679,0.00509028568237679,0.00510581047550427,0.00510958946757298,0.00518019127009068,0.00522828632786526,0.00522828632786526,0.00522828632786526,0.00538563502270247,0.00541066808361739,0.00556440736382107,0.00563506912389723,0.00563506912389723,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00565435917262713,0.00587674283568318,0.00588688659526833,0.00596492696441453,0.00596492696441453,0.00605328361252761,0.00613779091141902,0.00613779091141902,0.00613779091141902,0.00620533519482642,0.00620533519482642,0.00620533519482642,0.00620533519482642,0.00620533519482642,0.00623781220402311,0.00648876580010727,0.00648918088622615,0.00650523503482875,0.00652894592810596,0.00655745600784175,0.00660579795355668,0.00663025239466735,0.00663025239466735,0.00666350826683067,0.00666494570383505,0.00675429206255604,0.00675429206255604,0.00680078335029646,0.00680078335029646,0.00680078335029646,0.00680078335029646,0.0068330954046863,0.00690304691310504,0.00700737220180284,0.00703217931710316,0.00708977260369797,0.00722024895773784,0.00727563513973092,0.00741582084897645,0.00741582084897645,0.00742111557870457,0.00742111557870457,0.00742111557870457,0.00742111557870457,0.00742770397693551,0.0076601855443547,0.00771415163655997,0.00773067600485505,0.00780158132923968,0.00785509451741489,0.00794356495860334,0.00801028374290165,0.00805323943502259,0.00805323943502259,0.00805323943502259,0.00806603638307588,0.00806603638307588,0.00806603638307588,0.00806603638307588,0.00806603638307588,0.00833072957789266,0.00833559815308679,0.0083678980020762,0.00866730343739786,0.00866730343739786,0.00873525276255317,0.0087850490914251,0.008853183501386,0.0092389062521431,0.00931125027087149,0.00942847419401281,0.00942847419401281,0.00942847419401281,0.00942847419401281,0.00942847419401281,0.00942847419401281,0.00942847419401281,0.00943934486008094,0.00960851252471516,0.00970128964171808,0.00985338297389037,0.0100520060925423,0.0101302080457824,0.0101454126138059,0.0101454126138059,0.0101454126138059,0.0101454126138059,0.0101659103817409,0.0105552620535347,0.0105554794556091,0.0106356191648467,0.0108235345089294,0.0108531381310706,0.010885782399624,0.010885782399624,0.0109330512995468,0.011030597465961,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0112770182986626,0.0113182017125176,0.0113314971895085,0.0115383577377113,0.0115459426234949,0.0116493003524883,0.0116493003524883,0.0116493003524883,0.0119367174113015,0.0121538269378894,0.0121965571408123,0.0122018882908176,0.0122133323267957,0.0122464200012832,0.0122737963566646,0.0123821996215576,0.0124356856788624,0.0124356856788624,0.0124356856788624,0.0124356856788624,0.0124600009830022,0.0125698486631751,0.0127270659292015,0.0127316662227774,0.012794512067779,0.0128784508384128,0.0132360066632522,0.0132446599728858,0.0132446599728858,0.0132446599728858,0.0132446599728858,0.0132446599728858,0.0132721258808024,0.0132721258808024,0.0135836854867148,0.01364772277712,0.0137186096587235,0.013831868047334,0.013831868047334,0.0140130999743162,0.0140759471987301,0.0142813873474821,0.0143381570337482,0.0144929024583997,0.0146992874834873,0.0146992874834873,0.0149292736730753,0.0149292736730753,0.0149292736730753,0.0149959036361188,0.0153671801378381,0.0155632985157417,0.0158043680477052,0.0158043680477052,0.0158043680477052,0.0158043680477052,0.0158152650222246,0.0161552008788196,0.0161721863017567,0.0164265125110551,0.0165357167629694,0.0167009612922222,0.0167333869663457,0.0167333869663457,0.0167333869663457,0.0168551241835872,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0168681535745957,0.0172833891035736,0.017618786676881,0.017618786676881,0.017618786676881,0.017618786676881,0.0176808475359514,0.0178509652866361,0.0179959320142112,0.0180424255941623,0.0181155240591622,0.0181713487691551,0.0182501272436049,0.018557579755538,0.018557579755538,0.018557579755538,0.018557579755538,0.0186577166671513,0.0190693842548437,0.0191571973476701,0.0195170783487193,0.0195170783487193,0.0195170783487193,0.0195499254077624,0.0195525255743916,0.0195525255743916,0.0195525255743916,0.019664049888763,0.019664049888763,0.019664049888763,0.0197963246255939,0.0198128017907778,0.0202628322892186,0.0202628322892186,0.020320140735475,0.0203851610825071,0.0204970225268032,0.0207622008198176,0.0209401088716343,0.0209916031713257,0.0209916031713257,0.0210455427771182,0.0212288820077882,0.0213619639591695,0.0213862577843312,0.0214971545933183,0.0214971545933183,0.0214971545933183,0.0214971545933183,0.0217605787410797,0.0221146867840493,0.0223090275075439,0.0223090275075439,0.0223090275075439,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0224279402260424,0.0225172190683562,0.0225172190683562,0.0228682902840975,0.02288347584028,0.0229947051667775,0.0232996071309866,0.0234378291100043,0.0235569626720975,0.0235569626720975,0.0235569626720975,0.0235569626720975,0.0235569626720975,0.0236684052077515,0.0237864096535901,0.0239846526318488,0.0239846526318488,0.0244695450956315,0.0244695450956315,0.0246161343084512,0.0246161343084512,0.0246161343084512,0.0246161343084512,0.0246161343084512,0.0246914036123791,0.0249366517718917,0.0253368881840242,0.0253569886057642,0.0253871972359048,0.0255178811269692,0.0256944850488053,0.0256944850488053,0.0257017937558746,0.026426024407789,0.0264932794506646,0.0265527040999762,0.0267917681158904,0.0267917681158904,0.0268558447680782,0.0269250128994275,0.0269708722745415,0.0270966544833905,0.027228912987195,0.0272989880974085,0.0273250038842478,0.0273510021456713,0.0275711077119935,0.027600906631788,0.0276085542567878,0.0279077388677525,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0279565525130469,0.0281053867298406,0.0281053867298406,0.0281053867298406,0.0281527479142566,0.028359970078565,0.0285241246908141,0.028616120742572,0.0287776477816816,0.0290194025919701,0.0290421547818366,0.0294098928718285,0.0294351987253764,0.0294845800408121,0.0298390726005459,0.0300990078199116,0.0301947754391795,0.0305365119847792,0.0305365119847792,0.0309258793713557,0.0310008768673336,0.0310008768673336,0.0313506152053274,0.0313653625087113,0.0313653625087113,0.0313653625087113,0.0313653625087113,0.0313653625087113,0.031942101533984,0.0320121536693044,0.0324189699326855,0.0325536797316642,0.0325536797316642,0.0325536797316642,0.0325536797316642,0.0325536797316642,0.0325536797316642,0.0325772420212301,0.0333571140829516,0.0333571140829516,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033454163735359,0.033693938485293,0.0337594929060894,0.0338212501667683,0.0338945835582123,0.0339620966773449,0.0340405128607009,0.0342293629411025,0.0343682405510756,0.0343966403143687,0.034468888676547,0.034584722705889,0.0346153747990279,0.0347673703865855,0.034866024479097,0.0349825698714796,0.0349825698714796,0.0349825698714796,0.0349825698714796,0.0354290851382555,0.0356402493334979,0.0356402493334979,0.035874172562302,0.0361739979164421,0.0362226804934974,0.0362226804934974,0.0362226804934974,0.0362226804934974,0.0362226804934974,0.0362226804934974,0.0362226804934974,0.0363845417454479,0.0363845417454479,0.0364807042468125,0.0368466003481227,0.0368991106753669,0.036959502237616,0.0369865469127515,0.0374795966488097,0.0374795966488097,0.0375661756613165,0.0377112830519253,0.0379718671796374,0.0382182782648527,0.0384735245494323,0.038706438484883,0.0387530922100255,0.0387530922100255,0.0387530922100255,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389209462376777,0.0389692651179238,0.0389875248126118,0.0389994199323486,0.0392462828888066,0.0395067020091081,0.0397099774134085,0.0397679869266592,0.0397679869266592,0.0399945253579149,0.0400026771461371,0.0400429430307376,0.0400429430307376,0.0400429430307376,0.0400429430307376,0.0406612297450995,0.0413489269306676,0.0413489269306676,0.0413489269306676,0.0417194728364106,0.0420340764476604,0.0423931700112184,0.0426708236809135,0.0426708236809135,0.0426708236809135,0.0426708236809135,0.0428036411693761,0.0428036411693761,0.0428369604877565,0.0431599637212532,0.0433411115074804,0.0439712848600256,0.0440084149892984,0.0440084149892984,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0443570714148682,0.0444758009079544,0.0451546751606268,0.0459524002127382,0.0462866731264747,0.0462904189945805,0.046729817708204,0.046729817708204,0.046729817708204,0.0467729629338509,0.0469157801112817,0.0471539735876284,0.0471719464769759,0.0473768857381413,0.0474752082864341,0.0477904564915794,0.0477904564915794,0.0480739024393591,0.0481132020875486,0.0481132020875486,0.0481132020875486,0.0481132020875486,0.0481132020875486,0.0483874256141338,0.0483874256141338,0.0492129786143436,0.0492838789023564,0.0495114269340794,0.049762709717153,0.049762709717153,0.049762709717153,0.049762709717153,0.049762709717153,0.049762709717153,0.049762709717153,0.049762709717153,0.049762709717153,0.049762709717153,0.049762709717153,0.049762709717153,0.049762709717153,0.049762709717153,0.049762709717153,0.049762709717153,0.049762709717153,0.049762709717153,0.049762709717153,0.049762709717153,0.049762709717153,0.049762709717153,0.049762709717153,0.049762709717153]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Term<\/th>\n      <th>Ont<\/th>\n      <th>N<\/th>\n      <th>DE<\/th>\n      <th>P.DE<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"targets":5,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 9, 3, \",\", \".\");\n  }"},{"className":"dt-right","targets":[3,4,5]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render"],"jsHooks":[]}</script>
</div>
<p>Figure <a href="#fig:scatterplot-and-ma-plot">8</a> plots the relationship between UMI counts and read counts for all genes and samples. There are a few notable points about this figure:</p>
<ol type="1">
<li>The UMI counts and read counts are very strongly associated.</li>
<li>The log-fold change between the read counts and the UMI counts can be large, indicating the UMI counts are likely removing substantial PCR amplification noise.</li>
<li>The log-fold changes between the read counts and the UMI counts are larger on average for more highly expressed genes. The variation in the log-fold change is larger for the lowly expressed genes.</li>
<li>The hemoglobin genes stand out. These genes are extremely highly expressed, meaning that they likely saturate the 6 bp UMIs and which leads to the strange relationship between the UMI counts and the read counts for these genes.</li>
</ol>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>x</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log2</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/vector.html'>as.vector</a></span><span class='op'>(</span><span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"UMI_counts"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> <span class='fl'>1</span><span class='op'>)</span>
<span class='va'>y</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log2</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/vector.html'>as.vector</a></span><span class='op'>(</span><span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"read_counts"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> <span class='fl'>1</span><span class='op'>)</span>
<span class='va'>m</span> <span class='op'>&lt;-</span> <span class='op'>(</span><span class='va'>x</span> <span class='op'>+</span> <span class='va'>y</span><span class='op'>)</span> <span class='op'>/</span> <span class='fl'>2</span>
<span class='va'>d</span> <span class='op'>&lt;-</span> <span class='va'>y</span> <span class='op'>-</span> <span class='va'>x</span>
<span class='va'>g</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>ncol</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>)</span>
<span class='co'># NOTE: Only showing unique points for plotting speed.</span>
<span class='va'>i</span> <span class='op'>&lt;-</span> <span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/duplicated.html'>duplicated</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='va'>x</span>, <span class='st'>"."</span>, <span class='va'>y</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>X</span> <span class='op'>&lt;-</span> <span class='fu'>edgeR</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/cpm.html'>cpm</a></span><span class='op'>(</span><span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"UMI_counts"</span><span class='op'>)</span>, log <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='va'>Y</span> <span class='op'>&lt;-</span> <span class='fu'>edgeR</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/cpm.html'>cpm</a></span><span class='op'>(</span><span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"read_counts"</span><span class='op'>)</span>, log <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='va'>M</span> <span class='op'>&lt;-</span> <span class='op'>(</span><span class='va'>X</span> <span class='op'>+</span> <span class='va'>Y</span><span class='op'>)</span> <span class='op'>/</span> <span class='fl'>2</span>
<span class='va'>D</span> <span class='op'>&lt;-</span> <span class='va'>y</span> <span class='op'>-</span> <span class='va'>x</span>
<span class='va'>G</span> <span class='op'>&lt;-</span> <span class='va'>g</span>
<span class='co'># NOTE: Only showing unique points for plotting speed.</span>
<span class='va'>I</span> <span class='op'>&lt;-</span> <span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/duplicated.html'>duplicated</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='va'>x</span>, <span class='st'>"."</span>, <span class='va'>y</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/par.html'>par</a></span><span class='op'>(</span>mfrow <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>2</span>, <span class='fl'>2</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/limma/man/plotWithHighlights.html'>plotWithHighlights</a></span><span class='op'>(</span>
  x <span class='op'>=</span> <span class='va'>x</span><span class='op'>[</span><span class='va'>i</span><span class='op'>]</span>,
  y <span class='op'>=</span> <span class='va'>y</span><span class='op'>[</span><span class='va'>i</span><span class='op'>]</span>,
  status <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span><span class='va'>g</span><span class='op'>[</span><span class='va'>i</span><span class='op'>]</span> <span class='op'>%in%</span> <span class='va'>ribo_set</span>, <span class='fl'>1</span>, <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span><span class='va'>g</span><span class='op'>[</span><span class='va'>i</span><span class='op'>]</span> <span class='op'>%in%</span> <span class='va'>mito_set</span>, <span class='op'>-</span><span class='fl'>1</span>, <span class='fl'>0</span><span class='op'>)</span><span class='op'>)</span>,
  legend <span class='op'>=</span> <span class='cn'>FALSE</span>,
  bg.pch <span class='op'>=</span> <span class='st'>"."</span>,
  hl.pch <span class='op'>=</span> <span class='st'>"."</span>,
  hl.col <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"orange"</span>, <span class='st'>"dodgerBlue"</span><span class='op'>)</span>,
  hl.cex <span class='op'>=</span> <span class='fl'>3</span>,
  sub <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"r = "</span>, <span class='fu'><a href='https://scales.r-lib.org/reference/number.html'>number</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/stats/cor.html'>cor</a></span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>y</span><span class='op'>)</span>, accuracy <span class='op'>=</span> <span class='fl'>0.001</span><span class='op'>)</span><span class='op'>)</span>,
  xlab <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/expression.html'>expression</a></span><span class='op'>(</span><span class='va'>log</span><span class='op'>[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>(</span><span class='va'>UMIs</span> <span class='op'>+</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>,
  ylab <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/expression.html'>expression</a></span><span class='op'>(</span><span class='va'>log</span><span class='op'>[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>(</span><span class='va'>reads</span> <span class='op'>+</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/abline.html'>abline</a></span><span class='op'>(</span>a <span class='op'>=</span> <span class='fl'>0</span>, b <span class='op'>=</span> <span class='fl'>1</span>, lty <span class='op'>=</span> <span class='fl'>2</span>, col <span class='op'>=</span> <span class='st'>"grey"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/limma/man/plotWithHighlights.html'>plotWithHighlights</a></span><span class='op'>(</span>
  x <span class='op'>=</span> <span class='va'>m</span><span class='op'>[</span><span class='va'>i</span><span class='op'>]</span> <span class='op'>/</span> <span class='fl'>2</span>,
  y <span class='op'>=</span> <span class='va'>d</span><span class='op'>[</span><span class='va'>i</span><span class='op'>]</span>,
  pch <span class='op'>=</span> <span class='st'>"."</span>,
  status <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span><span class='va'>g</span><span class='op'>[</span><span class='va'>i</span><span class='op'>]</span> <span class='op'>%in%</span> <span class='va'>ribo_set</span>, <span class='fl'>1</span>, <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span><span class='va'>g</span><span class='op'>[</span><span class='va'>i</span><span class='op'>]</span> <span class='op'>%in%</span> <span class='va'>mito_set</span>, <span class='op'>-</span><span class='fl'>1</span>, <span class='fl'>0</span><span class='op'>)</span><span class='op'>)</span>,
  legend <span class='op'>=</span> <span class='cn'>FALSE</span>,
  bg.pch <span class='op'>=</span> <span class='st'>"."</span>,
  hl.pch <span class='op'>=</span> <span class='st'>"."</span>,
  hl.col <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"orange"</span>, <span class='st'>"dodgerBlue"</span><span class='op'>)</span>,
  hl.cex <span class='op'>=</span> <span class='fl'>3</span>,
  xlab <span class='op'>=</span> <span class='st'>"Average log-(count+1)"</span>,
  ylab <span class='op'>=</span> <span class='st'>"logFC (reads vs. UMIs)"</span>,
  ylim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/MathFun.html'>abs</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='va'>d</span>, <span class='va'>D</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/MathFun.html'>abs</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='va'>d</span>, <span class='va'>D</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/abline.html'>abline</a></span><span class='op'>(</span>h <span class='op'>=</span> <span class='fl'>0</span>, lty <span class='op'>=</span> <span class='fl'>2</span>, col <span class='op'>=</span> <span class='st'>"grey"</span><span class='op'>)</span>
<span class='va'>l</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/lowess.html'>lowess</a></span><span class='op'>(</span><span class='va'>m</span>, <span class='va'>d</span>, f <span class='op'>=</span> <span class='fl'>0.3</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/lines.html'>lines</a></span><span class='op'>(</span><span class='va'>l</span>, col <span class='op'>=</span> <span class='st'>"red"</span>, lwd <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/limma/man/plotWithHighlights.html'>plotWithHighlights</a></span><span class='op'>(</span>
  x <span class='op'>=</span> <span class='va'>X</span><span class='op'>[</span><span class='va'>I</span><span class='op'>]</span>,
  y <span class='op'>=</span> <span class='va'>Y</span><span class='op'>[</span><span class='va'>I</span><span class='op'>]</span>,
  status <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span><span class='va'>G</span><span class='op'>[</span><span class='va'>I</span><span class='op'>]</span> <span class='op'>%in%</span> <span class='va'>ribo_set</span>, <span class='fl'>1</span>, <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span><span class='va'>G</span><span class='op'>[</span><span class='va'>I</span><span class='op'>]</span> <span class='op'>%in%</span> <span class='va'>mito_set</span>, <span class='op'>-</span><span class='fl'>1</span>, <span class='fl'>0</span><span class='op'>)</span><span class='op'>)</span>,
  legend <span class='op'>=</span> <span class='cn'>FALSE</span>,
  bg.pch <span class='op'>=</span> <span class='st'>"."</span>,
  hl.pch <span class='op'>=</span> <span class='st'>"."</span>,
  hl.col <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"orange"</span>, <span class='st'>"dodgerBlue"</span><span class='op'>)</span>,
  hl.cex <span class='op'>=</span> <span class='fl'>3</span>,
  sub <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span>
    <span class='st'>"r = "</span>, 
    <span class='fu'><a href='https://scales.r-lib.org/reference/number.html'>number</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/stats/cor.html'>cor</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/vector.html'>as.vector</a></span><span class='op'>(</span><span class='va'>X</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/vector.html'>as.vector</a></span><span class='op'>(</span><span class='va'>Y</span><span class='op'>)</span><span class='op'>)</span>, accuracy <span class='op'>=</span> <span class='fl'>0.001</span><span class='op'>)</span><span class='op'>)</span>,
  xlab <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/expression.html'>expression</a></span><span class='op'>(</span><span class='fu'>logCPM</span><span class='op'>(</span><span class='va'>UMIs</span><span class='op'>)</span><span class='op'>)</span>,
  ylab <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/expression.html'>expression</a></span><span class='op'>(</span><span class='fu'>logCPM</span><span class='op'>(</span><span class='va'>reads</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/abline.html'>abline</a></span><span class='op'>(</span>a <span class='op'>=</span> <span class='fl'>0</span>, b <span class='op'>=</span> <span class='fl'>1</span>, lty <span class='op'>=</span> <span class='fl'>2</span>, col <span class='op'>=</span> <span class='st'>"grey"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/limma/man/plotWithHighlights.html'>plotWithHighlights</a></span><span class='op'>(</span>
  x <span class='op'>=</span> <span class='va'>M</span><span class='op'>[</span><span class='va'>I</span><span class='op'>]</span>,
  y <span class='op'>=</span> <span class='va'>D</span><span class='op'>[</span><span class='va'>i</span><span class='op'>]</span>,
  status <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span><span class='va'>G</span><span class='op'>[</span><span class='va'>I</span><span class='op'>]</span> <span class='op'>%in%</span> <span class='va'>ribo_set</span>, <span class='fl'>1</span>, <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span><span class='va'>G</span><span class='op'>[</span><span class='va'>I</span><span class='op'>]</span> <span class='op'>%in%</span> <span class='va'>mito_set</span>, <span class='op'>-</span><span class='fl'>1</span>, <span class='fl'>0</span><span class='op'>)</span><span class='op'>)</span>,
  legend <span class='op'>=</span> <span class='cn'>FALSE</span>,
  bg.pch <span class='op'>=</span> <span class='st'>"."</span>,
  hl.pch <span class='op'>=</span> <span class='st'>"."</span>,
  hl.col <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"orange"</span>, <span class='st'>"dodgerBlue"</span><span class='op'>)</span>,
  hl.cex <span class='op'>=</span> <span class='fl'>3</span>,
  xlab <span class='op'>=</span> <span class='st'>"Average logCPM"</span>,
  ylab <span class='op'>=</span> <span class='st'>"logFC (reads vs. UMIs)"</span>,
  ylim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/MathFun.html'>abs</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='va'>d</span>, <span class='va'>D</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/MathFun.html'>abs</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='va'>d</span>, <span class='va'>D</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/abline.html'>abline</a></span><span class='op'>(</span>h <span class='op'>=</span> <span class='fl'>0</span>, lty <span class='op'>=</span> <span class='fl'>2</span>, col <span class='op'>=</span> <span class='st'>"grey"</span><span class='op'>)</span>
<span class='va'>l</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/lowess.html'>lowess</a></span><span class='op'>(</span><span class='va'>M</span>, <span class='va'>D</span>, f <span class='op'>=</span> <span class='fl'>0.3</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/lines.html'>lines</a></span><span class='op'>(</span><span class='va'>l</span>, col <span class='op'>=</span> <span class='st'>"red"</span>, lwd <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:scatterplot-and-ma-plot"></span>
<img src="C094_Pellicci.mini-bulk.preprocess_files/scatterplot-and-ma-plot-1.png" alt="Scatterplot of the logarithm of the UMI and read counts (top) and logCPM of the UMI and read counts (bottom) for all genes and samples (left) and mean-difference plot of the same data (right). The correlation in the scatterplot is reported below the x-axis label. Genes that are ribosomal protein genes subunits are highlighted in orange and mitochondrial genes in blue. The red line is a trend fitted to the mean-differences." width="624" />
<p class="caption">
Figure 8: Scatterplot of the logarithm of the UMI and read counts (top) and logCPM of the UMI and read counts (bottom) for all genes and samples (left) and mean-difference plot of the same data (right). The correlation in the scatterplot is reported below the x-axis label. Genes that are ribosomal protein genes subunits are highlighted in orange and mitochondrial genes in blue. The red line is a trend fitted to the mean-differences.
</p>
</div>
</div>
<p>Rather than looking at all samples and all genes at once, can also examine the relationship between UMI counts and read counts within a single sample across genes (Figure <a href="#fig:within-sample-cors">9</a>) and within a single gene across samples (Figure <a href="#fig:within-gene-cors">10</a>).</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>X</span> <span class='op'>&lt;-</span> <span class='fu'>edgeR</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/cpm.html'>cpm</a></span><span class='op'>(</span><span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"UMI_counts"</span><span class='op'>)</span>, log <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='va'>Y</span> <span class='op'>&lt;-</span> <span class='fu'>edgeR</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/cpm.html'>cpm</a></span><span class='op'>(</span><span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"read_counts"</span><span class='op'>)</span>, log <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='va'>cor_j</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/diag.html'>diag</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/stats/cor.html'>cor</a></span><span class='op'>(</span><span class='va'>X</span>, <span class='va'>Y</span><span class='op'>)</span><span class='op'>)</span>
<span class='co'># NOTE: Splitting up the cor() computations is a hack to improve speed whilst </span>
<span class='co'>#       keeping memory usage somewhat under control.</span>
<span class='va'>b</span> <span class='op'>&lt;-</span> <span class='fl'>1000</span>
<span class='va'>s</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq</a></span><span class='op'>(</span><span class='fl'>1</span>, by <span class='op'>=</span> <span class='va'>b</span>, to <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>e</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>pmin</a></span><span class='op'>(</span><span class='va'>s</span> <span class='op'>+</span> <span class='va'>b</span> <span class='op'>-</span> <span class='fl'>1</span>, <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>cor_i</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/unlist.html'>unlist</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/base/funprog.html'>Map</a></span><span class='op'>(</span>
    f <span class='op'>=</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>ss</span>, <span class='va'>ee</span> <span class='op'>)</span> <span class='fu'><a href='https://rdrr.io/r/base/diag.html'>diag</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/stats/cor.html'>cor</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span><span class='va'>X</span><span class='op'>[</span><span class='va'>ss</span><span class='op'>:</span><span class='va'>ee</span>, <span class='op'>]</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span><span class='va'>Y</span><span class='op'>[</span><span class='va'>ss</span><span class='op'>:</span><span class='va'>ee</span>, <span class='op'>]</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>,
    ss <span class='op'>=</span> <span class='va'>s</span>,
    ee <span class='op'>=</span> <span class='va'>e</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>These figures show that the UMI counts and read counts are highly correlated across both samples (median = 0.896, median absolute deviation = 0.026) and genes (median = 0.831, median absolute deviation = 0.136).</p>
<aside>
The genes with the lowest correlations include the aforementioned highly-expressed hemoglobin subunits.
</aside>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/graphics/par.html'>par</a></span><span class='op'>(</span>mfrow <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>3</span>, <span class='fl'>3</span><span class='op'>)</span>, cex <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span>
<span class='va'>jj</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/sort.html'>sort</a></span><span class='op'>(</span><span class='va'>cor_j</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>3</span>, <span class='fl'>60</span><span class='op'>:</span><span class='fl'>61</span>, <span class='fl'>119</span><span class='op'>:</span><span class='fl'>121</span><span class='op'>)</span><span class='op'>]</span><span class='op'>)</span>
<span class='kw'>for</span> <span class='op'>(</span><span class='va'>j</span> <span class='kw'>in</span> <span class='va'>jj</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='va'>g</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span>
    x <span class='op'>=</span> <span class='va'>X</span><span class='op'>[</span>, <span class='va'>j</span><span class='op'>]</span>,
    y <span class='op'>=</span> <span class='va'>Y</span><span class='op'>[</span>, <span class='va'>j</span><span class='op'>]</span>,
    pch <span class='op'>=</span> <span class='st'>"."</span>,
    cex <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span><span class='va'>g</span> <span class='op'>%in%</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='va'>ribo_set</span>, <span class='va'>mito_set</span><span class='op'>)</span>, <span class='fl'>5</span>, <span class='fl'>1</span><span class='op'>)</span>,
    col <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span>
      <span class='va'>g</span> <span class='op'>%in%</span> <span class='va'>ribo_set</span>, <span class='st'>"orange"</span>,
      <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span><span class='va'>g</span> <span class='op'>%in%</span> <span class='va'>mito_set</span>, <span class='st'>"dodgerBlue"</span>, <span class='st'>"black"</span><span class='op'>)</span><span class='op'>)</span>,
    sub <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"r = "</span>, <span class='fu'><a href='https://scales.r-lib.org/reference/number.html'>number</a></span><span class='op'>(</span><span class='va'>cor_j</span><span class='op'>[</span><span class='va'>j</span><span class='op'>]</span>, accuracy <span class='op'>=</span> <span class='fl'>0.001</span><span class='op'>)</span><span class='op'>)</span>,
    xlab <span class='op'>=</span> <span class='st'>"logCPM(UMIs)"</span>,
    ylab <span class='op'>=</span> <span class='st'>"logCPM(reads)"</span>,
    main <span class='op'>=</span> <span class='va'>j</span>,
    xlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/range.html'>range</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>,
    ylim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/range.html'>range</a></span><span class='op'>(</span><span class='va'>y</span><span class='op'>)</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/graphics/abline.html'>abline</a></span><span class='op'>(</span>a <span class='op'>=</span> <span class='fl'>0</span>, b <span class='op'>=</span> <span class='fl'>1</span>, col <span class='op'>=</span> <span class='st'>"grey"</span>, lty <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>
<span class='op'>}</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/stats/density.html'>density</a></span><span class='op'>(</span><span class='va'>cor_j</span>, na.rm <span class='op'>=</span> <span class='cn'>TRUE</span>, from <span class='op'>=</span> <span class='op'>-</span><span class='fl'>1</span>, to <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span>,
  main <span class='op'>=</span> <span class='st'>"All samples"</span>, 
  xlab <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/expression.html'>expression</a></span><span class='op'>(</span><span class='va'>r</span><span class='op'>[</span><span class='va'>p</span><span class='op'>]</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:within-sample-cors"></span>
<img src="C094_Pellicci.mini-bulk.preprocess_files/within-sample-cors-1.png" alt="Association between UMI counts and read counts for each gene within each sample. The first 8 panels are scatter plots of the UMI counts and read counts for each gene for the samples with the 3 lowest correlations, 2 middlemost correlations, and 3 highest correlations. The correlation in the scatterplot is reported below each panel. Genes that are ribosomal protein genes subunits are highlighted in orange and mitochondrial genes in blue. The final panel is the distribution of these correlations across all samples." width="624" />
<p class="caption">
Figure 9: Association between UMI counts and read counts for each gene within each sample. The first 8 panels are scatter plots of the UMI counts and read counts for each gene for the samples with the 3 lowest correlations, 2 middlemost correlations, and 3 highest correlations. The correlation in the scatterplot is reported below each panel. Genes that are ribosomal protein genes subunits are highlighted in orange and mitochondrial genes in blue. The final panel is the distribution of these correlations across all samples.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/graphics/par.html'>par</a></span><span class='op'>(</span>mfrow <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>3</span>, <span class='fl'>3</span><span class='op'>)</span>, cex <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span>
<span class='co'># NOTE: There are some NA correlations, so this changes ii.</span>
<span class='va'>ii</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/sort.html'>sort</a></span><span class='op'>(</span><span class='va'>cor_i</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>3</span>, <span class='fl'>16506</span><span class='op'>:</span><span class='fl'>16507</span>, <span class='fl'>33011</span><span class='op'>:</span><span class='fl'>33013</span><span class='op'>)</span><span class='op'>]</span><span class='op'>)</span>
<span class='kw'>for</span> <span class='op'>(</span><span class='va'>i</span> <span class='kw'>in</span> <span class='va'>ii</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='va'>g</span> <span class='op'>&lt;-</span> <span class='va'>i</span>
  <span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span>
    x <span class='op'>=</span> <span class='va'>X</span><span class='op'>[</span><span class='va'>i</span>, <span class='op'>]</span>,
    y <span class='op'>=</span> <span class='va'>Y</span><span class='op'>[</span><span class='va'>i</span>, <span class='op'>]</span>,
    pch <span class='op'>=</span> <span class='fl'>16</span>,
    col <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span>
      <span class='va'>g</span> <span class='op'>%in%</span> <span class='va'>ribo_set</span>, <span class='st'>"orange"</span>,
      <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span><span class='va'>g</span> <span class='op'>%in%</span> <span class='va'>mito_set</span>, <span class='st'>"dodgerBlue"</span>, <span class='st'>"black"</span><span class='op'>)</span><span class='op'>)</span>,
    sub <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"r = "</span>, <span class='fu'><a href='https://scales.r-lib.org/reference/number.html'>number</a></span><span class='op'>(</span><span class='va'>cor_i</span><span class='op'>[</span><span class='va'>i</span><span class='op'>]</span>, accuracy <span class='op'>=</span> <span class='fl'>0.001</span><span class='op'>)</span><span class='op'>)</span>,
    xlab <span class='op'>=</span> <span class='st'>"logCPM(UMIs)"</span>,
    ylab <span class='op'>=</span> <span class='st'>"logCPM(reads)"</span>,
    main <span class='op'>=</span> <span class='va'>i</span>,
    xlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/range.html'>range</a></span><span class='op'>(</span><span class='va'>X</span><span class='op'>)</span>,
    ylim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/range.html'>range</a></span><span class='op'>(</span><span class='va'>Y</span><span class='op'>)</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/graphics/abline.html'>abline</a></span><span class='op'>(</span>a <span class='op'>=</span> <span class='fl'>0</span>, b <span class='op'>=</span> <span class='fl'>1</span>, col <span class='op'>=</span> <span class='st'>"grey"</span>, lty <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>
<span class='op'>}</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/stats/density.html'>density</a></span><span class='op'>(</span><span class='va'>cor_i</span>, na.rm <span class='op'>=</span> <span class='cn'>TRUE</span>, from <span class='op'>=</span> <span class='op'>-</span><span class='fl'>1</span>, to <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span>,
  main <span class='op'>=</span> <span class='st'>"All genes"</span>, 
  xlab <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/expression.html'>expression</a></span><span class='op'>(</span><span class='va'>r</span><span class='op'>[</span><span class='va'>p</span><span class='op'>]</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:within-gene-cors"></span>
<img src="C094_Pellicci.mini-bulk.preprocess_files/within-gene-cors-1.png" alt="Association between UMI counts and read counts for each sample within each gene. The first 8 panels are scatter plots of the UMI counts and read counts for each sample for the genes with the 3 lowest correlations, 2 middlemost correlations, and 3 highest correlations. The correlation in the scatterplot is reported below each panel. Genes that are ribosomal protein genes subunits are highlighted in orange and mitochondrial genes in blue. The final panel is the distribution of these correlations across all genes." width="624" />
<p class="caption">
Figure 10: Association between UMI counts and read counts for each sample within each gene. The first 8 panels are scatter plots of the UMI counts and read counts for each sample for the genes with the 3 lowest correlations, 2 middlemost correlations, and 3 highest correlations. The correlation in the scatterplot is reported below each panel. Genes that are ribosomal protein genes subunits are highlighted in orange and mitochondrial genes in blue. The final panel is the distribution of these correlations across all genes.
</p>
</div>
</div>
<h2 id="summary-2">Summary</h2>
<p>The ultimate aim of this analysis is to perform a differential expression (DE) analysis. On the one hand, a DE analysis is more powerful when there are more counts available, which is where the read counts are favoured. On the other hand, a DE analysis may be biased if these counts arise through PCR duplicates, which is where the deduplicated UMI counts are favoured. The choice between using the read counts or the UMI counts in the DE analysis is therefore a trade-off between these two choices.</p>
<p><strong>For now, we defer our choice of read counts vs.UMI counts, but the remainder of this preprocessing report uses (exonic) read counts</strong>.</p>
<h1 id="quality-control-of-cells">Quality control of cells</h1>
<h2 id="defining-the-quality-control-metrics">Defining the quality control metrics</h2>
<p>Low-quality cells need to be removed to ensure that technical effects do not distort downstream analysis results. We use several quality control (QC) metrics to measure the quality of the cells:</p>
<ul>
<li><code>sum</code>: This measures the library size of the samples, which is the total sum of counts across genes. We want samples to have high library sizes as this means more RNA has been successfully captured during library preparation.</li>
<li><code>detected</code>: This is the number of expressed features<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a> in each cell. Cells with few expressed features are likely to be of poor quality, as the diverse transcript population has not been successful captured.</li>
<li><code>altexps_ERCC_detected</code>: This measures the proportion of reads which are mapped to spike-in transcripts relative to the library size of each cell. High proportions are indicative of poor-quality cells, where endogenous RNA has been lost during processing (e.g., due to cell lysis or RNA degradation). The same amount of spike-in RNA to each cell, so an enrichment in spike-in counts is symptomatic of loss of endogenous RNA.</li>
<li><code>subsets_Mito_percent</code>: This measures the proportion of reads which are mapped to mitochondrial RNA. If there is a higher than expected proportion of mitochondrial RNA this is often symptomatic of a cell which is under stress and is therefore of low quality and will not be used for the analysis.</li>
<li><code>subsets_Ribo_percent</code>: This measures the proportion of UMIs which are mapped to ribosomal protein genes. If there is a higher than expected proportion of ribosomal protein gene expression this is often symptomatic of a cell which is of compromised quality and we may want to exclude it from the analysis.</li>
</ul>
<p>For CEL-Seq2 data, we typically observe library sizes that are in the tens of thousands<a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a> and several thousand expressed genes per cell.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>is_mito</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>%in%</span> <span class='va'>mito_set</span>
<span class='fu'><a href='https://rdrr.io/r/base/summary.html'>summary</a></span><span class='op'>(</span><span class='va'>is_mito</span><span class='op'>)</span>
<span class='va'>is_ribo</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>%in%</span> <span class='va'>ribo_set</span>
<span class='fu'><a href='https://rdrr.io/r/base/summary.html'>summary</a></span><span class='op'>(</span><span class='va'>is_ribo</span><span class='op'>)</span>
<span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='fu'>addPerCellQC</span><span class='op'>(</span>
  <span class='va'>sce</span>, 
  subsets <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>Mito <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span><span class='va'>is_mito</span><span class='op'>)</span>, Ribo <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span><span class='va'>is_ribo</span><span class='op'>)</span><span class='op'>)</span>,
  use.altexps <span class='op'>=</span> <span class='st'>"ERCC"</span>,
  assay.type <span class='op'>=</span> <span class='st'>"read_counts"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<h2 id="visualizing-the-qc-metrics">Visualizing the QC metrics</h2>
<p>Figure <a href="#fig:qcplot">11</a> compares the QC metrics of the single-cell samples across the different <code>sample</code> (combination of <code>tissue</code> and <code>donor</code>). The number of UMIs, number of gene detected, and percentage of UMIs that are mapped to ERCC spike-ins are all higher in the experimental samples than in the control samples This likely reflects that the experimental cells, which are  T cells, have a smaller transcriptome than the control cells (rather than this result being due to technical factors, although we cannot rule this out entirely). Together, these figures show that the vast majority of single-cell samples are good-quality<a href="#fn6" class="footnote-ref" id="fnref6" role="doc-noteref"><sup>6</sup></a>:</p>
<ul>
<li>The median library size is around 1,034,000.</li>
<li>The median number of genes detected is around 11,000.</li>
<li>The median percentage of reads that are mapped to mitochondrial RNA is around 17%.</li>
<li>The median percentage of reads that are mapped to ribosomal protein genes is around 11%.</li>
<li>The median percentage of reads that are mapped to ERCC spike-ins is around 0%.</li>
</ul>
<p>Figure <a href="#fig:qcplot">11</a> also highlights that the <code>Blood.1</code>, <code>Blood.2</code>, and <code>Blood.3</code> samples on plate <code>LCE515</code> have fewer reads and genes detected than other samples, but that the that QC metrics are otherwise generally consistent across samples (and plates).</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce</span>,
  <span class='st'>"sum"</span>,
  x <span class='op'>=</span> <span class='st'>"donor"</span>,
  other_fields <span class='op'>=</span> <span class='st'>"tissue"</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span>,
  colour_by <span class='op'>=</span> <span class='st'>"donor"</span>,
  shape_by <span class='op'>=</span> <span class='st'>"plate_number"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_continuous.html'>scale_y_log10</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/annotation_logticks.html'>annotation_logticks</a></span><span class='op'>(</span>
    sides <span class='op'>=</span> <span class='st'>"l"</span>,
    short <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grid/unit.html'>unit</a></span><span class='op'>(</span><span class='fl'>0.03</span>, <span class='st'>"cm"</span><span class='op'>)</span>,
    mid <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grid/unit.html'>unit</a></span><span class='op'>(</span><span class='fl'>0.06</span>, <span class='st'>"cm"</span><span class='op'>)</span>,
    long <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grid/unit.html'>unit</a></span><span class='op'>(</span><span class='fl'>0.09</span>, <span class='st'>"cm"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>donor_colours</span>, name <span class='op'>=</span> <span class='st'>"donor"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>""</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_grid.html'>facet_grid</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>tissue</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guide_legend.html'>guide_legend</a></span><span class='op'>(</span>override.aes <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>4</span>, alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce</span>,
  <span class='st'>"detected"</span>,
  x <span class='op'>=</span> <span class='st'>"donor"</span>,
  other_fields <span class='op'>=</span> <span class='st'>"tissue"</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span>,
  colour_by <span class='op'>=</span> <span class='st'>"donor"</span>,
  shape_by <span class='op'>=</span> <span class='st'>"plate_number"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>donor_colours</span>, name <span class='op'>=</span> <span class='st'>"donor"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>""</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_grid.html'>facet_grid</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>tissue</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guide_legend.html'>guide_legend</a></span><span class='op'>(</span>override.aes <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>4</span>, alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>p3</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce</span>,
  <span class='st'>"subsets_Mito_percent"</span>,
  x <span class='op'>=</span> <span class='st'>"donor"</span>,
  other_fields <span class='op'>=</span> <span class='st'>"tissue"</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span>,
  colour_by <span class='op'>=</span> <span class='st'>"donor"</span>,
  shape_by <span class='op'>=</span> <span class='st'>"plate_number"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/lims.html'>ylim</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='cn'>NA</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>donor_colours</span>, name <span class='op'>=</span> <span class='st'>"donor"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>""</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_grid.html'>facet_grid</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>tissue</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guide_legend.html'>guide_legend</a></span><span class='op'>(</span>override.aes <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>4</span>, alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>p4</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce</span>,
  <span class='st'>"subsets_Ribo_percent"</span>,
  x <span class='op'>=</span> <span class='st'>"donor"</span>,
  other_fields <span class='op'>=</span> <span class='st'>"tissue"</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span>,
  colour_by <span class='op'>=</span> <span class='st'>"donor"</span>,
  shape_by <span class='op'>=</span> <span class='st'>"plate_number"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/lims.html'>ylim</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='cn'>NA</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>donor_colours</span>, name <span class='op'>=</span> <span class='st'>"donor"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>""</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_grid.html'>facet_grid</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>tissue</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guide_legend.html'>guide_legend</a></span><span class='op'>(</span>override.aes <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>4</span>, alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>p5</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce</span>,
  <span class='st'>"altexps_ERCC_percent"</span>,
  x <span class='op'>=</span> <span class='st'>"donor"</span>,
  other_fields <span class='op'>=</span> <span class='st'>"tissue"</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span>,
  colour_by <span class='op'>=</span> <span class='st'>"donor"</span>,
  shape_by <span class='op'>=</span> <span class='st'>"plate_number"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/lims.html'>ylim</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='cn'>NA</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>donor_colours</span>, name <span class='op'>=</span> <span class='st'>"donor"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>""</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_grid.html'>facet_grid</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>tissue</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guide_legend.html'>guide_legend</a></span><span class='op'>(</span>override.aes <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>4</span>, alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>p1</span> <span class='op'>+</span> <span class='va'>p2</span> <span class='op'>+</span> <span class='va'>p3</span> <span class='op'>+</span> <span class='va'>p4</span> <span class='op'>+</span> <span class='va'>p5</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://patchwork.data-imaginist.com/reference/plot_layout.html'>plot_layout</a></span><span class='op'>(</span>ncol <span class='op'>=</span> <span class='fl'>2</span>, guides <span class='op'>=</span> <span class='st'>"collect"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:qcplot"></span>
<img src="C094_Pellicci.mini-bulk.preprocess_files/qcplot-1.png" alt="Distributions of various QC metrics for all samples in the dataset." width="624" />
<p class="caption">
Figure 11: Distributions of various QC metrics for all samples in the dataset.
</p>
</div>
</div>
<h2 id="identifying-outliers-by-each-metric">Identifying outliers by each metric</h2>
<p>Outliers are defined based on the median absolute deviation (MADs) from the median value of each metric across all cells. We remove small and large outliers for the library size and the number of expressed features, and large outliers for the spike-in proportions and mitochondrial gene expression proportions. Removal of low-quality cells is then performed by combining the filters for all of the metrics.</p>
<p>Due to the differences in the QC metrics by <code>sample</code>, we will compute the separate outlier thresholds for each <code>sample</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>sample</span>
<span class='va'>libsize_drop</span> <span class='op'>&lt;-</span> <span class='fu'>isOutlier</span><span class='op'>(</span>
  metric <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>sum</span>, 
  nmads <span class='op'>=</span> <span class='fl'>3</span>,
  type <span class='op'>=</span> <span class='st'>"lower"</span>, 
  log <span class='op'>=</span> <span class='cn'>TRUE</span>,
  batch <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span><span class='op'>)</span>
<span class='va'>feature_drop</span> <span class='op'>&lt;-</span> <span class='fu'>isOutlier</span><span class='op'>(</span>
  metric <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>detected</span>,
  nmads <span class='op'>=</span> <span class='fl'>3</span>, 
  type <span class='op'>=</span> <span class='st'>"lower"</span>, 
  log <span class='op'>=</span> <span class='cn'>TRUE</span>,
  batch <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span><span class='op'>)</span>
<span class='va'>spike_drop</span> <span class='op'>&lt;-</span> <span class='fu'>isOutlier</span><span class='op'>(</span>
  metric <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>altexps_ERCC_percent</span>, 
  nmads <span class='op'>=</span> <span class='fl'>3</span>, 
  type <span class='op'>=</span> <span class='st'>"higher"</span>,
  batch <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span><span class='op'>)</span>
<span class='va'>mito_drop</span> <span class='op'>&lt;-</span> <span class='fu'>isOutlier</span><span class='op'>(</span>
  metric <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>subsets_Mito_percent</span>, 
  nmads <span class='op'>=</span> <span class='fl'>3</span>, 
  type <span class='op'>=</span> <span class='st'>"higher"</span>,
  batch <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span><span class='op'>)</span>
<span class='va'>ribo_drop</span> <span class='op'>&lt;-</span> <span class='fu'>isOutlier</span><span class='op'>(</span>
  metric <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>subsets_Ribo_percent</span>, 
  nmads <span class='op'>=</span> <span class='fl'>3</span>, 
  type <span class='op'>=</span> <span class='st'>"higher"</span>,
  batch <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>The following table summarises the QC cutoffs:</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>libsize_drop_df</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
  batch <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/attributes.html'>attributes</a></span><span class='op'>(</span><span class='va'>libsize_drop</span><span class='op'>)</span><span class='op'>$</span><span class='va'>thresholds</span><span class='op'>)</span>,
  cutoff <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/attributes.html'>attributes</a></span><span class='op'>(</span><span class='va'>libsize_drop</span><span class='op'>)</span><span class='op'>$</span><span class='va'>thresholds</span><span class='op'>[</span><span class='st'>"lower"</span>, <span class='op'>]</span><span class='op'>)</span>
<span class='va'>feature_drop_df</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
  batch <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/attributes.html'>attributes</a></span><span class='op'>(</span><span class='va'>feature_drop</span><span class='op'>)</span><span class='op'>$</span><span class='va'>thresholds</span><span class='op'>)</span>,
  cutoff <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/attributes.html'>attributes</a></span><span class='op'>(</span><span class='va'>feature_drop</span><span class='op'>)</span><span class='op'>$</span><span class='va'>thresholds</span><span class='op'>[</span><span class='st'>"lower"</span>, <span class='op'>]</span><span class='op'>)</span>
<span class='va'>spike_drop_df</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
  batch <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/attributes.html'>attributes</a></span><span class='op'>(</span><span class='va'>spike_drop</span><span class='op'>)</span><span class='op'>$</span><span class='va'>thresholds</span><span class='op'>)</span>,
  cutoff <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/attributes.html'>attributes</a></span><span class='op'>(</span><span class='va'>spike_drop</span><span class='op'>)</span><span class='op'>$</span><span class='va'>thresholds</span><span class='op'>[</span><span class='st'>"higher"</span>, <span class='op'>]</span><span class='op'>)</span>
<span class='va'>mito_drop_df</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
  batch <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/attributes.html'>attributes</a></span><span class='op'>(</span><span class='va'>mito_drop</span><span class='op'>)</span><span class='op'>$</span><span class='va'>thresholds</span><span class='op'>)</span>,
  cutoff <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/attributes.html'>attributes</a></span><span class='op'>(</span><span class='va'>mito_drop</span><span class='op'>)</span><span class='op'>$</span><span class='va'>thresholds</span><span class='op'>[</span><span class='st'>"higher"</span>, <span class='op'>]</span><span class='op'>)</span>
<span class='va'>ribo_drop_df</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
  batch <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/attributes.html'>attributes</a></span><span class='op'>(</span><span class='va'>ribo_drop</span><span class='op'>)</span><span class='op'>$</span><span class='va'>thresholds</span><span class='op'>)</span>,
  cutoff <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/attributes.html'>attributes</a></span><span class='op'>(</span><span class='va'>ribo_drop</span><span class='op'>)</span><span class='op'>$</span><span class='va'>thresholds</span><span class='op'>[</span><span class='st'>"higher"</span>, <span class='op'>]</span><span class='op'>)</span>
<span class='va'>qc_cutoffs_df</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/funprog.html'>Reduce</a></span><span class='op'>(</span>
  <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>y</span><span class='op'>)</span> <span class='fu'>inner_join</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>y</span>, by <span class='op'>=</span> <span class='st'>"batch"</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
    <span class='va'>libsize_drop_df</span>,
    <span class='va'>feature_drop_df</span>, 
    <span class='va'>spike_drop_df</span>, 
    <span class='va'>mito_drop_df</span>, 
    <span class='va'>ribo_drop_df</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>qc_cutoffs_df</span><span class='op'>$</span><span class='va'>batch</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>qc_cutoffs_df</span><span class='op'>$</span><span class='va'>batch</span>, <span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>qc_cutoffs_df</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>
  <span class='st'>"batch"</span>, <span class='st'>"total counts"</span>, <span class='st'>"total features"</span>, <span class='st'>"%ERCC"</span>, <span class='st'>"%mito"</span>, <span class='st'>"%ribo"</span><span class='op'>)</span>
<span class='va'>qc_cutoffs_df</span> <span class='op'>%&gt;%</span>
  <span class='fu'>arrange</span><span class='op'>(</span><span class='va'>batch</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>knitr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/knitr/man/kable.html'>kable</a></span><span class='op'>(</span>caption <span class='op'>=</span> <span class='st'>"QC cutoffs"</span>, digits <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<table>
<caption><span id="tab:unnamed-chunk-15">Table 1: </span>QC cutoffs</caption>
<thead>
<tr class="header">
<th style="text-align: left;">batch</th>
<th style="text-align: right;">total counts</th>
<th style="text-align: right;">total features</th>
<th style="text-align: right;">%ERCC</th>
<th style="text-align: right;">%mito</th>
<th style="text-align: right;">%ribo</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Blood.1</td>
<td style="text-align: right;">84066.2</td>
<td style="text-align: right;">4162.7</td>
<td style="text-align: right;">0.4</td>
<td style="text-align: right;">19.6</td>
<td style="text-align: right;">11.1</td>
</tr>
<tr class="even">
<td style="text-align: left;">Blood.2</td>
<td style="text-align: right;">23377.7</td>
<td style="text-align: right;">2306.5</td>
<td style="text-align: right;">0.4</td>
<td style="text-align: right;">17.7</td>
<td style="text-align: right;">10.7</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Blood.3</td>
<td style="text-align: right;">46396.3</td>
<td style="text-align: right;">2397.6</td>
<td style="text-align: right;">0.3</td>
<td style="text-align: right;">21.0</td>
<td style="text-align: right;">9.6</td>
</tr>
<tr class="even">
<td style="text-align: left;">Blood.4</td>
<td style="text-align: right;">74696.4</td>
<td style="text-align: right;">9801.9</td>
<td style="text-align: right;">0.7</td>
<td style="text-align: right;">41.7</td>
<td style="text-align: right;">15.9</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Blood.5</td>
<td style="text-align: right;">332219.7</td>
<td style="text-align: right;">9740.5</td>
<td style="text-align: right;">1.0</td>
<td style="text-align: right;">39.2</td>
<td style="text-align: right;">15.2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Blood.6</td>
<td style="text-align: right;">16362.9</td>
<td style="text-align: right;">4494.8</td>
<td style="text-align: right;">0.3</td>
<td style="text-align: right;">20.9</td>
<td style="text-align: right;">13.9</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Blood.7</td>
<td style="text-align: right;">313085.1</td>
<td style="text-align: right;">10542.9</td>
<td style="text-align: right;">1.1</td>
<td style="text-align: right;">16.7</td>
<td style="text-align: right;">13.8</td>
</tr>
<tr class="even">
<td style="text-align: left;">Blood.8</td>
<td style="text-align: right;">309007.5</td>
<td style="text-align: right;">11629.6</td>
<td style="text-align: right;">0.9</td>
<td style="text-align: right;">23.9</td>
<td style="text-align: right;">14.7</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Thymus.1</td>
<td style="text-align: right;">1360099.4</td>
<td style="text-align: right;">8918.1</td>
<td style="text-align: right;">1.1</td>
<td style="text-align: right;">23.9</td>
<td style="text-align: right;">11.3</td>
</tr>
<tr class="even">
<td style="text-align: left;">Thymus.2</td>
<td style="text-align: right;">479118.4</td>
<td style="text-align: right;">8922.1</td>
<td style="text-align: right;">1.1</td>
<td style="text-align: right;">15.7</td>
<td style="text-align: right;">13.1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Thymus.3</td>
<td style="text-align: right;">1685651.5</td>
<td style="text-align: right;">5981.0</td>
<td style="text-align: right;">0.6</td>
<td style="text-align: right;">16.0</td>
<td style="text-align: right;">10.5</td>
</tr>
<tr class="even">
<td style="text-align: left;">Thymus.4</td>
<td style="text-align: right;">91965.1</td>
<td style="text-align: right;">7675.5</td>
<td style="text-align: right;">1.1</td>
<td style="text-align: right;">36.6</td>
<td style="text-align: right;">12.4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Thymus.5</td>
<td style="text-align: right;">73788.6</td>
<td style="text-align: right;">7318.9</td>
<td style="text-align: right;">0.8</td>
<td style="text-align: right;">41.3</td>
<td style="text-align: right;">15.5</td>
</tr>
<tr class="even">
<td style="text-align: left;">Thymus.6</td>
<td style="text-align: right;">650569.1</td>
<td style="text-align: right;">10944.6</td>
<td style="text-align: right;">0.6</td>
<td style="text-align: right;">20.5</td>
<td style="text-align: right;">14.0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Thymus.7</td>
<td style="text-align: right;">308618.9</td>
<td style="text-align: right;">9645.2</td>
<td style="text-align: right;">1.2</td>
<td style="text-align: right;">19.6</td>
<td style="text-align: right;">15.4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Thymus.8</td>
<td style="text-align: right;">929835.5</td>
<td style="text-align: right;">11961.8</td>
<td style="text-align: right;">0.7</td>
<td style="text-align: right;">16.0</td>
<td style="text-align: right;">15.8</td>
</tr>
</tbody>
</table>
</div>
<p><strong>We would typically remove outliers (e.g., small outliers for the library size and the number of expressed features) at this stage.</strong> However, we opt to retain all samples for the time being. The following table highlights samples that would typically exclude as outliers by this criteria:</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>keep</span> <span class='op'>&lt;-</span> <span class='op'>!</span><span class='op'>(</span><span class='va'>libsize_drop</span> <span class='op'>|</span> <span class='va'>libsize_drop</span> <span class='op'>|</span> <span class='va'>mito_drop</span><span class='op'>)</span>
<span class='va'>x</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/limma/man/cbind.html'>cbind</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>sample <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>, keep <span class='op'>=</span> <span class='va'>keep</span><span class='op'>)</span>,
  <span class='fu'>colData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"plate_number"</span>, <span class='st'>"sample"</span>, <span class='st'>"sum"</span>, <span class='st'>"detected"</span><span class='op'>)</span><span class='op'>]</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/row.names.html'>row.names</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='cn'>NULL</span>
<span class='fu'>DT</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/DT/man/datatable.html'>datatable</a></span><span class='op'>(</span>
  <span class='va'>x</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span><span class='va'>keep</span><span class='op'>)</span>, <span class='op'>]</span>,
  caption <span class='op'>=</span> <span class='st'>"Quality control metrics including library size (`sum`) and number of genes detected (`keep`) for each replicate. Each replicate is annotated by whether it would be kept or excluded by the typical quality control procedures (`keep`)."</span>,
  rownames <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div id="htmlwidget-ba2140ebbea5cd027109" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-ba2140ebbea5cd027109">{"x":{"filter":"none","caption":"<caption>Quality control metrics including library size (`sum`) and number of genes detected (`keep`) for each replicate. Each replicate is annotated by whether it would be kept or excluded by the typical quality control procedures (`keep`).<\/caption>","data":[["LCE515_A12","LCE515_A15","LCE515_A18","LCE515_A21","LCE515_A6","LCE515_B8","LCE516_B15","LCE516_B16","LCE516_B17","LCE516_E5","LCE531_C11","LCE515_A10","LCE515_A11","LCE515_A13","LCE515_A14","LCE515_A16","LCE515_A17","LCE515_A19","LCE515_A20","LCE515_A22","LCE515_A5","LCE515_A7","LCE515_A8","LCE515_A9","LCE515_B10","LCE515_B11","LCE515_B12","LCE515_B3","LCE515_B4","LCE515_B5","LCE515_B6","LCE515_B7","LCE515_B9","LCE516_B10","LCE516_B11","LCE516_B3","LCE516_B4","LCE516_B5","LCE516_B9","LCE516_C10","LCE516_C11","LCE516_C12","LCE516_E10","LCE516_E11","LCE516_E15","LCE516_E16","LCE516_E17","LCE516_E3","LCE516_E4","LCE516_E9","LCE516_F10","LCE516_F11","LCE516_F12","LCE531_C13","LCE531_C3","LCE531_C4","LCE531_C5","LCE531_C6","LCE531_C7","LCE531_C8","LCE531_D3","LCE531_D4","LCE531_D5","LCE531_F3","LCE531_F4","LCE531_F5","LCE531_F6","LCE531_F7","LCE531_F8","LCE531_G3","LCE531_G4","LCE531_G5","LCE531_I3","LCE531_I4","LCE531_I5","LCE531_I6","LCE531_I7","LCE531_I8","LCE531_J3","LCE531_J4","LCE531_J5","LCE516_B12","LCE516_B13","LCE516_B14","LCE516_B18","LCE516_B19","LCE516_B20","LCE516_B6","LCE516_B7","LCE516_B8","LCE516_C13","LCE516_C14","LCE516_C15","LCE516_E12","LCE516_E13","LCE516_E14","LCE516_E18","LCE516_E19","LCE516_E20","LCE516_E6","LCE516_E7","LCE516_E8","LCE516_F13","LCE516_F14","LCE516_F15","LCE531_C10","LCE531_C9","LCE531_D6","LCE531_D7","LCE531_D8","LCE531_F10","LCE531_F11","LCE531_F12","LCE531_F9","LCE531_G6","LCE531_G7","LCE531_G8","LCE531_I9","LCE531_J6","LCE531_J7","LCE531_J8"],[false,false,false,false,false,false,false,false,false,false,false,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true],["LCE515","LCE515","LCE515","LCE515","LCE515","LCE515","LCE516","LCE516","LCE516","LCE516","LCE531","LCE515","LCE515","LCE515","LCE515","LCE515","LCE515","LCE515","LCE515","LCE515","LCE515","LCE515","LCE515","LCE515","LCE515","LCE515","LCE515","LCE515","LCE515","LCE515","LCE515","LCE515","LCE515","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE516","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531","LCE531"],["Thymus.2","Thymus.2","Thymus.2","Thymus.2","Thymus.1","Thymus.3","Thymus.4","Thymus.4","Thymus.4","Thymus.5","Thymus.6","Blood.1","Thymus.2","Thymus.2","Thymus.2","Thymus.2","Thymus.2","Thymus.2","Thymus.2","Thymus.2","Thymus.1","Thymus.1","Blood.1","Blood.1","Blood.3","Blood.3","Blood.3","Blood.2","Blood.2","Blood.2","Blood.2","Thymus.3","Thymus.3","Thymus.4","Thymus.4","Thymus.4","Thymus.4","Thymus.4","Thymus.4","Blood.4","Blood.4","Blood.4","Thymus.5","Thymus.5","Thymus.5","Thymus.5","Thymus.5","Thymus.5","Thymus.5","Thymus.5","Blood.5","Blood.5","Blood.5","Thymus.6","Thymus.6","Thymus.6","Thymus.6","Thymus.6","Thymus.6","Thymus.6","Blood.6","Blood.6","Blood.6","Thymus.7","Thymus.7","Thymus.7","Thymus.7","Thymus.7","Thymus.7","Blood.7","Blood.7","Blood.7","Thymus.8","Thymus.8","Thymus.8","Thymus.8","Thymus.8","Thymus.8","Blood.8","Blood.8","Blood.8","Thymus.4","Thymus.4","Thymus.4","Thymus.4","Thymus.4","Thymus.4","Thymus.4","Thymus.4","Thymus.4","Blood.4","Blood.4","Blood.4","Thymus.5","Thymus.5","Thymus.5","Thymus.5","Thymus.5","Thymus.5","Thymus.5","Thymus.5","Thymus.5","Blood.5","Blood.5","Blood.5","Thymus.6","Thymus.6","Blood.6","Blood.6","Blood.6","Thymus.7","Thymus.7","Thymus.7","Thymus.7","Blood.7","Blood.7","Blood.7","Thymus.8","Blood.8","Blood.8","Blood.8"],[1372900,1305852,1568994,1016681,1381919,523320,1652853,1555733,1588530,3245,547183,120641,2430027,1732263,3033470,1843804,3124035,2533282,4651568,2880820,1376983,1716338,150375,133966,118388,266570,95905,249802,829664,90478,200319,1951189,1899492,542211,839835,1121242,810636,890186,951925,2772171,2536879,1831875,1250972,1130922,2055044,331276,1213393,942964,952360,1011627,1531076,1038702,1888530,1934130,4251176,3072274,2475733,2338002,1994646,2243920,1466074,1816700,336552,2117283,1971886,2233947,1867728,1857667,1033991,1421637,1774080,2411069,2414149,2778358,1703202,1754545,2026303,2487620,2194206,2004536,2337874,329035,406008,561083,492688,461230,422681,472374,420963,332548,323528,702630,919764,595767,723718,194476,694286,423124,137514,621577,499389,320490,691805,684101,695011,1245841,785064,166162,130090,229533,839936,525478,1242536,776999,1010677,689050,970072,959859,1074033,1142609,906095],[10964,9057,10368,8657,9649,6536,10636,10245,10222,544,10850,4046,12835,11484,13242,11951,13092,12575,14673,12959,9872,10645,5328,5092,5071,6801,4285,6663,8818,3860,5674,12667,11037,8067,9697,11435,9579,9280,9281,12183,12535,11633,10766,10264,12079,7088,10119,9774,10396,9951,10620,10131,11200,12166,14593,13672,12956,12635,12389,12770,10437,11575,8827,11845,12488,12967,11617,12976,10458,11497,11923,12737,13125,13347,12361,12518,12933,12708,11990,12190,12473,9029,9629,10613,10566,10484,9812,10155,10756,8702,9394,11196,12026,11371,11273,7820,12056,10488,7157,10570,10176,9052,11042,11607,11309,13287,12805,7803,6807,8345,12085,11174,13507,12674,12769,11915,12543,13096,12135,12234,11819]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>sample<\/th>\n      <th>keep<\/th>\n      <th>plate_number<\/th>\n      <th>sample<\/th>\n      <th>sum<\/th>\n      <th>detected<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[4,5]}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<h2 id="summary-3">Summary</h2>
<p>We have opted not to exclude any replicates at this stage. However, we have identified some replicates with unusually small library sizes and number of genes detected compared to the other samples. We will keep an eye on these replicates and may later remove some of these.</p>
<h1 id="examining-gene-level-metrics">Examining gene-level metrics</h1>
<h2 id="inspecting-the-most-highly-expressed-genes">Inspecting the most highly expressed genes</h2>
<p>Figure <a href="#fig:topgenes">12</a> shows the most highly expressed genes across the cell population in the combined data set. Some of them are mitochondrial genes, matching what weve already seen in the QC metrics, and ribosomal protein genes, which commonly are amongst the most highly expressed genes in scRNA-seq data.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>plotHighestExprs</span><span class='op'>(</span><span class='va'>sce</span>, n <span class='op'>=</span> <span class='fl'>50</span>, exprs_values <span class='op'>=</span> <span class='st'>"read_counts"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:topgenes"></span>
<img src="C094_Pellicci.mini-bulk.preprocess_files/topgenes-1.png" alt="Percentage of total counts assigned to the top 50 most highly-abundant features in the combined data set. For each feature, each bar represents the percentage assigned to that feature for a single cell, while the circle represents the average across all cells." width="624" />
<p class="caption">
Figure 12: Percentage of total counts assigned to the top 50 most highly-abundant features in the combined data set. For each feature, each bar represents the percentage assigned to that feature for a single cell, while the circle represents the average across all cells.
</p>
</div>
</div>
<p>Figure <a href="#fig:topgenes-filtered">13</a> shows the most highly expressed endogenous genes after excluding the mitochondrial and ribosomal protein genes.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>plotHighestExprs</span><span class='op'>(</span>
  <span class='va'>sce</span>,
  drop_features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>%in%</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='va'>mito_set</span>, <span class='va'>ribo_set</span><span class='op'>)</span>,
  n <span class='op'>=</span> <span class='fl'>50</span>,
  exprs_values <span class='op'>=</span> <span class='st'>"read_counts"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:topgenes-filtered"></span>
<img src="C094_Pellicci.mini-bulk.preprocess_files/topgenes-filtered-1.png" alt="Percentage of total counts assigned to the top 50 most highly-abundant features in the combined data set. For each feature, each bar represents the percentage assigned to that feature for a single cell, while the circle represents the average across all cells. Bars are coloured by the total number of expressed features in each cell, while circles are coloured according to whether the feature is labelled as a control feature." width="624" />
<p class="caption">
Figure 13: Percentage of total counts assigned to the top 50 most highly-abundant features in the combined data set. For each feature, each bar represents the percentage assigned to that feature for a single cell, while the circle represents the average across all cells. Bars are coloured by the total number of expressed features in each cell, while circles are coloured according to whether the feature is labelled as a control feature.
</p>
</div>
</div>
<h2 id="sex-check">Sex check</h2>
<h3 id="using-xist-expression">Using <em>XIST</em> expression</h3>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"logCPM_reads"</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'>edgeR</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/cpm.html'>cpm</a></span><span class='op'>(</span><span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"read_counts"</span><span class='op'>)</span>, log <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/stopifnot.html'>stopifnot</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/base/identical.html'>identical</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/which.min.html'>which.min</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>nreads</span><span class='op'>)</span><span class='op'>)</span>, 
    <span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/which.min.html'>which.min</a></span><span class='op'>(</span><span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"logCPM_reads"</span><span class='op'>)</span><span class='op'>[</span><span class='st'>"XIST"</span>, <span class='op'>]</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>Figure <a href="#fig:xist-expression">14</a> visualises the expression of <em>XIST</em>, a female-specific gene. This can be used to confirm the reported sex of the sample and to detect sample mix-ups.</p>
<p>The results are fairly clear for the <code>Thymus</code> samples with donors 3, 4, and 6 having lower <em>XIST</em> expression (i.e.male). There is also one donor 5 replicate with low <em>XIST</em> expression (LCE516_E5), but this is also the technical replicate with the fewest reads (<span class="math inline">\(n_{reads} =\)</span> 24,379).</p>
<p>The results are less clear for the <code>Blood</code> samples with some of the <code>Blood.3</code> technical replicates having comparable <em>XIST</em> expression to replicates from nominally female donors 1 and 2.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>plotExpression</span><span class='op'>(</span>
  <span class='va'>sce</span>,
  <span class='st'>"XIST"</span>,
  x <span class='op'>=</span> <span class='st'>"donor"</span>,
  exprs_values <span class='op'>=</span> <span class='st'>"logCPM_reads"</span>,
  colour_by <span class='op'>=</span> <span class='st'>"donor"</span>,
  other_fields <span class='op'>=</span> <span class='st'>"tissue"</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>donor_colours</span>, name <span class='op'>=</span> <span class='st'>"donor"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>fill <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"logCPM(read counts)"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/lims.html'>ylim</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='cn'>NA</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_grid.html'>facet_grid</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>tissue</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>""</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ggtitle</a></span><span class='op'>(</span><span class='st'>"XIST"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:xist-expression"></span>
<img src="C094_Pellicci.mini-bulk.preprocess_files/xist-expression-1.png" alt="Expression of *XIST* by donor and tissue on a 'logCPM' scale using the read counts." width="624" />
<p class="caption">
Figure 14: Expression of <em>XIST</em> by donor and tissue on a logCPM scale using the read counts.
</p>
</div>
</div>
<h3 id="using-y-chromosome-gene-expression">Using Y chromosome gene expression</h3>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>is_Y</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/any.html'>any</a></span><span class='op'>(</span><span class='fu'>rowData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>$</span><span class='va'>ENSEMBL.SEQNAME</span> <span class='op'>==</span> <span class='st'>"Y"</span><span class='op'>)</span>
<span class='va'>sce_Y</span> <span class='op'>&lt;-</span> <span class='fu'>addPerCellQC</span><span class='op'>(</span>
  <span class='va'>sce</span>, 
  subsets <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>Y <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span><span class='va'>is_Y</span><span class='op'>)</span><span class='op'>)</span>,
  use.altexps <span class='op'>=</span> <span class='st'>"ERCC"</span>,
  assay.type <span class='op'>=</span> <span class='st'>"UMI_counts"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>Figure <a href="#fig:chr-Y-expression">15</a> visualises the proportion of read counts coming from Y chromosome genes. This can be used to confirm the reported sex of the sample and to detect sample mix-ups. The results in Figure <a href="#fig:chr-Y-expression">15</a> are consistent with those in Figure <a href="#fig:xist-expression">14</a>.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce_Y</span>,
  <span class='st'>"subsets_Y_percent"</span>,
  x <span class='op'>=</span> <span class='st'>"donor"</span>,
  other_fields <span class='op'>=</span> <span class='st'>"tissue"</span>,
  colour_by <span class='op'>=</span> <span class='st'>"donor"</span>,
  point_alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/lims.html'>ylim</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='cn'>NA</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>donor_colours</span>, name <span class='op'>=</span> <span class='st'>"donor"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>""</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_grid.html'>facet_grid</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>tissue</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guide_legend.html'>guide_legend</a></span><span class='op'>(</span>override.aes <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>4</span>, alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:chr-Y-expression"></span>
<img src="C094_Pellicci.mini-bulk.preprocess_files/chr-Y-expression-1.png" alt="Proportion of read counts coming from Y chromosome genes by donor and tissue." width="624" />
<p class="caption">
Figure 15: Proportion of read counts coming from Y chromosome genes by donor and tissue.
</p>
</div>
</div>
<h1 id="concluding-remarks">Concluding remarks</h1>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># Remove some unneeded assays</span>
<span class='fu'>assay</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"logCPM_reads"</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='cn'>NULL</span>

<span class='co'># Save</span>
<span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span>
  <span class='va'>sce</span>,
  <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"data"</span>, <span class='st'>"SCEs"</span>, <span class='st'>"C094_Pellicci.mini-bulk.preprocessed.SCE.rds"</span><span class='op'>)</span>,
  compress <span class='op'>=</span> <span class='st'>"xz"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>The processed <em>SingleCellExperiment</em> object is available (see <a href="../data/SCEs/C094_Pellicci.mini-bulk.preprocessed.SCE.rds"><code>data/SCEs/C094_Pellicci.mini-bulk.preprocessed.SCE.rds</code></a>). This will be used in downstream analyses, i.e.differential expression analysis.</p>
<h1 class="appendix" id="additional-information">Additional information</h1>
<p>The following are available on request:</p>
<ul>
<li>Full CSV tables of any data presented.</li>
<li>PDF/PNG files of any static plots.</li>
</ul>
<h1 class="appendix" id="session-info">Session info</h1>
<summary>
The analysis and this document were prepared using the following software (click triangle to expand)
</summary>
<details>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>sessioninfo</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/sessioninfo/man/session_info.html'>session_info</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<pre><code> Session info 
 setting  value                       
 version  R version 4.0.3 (2020-10-10)
 os       CentOS Linux 7 (Core)       
 system   x86_64, linux-gnu           
 ui       X11                         
 language (EN)                        
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 
 tz       Australia/Melbourne         
 date     2021-08-23                  

 Packages 
 ! package                           * version  date       lib
 P AnnotationDbi                     * 1.52.0   2020-10-27 [?]
 P AnnotationFilter                  * 1.14.0   2020-10-27 [?]
 P AnnotationHub                     * 2.22.0   2020-10-27 [?]
 P askpass                             1.1      2019-01-13 [?]
 P assertthat                          0.2.1    2019-03-21 [?]
 P beachmat                            2.6.4    2020-12-20 [?]
 P beeswarm                            0.3.1    2021-03-07 [?]
 P Biobase                           * 2.50.0   2020-10-27 [?]
 P BiocFileCache                     * 1.14.0   2020-10-27 [?]
 P BiocGenerics                      * 0.36.0   2020-10-27 [?]
 P BiocManager                         1.30.12  2021-03-28 [?]
 P BiocNeighbors                       1.8.2    2020-12-07 [?]
 P BiocParallel                        1.24.1   2020-11-06 [?]
 P BiocSingular                        1.6.0    2020-10-27 [?]
 P BiocStyle                         * 2.18.1   2020-11-24 [?]
 P BiocVersion                         3.12.0   2020-04-27 [?]
 P biomaRt                             2.46.3   2021-02-09 [?]
 P Biostrings                          2.58.0   2020-10-27 [?]
 P bit                                 4.0.4    2020-08-04 [?]
 P bit64                               4.0.5    2020-08-30 [?]
 P bitops                              1.0-6    2013-08-17 [?]
 P blob                                1.2.1    2020-01-20 [?]
 P bslib                               0.2.4    2021-01-25 [?]
 P cachem                              1.0.4    2021-02-13 [?]
 P cellranger                          1.1.0    2016-07-27 [?]
 P cli                                 2.4.0    2021-04-05 [?]
 P colorspace                          2.0-0    2020-11-11 [?]
 P cowplot                           * 1.1.1    2020-12-30 [?]
 P crayon                              1.4.1    2021-02-08 [?]
 P crosstalk                           1.1.1    2021-01-12 [?]
 P curl                                4.3      2019-12-02 [?]
 P DBI                                 1.1.1    2021-01-15 [?]
 P dbplyr                            * 2.1.0    2021-02-03 [?]
 P DelayedArray                        0.16.3   2021-03-24 [?]
 P DelayedMatrixStats                  1.12.3   2021-02-03 [?]
 P DEoptimR                            1.0-8    2016-11-19 [?]
 P digest                              0.6.27   2020-10-24 [?]
 P distill                             1.2      2021-01-13 [?]
 P downlit                             0.2.1    2020-11-04 [?]
 P dplyr                             * 1.0.5    2021-03-05 [?]
 P DT                                  0.17     2021-01-06 [?]
 P edgeR                               3.32.1   2021-01-14 [?]
 P ellipsis                            0.3.1    2020-05-15 [?]
 P ensembldb                         * 2.14.0   2020-10-27 [?]
 P evaluate                            0.14     2019-05-28 [?]
 P fansi                               0.4.2    2021-01-15 [?]
 P farver                              2.1.0    2021-02-28 [?]
 P fastmap                             1.1.0    2021-01-25 [?]
 P generics                            0.1.0    2020-10-31 [?]
 P GenomeInfoDb                      * 1.26.4   2021-03-10 [?]
 P GenomeInfoDbData                    1.2.4    2020-12-07 [?]
 P GenomicAlignments                   1.26.0   2020-10-27 [?]
 P GenomicFeatures                   * 1.42.3   2021-04-01 [?]
 P GenomicRanges                     * 1.42.0   2020-10-27 [?]
 P GGally                              2.1.1    2021-03-08 [?]
 P ggbeeswarm                          0.6.0    2017-08-07 [?]
 P ggplot2                           * 3.3.3    2020-12-30 [?]
 P glue                                1.4.2    2020-08-27 [?]
 P GO.db                             * 3.12.1   2021-01-29 [?]
 P graph                               1.68.0   2020-10-27 [?]
 P gridExtra                           2.3      2017-09-09 [?]
 P gtable                              0.3.0    2019-03-25 [?]
 P here                              * 1.0.1    2020-12-13 [?]
   highr                               0.9      2021-04-16 [1]
 P hms                                 1.0.0    2021-01-13 [?]
 P Homo.sapiens                      * 1.3.1    2021-02-18 [?]
 P htmltools                           0.5.1.1  2021-01-22 [?]
 P htmlwidgets                         1.5.3    2020-12-10 [?]
 P httpuv                              1.5.5    2021-01-13 [?]
 P httr                                1.4.2    2020-07-20 [?]
 P interactiveDisplayBase              1.28.0   2020-10-27 [?]
 P IRanges                           * 2.24.1   2020-12-12 [?]
 P irlba                               2.3.3    2019-02-05 [?]
 P janitor                           * 2.1.0    2021-01-05 [?]
 P jquerylib                           0.1.3    2020-12-17 [?]
 P jsonlite                            1.7.2    2020-12-09 [?]
   knitr                               1.33     2021-04-24 [1]
 P labeling                            0.4.2    2020-10-20 [?]
 P later                               1.1.0.1  2020-06-05 [?]
   lattice                             0.20-41  2020-04-02 [2]
 P lazyeval                            0.2.2    2019-03-15 [?]
 P lifecycle                           1.0.0    2021-02-15 [?]
 P limma                             * 3.46.0   2020-10-27 [?]
 P locfit                              1.5-9.4  2020-03-25 [?]
 P lubridate                           1.7.10   2021-02-26 [?]
 P magrittr                            2.0.1    2020-11-17 [?]
   Matrix                              1.2-18   2019-11-27 [2]
 P MatrixGenerics                    * 1.2.1    2021-01-30 [?]
 P matrixStats                       * 0.58.0   2021-01-29 [?]
 P mclust                              5.4.7    2020-11-20 [?]
 P memoise                             2.0.0    2021-01-26 [?]
   mime                                0.11     2021-06-23 [1]
 P msigdbr                           * 7.2.1    2020-10-02 [?]
 P munsell                             0.5.0    2018-06-12 [?]
 P openssl                             1.4.3    2020-09-18 [?]
 P org.Hs.eg.db                      * 3.12.0   2020-12-08 [?]
 P org.Mm.eg.db                        3.12.0   2020-12-08 [?]
 P OrganismDbi                       * 1.32.0   2020-10-27 [?]
 P patchwork                         * 1.1.1    2020-12-17 [?]
 P pillar                              1.5.1    2021-03-05 [?]
 P pkgconfig                           2.0.3    2019-09-22 [?]
 P plyr                                1.8.6    2020-03-03 [?]
 P prettyunits                         1.1.1    2020-01-24 [?]
 P progress                            1.2.2    2019-05-16 [?]
 P promises                            1.2.0.1  2021-02-11 [?]
 P ProtGenerics                        1.22.0   2020-10-27 [?]
 P purrr                               0.3.4    2020-04-17 [?]
 P R6                                  2.5.0    2020-10-28 [?]
 P rappdirs                            0.3.3    2021-01-31 [?]
 P RBGL                                1.66.0   2020-10-27 [?]
 P RColorBrewer                        1.1-2    2014-12-07 [?]
 P Rcpp                                1.0.6    2021-01-15 [?]
 P RCurl                               1.98-1.3 2021-03-16 [?]
 P readxl                            * 1.3.1    2019-03-13 [?]
 P reshape                             0.8.8    2018-10-23 [?]
 P Rhtslib                             1.22.0   2020-10-27 [?]
   rlang                               0.4.11   2021-04-30 [1]
 P rmarkdown                           2.7      2021-02-19 [?]
 P robustbase                          0.93-7   2021-01-04 [?]
 P rprojroot                           2.0.2    2020-11-15 [?]
 P Rsamtools                           2.6.0    2020-10-27 [?]
 P RSQLite                             2.2.5    2021-03-27 [?]
 P rstudioapi                          0.13     2020-11-12 [?]
 P rsvd                                1.0.3    2020-02-17 [?]
 P rtracklayer                         1.50.0   2020-10-27 [?]
 P S4Vectors                         * 0.28.1   2020-12-09 [?]
 P sass                                0.3.1    2021-01-24 [?]
 P scales                            * 1.1.1    2020-05-11 [?]
 P scater                            * 1.18.6   2021-02-26 [?]
 P scPipe                            * 1.12.0   2020-10-27 [?]
 P scuttle                             1.0.4    2020-12-17 [?]
 P sessioninfo                         1.1.1    2018-11-05 [?]
 P shiny                               1.6.0    2021-01-25 [?]
 P SingleCellExperiment              * 1.12.0   2020-10-27 [?]
 P snakecase                           0.11.0   2019-05-25 [?]
 P sparseMatrixStats                   1.2.1    2021-02-02 [?]
   stringi                             1.7.3    2021-07-16 [1]
 P stringr                             1.4.0    2019-02-10 [?]
 P SummarizedExperiment              * 1.20.0   2020-10-27 [?]
 P tibble                              3.1.0    2021-02-25 [?]
 P tidyselect                          1.1.0    2020-05-11 [?]
 P TxDb.Hsapiens.UCSC.hg19.knownGene * 3.2.2    2021-02-18 [?]
 P utf8                                1.2.1    2021-03-12 [?]
 P vctrs                               0.3.7    2021-03-29 [?]
 P vipor                               0.4.5    2017-03-22 [?]
 P viridis                             0.5.1    2018-03-29 [?]
 P viridisLite                         0.3.0    2018-02-01 [?]
 P withr                               2.4.1    2021-01-26 [?]
   xfun                                0.24     2021-06-15 [1]
 P XML                                 3.99-0.6 2021-03-16 [?]
 P xml2                                1.3.2    2020-04-23 [?]
 P xtable                              1.8-4    2019-04-21 [?]
 P XVector                             0.30.0   2020-10-27 [?]
 P yaml                                2.2.1    2020-02-01 [?]
 P zlibbioc                            1.36.0   2020-10-27 [?]
 source        
 Bioconductor  
 Bioconductor  
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 Bioconductor  
 CRAN (R 4.0.2)
 Bioconductor  
 Bioconductor  
 Bioconductor  
 CRAN (R 4.0.3)
 Bioconductor  
 Bioconductor  
 Bioconductor  
 Bioconductor  
 Bioconductor  
 Bioconductor  
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 CRAN (R 4.0.2)
 CRAN (R 4.0.0)
 CRAN (R 4.0.3)
 CRAN (R 4.0.2)
 CRAN (R 4.0.0)
 CRAN (R 4.0.3)
 CRAN (R 4.0.2)
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 CRAN (R 4.0.2)
 Bioconductor  
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 CRAN (R 4.0.2)
 Bioconductor  
 CRAN (R 4.0.0)
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 CRAN (R 4.0.2)
 CRAN (R 4.0.2)
 CRAN (R 4.0.0)
 Bioconductor  
 Bioconductor  
 Bioconductor  
 Bioconductor  
 Bioconductor  
 CRAN (R 4.0.3)
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 CRAN (R 4.0.0)
 Bioconductor  
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.3)
 CRAN (R 4.0.2)
 Bioconductor  
 CRAN (R 4.0.2)
 CRAN (R 4.0.2)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 Bioconductor  
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.3)
 CRAN (R 4.0.2)
 CRAN (R 4.0.2)
 CRAN (R 4.0.3)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.3)
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.3)
 CRAN (R 4.0.2)
 CRAN (R 4.0.3)
 Bioconductor  
 CRAN (R 4.0.2)
 CRAN (R 4.0.2)
 CRAN (R 4.0.2)
 CRAN (R 4.0.3)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 Bioconductor  
 Bioconductor  
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 CRAN (R 4.0.5)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 Bioconductor  
 CRAN (R 4.0.3)
 CRAN (R 4.0.2)
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 Bioconductor  
 CRAN (R 4.0.3)
 CRAN (R 4.0.2)
 CRAN (R 4.0.0)
 Bioconductor  
 Bioconductor  
 CRAN (R 4.0.2)
 CRAN (R 4.0.0)
 Bioconductor  
 Bioconductor  
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 Bioconductor  
 CRAN (R 4.0.0)
 Bioconductor  
 CRAN (R 4.0.3)
 CRAN (R 4.0.0)
 Bioconductor  
 CRAN (R 4.0.2)
 CRAN (R 4.0.0)
 Bioconductor  
 CRAN (R 4.0.2)
 CRAN (R 4.0.3)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 CRAN (R 4.0.3)
 CRAN (R 4.0.5)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 Bioconductor  
 CRAN (R 4.0.0)
 Bioconductor  

[1] /stornext/Home/data/allstaff/h/ho.w/C094_Pellicci/renv/library/R-4.0/x86_64-pc-linux-gnu
[2] /stornext/System/data/apps/R/R-4.0.3/lib64/R/library

 P  Loaded and on-disk path mismatch.</code></pre>
</div>
</details>
<div class="sourceCode" id="cb2"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<section class="footnotes" role="doc-endnotes">
<hr />
<ol>
<li id="fn1" role="doc-endnote"><p>Some care is taken to account for missing and duplicate gene symbols; missing symbols are replaced with the Ensembl identifier and duplicated symbols are concatenated with the (unique) Ensembl identifiers.<a href="#fnref1" class="footnote-back" role="doc-backlink"></a></p></li>
<li id="fn2" role="doc-endnote"><p>I.e. the library size when using read counts<a href="#fnref2" class="footnote-back" role="doc-backlink"></a></p></li>
<li id="fn3" role="doc-endnote"><p>The choice of a 6 bp UMI in the mini-bulk protocol is because it is adapted from SCOREs CEL-Seq2 single-cell RNA-sequencing protocol. With scRNA-seq data it is very rare to observe more than <span class="math inline">\(4^k = 4^6 = 4096\)</span> unique molecules for any one gene.<a href="#fnref3" class="footnote-back" role="doc-backlink"></a></p></li>
<li id="fn4" role="doc-endnote"><p>The number of expressed genes refers to the number of genes which have non-zero counts (ie. they have been identified in the cell at least once)<a href="#fnref4" class="footnote-back" role="doc-backlink"></a></p></li>
<li id="fn5" role="doc-endnote"><p>This is consistent with the use of UMI counts rather than read counts, as each transcript molecule can only produce one UMI count but can yield many reads after fragmentation.<a href="#fnref5" class="footnote-back" role="doc-backlink"></a></p></li>
<li id="fn6" role="doc-endnote"><p>Bearing in mind that the experimental cells, which are  T cells, have a smaller transcriptome than the control cells.<a href="#fnref6" class="footnote-back" role="doc-backlink"></a></p></li>
</ol>
</section>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
<h3 id="references">References</h3>
<div id="references-listing"></div>
</div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
