<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #20794d; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007ba5; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #007ba5; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #007ba5; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #20794d; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

  <!--radix_placeholder_meta_tags-->
  <title>C094: Preprocessing the Pellicci gamma-delta T-cell single-cell data set</title>



  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2021-08-05"/>
  <meta property="article:created" itemprop="dateCreated" content="2021-08-05"/>
  <meta name="article:author" content="Peter Hickey"/>

  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="C094: Preprocessing the Pellicci gamma-delta T-cell single-cell data set"/>
  <meta property="og:type" content="article"/>
  <meta property="og:locale" content="en_US"/>
  <meta property="og:site_name" content="C094"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="C094: Preprocessing the Pellicci gamma-delta T-cell single-cell data set"/>

  <!--/radix_placeholder_meta_tags-->
  
  <meta name="citation_reference" content="citation_title=Computational and analytical challenges in single-cell transcriptomics;citation_publication_date=2015;citation_volume=16;citation_author=O. Stegle;citation_author=S. A. Teichmann;citation_author=J. C. Marioni"/>
  <meta name="citation_reference" content="citation_title=A scaling normalization method for differential expression analysis of RNA-seq data;citation_publication_date=2010;citation_volume=11;citation_author=M. D. Robinson;citation_author=A. Oshlack"/>
  <meta name="citation_reference" content="citation_publication_date=2016;citation_volume=17;citation_author=A. T. Lun;citation_author=K. Bach;citation_author=J. C. Marioni"/>
  <meta name="citation_reference" content="citation_title=Independent filtering increases detection power for high-throughput experiments;citation_publication_date=2010;citation_volume=107;citation_author=R. Bourgon;citation_author=R. Gentleman;citation_author=W. Huber"/>
  <meta name="citation_reference" content="citation_title=Assessing the reliability of spike-in normalization for analyses of single-cell RNA sequencing data;citation_publication_date=2017;citation_volume=27;citation_author=A. T. L. Lun;citation_author=F. J. Calero-Nieto;citation_author=L. Haim-Vilmovsky;citation_author=B. Gottgens;citation_author=J. C. Marioni"/>
  <meta name="citation_reference" content="citation_title=UMAP: Uniform manifold approximation and projection for dimension reduction;citation_publication_date=2018;citation_author=Leland McInnes;citation_author=John Healy;citation_author=James Melville"/>
  <meta name="citation_reference" content="citation_title=A step-by-step workflow for low-level analysis of single-cell RNA-seq data;citation_publication_date=2016;citation_volume=5;citation_author=A. T. L. Lun;citation_author=D. J. McCarthy;citation_author=J. C. Marioni"/>
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","author","date","editor_options","bibliography"]}},"value":[{"type":"character","attributes":{},"value":["Preprocessing the Pellicci gamma-delta T-cell single-cell data set"]},{"type":"NULL"},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","url","affiliation","affiliation_url"]}},"value":[{"type":"character","attributes":{},"value":["Peter Hickey"]},{"type":"character","attributes":{},"value":["https://peterhickey.org"]},{"type":"character","attributes":{},"value":["Single Cell Open Research Endeavour (SCORE), WEHI"]},{"type":"character","attributes":{},"value":["https://www.wehi.edu.au/people/rory-bowden/4536/wehi-advanced-genomics-facility"]}]}]},{"type":"character","attributes":{},"value":["2021-08-05"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["chunk_output_type"]}},"value":[{"type":"character","attributes":{},"value":["console"]}]},{"type":"character","attributes":{},"value":["ref.bib"]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  <!--radix_placeholder_navigation_in_header-->
  <meta name="distill:offset" content=""/>

  <script type="application/javascript">

    window.headroom_prevent_pin = false;

    window.document.addEventListener("DOMContentLoaded", function (event) {

      // initialize headroom for banner
      var header = $('header').get(0);
      var headerHeight = header.offsetHeight;
      var headroom = new Headroom(header, {
        tolerance: 5,
        onPin : function() {
          if (window.headroom_prevent_pin) {
            window.headroom_prevent_pin = false;
            headroom.unpin();
          }
        }
      });
      headroom.init();
      if(window.location.hash)
        headroom.unpin();
      $(header).addClass('headroom--transition');

      // offset scroll location for banner on hash change
      // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
      window.addEventListener("hashchange", function(event) {
        window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
      });

      // responsive menu
      $('.distill-site-header').each(function(i, val) {
        var topnav = $(this);
        var toggle = topnav.find('.nav-toggle');
        toggle.on('click', function() {
          topnav.toggleClass('responsive');
        });
      });

      // nav dropdowns
      $('.nav-dropbtn').click(function(e) {
        $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
        $(this).parent().siblings('.nav-dropdown')
           .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $("body").click(function(e){
        $('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $(".nav-dropdown").click(function(e){
        e.stopPropagation();
      });
    });
  </script>

  <style type="text/css">

  /* Theme (user-documented overrideables for nav appearance) */

  .distill-site-nav {
    color: rgba(255, 255, 255, 0.8);
    background-color: #0F2E3D;
    font-size: 15px;
    font-weight: 300;
  }

  .distill-site-nav a {
    color: inherit;
    text-decoration: none;
  }

  .distill-site-nav a:hover {
    color: white;
  }

  @media print {
    .distill-site-nav {
      display: none;
    }
  }

  .distill-site-header {

  }

  .distill-site-footer {

  }


  /* Site Header */

  .distill-site-header {
    width: 100%;
    box-sizing: border-box;
    z-index: 3;
  }

  .distill-site-header .nav-left {
    display: inline-block;
    margin-left: 8px;
  }

  @media screen and (max-width: 768px) {
    .distill-site-header .nav-left {
      margin-left: 0;
    }
  }


  .distill-site-header .nav-right {
    float: right;
    margin-right: 8px;
  }

  .distill-site-header a,
  .distill-site-header .title {
    display: inline-block;
    text-align: center;
    padding: 14px 10px 14px 10px;
  }

  .distill-site-header .title {
    font-size: 18px;
    min-width: 150px;
  }

  .distill-site-header .logo {
    padding: 0;
  }

  .distill-site-header .logo img {
    display: none;
    max-height: 20px;
    width: auto;
    margin-bottom: -4px;
  }

  .distill-site-header .nav-image img {
    max-height: 18px;
    width: auto;
    display: inline-block;
    margin-bottom: -3px;
  }



  @media screen and (min-width: 1000px) {
    .distill-site-header .logo img {
      display: inline-block;
    }
    .distill-site-header .nav-left {
      margin-left: 20px;
    }
    .distill-site-header .nav-right {
      margin-right: 20px;
    }
    .distill-site-header .title {
      padding-left: 12px;
    }
  }


  .distill-site-header .nav-toggle {
    display: none;
  }

  .nav-dropdown {
    display: inline-block;
    position: relative;
  }

  .nav-dropdown .nav-dropbtn {
    border: none;
    outline: none;
    color: rgba(255, 255, 255, 0.8);
    padding: 16px 10px;
    background-color: transparent;
    font-family: inherit;
    font-size: inherit;
    font-weight: inherit;
    margin: 0;
    margin-top: 1px;
    z-index: 2;
  }

  .nav-dropdown-content {
    display: none;
    position: absolute;
    background-color: white;
    min-width: 200px;
    border: 1px solid rgba(0,0,0,0.15);
    border-radius: 4px;
    box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
    z-index: 1;
    margin-top: 2px;
    white-space: nowrap;
    padding-top: 4px;
    padding-bottom: 4px;
  }

  .nav-dropdown-content hr {
    margin-top: 4px;
    margin-bottom: 4px;
    border: none;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .nav-dropdown-active {
    display: block;
  }

  .nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
    color: black;
    padding: 6px 24px;
    text-decoration: none;
    display: block;
    text-align: left;
  }

  .nav-dropdown-content .nav-dropdown-header {
    display: block;
    padding: 5px 24px;
    padding-bottom: 0;
    text-transform: uppercase;
    font-size: 14px;
    color: #999999;
    white-space: nowrap;
  }

  .nav-dropdown:hover .nav-dropbtn {
    color: white;
  }

  .nav-dropdown-content a:hover {
    background-color: #ddd;
    color: black;
  }

  .nav-right .nav-dropdown-content {
    margin-left: -45%;
    right: 0;
  }

  @media screen and (max-width: 768px) {
    .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
    .distill-site-header a.nav-toggle {
      float: right;
      display: block;
    }
    .distill-site-header .title {
      margin-left: 0;
    }
    .distill-site-header .nav-right {
      margin-right: 0;
    }
    .distill-site-header {
      overflow: hidden;
    }
    .nav-right .nav-dropdown-content {
      margin-left: 0;
    }
  }


  @media screen and (max-width: 768px) {
    .distill-site-header.responsive {position: relative; min-height: 500px; }
    .distill-site-header.responsive a.nav-toggle {
      position: absolute;
      right: 0;
      top: 0;
    }
    .distill-site-header.responsive a,
    .distill-site-header.responsive .nav-dropdown {
      display: block;
      text-align: left;
    }
    .distill-site-header.responsive .nav-left,
    .distill-site-header.responsive .nav-right {
      width: 100%;
    }
    .distill-site-header.responsive .nav-dropdown {float: none;}
    .distill-site-header.responsive .nav-dropdown-content {position: relative;}
    .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
      display: block;
      width: 100%;
      text-align: left;
    }
  }

  /* Site Footer */

  .distill-site-footer {
    width: 100%;
    overflow: hidden;
    box-sizing: border-box;
    z-index: 3;
    margin-top: 30px;
    padding-top: 30px;
    padding-bottom: 30px;
    text-align: center;
  }

  /* Headroom */

  d-title {
    padding-top: 6rem;
  }

  @media print {
    d-title {
      padding-top: 4rem;
    }
  }

  .headroom {
    z-index: 1000;
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
  }

  .headroom--transition {
    transition: all .4s ease-in-out;
  }

  .headroom--unpinned {
    top: -100px;
  }

  .headroom--pinned {
    top: 0;
  }

  /* adjust viewport for navbar height */
  /* helps vertically center bootstrap (non-distill) content */
  .min-vh-100 {
    min-height: calc(100vh - 100px) !important;
  }

  </style>

  <script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
  <link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet"/>
  <link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet"/>
  <script src="site_libs/headroom-0.9.4/headroom.min.js"></script>
  <script src="site_libs/autocomplete-0.37.1/autocomplete.min.js"></script>
  <script src="site_libs/fuse-6.4.1/fuse.min.js"></script>

  <script type="application/javascript">

  function getMeta(metaName) {
    var metas = document.getElementsByTagName('meta');
    for (let i = 0; i < metas.length; i++) {
      if (metas[i].getAttribute('name') === metaName) {
        return metas[i].getAttribute('content');
      }
    }
    return '';
  }

  function offsetURL(url) {
    var offset = getMeta('distill:offset');
    return offset ? offset + '/' + url : url;
  }

  function createFuseIndex() {

    // create fuse index
    var options = {
      keys: [
        { name: 'title', weight: 20 },
        { name: 'categories', weight: 15 },
        { name: 'description', weight: 10 },
        { name: 'contents', weight: 5 },
      ],
      ignoreLocation: true,
      threshold: 0
    };
    var fuse = new window.Fuse([], options);

    // fetch the main search.json
    return fetch(offsetURL('search.json'))
      .then(function(response) {
        if (response.status == 200) {
          return response.json().then(function(json) {
            // index main articles
            json.articles.forEach(function(article) {
              fuse.add(article);
            });
            // download collections and index their articles
            return Promise.all(json.collections.map(function(collection) {
              return fetch(offsetURL(collection)).then(function(response) {
                if (response.status === 200) {
                  return response.json().then(function(articles) {
                    articles.forEach(function(article) {
                      fuse.add(article);
                    });
                  })
                } else {
                  return Promise.reject(
                    new Error('Unexpected status from search index request: ' +
                              response.status)
                  );
                }
              });
            })).then(function() {
              return fuse;
            });
          });

        } else {
          return Promise.reject(
            new Error('Unexpected status from search index request: ' +
                        response.status)
          );
        }
      });
  }

  window.document.addEventListener("DOMContentLoaded", function (event) {

    // get search element (bail if we don't have one)
    var searchEl = window.document.getElementById('distill-search');
    if (!searchEl)
      return;

    createFuseIndex()
      .then(function(fuse) {

        // make search box visible
        searchEl.classList.remove('hidden');

        // initialize autocomplete
        var options = {
          autoselect: true,
          hint: false,
          minLength: 2,
        };
        window.autocomplete(searchEl, options, [{
          source: function(query, callback) {
            const searchOptions = {
              isCaseSensitive: false,
              shouldSort: true,
              minMatchCharLength: 2,
              limit: 10,
            };
            var results = fuse.search(query, searchOptions);
            callback(results
              .map(function(result) { return result.item; })
            );
          },
          templates: {
            suggestion: function(suggestion) {
              var img = suggestion.preview && Object.keys(suggestion.preview).length > 0
                ? `<img src="${offsetURL(suggestion.preview)}"</img>`
                : '';
              var html = `
                <div class="search-item">
                  <h3>${suggestion.title}</h3>
                  <div class="search-item-description">
                    ${suggestion.description || ''}
                  </div>
                  <div class="search-item-preview">
                    ${img}
                  </div>
                </div>
              `;
              return html;
            }
          }
        }]).on('autocomplete:selected', function(event, suggestion) {
          window.location.href = offsetURL(suggestion.path);
        });
        // remove inline display style on autocompleter (we want to
        // manage responsive display via css)
        $('.algolia-autocomplete').css("display", "");
      })
      .catch(function(error) {
        console.log(error);
      });

  });

  </script>

  <style type="text/css">

  .nav-search {
    font-size: x-small;
  }

  /* Algolioa Autocomplete */

  .algolia-autocomplete {
    display: inline-block;
    margin-left: 10px;
    vertical-align: sub;
    background-color: white;
    color: black;
    padding: 6px;
    padding-top: 8px;
    padding-bottom: 0;
    border-radius: 6px;
    border: 1px #0F2E3D solid;
    width: 180px;
  }


  @media screen and (max-width: 768px) {
    .distill-site-nav .algolia-autocomplete {
      display: none;
      visibility: hidden;
    }
    .distill-site-nav.responsive .algolia-autocomplete {
      display: inline-block;
      visibility: visible;
    }
    .distill-site-nav.responsive .algolia-autocomplete .aa-dropdown-menu {
      margin-left: 0;
      width: 400px;
      max-height: 400px;
    }
  }

  .algolia-autocomplete .aa-input, .algolia-autocomplete .aa-hint {
    width: 90%;
    outline: none;
    border: none;
  }

  .algolia-autocomplete .aa-hint {
    color: #999;
  }
  .algolia-autocomplete .aa-dropdown-menu {
    width: 550px;
    max-height: 70vh;
    overflow-x: visible;
    overflow-y: scroll;
    padding: 5px;
    margin-top: 3px;
    margin-left: -150px;
    background-color: #fff;
    border-radius: 5px;
    border: 1px solid #999;
    border-top: none;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion {
    cursor: pointer;
    padding: 5px 4px;
    border-bottom: 1px solid #eee;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion:last-of-type {
    border-bottom: none;
    margin-bottom: 2px;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item {
    overflow: hidden;
    font-size: 0.8em;
    line-height: 1.4em;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item h3 {
    font-size: 1rem;
    margin-block-start: 0;
    margin-block-end: 5px;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-description {
    display: inline-block;
    overflow: hidden;
    height: 2.8em;
    width: 80%;
    margin-right: 4%;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview {
    display: inline-block;
    width: 15%;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview img {
    height: 3em;
    width: auto;
    display: none;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview img[src] {
    display: initial;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion.aa-cursor {
    background-color: #eee;
  }
  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion em {
    font-weight: bold;
    font-style: normal;
  }

  </style>


  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: auto;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  d-article details {
    grid-column: text;
    margin-bottom: 0.8em;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure img {
    width: 100%;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }

  /* Citation hover box */

  .tippy-box[data-theme~=light-border] {
    background-color: rgba(250, 250, 250, 0.95);
  }

  .tippy-content > p {
    margin-bottom: 0;
    padding: 2px;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('details, div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme
    $('code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // hoverable references
      $('span.citation[data-cites]').each(function() {
        var refHtml = $('#ref-' + $(this).attr('data-cites')).html();
        window.tippy(this, {
          allowHTML: true,
          content: refHtml,
          maxWidth: 500,
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        });
      });

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace("index.html", "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <!--/radix_placeholder_distill-->
  <script src="site_libs/header-attrs-2.7/header-attrs.js"></script>
  <script src="site_libs/popper-2.6.0/popper.min.js"></script>
  <link href="site_libs/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="site_libs/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="site_libs/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="site_libs/anchor-4.2.2/anchor.min.js"></script>
  <script src="site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <!--/radix_placeholder_site_in_header-->


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"Preprocessing the Pellicci gamma-delta T-cell single-cell data set","authors":[{"author":"Peter Hickey","authorURL":"https://peterhickey.org","affiliation":"Single Cell Open Research Endeavour (SCORE), WEHI","affiliationURL":"https://www.wehi.edu.au/people/rory-bowden/4536/wehi-advanced-genomics-facility","orcidID":""}],"publishedDate":"2021-08-05T00:00:00.000+10:00","citationText":"Hickey, 2021"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a href="index.html" class="title">C094</a>
<a href="index.html">Introduction</a>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Preprocessing
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="C094_Pellicci.single-cell.preprocess.html">Single cell</a>
<a href="C094_Pellicci.mini-bulk.preprocess.html">Mini bulk</a>
</div>
</div>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Selecting biologically relevant cells
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="C094_Pellicci.single-cell.cell_selection.html">Single cell</a>
</div>
</div>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Merging cells
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="C094_Pellicci.single-cell.merge.whole_cell.html">Single cell (whole cell)</a>
<a href="C094_Pellicci.single-cell.merge.thymus_only.html">Single cell (thymus only)</a>
<a href="C094_Pellicci.single-cell.merge.S3_only.html">Single cell (S3 only)</a>
</div>
</div>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Annotating cells
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="C094_Pellicci.single-cell.annotate.whole_cell.html">Single cell (whole cell)</a>
<a href="C094_Pellicci.single-cell.annotate.thymus_only.html">Single cell (thymus only)</a>
<a href="C094_Pellicci.single-cell.annotate.S3_only.html">Single cell (S3 only)</a>
</div>
</div>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Mini-bulk differential analyses
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="C094_Pellicci.mini-bulk.multi-sample_comparisons.all_samples.html">All samples</a>
<a href="C094_Pellicci.mini-bulk.multi-sample_comparisons.excluding_blood_1-3.html">Excluding Blood.1 - Blood.3</a>
<a href="C094_Pellicci.mini-bulk.multi-sample_comparisons.excluding_donors_1-3.html">Excluding donors 1-3</a>
</div>
</div>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Supplementary files
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="../output/scPipe/">scPipe reports</a>
<a href="../output/DEGs/">Differentially expressed genes</a>
<a href="../output/glimma-plots/">Glimma plots</a>
</div>
</div>
</div>
<div class="nav-right">
<a href="https://github.com/WEHISCORE/C094_Pellicci">
<i class="fa fa-github" aria-hidden="true"></i>
</a>
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>Preprocessing the Pellicci gamma-delta T-cell single-cell data set</h1>
<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->

</div>

<div class="d-byline">
  Peter Hickey <a href="https://peterhickey.org" class="uri">https://peterhickey.org</a> (Single Cell Open Research Endeavour (SCORE), WEHI)<a href="https://www.wehi.edu.au/people/rory-bowden/4536/wehi-advanced-genomics-facility" class="uri">https://www.wehi.edu.au/people/rory-bowden/4536/wehi-advanced-genomics-facility</a>
  
<br/>2021-08-05
</div>

<div class="d-article">
<div class="d-contents d-contents-float">
<nav class="l-text toc figcaption" id="TOC">
<h3>Contents</h3>
<ul>
<li><a href="#setting-up-the-data">Setting up the data</a>
<ul>
<li><a href="#post-hoc-gating-into-developmental-stages"><em>Post-hoc</em> gating into developmental stages</a></li>
<li><a href="#incorporating-cell-based-annotations">Incorporating cell-based annotations</a></li>
<li><a href="#incorporating-facs-data">Incorporating FACS data</a></li>
<li><a href="#incorporating-gene-based-annotation">Incorporating gene-based annotation</a></li>
</ul></li>
<li><a href="#plate-alignment-issue-lce511">Plate alignment issue (LCE511)</a>
<ul>
<li><a href="#motivation">Motivation</a></li>
<li><a href="#analysis">Analysis</a></li>
<li><a href="#summary">Summary</a></li>
</ul></li>
<li><a href="#quality-control-of-cells">Quality control of cells</a>
<ul>
<li><a href="#defining-the-quality-control-metrics">Defining the quality control metrics</a></li>
<li><a href="#visualizing-the-qc-metrics">Visualizing the QC metrics</a></li>
<li><a href="#identifying-outliers-by-each-metric">Identifying outliers by each metric</a>
<ul>
<li><a href="#checking-for-removal-of-biologically-relevant-subpopulations">Checking for removal of biologically relevant subpopulations</a></li>
</ul></li>
<li><a href="#summary-1">Summary</a></li>
</ul></li>
<li><a href="#examining-gene-level-metrics">Examining gene-level metrics</a>
<ul>
<li><a href="#inspecting-the-most-highly-expressed-genes">Inspecting the most highly expressed genes</a></li>
<li><a href="#filtering-out-low-abundance-genes">Filtering out low-abundance genes</a></li>
</ul></li>
<li><a href="#normalization">Normalization</a>
<ul>
<li><a href="#using-the-deconvolution-method-to-deal-with-zero-counts">Using the deconvolution method to deal with zero counts</a></li>
<li><a href="#applying-the-size-factors-to-normalize-gene-expression">Applying the size factors to normalize gene expression</a></li>
<li><a href="#computing-separate-size-factors-for-spike-in-transcripts">Computing separate size factors for spike-in transcripts</a></li>
</ul></li>
<li><a href="#feature-selection">Feature selection</a>
<ul>
<li><a href="#motivation-1">Motivation</a></li>
<li><a href="#quantifying-per-gene-variation">Quantifying per-gene variation</a>
<ul>
<li><a href="#variance-of-the-log-counts">Variance of the log-counts</a></li>
<li><a href="#quantifying-technical-noise">Quantifying technical noise</a></li>
<li><a href="#accounting-for-blocking-factors">Accounting for blocking factors</a></li>
</ul></li>
<li><a href="#selecting-highly-variable-genes-hvgs">Selecting highly variable genes (HVGs)</a>
<ul>
<li><a href="#exclusion-of-gene-sets-from-hvgs">Exclusion of gene sets from HVGs</a></li>
</ul></li>
<li><a href="#summary-2">Summary</a></li>
</ul></li>
<li><a href="#dimensionality-reduction">Dimensionality reduction</a>
<ul>
<li><a href="#principal-component-analysis">Principal component analysis</a></li>
<li><a href="#dimensionality-reduction-for-visualization">Dimensionality reduction for visualization</a></li>
<li><a href="#summary-3">Summary</a></li>
</ul></li>
<li><a href="#concluding-remarks">Concluding remarks</a></li>
<li><a href="#additional-information">Additional information</a></li>
<li><a href="#session-info">Session info</a></li>
</ul>
</nav>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://here.r-lib.org/'>here</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/Bioconductor/BiocStyle'>BiocStyle</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://ggplot2.tidyverse.org'>ggplot2</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://wilkelab.org/cowplot/'>cowplot</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://patchwork.data-imaginist.com'>patchwork</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://scales.r-lib.org'>scales</a></span><span class='op'>)</span>

<span class='co'># NOTE: Using &gt;= 4 cores seizes up my laptop. Can use more on unix box.</span>
<span class='fu'><a href='https://rdrr.io/r/base/options.html'>options</a></span><span class='op'>(</span><span class='st'>"mc.cores"</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/Sys.info.html'>Sys.info</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>[[</span><span class='st'>"nodename"</span><span class='op'>]</span><span class='op'>]</span> <span class='op'>==</span> <span class='st'>"PC1331"</span>, <span class='fl'>2L</span>, <span class='fl'>3L</span><span class='op'>)</span><span class='op'>)</span>

<span class='fu'>knitr</span><span class='fu'>::</span><span class='va'><a href='https://rdrr.io/pkg/knitr/man/opts_chunk.html'>opts_chunk</a></span><span class='op'>$</span><span class='fu'>set</span><span class='op'>(</span>
  fig.path <span class='op'>=</span> <span class='st'>"C094_Pellicci.single-cell.preprocess_files/"</span><span class='op'>)</span>

<span class='kw'><a href='https://rdrr.io/r/base/source.html'>source</a></span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"code"</span>, <span class='st'>"helper_functions.R"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<h1 id="setting-up-the-data">Setting up the data</h1>
<p>The count data were processed using <em><a href="https://bioconductor.org/packages/3.12/scPipe">scPipe</a></em> and available in a <em>SingleCellExperiment</em> object, along with their metadata.</p>
<aside>
See <a href="../data/SCEs/C094_Pellicci.scPipe.SCE.rds"><code>data/SCEs/C094_Pellicci.scPipe.SCE.rds</code></a>.
</aside>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"data"</span>, <span class='st'>"SCEs"</span>, <span class='st'>"C094_Pellicci.scPipe.SCE.rds"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>This data was processed to clean up the sample metadata, tidy the FACS index sorting data, and attach the gene metadata.</p>
<aside>
See <a href="../analysis/C094_Pellicci.preprocess.R"><code>analysis/C094_Pellicci.preprocess.R</code></a>.
</aside>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/source.html'>source</a></span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"analysis"</span>, <span class='st'>"C094_Pellicci.preprocess.R"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p><strong>We then filter the data to only include the single-cells</strong>.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>sample_type</span> <span class='op'>==</span> <span class='st'>"Single cell"</span><span class='op'>]</span>
<span class='fu'>colData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/droplevels.html'>droplevels</a></span><span class='op'>(</span><span class='fu'>colData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># Only retain genes with non-zero counts in at least 1 cell.</span>
<span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/colSums.html'>rowSums</a></span><span class='op'>(</span><span class='fu'>counts</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>&gt;</span> <span class='fl'>0</span>, <span class='op'>]</span>
</code></pre>
</div>
</details>
</div>
<h2 id="post-hoc-gating-into-developmental-stages"><em>Post-hoc</em> gating into developmental stages</h2>
<p>Gamma-delta T-cells are classified into 3 developmental stages based on the expression of the cell surface markers <em>CD4</em> and <em>CD161</em>:</p>
<ul>
<li><code>S1</code>: <em>CD4+</em>/<em>CD161-</em></li>
<li><code>S2</code>: <em>CD4-</em>/<em>CD161-</em></li>
<li><code>S3</code>: <em>CD4-</em>/<em>CD161+</em></li>
</ul>
<p>Figure <a href="#fig:facs">1</a> shows the expression of <em>CD4</em> and <em>CD161</em>, as measured by FACS, for all single-cells.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>x</span> <span class='op'>&lt;-</span> <span class='fu'>makePerCellDF</span><span class='op'>(</span>
  <span class='va'>sce</span>, 
  features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"V525_50_A_CD4_BV510"</span>, <span class='st'>"B530_30_A_CD161_FITC"</span><span class='op'>)</span>,
  assay.type <span class='op'>=</span> <span class='st'>"pseudolog"</span><span class='op'>)</span>

<span class='fu'><a href='https://ggplot2.tidyverse.org/reference/ggplot.html'>ggplot</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>V525_50_A_CD4_BV510</span>, y <span class='op'>=</span> <span class='va'>B530_30_A_CD161_FITC</span><span class='op'>)</span>,
  data <span class='op'>=</span> <span class='va'>x</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span>data <span class='op'>=</span> <span class='va'>x</span>, colour <span class='op'>=</span> <span class='st'>"black"</span>, size <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>10</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>"CD4"</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"CD161"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_grid.html'>facet_grid</a></span><span class='op'>(</span><span class='va'>tissue</span> <span class='op'>~</span> <span class='va'>donor</span>, labeller <span class='op'>=</span> <span class='va'>label_both</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:facs"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/facs-1.png" alt="Expression of *CD4* and *CD161*, as measured by FACS, for each sample." width="624" />
<p class="caption">
Figure 1: Expression of <em>CD4</em> and <em>CD161</em>, as measured by FACS, for each sample.
</p>
</div>
</div>
<p>Casey ‘<em>post-hoc</em> gated’ the single-cells into <code>S1</code>, <code>S2</code>, or <code>S3</code> by transplanting the gating strategy used for the mini-bulk samples to the single-cells. Not all single-cells can be classified using these gates (e.g., some cells are seemingly <em>CD4+</em>/<em>CD161+</em>); these cells are labelled as <code>Unknown</code>.</p>
<p>Figure <a href="#fig:facs-stages">2</a> re-plots the same data and colours each cell according to the <em>post-hoc</em> <code>stage</code>. We observe that boundaries between <code>S1</code>, <code>S2</code>, and <code>S3</code> appear rather arbitrary<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> and that the <code>Unknown</code> cells are a <em>potpourri</em> that includes cells that are intermediate between <code>S1</code>, <code>S2</code>, and <code>S3</code> as well as a handful of <em>CD4+</em>/<em>CD161+</em> cells.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>bg</span> <span class='op'>&lt;-</span> <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/select.html'>select</a></span><span class='op'>(</span><span class='va'>x</span>, <span class='op'>-</span><span class='va'>donor</span>, <span class='op'>-</span><span class='va'>tissue</span><span class='op'>)</span>
<span class='fu'><a href='https://ggplot2.tidyverse.org/reference/ggplot.html'>ggplot</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>V525_50_A_CD4_BV510</span>, y <span class='op'>=</span> <span class='va'>B530_30_A_CD161_FITC</span><span class='op'>)</span>, data <span class='op'>=</span> <span class='va'>x</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span>data <span class='op'>=</span> <span class='va'>bg</span>, colour <span class='op'>=</span> <span class='fu'>scales</span><span class='fu'>::</span><span class='fu'><a href='https://scales.r-lib.org/reference/alpha.html'>alpha</a></span><span class='op'>(</span><span class='st'>"grey"</span>, <span class='fl'>0.5</span><span class='op'>)</span>, size <span class='op'>=</span> <span class='fl'>0.125</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='va'>stage</span><span class='op'>)</span>, alpha <span class='op'>=</span> <span class='fl'>1</span>, size <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>
    values <span class='op'>=</span> <span class='va'>stage_colours</span>,
    name <span class='op'>=</span> <span class='st'>"stage"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>
    values <span class='op'>=</span> <span class='va'>stage_colours</span>,
    name <span class='op'>=</span> <span class='st'>"stage"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>10</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_grid.html'>facet_grid</a></span><span class='op'>(</span><span class='va'>tissue</span> <span class='op'>~</span> <span class='va'>donor</span>, labeller <span class='op'>=</span> <span class='va'>label_both</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guide_legend.html'>guide_legend</a></span><span class='op'>(</span>override.aes <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>2</span>, alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>"CD4"</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"CD161"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:facs-stages"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/facs-stages-1.png" alt="Expression of *CD4* and *CD161*, as measured by FACS, for each sample and colouring each cell by its *post-hoc* `stage`." width="624" />
<p class="caption">
Figure 2: Expression of <em>CD4</em> and <em>CD161</em>, as measured by FACS, for each sample and colouring each cell by its <em>post-hoc</em> <code>stage</code>.
</p>
</div>
</div>
<h2 id="incorporating-cell-based-annotations">Incorporating cell-based annotations</h2>
<p>Cell-based annotations are included in the <em>colData</em> of the <em>SingleCellExperiment</em>. Figure <a href="#fig:coldata">3</a> is a visual summary of the cell-based metadata.</p>
<aside>
The ‘SKW-3’ sample is a control cell line, originally thought to originate from the peripheral blood of a 61 year old male patient with T cell chronic lymphocytic leukemia. (although <a href="https://web.expasy.org/cellosaurus/CVCL_2197">this is known to be problematic</a>).
</aside>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='fu'>ggcells</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>plate_number</span>, fill <span class='op'>=</span> <span class='va'>tissue</span><span class='op'>)</span>,
    position <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/position_stack.html'>position_fill</a></span><span class='op'>(</span>reverse <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Frequency"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>9</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>tissue_colours</span><span class='op'>)</span>
<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='fu'>ggcells</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>plate_number</span>, fill <span class='op'>=</span> <span class='va'>donor</span><span class='op'>)</span>,
    position <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/position_stack.html'>position_fill</a></span><span class='op'>(</span>reverse <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Frequency"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>9</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>donor_colours</span><span class='op'>)</span>
<span class='va'>p3</span> <span class='op'>&lt;-</span>  <span class='fu'>ggcells</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>plate_number</span>, fill <span class='op'>=</span> <span class='va'>stage</span><span class='op'>)</span>,
    position <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/position_stack.html'>position_fill</a></span><span class='op'>(</span>reverse <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Frequency"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>9</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>stage_colours</span><span class='op'>)</span>
<span class='va'>p4</span> <span class='op'>&lt;-</span> <span class='fu'>ggcells</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>plate_number</span>, fill <span class='op'>=</span> <span class='va'>plate_number</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Number of single cells"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>9</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>plate_number_colours</span><span class='op'>)</span>
<span class='va'>p1</span> <span class='op'>+</span> <span class='va'>p2</span> <span class='op'>+</span> <span class='va'>p3</span> <span class='op'>+</span> <span class='va'>p4</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:coldata"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/coldata-1.png" alt="Breakdown of experiment by plate." width="624" />
<p class="caption">
Figure 3: Breakdown of experiment by plate.
</p>
</div>
</div>
<p>Figure <a href="#fig:plate-map">4</a> shows the design of each plate.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># TODO: Fix legend</span>
<span class='va'>p</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>lapply</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span><span class='op'>)</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>p</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='va'>z</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span> <span class='op'>==</span> <span class='va'>p</span><span class='op'>]</span>
  <span class='fu'>plotPlatePosition</span><span class='op'>(</span>
    <span class='va'>z</span>,
    <span class='fu'><a href='https://rdrr.io/r/base/character.html'>as.character</a></span><span class='op'>(</span><span class='va'>z</span><span class='op'>$</span><span class='va'>well_position</span><span class='op'>)</span>,
    point_size <span class='op'>=</span> <span class='fl'>2</span>,
    point_alpha <span class='op'>=</span> <span class='fl'>1</span>,
    theme_size <span class='op'>=</span> <span class='fl'>5</span>,
    colour_by <span class='op'>=</span> <span class='st'>"donor"</span>,
    shape_by <span class='op'>=</span> <span class='st'>"tissue"</span><span class='op'>)</span> <span class='op'>+</span>
    <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>donor_colours</span>, name <span class='op'>=</span> <span class='st'>"donor"</span><span class='op'>)</span> <span class='op'>+</span>
    <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ggtitle</a></span><span class='op'>(</span><span class='va'>p</span><span class='op'>)</span> <span class='op'>+</span>
    <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>
      legend.text <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span>,
      legend.title <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
    <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>
      colour <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guide_legend.html'>guide_legend</a></span><span class='op'>(</span>override.aes <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span><span class='op'>)</span>,
      shape <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guide_legend.html'>guide_legend</a></span><span class='op'>(</span>override.aes <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='op'>}</span><span class='op'>)</span>

<span class='fu'><a href='https://patchwork.data-imaginist.com/reference/wrap_plots.html'>wrap_plots</a></span><span class='op'>(</span><span class='va'>p</span>, ncol <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span> <span class='op'>+</span> <span class='fu'><a href='https://patchwork.data-imaginist.com/reference/plot_layout.html'>plot_layout</a></span><span class='op'>(</span>guides <span class='op'>=</span> <span class='st'>"collect"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:plate-map"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/plate-map-1.png" alt="Plate maps." width="624" />
<p class="caption">
Figure 4: Plate maps.
</p>
</div>
</div>
<h2 id="incorporating-facs-data">Incorporating FACS data</h2>
<p>We store the values from the FACS measurements as an ‘alternative experiment.’ We store both the ‘raw’ data and the ‘pseudolog-transformed’ data. The table below is a <em><a href="https://CRAN.R-project.org/package=skimr">skimr</a></em> summary of the ‘raw’ FACS measurements.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span><span class='fu'>assay</span><span class='op'>(</span><span class='fu'>altExp</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"FACS"</span><span class='op'>)</span>, <span class='st'>"raw"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>skimr</span><span class='fu'>::</span><span class='fu'><a href='https://docs.ropensci.org/skimr/reference/skim.html'>skim</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<table>
<caption><span id="tab:unnamed-chunk-4">Table 1: </span>Data summary</caption>
<tbody>
<tr class="odd">
<td style="text-align: left;">Name</td>
<td style="text-align: left;">Piped data</td>
</tr>
<tr class="even">
<td style="text-align: left;">Number of rows</td>
<td style="text-align: left;">2256</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Number of columns</td>
<td style="text-align: left;">11</td>
</tr>
<tr class="even">
<td style="text-align: left;">_______________________</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Column type frequency:</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">numeric</td>
<td style="text-align: left;">11</td>
</tr>
<tr class="odd">
<td style="text-align: left;">________________________</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">Group variables</td>
<td style="text-align: left;">None</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: numeric</strong></p>
<table style="width:100%;">
<colgroup>
<col style="width: 21%" />
<col style="width: 8%" />
<col style="width: 11%" />
<col style="width: 8%" />
<col style="width: 7%" />
<col style="width: 7%" />
<col style="width: 7%" />
<col style="width: 8%" />
<col style="width: 8%" />
<col style="width: 8%" />
<col style="width: 4%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">skim_variable</th>
<th style="text-align: right;">n_missing</th>
<th style="text-align: right;">complete_rate</th>
<th style="text-align: right;">mean</th>
<th style="text-align: right;">sd</th>
<th style="text-align: right;">p0</th>
<th style="text-align: right;">p25</th>
<th style="text-align: right;">p50</th>
<th style="text-align: right;">p75</th>
<th style="text-align: right;">p100</th>
<th style="text-align: left;">hist</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">FSC_A</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">107758.76</td>
<td style="text-align: right;">17036.49</td>
<td style="text-align: right;">62976.00</td>
<td style="text-align: right;">96960.00</td>
<td style="text-align: right;">106816.01</td>
<td style="text-align: right;">116656.00</td>
<td style="text-align: right;">198080.01</td>
<td style="text-align: left;">▂▇▃▁▁</td>
</tr>
<tr class="even">
<td style="text-align: left;">SSC_A</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">41675.87</td>
<td style="text-align: right;">16044.46</td>
<td style="text-align: right;">12096.00</td>
<td style="text-align: right;">30528.00</td>
<td style="text-align: right;">38272.00</td>
<td style="text-align: right;">48432.00</td>
<td style="text-align: right;">169344.00</td>
<td style="text-align: left;">▇▃▁▁▁</td>
</tr>
<tr class="odd">
<td style="text-align: left;">B530_30_A_CD161_FITC</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1901.25</td>
<td style="text-align: right;">7741.33</td>
<td style="text-align: right;">-52.01</td>
<td style="text-align: right;">152.49</td>
<td style="text-align: right;">603.57</td>
<td style="text-align: right;">1399.51</td>
<td style="text-align: right;">129036.06</td>
<td style="text-align: left;">▇▁▁▁▁</td>
</tr>
<tr class="even">
<td style="text-align: left;">B576_26_A_CD38_PE</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">214.45</td>
<td style="text-align: right;">5474.18</td>
<td style="text-align: right;">-110.59</td>
<td style="text-align: right;">35.65</td>
<td style="text-align: right;">91.80</td>
<td style="text-align: right;">153.55</td>
<td style="text-align: right;">259922.41</td>
<td style="text-align: left;">▇▁▁▁▁</td>
</tr>
<tr class="odd">
<td style="text-align: left;">B710_50_A_CD3_Per_CP</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">3755.17</td>
<td style="text-align: right;">1768.87</td>
<td style="text-align: right;">-110.57</td>
<td style="text-align: right;">2687.26</td>
<td style="text-align: right;">3658.23</td>
<td style="text-align: right;">4702.36</td>
<td style="text-align: right;">40131.52</td>
<td style="text-align: left;">▇▁▁▁▁</td>
</tr>
<tr class="even">
<td style="text-align: left;">B780_60_A_Vd2_PE_Cy7</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">9081.48</td>
<td style="text-align: right;">5661.25</td>
<td style="text-align: right;">-95.62</td>
<td style="text-align: right;">4950.54</td>
<td style="text-align: right;">7867.58</td>
<td style="text-align: right;">11901.29</td>
<td style="text-align: right;">38459.43</td>
<td style="text-align: left;">▇▆▂▁▁</td>
</tr>
<tr class="odd">
<td style="text-align: left;">R780_60_A_L_D_Dump_APC_Cy7</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">236.57</td>
<td style="text-align: right;">205.22</td>
<td style="text-align: right;">-110.57</td>
<td style="text-align: right;">81.70</td>
<td style="text-align: right;">217.43</td>
<td style="text-align: right;">371.38</td>
<td style="text-align: right;">939.13</td>
<td style="text-align: left;">▆▇▅▂▁</td>
</tr>
<tr class="even">
<td style="text-align: left;">V450_50_A_Vg9_BV421</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">980.46</td>
<td style="text-align: right;">1162.30</td>
<td style="text-align: right;">-110.60</td>
<td style="text-align: right;">481.31</td>
<td style="text-align: right;">663.11</td>
<td style="text-align: right;">1037.90</td>
<td style="text-align: right;">31046.83</td>
<td style="text-align: left;">▇▁▁▁▁</td>
</tr>
<tr class="odd">
<td style="text-align: left;">V525_50_A_CD4_BV510</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">327.75</td>
<td style="text-align: right;">967.93</td>
<td style="text-align: right;">-110.59</td>
<td style="text-align: right;">42.47</td>
<td style="text-align: right;">136.91</td>
<td style="text-align: right;">279.28</td>
<td style="text-align: right;">26281.07</td>
<td style="text-align: left;">▇▁▁▁▁</td>
</tr>
<tr class="even">
<td style="text-align: left;">V670_40_A_CD45Ra_BV711</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1419.47</td>
<td style="text-align: right;">3656.05</td>
<td style="text-align: right;">-110.60</td>
<td style="text-align: right;">-110.45</td>
<td style="text-align: right;">-110.30</td>
<td style="text-align: right;">632.95</td>
<td style="text-align: right;">23668.12</td>
<td style="text-align: left;">▇▁▁▁▁</td>
</tr>
<tr class="odd">
<td style="text-align: left;">V710_50_A_CD1a_BV786</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2705.58</td>
<td style="text-align: right;">5291.19</td>
<td style="text-align: right;">-110.59</td>
<td style="text-align: right;">250.10</td>
<td style="text-align: right;">667.30</td>
<td style="text-align: right;">1660.13</td>
<td style="text-align: right;">43694.87</td>
<td style="text-align: left;">▇▁▁▁▁</td>
</tr>
</tbody>
</table>
</div>
<p>The table below is a <em><a href="https://CRAN.R-project.org/package=skimr">skimr</a></em> summary of the ‘pseudolog-transformed’ FACS measurements.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span><span class='fu'>assay</span><span class='op'>(</span><span class='fu'>altExp</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"FACS"</span><span class='op'>)</span>, <span class='st'>"pseudolog"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>skimr</span><span class='fu'>::</span><span class='fu'><a href='https://docs.ropensci.org/skimr/reference/skim.html'>skim</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<table>
<caption><span id="tab:unnamed-chunk-5">Table 2: </span>Data summary</caption>
<tbody>
<tr class="odd">
<td style="text-align: left;">Name</td>
<td style="text-align: left;">Piped data</td>
</tr>
<tr class="even">
<td style="text-align: left;">Number of rows</td>
<td style="text-align: left;">2256</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Number of columns</td>
<td style="text-align: left;">11</td>
</tr>
<tr class="even">
<td style="text-align: left;">_______________________</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Column type frequency:</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">numeric</td>
<td style="text-align: left;">11</td>
</tr>
<tr class="odd">
<td style="text-align: left;">________________________</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">Group variables</td>
<td style="text-align: left;">None</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: numeric</strong></p>
<table>
<thead>
<tr class="header">
<th style="text-align: left;">skim_variable</th>
<th style="text-align: right;">n_missing</th>
<th style="text-align: right;">complete_rate</th>
<th style="text-align: right;">mean</th>
<th style="text-align: right;">sd</th>
<th style="text-align: right;">p0</th>
<th style="text-align: right;">p25</th>
<th style="text-align: right;">p50</th>
<th style="text-align: right;">p75</th>
<th style="text-align: right;">p100</th>
<th style="text-align: left;">hist</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">FSC_A</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">7.26</td>
<td style="text-align: right;">0.16</td>
<td style="text-align: right;">6.73</td>
<td style="text-align: right;">7.16</td>
<td style="text-align: right;">7.26</td>
<td style="text-align: right;">7.35</td>
<td style="text-align: right;">7.88</td>
<td style="text-align: left;">▁▃▇▂▁</td>
</tr>
<tr class="even">
<td style="text-align: left;">SSC_A</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">6.26</td>
<td style="text-align: right;">0.35</td>
<td style="text-align: right;">5.08</td>
<td style="text-align: right;">6.01</td>
<td style="text-align: right;">6.23</td>
<td style="text-align: right;">6.47</td>
<td style="text-align: right;">7.72</td>
<td style="text-align: left;">▁▆▇▂▁</td>
</tr>
<tr class="odd">
<td style="text-align: left;">B530_30_A_CD161_FITC</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2.02</td>
<td style="text-align: right;">1.28</td>
<td style="text-align: right;">-0.34</td>
<td style="text-align: right;">0.89</td>
<td style="text-align: right;">2.10</td>
<td style="text-align: right;">2.93</td>
<td style="text-align: right;">7.45</td>
<td style="text-align: left;">▆▇▆▁▁</td>
</tr>
<tr class="even">
<td style="text-align: left;">B576_26_A_CD38_PE</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0.56</td>
<td style="text-align: right;">0.52</td>
<td style="text-align: right;">-0.68</td>
<td style="text-align: right;">0.24</td>
<td style="text-align: right;">0.58</td>
<td style="text-align: right;">0.90</td>
<td style="text-align: right;">8.15</td>
<td style="text-align: left;">▇▁▁▁▁</td>
</tr>
<tr class="odd">
<td style="text-align: left;">B710_50_A_CD3_Per_CP</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">3.75</td>
<td style="text-align: right;">0.78</td>
<td style="text-align: right;">-0.68</td>
<td style="text-align: right;">3.58</td>
<td style="text-align: right;">3.89</td>
<td style="text-align: right;">4.14</td>
<td style="text-align: right;">6.28</td>
<td style="text-align: left;">▁▁▂▇▁</td>
</tr>
<tr class="even">
<td style="text-align: left;">B780_60_A_Vd2_PE_Cy7</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">4.55</td>
<td style="text-align: right;">0.89</td>
<td style="text-align: right;">-0.60</td>
<td style="text-align: right;">4.19</td>
<td style="text-align: right;">4.65</td>
<td style="text-align: right;">5.07</td>
<td style="text-align: right;">6.24</td>
<td style="text-align: left;">▁▁▁▇▅</td>
</tr>
<tr class="odd">
<td style="text-align: left;">R780_60_A_L_D_Dump_APC_Cy7</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1.04</td>
<td style="text-align: right;">0.77</td>
<td style="text-align: right;">-0.68</td>
<td style="text-align: right;">0.52</td>
<td style="text-align: right;">1.17</td>
<td style="text-align: right;">1.64</td>
<td style="text-align: right;">2.53</td>
<td style="text-align: left;">▃▃▆▇▃</td>
</tr>
<tr class="even">
<td style="text-align: left;">V450_50_A_Vg9_BV421</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2.27</td>
<td style="text-align: right;">0.79</td>
<td style="text-align: right;">-0.68</td>
<td style="text-align: right;">1.88</td>
<td style="text-align: right;">2.19</td>
<td style="text-align: right;">2.63</td>
<td style="text-align: right;">6.03</td>
<td style="text-align: left;">▁▅▇▁▁</td>
</tr>
<tr class="odd">
<td style="text-align: left;">V525_50_A_CD4_BV510</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0.92</td>
<td style="text-align: right;">0.96</td>
<td style="text-align: right;">-0.68</td>
<td style="text-align: right;">0.28</td>
<td style="text-align: right;">0.82</td>
<td style="text-align: right;">1.38</td>
<td style="text-align: right;">5.86</td>
<td style="text-align: left;">▆▇▂▁▁</td>
</tr>
<tr class="even">
<td style="text-align: left;">V670_40_A_CD45Ra_BV711</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0.81</td>
<td style="text-align: right;">1.97</td>
<td style="text-align: right;">-0.68</td>
<td style="text-align: right;">-0.68</td>
<td style="text-align: right;">-0.68</td>
<td style="text-align: right;">2.15</td>
<td style="text-align: right;">5.75</td>
<td style="text-align: left;">▇▂▂▁▁</td>
</tr>
<tr class="odd">
<td style="text-align: left;">V710_50_A_CD1a_BV786</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2.28</td>
<td style="text-align: right;">1.64</td>
<td style="text-align: right;">-0.68</td>
<td style="text-align: right;">1.28</td>
<td style="text-align: right;">2.20</td>
<td style="text-align: right;">3.10</td>
<td style="text-align: right;">6.37</td>
<td style="text-align: left;">▅▇▇▃▂</td>
</tr>
</tbody>
</table>
</div>
<h2 id="incorporating-gene-based-annotation">Incorporating gene-based annotation</h2>
<p>I used the <em><a href="https://bioconductor.org/packages/3.12/Homo.sapiens">Homo.sapiens</a></em> package and the <code>EnsDb.Hsapiens.v98</code> resource, which respectively cover the NCBI/RefSeq and Ensembl databases, to obtain gene-based annotations, such as the chromosome and gene symbol.</p>
<p>Having quantified gene expression against the GENCODE gene annotation, we have Ensembl-style identifiers for the genes. These identifiers are used as they are unambiguous and highly stable. However, they are difficult to interpret compared to the gene symbols which are more commonly used in the literature. Henceforth, we will use gene symbols (where available) to refer to genes in our analysis and otherwise use the Ensembl-style gene identifiers<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>.</p>
<p>We also construct some useful gene sets: mitochondrial genes, ribosomal protein genes, sex chromosome genes, and pseudogenes.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># Some useful gene sets</span>
<span class='va'>mito_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/any.html'>any</a></span><span class='op'>(</span><span class='fu'>rowData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>$</span><span class='va'>ENSEMBL.SEQNAME</span> <span class='op'>==</span> <span class='st'>"MT"</span><span class='op'>)</span><span class='op'>]</span>
<span class='va'>ribo_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/grep.html'>grep</a></span><span class='op'>(</span><span class='st'>"^RP(S|L)"</span>, <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>, value <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='co'># NOTE: A more curated approach for identifying ribosomal protein genes </span>
<span class='co'>#       (https://github.com/Bioconductor/OrchestratingSingleCellAnalysis-base/blob/ae201bf26e3e4fa82d9165d8abf4f4dc4b8e5a68/feature-selection.Rmd#L376-L380)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://igordot.github.io/msigdbr'>msigdbr</a></span><span class='op'>)</span>
<span class='va'>c2_sets</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://igordot.github.io/msigdbr/reference/msigdbr.html'>msigdbr</a></span><span class='op'>(</span>species <span class='op'>=</span> <span class='st'>"Homo sapiens"</span>, category <span class='op'>=</span> <span class='st'>"C2"</span><span class='op'>)</span>
<span class='va'>ribo_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/sets.html'>union</a></span><span class='op'>(</span>
  <span class='va'>ribo_set</span>,
  <span class='va'>c2_sets</span><span class='op'>[</span><span class='va'>c2_sets</span><span class='op'>$</span><span class='va'>gs_name</span> <span class='op'>==</span> <span class='st'>"KEGG_RIBOSOME"</span>, <span class='op'>]</span><span class='op'>$</span><span class='va'>gene_symbol</span><span class='op'>)</span>
<span class='va'>ribo_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/sets.html'>intersect</a></span><span class='op'>(</span><span class='va'>ribo_set</span>, <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>sex_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/any.html'>any</a></span><span class='op'>(</span><span class='fu'>rowData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>$</span><span class='va'>ENSEMBL.SEQNAME</span> <span class='op'>%in%</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"X"</span>, <span class='st'>"Y"</span><span class='op'>)</span><span class='op'>)</span><span class='op'>]</span>
<span class='va'>pseudogene_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span>
  <span class='fu'><a href='https://rdrr.io/r/base/any.html'>any</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/grep.html'>grepl</a></span><span class='op'>(</span><span class='st'>"pseudogene"</span>, <span class='fu'>rowData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>$</span><span class='va'>ENSEMBL.GENEBIOTYPE</span><span class='op'>)</span><span class='op'>)</span><span class='op'>]</span>
</code></pre>
</div>
</details>
</div>
<h1 id="plate-alignment-issue-lce511">Plate alignment issue (LCE511)</h1>
<h2 id="motivation">Motivation</h2>
<p>Casey and Dan reported that plate <code>LCE511</code> came out alignment during the sort, specifically while sorting the final row (row <code>P</code>), which contains the <code>Blood 2</code> cells (see Figure <a href="#fig:plate-map">4</a>). Depending on how exactly the plate came out of alignment, this may have resulted in the wells in the adjacent row (row O) containing 2 cells (i.e. doublets with 1 cell from those intended for row O and 1 cells from those intended for row P).</p>
<p>We want to check:</p>
<ol type="1">
<li>Do the wells in row P appear to be missing a cell?</li>
<li>Do the wells in the previous row (row O) contain 2 cells (i.e. do they contain doublets)?</li>
</ol>
<h2 id="analysis">Analysis</h2>
<p>Figure <a href="#fig:qc-lce511">5</a> plots the number of UMIs (<code>sum</code>) and number of genes detected (<code>detected</code>) for wells in each row of place <code>LCE511</code>. We observe substantially fewer UMIs and genes detected for the wells in row P, confirming that these wells appear to be missing a cell. However, we do not observe any strong differences in the number of UMIs or genes detected for the wells in row O compared to wells in rows L - N.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>z</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span> <span class='op'>==</span> <span class='st'>"LCE511"</span><span class='op'>]</span>
<span class='va'>z</span> <span class='op'>&lt;-</span> <span class='fu'>addPerCellQC</span><span class='op'>(</span><span class='va'>z</span>, use.altexps <span class='op'>=</span> <span class='st'>"ERCC"</span><span class='op'>)</span>

<span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>z</span>, 
  y <span class='op'>=</span> <span class='st'>"sum"</span>,
  x <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/AsIs.html'>I</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://rdrr.io/r/base/interaction.html'>interaction</a></span><span class='op'>(</span>
      <span class='fu'><a href='https://rdrr.io/r/base/substr.html'>substr</a></span><span class='op'>(</span><span class='va'>z</span><span class='op'>$</span><span class='va'>well_position</span>, <span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span>, 
      <span class='va'>z</span><span class='op'>$</span><span class='va'>tissue</span>,
      <span class='va'>z</span><span class='op'>$</span><span class='va'>donor</span>,
      drop <span class='op'>=</span> <span class='cn'>TRUE</span>, 
      lex.order <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span>, 
  colour_by <span class='op'>=</span> <span class='st'>"donor"</span>,
  shape_by <span class='op'>=</span> <span class='st'>"tissue"</span>,
  point_alpha <span class='op'>=</span> <span class='fl'>1</span>,
  theme_size <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_continuous.html'>scale_y_log10</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>donor_colours</span>, name <span class='op'>=</span> <span class='st'>"donor"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>"Row.tissue.donor"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>angle <span class='op'>=</span> <span class='fl'>45</span>, hjust <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>z</span>, 
  y <span class='op'>=</span> <span class='st'>"detected"</span>,
  x <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/AsIs.html'>I</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://rdrr.io/r/base/interaction.html'>interaction</a></span><span class='op'>(</span>
      <span class='fu'><a href='https://rdrr.io/r/base/substr.html'>substr</a></span><span class='op'>(</span><span class='va'>z</span><span class='op'>$</span><span class='va'>well_position</span>, <span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span>, 
      <span class='va'>z</span><span class='op'>$</span><span class='va'>tissue</span>,
      <span class='va'>z</span><span class='op'>$</span><span class='va'>donor</span>,
      drop <span class='op'>=</span> <span class='cn'>TRUE</span>, 
      lex.order <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span>, 
  colour_by <span class='op'>=</span> <span class='st'>"donor"</span>,
  shape_by <span class='op'>=</span> <span class='st'>"tissue"</span>,
  point_alpha <span class='op'>=</span> <span class='fl'>1</span>,
  theme_size <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>donor_colours</span>, name <span class='op'>=</span> <span class='st'>"donor"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>"Row.tissue.donor"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>angle <span class='op'>=</span> <span class='fl'>45</span>, hjust <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>p1</span> <span class='op'>+</span> <span class='va'>p2</span> <span class='op'>+</span>
  <span class='fu'><a href='https://patchwork.data-imaginist.com/reference/plot_layout.html'>plot_layout</a></span><span class='op'>(</span>ncol <span class='op'>=</span> <span class='fl'>2</span>, guides <span class='op'>=</span> <span class='st'>"collect"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://patchwork.data-imaginist.com/reference/plot_annotation.html'>plot_annotation</a></span><span class='op'>(</span><span class='st'>"LCE511"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:qc-lce511"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/qc-lce511-1.png" alt="Distributions of various QC metrics for all samples on plate `LCE511`." width="624" />
<p class="caption">
Figure 5: Distributions of various QC metrics for all samples on plate <code>LCE511</code>.
</p>
</div>
</div>
<p>Thus far, we have no strong evidence that row O contains doublets. As another check, we look to see if the samples in these wells co-express mutually-exclusive tissue-marker genes. That is, since row O nominally contains <code>Blood</code> samples and row P nominally contained <code>Thymus</code> samples, we check if the samples in row <code>O</code> co-express <code>Blood</code> and <code>Thymus</code> marker genes<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a>, which would provide evidence that row O wells contain doublets.</p>
<p>Figure <a href="#fig:marker-genes-lce511">6</a> is a heatmap of <code>Blood</code> and <code>Thymus</code> marker genes for all wells on plate <code>LCE511</code>. Firstly, it shows that there aren’t any great markers of <code>Blood</code> vs. <code>Thymus</code> single-cells in this data. Secondly, and more importantly, there is some evidence that row O wells express some <code>Blood</code>-specific markers (e.g., <em>NKG7</em>, <em>CCL5</em>, <em>HLA-B</em>, <em>HLA-C</em>). However, these same markers are also expressed by the <code>Thymus 1</code> samples in row J. There is no strong expression of the <code>Thymus</code>-specific markers in row O wells, but nor is there in any of the <code>Thymus</code> samples on this plate.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># TODO: Fix figure size and legend size.</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>scran</span><span class='op'>)</span>

<span class='va'>tmp</span> <span class='op'>&lt;-</span> <span class='fu'>logNormCounts</span><span class='op'>(</span>
  <span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span> <span class='op'>!=</span> <span class='st'>"LCE511"</span> <span class='op'>&amp;</span> 
        <span class='va'>sce</span><span class='op'>$</span><span class='va'>tissue</span> <span class='op'>%in%</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Blood"</span>, <span class='st'>"Thymus"</span><span class='op'>)</span><span class='op'>]</span><span class='op'>)</span>
<span class='fu'>colData</span><span class='op'>(</span><span class='va'>tmp</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/droplevels.html'>droplevels</a></span><span class='op'>(</span><span class='fu'>colData</span><span class='op'>(</span><span class='va'>tmp</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>w</span> <span class='op'>&lt;-</span> <span class='fu'>logNormCounts</span><span class='op'>(</span>
  <span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span> <span class='op'>==</span> <span class='st'>"LCE511"</span> <span class='op'>&amp;</span>
        <span class='va'>sce</span><span class='op'>$</span><span class='va'>tissue</span> <span class='op'>%in%</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Blood"</span>, <span class='st'>"Thymus"</span><span class='op'>)</span><span class='op'>]</span><span class='op'>)</span>
<span class='va'>w</span><span class='op'>$</span><span class='va'>row</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/substr.html'>substr</a></span><span class='op'>(</span><span class='va'>w</span><span class='op'>$</span><span class='va'>well_position</span>, <span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span>
<span class='fu'>colData</span><span class='op'>(</span><span class='va'>w</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/droplevels.html'>droplevels</a></span><span class='op'>(</span><span class='fu'>colData</span><span class='op'>(</span><span class='va'>w</span><span class='op'>)</span><span class='op'>)</span>

<span class='va'>markers</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scran/man/findMarkers.html'>findMarkers</a></span><span class='op'>(</span>
  <span class='va'>tmp</span>, 
  groups <span class='op'>=</span> <span class='va'>tmp</span><span class='op'>$</span><span class='va'>tissue</span>,
  direction <span class='op'>=</span> <span class='st'>"up"</span>,
  test.type <span class='op'>=</span> <span class='st'>"wilcox"</span>,
  block <span class='op'>=</span> <span class='va'>tmp</span><span class='op'>$</span><span class='va'>donor</span><span class='op'>)</span>

<span class='va'>features</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/unlist.html'>unlist</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>lapply</a></span><span class='op'>(</span><span class='va'>markers</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span> <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>[</span><span class='va'>x</span><span class='op'>$</span><span class='va'>FDR</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span>, <span class='op'>]</span><span class='op'>)</span>, <span class='fl'>10</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'>plotHeatmap</span><span class='op'>(</span>
  <span class='va'>w</span>, 
  <span class='va'>features</span>,
  order_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"tissue"</span>, <span class='st'>"row"</span>, <span class='st'>"donor"</span><span class='op'>)</span>,
  center <span class='op'>=</span> <span class='cn'>TRUE</span>,
  cluster_rows <span class='op'>=</span> <span class='cn'>FALSE</span>,
  symmetric <span class='op'>=</span> <span class='cn'>TRUE</span>,
  zlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>3</span>, <span class='fl'>3</span><span class='op'>)</span>,
  color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>hcl.colors</a></span><span class='op'>(</span><span class='fl'>101</span>, <span class='st'>"Blue-Red 3"</span><span class='op'>)</span>,
  annotation_row <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
    tissue <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/grep.html'>sub</a></span><span class='op'>(</span><span class='st'>"[0-9]+"</span>, <span class='st'>""</span>, <span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>features</span><span class='op'>)</span><span class='op'>)</span>,
    row.names <span class='op'>=</span> <span class='va'>features</span><span class='op'>)</span>,
  column_annotation_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
    tissue <span class='op'>=</span> <span class='va'>tissue_colours</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>w</span><span class='op'>$</span><span class='va'>tissue</span><span class='op'>)</span><span class='op'>]</span>,
    donor <span class='op'>=</span> <span class='va'>donor_colours</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>w</span><span class='op'>$</span><span class='va'>donor</span><span class='op'>)</span><span class='op'>]</span><span class='op'>)</span>,
  main <span class='op'>=</span> <span class='st'>"LCE511"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:marker-genes-lce511"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/marker-genes-lce511-1.png" alt="Expression of the top-10 `Blood` and `Thymus` marker genes for all wells on plate `LCE511` grouped by `Tissue` then `row`. We are looking to see if the wells in row `O` co-express these marker genes." width="624" />
<p class="caption">
Figure 6: Expression of the top-10 <code>Blood</code> and <code>Thymus</code> marker genes for all wells on plate <code>LCE511</code> grouped by <code>Tissue</code> then <code>row</code>. We are looking to see if the wells in row <code>O</code> co-express these marker genes.
</p>
</div>
</div>
<h2 id="summary">Summary</h2>
<p>Consistent with plate <code>LCE511</code> having come out of alignment during the sort, specifically while sorting the final row (row P), we find that wells in this row have an unusually small number of UMIs and genes detected. We therefore remove these wells from further analysis. Since row P is the only row on plate <code>LCE511</code> containing samples from <code>Blood 2</code>, this means that there are are no longer any single-cells from this sample on this plate.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>remove_p</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span> <span class='op'>==</span> <span class='st'>"LCE511"</span> <span class='op'>&amp;</span> 
  <span class='fu'><a href='https://rdrr.io/r/base/substr.html'>substr</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>well_position</span>, <span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span> <span class='op'>==</span> <span class='st'>"P"</span>
<span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>[</span>, <span class='op'>!</span><span class='va'>remove_p</span><span class='op'>]</span>
</code></pre>
</div>
</details>
</div>
<p>We have not found any strong evidence that the wells in the prior row (row O) have been contaminated by the sample intended for row P. However, since we have many additional single-cells from this sample elsewhere on this plate and on the other plates, we opt to also the row O wells on plate <code>LCE511</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>remove_o</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span> <span class='op'>==</span> <span class='st'>"LCE511"</span> <span class='op'>&amp;</span>
  <span class='fu'><a href='https://rdrr.io/r/base/substr.html'>substr</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>well_position</span>, <span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span> <span class='op'>==</span> <span class='st'>"O"</span>
<span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>[</span>, <span class='op'>!</span><span class='va'>remove_o</span><span class='op'>]</span>
<span class='fu'>colData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/droplevels.html'>droplevels</a></span><span class='op'>(</span><span class='fu'>colData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<h1 id="quality-control-of-cells">Quality control of cells</h1>
<h2 id="defining-the-quality-control-metrics">Defining the quality control metrics</h2>
<p>Low-quality cells need to be removed to ensure that technical effects do not distort downstream analysis results. We use several quality control (QC) metrics to measure the quality of the cells:</p>
<ul>
<li><code>sum</code>: This measures the library size of the cells, which is the total sum of UMI counts across both genes and spike-in transcripts. We want cells to have high library sizes as this means more RNA has been successfully captured during library preparation.</li>
<li><code>detected</code>: This is the number of expressed features<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a> in each cell. Cells with few expressed features are likely to be of poor quality, as the diverse transcript population has not been successful captured.</li>
<li><code>altexps_ERCC_detected</code>: This measures the proportion of reads which are mapped to spike-in transcripts relative to the library size of each cell. High proportions are indicative of poor-quality cells, where endogenous RNA has been lost during processing (e.g., due to cell lysis or RNA degradation). The same amount of spike-in RNA to each cell, so an enrichment in spike-in counts is symptomatic of loss of endogenous RNA.</li>
<li><code>subsets_Mito_percent</code>: This measures the proportion of reads which are mapped to mitochondrial RNA. If there is a higher than expected proportion of mitochondrial RNA this is often symptomatic of a cell which is under stress and is therefore of low quality and will not be used for the analysis.</li>
<li><code>subsets_Ribo_percent</code>: This measures the proportion of UMIs which are mapped to ribosomal protein genes. If there is a higher than expected proportion of ribosomal protein gene expression this is often symptomatic of a cell which is of compromised quality and we may want to exclude it from the analysis.</li>
</ul>
<p>For CEL-Seq2 data, we typically observe library sizes that are in the tens of thousands<a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a> and several thousand expressed genes per cell.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>is_mito</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>%in%</span> <span class='va'>mito_set</span>
<span class='fu'><a href='https://rdrr.io/r/base/summary.html'>summary</a></span><span class='op'>(</span><span class='va'>is_mito</span><span class='op'>)</span>
<span class='va'>is_ribo</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>%in%</span> <span class='va'>ribo_set</span>
<span class='fu'><a href='https://rdrr.io/r/base/summary.html'>summary</a></span><span class='op'>(</span><span class='va'>is_ribo</span><span class='op'>)</span>
<span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='fu'>addPerCellQC</span><span class='op'>(</span>
  <span class='va'>sce</span>, 
  subsets <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>Mito <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span><span class='va'>is_mito</span><span class='op'>)</span>, Ribo <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span><span class='va'>is_ribo</span><span class='op'>)</span><span class='op'>)</span>,
  use.altexps <span class='op'>=</span> <span class='st'>"ERCC"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<h2 id="visualizing-the-qc-metrics">Visualizing the QC metrics</h2>
<p>Figure <a href="#fig:qcplot">7</a> compares the QC metrics of the single-cell samples across the different <code>sample</code> (combination of <code>tissue</code> and <code>donor</code>) and Figure <a href="#fig:qcplot-by-plate-number">8</a> further stratifies by <code>plate_number</code>. The number of UMIs, number of gene detected, and percentage of UMIs that are mapped to ERCC spike-ins are all higher in the ‘experimental’ samples than in the ‘control’ samples This likely reflects that the ‘experimental’ cells, which are γδ T cells, have a ‘smaller’ transcriptome than the ‘control’ cells (rather than this result being due to technical factors, although we cannot rule this out entirely). Together, these figures show that the vast majority of single-cell samples are good-quality<a href="#fn6" class="footnote-ref" id="fnref6" role="doc-noteref"><sup>6</sup></a>:</p>
<ul>
<li>The median library size is around 1,700.</li>
<li>The median number of genes detected is around 1,000.</li>
<li>The median percentage of UMIs that are mapped to mitochondrial RNA is around 8%.</li>
<li>The median percentage of UMIs that are mapped to ribosomal protein genes is around 22%.</li>
<li>The median percentage of UMIs that are mapped to ERCC spike-ins is around 19%.</li>
</ul>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce</span>,
  <span class='st'>"sum"</span>,
  x <span class='op'>=</span> <span class='st'>"sample"</span>,
  other_fields <span class='op'>=</span> <span class='st'>"plate_number"</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span>,
  colour_by <span class='op'>=</span> <span class='st'>"sample"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_continuous.html'>scale_y_log10</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/annotation_logticks.html'>annotation_logticks</a></span><span class='op'>(</span>
    sides <span class='op'>=</span> <span class='st'>"l"</span>,
    short <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grid/unit.html'>unit</a></span><span class='op'>(</span><span class='fl'>0.03</span>, <span class='st'>"cm"</span><span class='op'>)</span>,
    mid <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grid/unit.html'>unit</a></span><span class='op'>(</span><span class='fl'>0.06</span>, <span class='st'>"cm"</span><span class='op'>)</span>,
    long <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grid/unit.html'>unit</a></span><span class='op'>(</span><span class='fl'>0.09</span>, <span class='st'>"cm"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>sample_colours</span>, name <span class='op'>=</span> <span class='st'>"sample"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>""</span><span class='op'>)</span>
<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce</span>,
  <span class='st'>"detected"</span>,
  x <span class='op'>=</span> <span class='st'>"sample"</span>,
  other_fields <span class='op'>=</span> <span class='st'>"plate_number"</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span>,
  colour_by <span class='op'>=</span> <span class='st'>"sample"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>sample_colours</span>, name <span class='op'>=</span> <span class='st'>"sample"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>""</span><span class='op'>)</span>
<span class='va'>p3</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce</span>,
  <span class='st'>"subsets_Mito_percent"</span>,
  x <span class='op'>=</span> <span class='st'>"sample"</span>,
  other_fields <span class='op'>=</span> <span class='st'>"plate_number"</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span>,
  colour_by <span class='op'>=</span> <span class='st'>"sample"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/lims.html'>ylim</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='cn'>NA</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>sample_colours</span>, name <span class='op'>=</span> <span class='st'>"sample"</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>""</span><span class='op'>)</span>
<span class='va'>p4</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce</span>,
  <span class='st'>"subsets_Ribo_percent"</span>,
  x <span class='op'>=</span> <span class='st'>"sample"</span>,
  other_fields <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"plate_number"</span>, <span class='st'>"sample_type"</span><span class='op'>)</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span>,
  colour_by <span class='op'>=</span> <span class='st'>"sample"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/lims.html'>ylim</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='cn'>NA</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>sample_colours</span>, name <span class='op'>=</span> <span class='st'>"sample"</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>""</span><span class='op'>)</span>
<span class='va'>p5</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce</span>,
  <span class='st'>"altexps_ERCC_percent"</span>,
  x <span class='op'>=</span> <span class='st'>"sample"</span>,
  other_fields <span class='op'>=</span> <span class='st'>"plate_number"</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span>,
  colour_by <span class='op'>=</span> <span class='st'>"sample"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/lims.html'>ylim</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='cn'>NA</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>sample_colours</span>, name <span class='op'>=</span> <span class='st'>"sample"</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>""</span><span class='op'>)</span>
<span class='va'>p1</span> <span class='op'>+</span> <span class='va'>p2</span> <span class='op'>+</span> <span class='va'>p3</span> <span class='op'>+</span> <span class='va'>p4</span> <span class='op'>+</span> <span class='va'>p5</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://patchwork.data-imaginist.com/reference/plot_layout.html'>plot_layout</a></span><span class='op'>(</span>ncol <span class='op'>=</span> <span class='fl'>2</span>, guides <span class='op'>=</span> <span class='st'>"collect"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:qcplot"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/qcplot-1.png" alt="Distributions of various QC metrics for all samples in the dataset." width="624" />
<p class="caption">
Figure 7: Distributions of various QC metrics for all samples in the dataset.
</p>
</div>
</div>
<p>Figure <a href="#fig:qcplot-by-plate-number">8</a> shows that QC metrics are generally consistent across plates.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='va'>p1</span> <span class='op'>+</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_wrap.html'>facet_wrap</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>plate_number</span>, ncol <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>
<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='va'>p2</span> <span class='op'>+</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_wrap.html'>facet_wrap</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>plate_number</span>, ncol <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>
<span class='va'>p3</span> <span class='op'>&lt;-</span> <span class='va'>p3</span> <span class='op'>+</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_wrap.html'>facet_wrap</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>plate_number</span>, ncol <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>
<span class='va'>p4</span> <span class='op'>&lt;-</span> <span class='va'>p4</span> <span class='op'>+</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_wrap.html'>facet_wrap</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>plate_number</span>, ncol <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>
<span class='va'>p5</span> <span class='op'>&lt;-</span> <span class='va'>p5</span> <span class='op'>+</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_wrap.html'>facet_wrap</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>plate_number</span>, ncol <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>
<span class='va'>p1</span> <span class='op'>+</span> <span class='va'>p2</span> <span class='op'>+</span> <span class='va'>p3</span> <span class='op'>+</span> <span class='va'>p4</span> <span class='op'>+</span> <span class='va'>p5</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://patchwork.data-imaginist.com/reference/plot_layout.html'>plot_layout</a></span><span class='op'>(</span>ncol <span class='op'>=</span> <span class='fl'>2</span>, guides <span class='op'>=</span> <span class='st'>"collect"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:qcplot-by-plate-number"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/qcplot-by-plate-number-1.png" alt="Distributions of various QC metrics for all samples in the dataset, stratified by `plate_number`." width="624" />
<p class="caption">
Figure 8: Distributions of various QC metrics for all samples in the dataset, stratified by <code>plate_number</code>.
</p>
</div>
</div>
<h2 id="identifying-outliers-by-each-metric">Identifying outliers by each metric</h2>
<p>Outliers are defined based on the median absolute deviation (MADs) from the median value of each metric across all cells. We remove small and large outliers for the library size and the number of expressed features, and large outliers for the spike-in proportions and mitochondrial gene expression proportions. Removal of low-quality cells is then performed by combining the filters for all of the metrics.</p>
<p>Due to the differences in the QC metrics by <code>sample</code>, we will compute the separate outlier thresholds for each <code>sample</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>sample</span>
<span class='va'>libsize_drop</span> <span class='op'>&lt;-</span> <span class='fu'>isOutlier</span><span class='op'>(</span>
  metric <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>sum</span>, 
  nmads <span class='op'>=</span> <span class='fl'>3</span>,
  type <span class='op'>=</span> <span class='st'>"lower"</span>, 
  log <span class='op'>=</span> <span class='cn'>TRUE</span>,
  batch <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span><span class='op'>)</span>
<span class='va'>feature_drop</span> <span class='op'>&lt;-</span> <span class='fu'>isOutlier</span><span class='op'>(</span>
  metric <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>detected</span>,
  nmads <span class='op'>=</span> <span class='fl'>3</span>, 
  type <span class='op'>=</span> <span class='st'>"lower"</span>, 
  log <span class='op'>=</span> <span class='cn'>TRUE</span>,
  batch <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span><span class='op'>)</span>
<span class='va'>spike_drop</span> <span class='op'>&lt;-</span> <span class='fu'>isOutlier</span><span class='op'>(</span>
  metric <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>altexps_ERCC_percent</span>, 
  nmads <span class='op'>=</span> <span class='fl'>3</span>, 
  type <span class='op'>=</span> <span class='st'>"higher"</span>,
  batch <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span><span class='op'>)</span>
<span class='va'>mito_drop</span> <span class='op'>&lt;-</span> <span class='fu'>isOutlier</span><span class='op'>(</span>
  metric <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>subsets_Mito_percent</span>, 
  nmads <span class='op'>=</span> <span class='fl'>3</span>, 
  type <span class='op'>=</span> <span class='st'>"higher"</span>,
  batch <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span><span class='op'>)</span>
<span class='va'>ribo_drop</span> <span class='op'>&lt;-</span> <span class='fu'>isOutlier</span><span class='op'>(</span>
  metric <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>subsets_Ribo_percent</span>, 
  nmads <span class='op'>=</span> <span class='fl'>3</span>, 
  type <span class='op'>=</span> <span class='st'>"higher"</span>,
  batch <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>The following table summarises the QC cutoffs:</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>libsize_drop_df</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
  batch <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/attributes.html'>attributes</a></span><span class='op'>(</span><span class='va'>libsize_drop</span><span class='op'>)</span><span class='op'>$</span><span class='va'>thresholds</span><span class='op'>)</span>,
  cutoff <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/attributes.html'>attributes</a></span><span class='op'>(</span><span class='va'>libsize_drop</span><span class='op'>)</span><span class='op'>$</span><span class='va'>thresholds</span><span class='op'>[</span><span class='st'>"lower"</span>, <span class='op'>]</span><span class='op'>)</span>
<span class='va'>feature_drop_df</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
  batch <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/attributes.html'>attributes</a></span><span class='op'>(</span><span class='va'>feature_drop</span><span class='op'>)</span><span class='op'>$</span><span class='va'>thresholds</span><span class='op'>)</span>,
  cutoff <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/attributes.html'>attributes</a></span><span class='op'>(</span><span class='va'>feature_drop</span><span class='op'>)</span><span class='op'>$</span><span class='va'>thresholds</span><span class='op'>[</span><span class='st'>"lower"</span>, <span class='op'>]</span><span class='op'>)</span>
<span class='va'>spike_drop_df</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
  batch <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/attributes.html'>attributes</a></span><span class='op'>(</span><span class='va'>spike_drop</span><span class='op'>)</span><span class='op'>$</span><span class='va'>thresholds</span><span class='op'>)</span>,
  cutoff <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/attributes.html'>attributes</a></span><span class='op'>(</span><span class='va'>spike_drop</span><span class='op'>)</span><span class='op'>$</span><span class='va'>thresholds</span><span class='op'>[</span><span class='st'>"higher"</span>, <span class='op'>]</span><span class='op'>)</span>
<span class='va'>mito_drop_df</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
  batch <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/attributes.html'>attributes</a></span><span class='op'>(</span><span class='va'>mito_drop</span><span class='op'>)</span><span class='op'>$</span><span class='va'>thresholds</span><span class='op'>)</span>,
  cutoff <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/attributes.html'>attributes</a></span><span class='op'>(</span><span class='va'>mito_drop</span><span class='op'>)</span><span class='op'>$</span><span class='va'>thresholds</span><span class='op'>[</span><span class='st'>"higher"</span>, <span class='op'>]</span><span class='op'>)</span>
<span class='va'>ribo_drop_df</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
  batch <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/attributes.html'>attributes</a></span><span class='op'>(</span><span class='va'>ribo_drop</span><span class='op'>)</span><span class='op'>$</span><span class='va'>thresholds</span><span class='op'>)</span>,
  cutoff <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/attributes.html'>attributes</a></span><span class='op'>(</span><span class='va'>ribo_drop</span><span class='op'>)</span><span class='op'>$</span><span class='va'>thresholds</span><span class='op'>[</span><span class='st'>"higher"</span>, <span class='op'>]</span><span class='op'>)</span>
<span class='va'>qc_cutoffs_df</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/funprog.html'>Reduce</a></span><span class='op'>(</span>
  <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>y</span><span class='op'>)</span> <span class='fu'>inner_join</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>y</span>, by <span class='op'>=</span> <span class='st'>"batch"</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
    <span class='va'>libsize_drop_df</span>,
    <span class='va'>feature_drop_df</span>, 
    <span class='va'>spike_drop_df</span>, 
    <span class='va'>mito_drop_df</span>, 
    <span class='va'>ribo_drop_df</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>qc_cutoffs_df</span><span class='op'>$</span><span class='va'>batch</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>qc_cutoffs_df</span><span class='op'>$</span><span class='va'>batch</span>, <span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>qc_cutoffs_df</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>
  <span class='st'>"batch"</span>, <span class='st'>"total counts"</span>, <span class='st'>"total features"</span>, <span class='st'>"%ERCC"</span>, <span class='st'>"%mito"</span>, <span class='st'>"%ribo"</span><span class='op'>)</span>
<span class='va'>qc_cutoffs_df</span> <span class='op'>%&gt;%</span>
  <span class='fu'>arrange</span><span class='op'>(</span><span class='va'>batch</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>knitr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/knitr/man/kable.html'>kable</a></span><span class='op'>(</span>caption <span class='op'>=</span> <span class='st'>"QC cutoffs"</span>, digits <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<table>
<caption><span id="tab:unnamed-chunk-11">Table 3: </span>QC cutoffs</caption>
<thead>
<tr class="header">
<th style="text-align: left;">batch</th>
<th style="text-align: right;">total counts</th>
<th style="text-align: right;">total features</th>
<th style="text-align: right;">%ERCC</th>
<th style="text-align: right;">%mito</th>
<th style="text-align: right;">%ribo</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Blood.1</td>
<td style="text-align: right;">630.5</td>
<td style="text-align: right;">454.8</td>
<td style="text-align: right;">26.9</td>
<td style="text-align: right;">15.1</td>
<td style="text-align: right;">33.4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Blood.2</td>
<td style="text-align: right;">624.0</td>
<td style="text-align: right;">489.4</td>
<td style="text-align: right;">30.5</td>
<td style="text-align: right;">14.6</td>
<td style="text-align: right;">32.5</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Blood.3</td>
<td style="text-align: right;">398.5</td>
<td style="text-align: right;">353.4</td>
<td style="text-align: right;">34.1</td>
<td style="text-align: right;">14.1</td>
<td style="text-align: right;">36.2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Thymus.1</td>
<td style="text-align: right;">493.5</td>
<td style="text-align: right;">391.3</td>
<td style="text-align: right;">33.0</td>
<td style="text-align: right;">15.8</td>
<td style="text-align: right;">35.3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Thymus.2</td>
<td style="text-align: right;">510.0</td>
<td style="text-align: right;">385.9</td>
<td style="text-align: right;">33.9</td>
<td style="text-align: right;">17.3</td>
<td style="text-align: right;">33.4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Thymus.3</td>
<td style="text-align: right;">464.1</td>
<td style="text-align: right;">381.5</td>
<td style="text-align: right;">34.7</td>
<td style="text-align: right;">15.0</td>
<td style="text-align: right;">32.8</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SKW3.NA</td>
<td style="text-align: right;">2766.0</td>
<td style="text-align: right;">1467.2</td>
<td style="text-align: right;">6.7</td>
<td style="text-align: right;">12.3</td>
<td style="text-align: right;">32.0</td>
</tr>
</tbody>
</table>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>sce_pre_QC_outlier_removal</span> <span class='op'>&lt;-</span> <span class='va'>sce</span>
<span class='va'>keep</span> <span class='op'>&lt;-</span> <span class='op'>!</span><span class='op'>(</span><span class='va'>libsize_drop</span> <span class='op'>|</span> <span class='va'>feature_drop</span> <span class='op'>|</span> <span class='va'>spike_drop</span> <span class='op'>|</span> <span class='va'>mito_drop</span><span class='op'>)</span>
<span class='va'>sce_pre_QC_outlier_removal</span><span class='op'>$</span><span class='va'>keep</span> <span class='op'>&lt;-</span> <span class='va'>keep</span>
<span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>[</span>, <span class='va'>keep</span><span class='op'>]</span>
</code></pre>
</div>
</details>
</div>
<p>The table below summarises the number of cells per plate left following removal of outliers based on the QC metrics. The vast majority of cells are retained for all <code>sample</code>.</p>
<p>More cells are removed due to high percentages of ERCC transcripts or mitochondrial RNA than due to low library size and number of expressed genes. In total, we remove 92 cells based on these QC metrics, and retain 2,120 cells.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>inner_join</span><span class='op'>(</span>
  <span class='fu'>distinct</span><span class='op'>(</span>
    <span class='fu'><a href='https://rdrr.io/r/base/as.data.frame.html'>as.data.frame</a></span><span class='op'>(</span><span class='fu'>colData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>)</span>, 
    <span class='va'>batch</span>, <span class='va'>sample_gate</span>, <span class='va'>sample</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
    ByLibSize <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/tapply.html'>tapply</a></span><span class='op'>(</span>
      <span class='va'>libsize_drop</span>, 
      <span class='va'>sce_pre_QC_outlier_removal</span><span class='op'>$</span><span class='va'>batch</span>,
      <span class='va'>sum</span><span class='op'>)</span>,
    ByFeature <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/tapply.html'>tapply</a></span><span class='op'>(</span>
      <span class='va'>feature_drop</span>, 
      <span class='va'>sce_pre_QC_outlier_removal</span><span class='op'>$</span><span class='va'>batch</span>,
      <span class='va'>sum</span><span class='op'>)</span>, 
    BySpike <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/tapply.html'>tapply</a></span><span class='op'>(</span>
      <span class='va'>spike_drop</span>, 
      <span class='va'>sce_pre_QC_outlier_removal</span><span class='op'>$</span><span class='va'>batch</span>,
      <span class='va'>sum</span>,
      na.rm <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>,
    ByMito <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/tapply.html'>tapply</a></span><span class='op'>(</span>
      <span class='va'>mito_drop</span>, 
      <span class='va'>sce_pre_QC_outlier_removal</span><span class='op'>$</span><span class='va'>batch</span>,
      <span class='va'>sum</span>,
      na.rm <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>,
    Remaining <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/vector.html'>as.vector</a></span><span class='op'>(</span>
      <span class='fu'><a href='https://rdrr.io/r/base/unname.html'>unname</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/table.html'>table</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>,
    PercRemaining <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Round.html'>round</a></span><span class='op'>(</span>
      <span class='fl'>100</span> <span class='op'>*</span> <span class='fu'><a href='https://rdrr.io/r/base/vector.html'>as.vector</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/unname.html'>unname</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/table.html'>table</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>/</span>
        <span class='fu'><a href='https://rdrr.io/r/base/vector.html'>as.vector</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/unname.html'>unname</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/table.html'>table</a></span><span class='op'>(</span><span class='va'>sce_pre_QC_outlier_removal</span><span class='op'>$</span><span class='va'>batch</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>,
      <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
    <span class='fu'>tibble</span><span class='fu'>::</span><span class='fu'><a href='https://tibble.tidyverse.org/reference/rownames.html'>rownames_to_column</a></span><span class='op'>(</span><span class='st'>"batch"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/select.html'>select</a></span><span class='op'>(</span><span class='op'>-</span><span class='va'>batch</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>arrange</span><span class='op'>(</span><span class='fu'>desc</span><span class='op'>(</span><span class='va'>PercRemaining</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>knitr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/knitr/man/kable.html'>kable</a></span><span class='op'>(</span>
    caption <span class='op'>=</span> <span class='st'>"Number of samples removed by each QC step and the number of samples remaining. Rows are ordered by the percentage of cells remaining in that group."</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<table>
<caption><span id="tab:unnamed-chunk-13">Table 4: </span>Number of samples removed by each QC step and the number of samples remaining. Rows are ordered by the percentage of cells remaining in that group.</caption>
<thead>
<tr class="header">
<th style="text-align: left;">sample_gate</th>
<th style="text-align: left;">sample</th>
<th style="text-align: right;">ByLibSize</th>
<th style="text-align: right;">ByFeature</th>
<th style="text-align: right;">BySpike</th>
<th style="text-align: right;">ByMito</th>
<th style="text-align: right;">Remaining</th>
<th style="text-align: right;">PercRemaining</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">P5</td>
<td style="text-align: left;">Blood.3</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">159</td>
<td style="text-align: right;">99.4</td>
</tr>
<tr class="even">
<td style="text-align: left;">P5</td>
<td style="text-align: left;">Blood.2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">137</td>
<td style="text-align: right;">97.9</td>
</tr>
<tr class="odd">
<td style="text-align: left;">P5</td>
<td style="text-align: left;">Thymus.2</td>
<td style="text-align: right;">8</td>
<td style="text-align: right;">8</td>
<td style="text-align: right;">18</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">525</td>
<td style="text-align: right;">95.8</td>
</tr>
<tr class="even">
<td style="text-align: left;">P5</td>
<td style="text-align: left;">Blood.1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">153</td>
<td style="text-align: right;">95.6</td>
</tr>
<tr class="odd">
<td style="text-align: left;">P5</td>
<td style="text-align: left;">Thymus.1</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">18</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">546</td>
<td style="text-align: right;">95.5</td>
</tr>
<tr class="even">
<td style="text-align: left;">P5</td>
<td style="text-align: left;">Thymus.3</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">24</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">545</td>
<td style="text-align: right;">95.3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">P5</td>
<td style="text-align: left;">SKW3.NA</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">55</td>
<td style="text-align: right;">91.7</td>
</tr>
</tbody>
</table>
</div>
<h3 id="checking-for-removal-of-biologically-relevant-subpopulations">Checking for removal of biologically relevant subpopulations</h3>
<p>The biggest practical concern during QC is whether an entire cell type is inadvertently discarded. There is always some risk of this occurring as the QC metrics are never fully independent of biological state. We can diagnose cell type loss by looking for systematic differences in gene expression between the discarded and retained cells.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>lost</span> <span class='op'>&lt;-</span> <span class='fu'>calculateAverage</span><span class='op'>(</span><span class='fu'>counts</span><span class='op'>(</span><span class='va'>sce_pre_QC_outlier_removal</span><span class='op'>)</span><span class='op'>[</span>, <span class='op'>!</span><span class='va'>keep</span><span class='op'>]</span><span class='op'>)</span>
<span class='va'>kept</span> <span class='op'>&lt;-</span> <span class='fu'>calculateAverage</span><span class='op'>(</span><span class='fu'>counts</span><span class='op'>(</span><span class='va'>sce_pre_QC_outlier_removal</span><span class='op'>)</span><span class='op'>[</span>, <span class='va'>keep</span><span class='op'>]</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://bioinf.wehi.edu.au/edgeR'>edgeR</a></span><span class='op'>)</span>
<span class='va'>logged</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/cpm.html'>cpm</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/cbind.html'>cbind</a></span><span class='op'>(</span><span class='va'>lost</span>, <span class='va'>kept</span><span class='op'>)</span>, log <span class='op'>=</span> <span class='cn'>TRUE</span>, prior.count <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>
<span class='va'>logFC</span> <span class='op'>&lt;-</span> <span class='va'>logged</span><span class='op'>[</span>, <span class='fl'>1</span><span class='op'>]</span> <span class='op'>-</span> <span class='va'>logged</span><span class='op'>[</span>, <span class='fl'>2</span><span class='op'>]</span>
<span class='va'>abundance</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colSums.html'>rowMeans</a></span><span class='op'>(</span><span class='va'>logged</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>If the discarded pool is enriched for a certain cell type, we should observe increased expression of the corresponding marker genes. Figure <a href="#fig:qc-md-plot">9</a> shows the result of this analysis, highlighting that there are few genes with a large <span class="math inline">\(logFC\)</span> between ‘lost’ and ‘kept’ cells (those few genes with larger <span class="math inline">\(logFC\)</span> have low average expression). This suggests that the QC step did not inadvertently filter out an entire biologically relevant subpopulation.</p>
<aside>
An interactive version of Figure <a href="#fig:qc-md-plot">9</a> is available from <a href="../output/glimma-plots/qc-md-plot.html"><code>output/glimma-plots/qc-md-plot.html</code></a>.
</aside>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>is_mito</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>%in%</span> <span class='va'>mito_set</span>
<span class='va'>is_ribo</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>%in%</span> <span class='va'>ribo_set</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/par.html'>par</a></span><span class='op'>(</span>mfrow <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span>
  <span class='va'>abundance</span>,
  <span class='va'>logFC</span>,
  xlab <span class='op'>=</span> <span class='st'>"Average expression"</span>,
  ylab <span class='op'>=</span> <span class='st'>"Log-FC (lost/kept)"</span>,
  pch <span class='op'>=</span> <span class='fl'>16</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/points.html'>points</a></span><span class='op'>(</span>
  <span class='va'>abundance</span><span class='op'>[</span><span class='va'>is_mito</span><span class='op'>]</span>,
  <span class='va'>logFC</span><span class='op'>[</span><span class='va'>is_mito</span><span class='op'>]</span>,
  col <span class='op'>=</span> <span class='st'>"dodgerblue"</span>,
  pch <span class='op'>=</span> <span class='fl'>16</span>,
  cex <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/points.html'>points</a></span><span class='op'>(</span>
  <span class='va'>abundance</span><span class='op'>[</span><span class='va'>is_ribo</span><span class='op'>]</span>,
  <span class='va'>logFC</span><span class='op'>[</span><span class='va'>is_ribo</span><span class='op'>]</span>,
  col <span class='op'>=</span> <span class='st'>"orange"</span>,
  pch <span class='op'>=</span> <span class='fl'>16</span>,
  cex <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/abline.html'>abline</a></span><span class='op'>(</span>h <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span>, col <span class='op'>=</span> <span class='st'>"red"</span>, lty <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:qc-md-plot"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/qc-md-plot-1.png" alt="Log-fold change in expression in the discarded cells compared to the retained cells. Each point represents a gene with mitochondrial transcripts in blue and ribosomal protein genes in orange. Dashed red lines indicate $|logFC| = 1" width="624" />
<p class="caption">
Figure 9: Log-fold change in expression in the discarded cells compared to the retained cells. Each point represents a gene with mitochondrial transcripts in blue and ribosomal protein genes in orange. Dashed red lines indicate $|logFC| = 1
</p>
</div>
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/hasaru-k/GlimmaV2'>Glimma</a></span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/Glimma/man/glMDPlot.html'>glMDPlot</a></span><span class='op'>(</span>
  x <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>abundance <span class='op'>=</span> <span class='va'>abundance</span>, logFC <span class='op'>=</span> <span class='va'>logFC</span><span class='op'>)</span>,
  xval <span class='op'>=</span> <span class='st'>"abundance"</span>, 
  yval <span class='op'>=</span> <span class='st'>"logFC"</span>,
  counts <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/cbind.html'>cbind</a></span><span class='op'>(</span><span class='va'>lost</span>, <span class='va'>kept</span><span class='op'>)</span>,
    anno <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/cbind.html'>cbind</a></span><span class='op'>(</span>
      <span class='fu'>flattenDF</span><span class='op'>(</span><span class='fu'>rowData</span><span class='op'>(</span><span class='va'>sce_pre_QC_outlier_removal</span><span class='op'>)</span><span class='op'>)</span>, 
      <span class='fu'>DataFrame</span><span class='op'>(</span>
        GeneID <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce_pre_QC_outlier_removal</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>,
  display.columns <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"ENSEMBL.SYMBOL"</span>, <span class='st'>"ENSEMBL.GENEID"</span>, <span class='st'>"ENSEMBL.SEQNAME"</span><span class='op'>)</span>,
  groups <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"lost"</span>, <span class='st'>"kept"</span><span class='op'>)</span><span class='op'>)</span>,
  samples <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"lost"</span>, <span class='st'>"kept"</span><span class='op'>)</span>,
  status <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/unname.html'>unname</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span><span class='va'>is_mito</span>, <span class='fl'>1</span>, <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span><span class='va'>is_ribo</span>, <span class='op'>-</span><span class='fl'>1</span>, <span class='fl'>0</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>,
  transform <span class='op'>=</span> <span class='cn'>FALSE</span>,
  main <span class='op'>=</span> <span class='st'>"lost vs. kept"</span>,
  side.ylab <span class='op'>=</span> <span class='st'>"Average count"</span>,
  cols <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"orange"</span>, <span class='st'>"black"</span>, <span class='st'>"dodgerBlue"</span><span class='op'>)</span>,
  path <span class='op'>=</span> <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"output"</span><span class='op'>)</span>,
  html <span class='op'>=</span> <span class='st'>"qc-md-plot"</span>,
  launch <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>Another concern is whether the cells removed during QC preferentially derive from particular experimental groups. Figure <a href="#fig:barplot-highlighting-outliers">10</a> shows that we exclude more <code>Thymus</code> than <code>Blood</code> samples, although this is roughly proportional to the abundance of samples from each tissue.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>ggcells</span><span class='op'>(</span><span class='va'>sce_pre_QC_outlier_removal</span>, exprs_values <span class='op'>=</span> <span class='st'>"counts"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>batch</span>, fill <span class='op'>=</span> <span class='va'>keep</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Number of cells"</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>angle <span class='op'>=</span> <span class='fl'>45</span>, hjust <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:barplot-highlighting-outliers"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/barplot-highlighting-outliers-1.png" alt="Cells removed during QC, stratified by experimental group. " width="624" />
<p class="caption">
Figure 10: Cells removed during QC, stratified by experimental group.
</p>
</div>
</div>
<p>Finally, Figure <a href="#fig:qcplot-highlighting-outliers">11</a> compares the QC metrics of the discarded and retained cells.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce_pre_QC_outlier_removal</span>,
  <span class='st'>"sum"</span>,
  x <span class='op'>=</span> <span class='st'>"sample"</span>,
  colour_by <span class='op'>=</span> <span class='st'>"keep"</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_continuous.html'>scale_y_log10</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>angle <span class='op'>=</span> <span class='fl'>45</span>, hjust <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/annotation_logticks.html'>annotation_logticks</a></span><span class='op'>(</span>
    sides <span class='op'>=</span> <span class='st'>"l"</span>,
    short <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grid/unit.html'>unit</a></span><span class='op'>(</span><span class='fl'>0.03</span>, <span class='st'>"cm"</span><span class='op'>)</span>,
    mid <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grid/unit.html'>unit</a></span><span class='op'>(</span><span class='fl'>0.06</span>, <span class='st'>"cm"</span><span class='op'>)</span>,
    long <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grid/unit.html'>unit</a></span><span class='op'>(</span><span class='fl'>0.09</span>, <span class='st'>"cm"</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce_pre_QC_outlier_removal</span>,
  <span class='st'>"detected"</span>,
  x <span class='op'>=</span> <span class='st'>"sample"</span>,
  colour_by <span class='op'>=</span> <span class='st'>"keep"</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>angle <span class='op'>=</span> <span class='fl'>45</span>, hjust <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>p3</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce_pre_QC_outlier_removal</span>,
  <span class='st'>"subsets_Mito_percent"</span>,
  x <span class='op'>=</span> <span class='st'>"sample"</span>,
  colour_by <span class='op'>=</span> <span class='st'>"keep"</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>angle <span class='op'>=</span> <span class='fl'>45</span>, hjust <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>p4</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce_pre_QC_outlier_removal</span>,
  <span class='st'>"subsets_Ribo_percent"</span>,
  x <span class='op'>=</span> <span class='st'>"sample"</span>,
  colour_by <span class='op'>=</span> <span class='st'>"keep"</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>angle <span class='op'>=</span> <span class='fl'>45</span>, hjust <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>p1</span> <span class='op'>+</span> <span class='va'>p2</span> <span class='op'>+</span> <span class='va'>p3</span> <span class='op'>+</span> <span class='va'>p4</span> <span class='op'>+</span> <span class='fu'><a href='https://patchwork.data-imaginist.com/reference/plot_layout.html'>plot_layout</a></span><span class='op'>(</span>guides <span class='op'>=</span> <span class='st'>"collect"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:qcplot-highlighting-outliers"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/qcplot-highlighting-outliers-1.png" alt="Distribution of QC metrics for each plate in the dataset. Each point represents a cell and is colored according to whether it was discarded during the QC process. Note that a cell will only be kept if it passes the relevant threshold for all QC metrics." width="624" />
<p class="caption">
Figure 11: Distribution of QC metrics for each plate in the dataset. Each point represents a cell and is colored according to whether it was discarded during the QC process. Note that a cell will only be kept if it passes the relevant threshold for all QC metrics.
</p>
</div>
</div>
<h2 id="summary-1">Summary</h2>
<p>To conclude, Figures <a href="#fig:qcplot-post-outlier-removal">12</a> and <a href="#fig:qcplot-post-outlier-removal-by-plate-number">13</a> shows that following QC that most samples have similar QC metrics, as is to be expected.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce</span>,
  <span class='st'>"sum"</span>,
  x <span class='op'>=</span> <span class='st'>"sample"</span>,
  other_fields <span class='op'>=</span> <span class='st'>"plate_number"</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span>,
  colour_by <span class='op'>=</span> <span class='st'>"sample"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_continuous.html'>scale_y_log10</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/annotation_logticks.html'>annotation_logticks</a></span><span class='op'>(</span>
    sides <span class='op'>=</span> <span class='st'>"l"</span>,
    short <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grid/unit.html'>unit</a></span><span class='op'>(</span><span class='fl'>0.03</span>, <span class='st'>"cm"</span><span class='op'>)</span>,
    mid <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grid/unit.html'>unit</a></span><span class='op'>(</span><span class='fl'>0.06</span>, <span class='st'>"cm"</span><span class='op'>)</span>,
    long <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grid/unit.html'>unit</a></span><span class='op'>(</span><span class='fl'>0.09</span>, <span class='st'>"cm"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>sample_colours</span>, name <span class='op'>=</span> <span class='st'>"sample"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>""</span><span class='op'>)</span>
<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce</span>,
  <span class='st'>"detected"</span>,
  x <span class='op'>=</span> <span class='st'>"sample"</span>,
  other_fields <span class='op'>=</span> <span class='st'>"plate_number"</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span>,
  colour_by <span class='op'>=</span> <span class='st'>"sample"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>""</span><span class='op'>)</span>
<span class='va'>p3</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce</span>,
  <span class='st'>"subsets_Mito_percent"</span>,
  x <span class='op'>=</span> <span class='st'>"sample"</span>,
  other_fields <span class='op'>=</span> <span class='st'>"plate_number"</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span>,
  colour_by <span class='op'>=</span> <span class='st'>"sample"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_continuous.html'>scale_y_log10</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/lims.html'>ylim</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='cn'>NA</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>""</span><span class='op'>)</span>
<span class='va'>p4</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce</span>,
  <span class='st'>"subsets_Ribo_percent"</span>,
  x <span class='op'>=</span> <span class='st'>"sample"</span>,
  other_fields <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"plate_number"</span>, <span class='st'>"sample_type"</span><span class='op'>)</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span>,
  colour_by <span class='op'>=</span> <span class='st'>"sample"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_continuous.html'>scale_y_log10</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/lims.html'>ylim</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='cn'>NA</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>""</span><span class='op'>)</span>
<span class='va'>p5</span> <span class='op'>&lt;-</span> <span class='fu'>plotColData</span><span class='op'>(</span>
  <span class='va'>sce</span>,
  <span class='st'>"altexps_ERCC_percent"</span>,
  x <span class='op'>=</span> <span class='st'>"sample"</span>,
  other_fields <span class='op'>=</span> <span class='st'>"plate_number"</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span>,
  colour_by <span class='op'>=</span> <span class='st'>"sample"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_continuous.html'>scale_y_log10</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/lims.html'>ylim</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='cn'>NA</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>""</span><span class='op'>)</span>

<span class='va'>p1</span> <span class='op'>+</span> <span class='va'>p2</span> <span class='op'>+</span> <span class='va'>p3</span> <span class='op'>+</span> <span class='va'>p4</span> <span class='op'>+</span> <span class='va'>p5</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://patchwork.data-imaginist.com/reference/plot_layout.html'>plot_layout</a></span><span class='op'>(</span>ncol <span class='op'>=</span> <span class='fl'>2</span>, guides <span class='op'>=</span> <span class='st'>"collect"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:qcplot-post-outlier-removal"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/qcplot-post-outlier-removal-1.png" alt="Distributions of various QC metrics for all samples in the dataset post-QC." width="624" />
<p class="caption">
Figure 12: Distributions of various QC metrics for all samples in the dataset post-QC.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='va'>p1</span> <span class='op'>+</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_wrap.html'>facet_wrap</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>plate_number</span>, ncol <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>
<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='va'>p2</span> <span class='op'>+</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_wrap.html'>facet_wrap</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>plate_number</span>, ncol <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>
<span class='va'>p3</span> <span class='op'>&lt;-</span> <span class='va'>p3</span> <span class='op'>+</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_wrap.html'>facet_wrap</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>plate_number</span>, ncol <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>
<span class='va'>p4</span> <span class='op'>&lt;-</span> <span class='va'>p4</span> <span class='op'>+</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_wrap.html'>facet_wrap</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>plate_number</span>, ncol <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>
<span class='va'>p5</span> <span class='op'>&lt;-</span> <span class='va'>p5</span> <span class='op'>+</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_wrap.html'>facet_wrap</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>plate_number</span>, ncol <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>

<span class='va'>p1</span> <span class='op'>+</span> <span class='va'>p2</span> <span class='op'>+</span> <span class='va'>p3</span> <span class='op'>+</span> <span class='va'>p4</span> <span class='op'>+</span> <span class='va'>p5</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://patchwork.data-imaginist.com/reference/plot_layout.html'>plot_layout</a></span><span class='op'>(</span>ncol <span class='op'>=</span> <span class='fl'>2</span>, guides <span class='op'>=</span> <span class='st'>"collect"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:qcplot-post-outlier-removal-by-plate-number"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/qcplot-post-outlier-removal-by-plate-number-1.png" alt="Distributions of various QC metrics for all **single-cell** samples in the dataset, stratified by `plate_number`." width="624" />
<p class="caption">
Figure 13: Distributions of various QC metrics for all <strong>single-cell</strong> samples in the dataset, stratified by <code>plate_number</code>.
</p>
</div>
</div>
<p>Figure <a href="#fig:coldata-postqc">14</a> summarises the experimental design following QC. We remind ourself that all <code>Blood 2</code> samples on plate number <code>LCE511</code> were excluded in <a href="#plate-alignment-issue-lce511">Plate alignment issue (LCE511)</a>.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='fu'>ggcells</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>plate_number</span>, fill <span class='op'>=</span> <span class='va'>sample</span><span class='op'>)</span>,
    position <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/position_stack.html'>position_fill</a></span><span class='op'>(</span>reverse <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Frequency"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>10</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>sample_colours</span><span class='op'>)</span>
<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='fu'>ggcells</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_bar.html'>geom_bar</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>plate_number</span>, fill <span class='op'>=</span> <span class='va'>plate_number</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"Number of samples"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>10</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>plate_number_colours</span><span class='op'>)</span>
<span class='va'>p1</span> <span class='op'>|</span> <span class='va'>p2</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:coldata-postqc"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/coldata-postqc-1.png" alt="Breakdown of experiment by plate." width="624" />
<p class="caption">
Figure 14: Breakdown of experiment by plate.
</p>
</div>
</div>
<h1 id="examining-gene-level-metrics">Examining gene-level metrics</h1>
<h2 id="inspecting-the-most-highly-expressed-genes">Inspecting the most highly expressed genes</h2>
<p>Figure <a href="#fig:topgenes">15</a> shows the most highly expressed genes across the cell population in the combined data set. Some of them are mitochondrial genes, matching what we’ve already seen in the QC metrics, and ribosomal protein genes, which commonly are amongst the most highly expressed genes in scRNA-seq data.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>plotHighestExprs</span><span class='op'>(</span><span class='va'>sce</span>, n <span class='op'>=</span> <span class='fl'>50</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:topgenes"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/topgenes-1.png" alt="Percentage of total counts assigned to the top 50 most highly-abundant features in the combined data set. For each feature, each bar represents the percentage assigned to that feature for a single cell, while the circle represents the average across all cells." width="624" />
<p class="caption">
Figure 15: Percentage of total counts assigned to the top 50 most highly-abundant features in the combined data set. For each feature, each bar represents the percentage assigned to that feature for a single cell, while the circle represents the average across all cells.
</p>
</div>
</div>
<p>Figure <a href="#fig:topgenes-filtered">16</a> shows the most highly expressed endogenous genes after excluding the mitochondrial and ribosomal protein genes.</p>
<aside>
A cursory analysis shows that many of these are lncRNAs. <strong>TODO</strong>: What does this mean?
</aside>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>plotHighestExprs</span><span class='op'>(</span>
  <span class='va'>sce</span>,
  drop_features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>%in%</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='va'>mito_set</span>, <span class='va'>ribo_set</span><span class='op'>)</span>,
  n <span class='op'>=</span> <span class='fl'>50</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:topgenes-filtered"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/topgenes-filtered-1.png" alt="Percentage of total counts assigned to the top 50 most highly-abundant features in the combined data set. For each feature, each bar represents the percentage assigned to that feature for a single cell, while the circle represents the average across all cells. Bars are coloured by the total number of expressed features in each cell, while circles are coloured according to whether the feature is labelled as a control feature." width="624" />
<p class="caption">
Figure 16: Percentage of total counts assigned to the top 50 most highly-abundant features in the combined data set. For each feature, each bar represents the percentage assigned to that feature for a single cell, while the circle represents the average across all cells. Bars are coloured by the total number of expressed features in each cell, while circles are coloured according to whether the feature is labelled as a control feature.
</p>
</div>
</div>
<h2 id="filtering-out-low-abundance-genes">Filtering out low-abundance genes</h2>
<p>Low-abundance genes are problematic as zero or near-zero counts do not contain much information for reliable statistical inference <span class="citation" data-cites="bourgon2010independent">(<a href="#ref-bourgon2010independent" role="doc-biblioref">Bourgon, Gentleman, and Huber 2010</a>)</span>. These genes typically do not provide enough evidence to reject the null hypothesis during testing, yet they still increase the severity of the multiple testing correction. In addition, the discreteness of the counts may interfere with statistical procedures, e.g., by compromising the accuracy of continuous approximations. Thus, low-abundance genes are often removed in many RNA-seq analysis pipelines before the application of downstream methods.</p>
<p>The ‘optimal’ choice of filtering strategy depends on the downstream application. A more aggressive filter is usually required to remove discreteness (e.g., for normalization) compared to that required for removing underpowered tests. For hypothesis testing, the filter statistic should also be independent of the test statistic under the null hypothesis. Thus, we (or the relevant function) will filter at each step as needed, rather than applying a single filter for the entire analysis.</p>
<p>Several metrics can be used to define low-abundance genes. The most obvious is the average count for each gene, computed across all cells in the data set. We typically observe a peak of moderately expressed genes following a plateau of lowly expressed genes (Figure <a href="#fig:abhist">17</a>).</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>ave_counts</span> <span class='op'>&lt;-</span> <span class='fu'>calculateAverage</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/par.html'>par</a></span><span class='op'>(</span>mfrow <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/hist.html'>hist</a></span><span class='op'>(</span>
  x <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log10</a></span><span class='op'>(</span><span class='va'>ave_counts</span><span class='op'>)</span>, 
  breaks <span class='op'>=</span> <span class='fl'>100</span>, 
  main <span class='op'>=</span> <span class='st'>""</span>, 
  col <span class='op'>=</span> <span class='st'>"grey"</span>,
  xlab <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/expression.html'>expression</a></span><span class='op'>(</span><span class='va'>Log</span><span class='op'>[</span><span class='fl'>10</span><span class='op'>]</span> <span class='op'>~</span> <span class='st'>"average count"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:abhist"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/abhist-1.png" alt="Histogram of log-average counts for all genes in the combined data set." width="624" />
<p class="caption">
Figure 17: Histogram of log-average counts for all genes in the combined data set.
</p>
</div>
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>to_keep</span> <span class='op'>&lt;-</span> <span class='va'>ave_counts</span> <span class='op'>&gt;</span> <span class='fl'>0</span>
<span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>[</span><span class='va'>to_keep</span>, <span class='op'>]</span>
</code></pre>
</div>
</details>
</div>
<p>We remove 139 genes that are not expressed in any cell. Such genes provide no information and would be removed by any filtering strategy. We retain 27,999 for downstream analysis.</p>
<h1 id="normalization">Normalization</h1>
<p>UMI counts are subject to differences in capture efficiency and sequencing depth between cells <span class="citation" data-cites="stegle2015computational">(<a href="#ref-stegle2015computational" role="doc-biblioref">Stegle, Teichmann, and Marioni 2015</a>)</span>. Normalization is required to eliminate these cell-specific biases prior to downstream quantitative analyses. This is often done by assuming that most genes are not differentially expressed (DE) between cells. Any systematic difference in count size across the non-DE majority of genes between two cells is assumed to represent bias and is removed by scaling. More specifically, ‘size factors’ are calculated that represent the extent to which counts should be scaled in each library.</p>
<h2 id="using-the-deconvolution-method-to-deal-with-zero-counts">Using the deconvolution method to deal with zero counts</h2>
<p>Size factors can be computed with several different approaches, e.g., using the <code>estimateSizeFactorsFromMatrix</code> function in the <em><a href="https://bioconductor.org/packages/3.12/DESeq2">DESeq2</a></em> package <span class="citation" data-cites="anders2010differential love2014moderated">(<a href="#ref-anders2010differential" role="doc-biblioref">Anders and Huber 2010</a>; <a href="#ref-love2014moderated" role="doc-biblioref">Love, Huber, and Anders 2014</a>)</span>, or with the <code>calcNormFactors</code> function <span class="citation" data-cites="robinson2010scaling">(<a href="#ref-robinson2010scaling" role="doc-biblioref">Robinson and Oshlack 2010</a>)</span> in the <em><a href="https://bioconductor.org/packages/3.12/edgeR">edgeR</a></em> package. However, single-cell data can be problematic for these bulk data-based methods due to the dominance of low and zero counts. To overcome this, we pool counts from many cells to increase the count size for accurate size factor estimation <span class="citation" data-cites="lun2016pooling">(<a href="#ref-lun2016pooling" role="doc-biblioref">A. T. Lun, Bach, and Marioni 2016</a>)</span>. Pool-based size factors are then ‘deconvolved’ into cell-based factors for cell-specific normalization. This removes scaling biases associated with cell-specific differences in capture efficiency, sequencing depth and composition biases.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>scran</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>9434887</span><span class='op'>)</span>
<span class='va'>clusters</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scran/man/quickCluster.html'>quickCluster</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/Round.html'>round</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/proportions.html'>prop.table</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/table.html'>table</a></span><span class='op'>(</span><span class='va'>clusters</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span><span class='op'>)</span>, <span class='fl'>1</span><span class='op'>)</span>, <span class='fl'>2</span><span class='op'>)</span>
<span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scran/man/computeSumFactors.html'>computeSumFactors</a></span><span class='op'>(</span><span class='va'>sce</span>, clusters <span class='op'>=</span> <span class='va'>clusters</span>, min.mean <span class='op'>=</span> <span class='fl'>0.1</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/summary.html'>summary</a></span><span class='op'>(</span><span class='fu'>sizeFactors</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>We check that the size factors are roughly aligned with the total library sizes (Figure <a href="#fig:normplot">18</a>). Deviations from the diagonal correspond to composition biases due to differential expression between cell subpopulations.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>xlim</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.1</span>, <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>sum</span><span class='op'>)</span> <span class='op'>/</span> <span class='fl'>1e3</span><span class='op'>)</span>
<span class='va'>ylim</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/range.html'>range</a></span><span class='op'>(</span><span class='fu'>sizeFactors</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/par.html'>par</a></span><span class='op'>(</span>mfrow <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>3</span>, <span class='fl'>3</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>lapply</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span><span class='op'>)</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>b</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span> <span class='op'>==</span> <span class='va'>b</span><span class='op'>]</span>
  <span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span>
    x <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>sum</span> <span class='op'>/</span> <span class='fl'>1e3</span>, 
    y <span class='op'>=</span> <span class='fu'>sizeFactors</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>, 
    log <span class='op'>=</span> <span class='st'>"xy"</span>,
    xlab <span class='op'>=</span> <span class='st'>"Library size (thousands)"</span>, 
    ylab <span class='op'>=</span> <span class='st'>"Size factor"</span>,
    xlim <span class='op'>=</span> <span class='va'>xlim</span>,
    ylim <span class='op'>=</span> <span class='va'>ylim</span>,
    main <span class='op'>=</span> <span class='va'>b</span>,
    pch <span class='op'>=</span> <span class='fl'>16</span>,
    cex <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span>
<span class='op'>}</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:normplot"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/normplot-1.png" alt="Size factors from deconvolution, plotted against library sizes for all cells in each dataset. Axes are shown on a log-scale." width="624" />
<p class="caption">
Figure 18: Size factors from deconvolution, plotted against library sizes for all cells in each dataset. Axes are shown on a log-scale.
</p>
</div>
</div>
<h2 id="applying-the-size-factors-to-normalize-gene-expression">Applying the size factors to normalize gene expression</h2>
<p>The count data are used to compute normalized log-expression values for use in downstream analyses. Each value is defined as the log<sub>2</sub>-ratio of each count to the size factor for the corresponding cell, after adding a prior count of 1 to avoid undefined values at zero counts. Division by the size factor ensures that any cell-specific biases are removed.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># NOTE: No need for multiBatchNorm() since the size factors were computed </span>
<span class='co'>#       jointly using all samples.</span>
<span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='fu'>logNormCounts</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<h2 id="computing-separate-size-factors-for-spike-in-transcripts">Computing separate size factors for spike-in transcripts</h2>
<p>Size factors computed from the counts for endogenous genes are usually not appropriate for normalizing the counts for spike-in transcripts. To ensure normalization is performed correctly, we compute a separate set of size factors for the spike-in set. For each cell, the spike-in-specific size factor is defined as the total count across all transcripts in the spike-in set. This assumes that none of the spike-in transcripts are differentially expressed, which is reasonable given that the same amount and composition of spike-in RNA should have been added to each cell <span class="citation" data-cites="lun2017assessing">(<a href="#ref-lun2017assessing" role="doc-biblioref">A. T. L. Lun et al. 2017</a>)</span>.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>sizeFactors</span><span class='op'>(</span><span class='fu'>altExp</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"ERCC"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'>librarySizeFactors</span><span class='op'>(</span><span class='fu'>altExp</span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"ERCC"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<h1 id="feature-selection">Feature selection</h1>
<h2 id="motivation-1">Motivation</h2>
<p>We often use scRNA-seq data in exploratory analyses to characterize heterogeneity across cells. Procedures like dimensionality reduction and clustering compare cells based on their gene expression profiles, which involves aggregating per-gene differences into a single (dis)similarity metric between a pair of cells. The choice of genes to use in this calculation has a major impact on the behaviour of the metric and the performance of downstream methods. We want to select genes that contain useful information about the biology of the system while removing genes that contain random noise. This aims to preserve interesting biological structure without the variance that obscures that structure. It also reduces the size of the dataset to improve computational efficiency of later steps.</p>
<h2 id="quantifying-per-gene-variation">Quantifying per-gene variation</h2>
<h3 id="variance-of-the-log-counts">Variance of the log-counts</h3>
<p>The simplest approach to quantifying per-gene variation is to simply compute the variance of the log-normalized expression values (referred to as “log-counts” for simplicity) for each gene across all cells in the population <span class="citation" data-cites="lun2016step">(<a href="#ref-lun2016step" role="doc-biblioref">A. T. L. Lun, McCarthy, and Marioni 2016</a>)</span>. This has an advantage in that the feature selection is based on the same log-values that are used for later downstream steps. In particular, genes with the largest variances in log-values will contribute the most to the Euclidean distances between cells. By using log-values here, we ensure that our quantitative definition of heterogeneity is consistent throughout the entire analysis.</p>
<p>Calculation of the per-gene variance is simple but feature selection requires modelling of the mean-variance relationship.</p>
<h3 id="quantifying-technical-noise">Quantifying technical noise</h3>
<p>To account for the mean-variance relationship, we fit a trend to the variance with respect to abundance across the ERCC spike-in transcripts. The premise here is that spike-ins should not be affected by biological variation, so the fitted value of the spike-in trend should represent a better estimate of the technical component for each gene.</p>
<h3 id="accounting-for-blocking-factors">Accounting for blocking factors</h3>
<p>Data containing multiple batches will often exhibit batch effects. We are usually not interested in highly variable genes (HVGs) that are driven by batch effects. Rather, we want to focus on genes that are highly variable within each batch. This is naturally achieved by performing trend fitting and variance decomposition separately for each batch. Here we treat each plate as a separate batch.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span>
<span class='va'>batch_colours</span> <span class='op'>&lt;-</span> <span class='va'>plate_number_colours</span>
<span class='va'>var_fit</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scran/man/modelGeneVarWithSpikes.html'>modelGeneVarWithSpikes</a></span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"ERCC"</span>, block <span class='op'>=</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>The use of a batch-specific trend fit is useful as it accommodates differences in the mean-variance trends between batches. This is especially important if batches exhibit systematic technical differences, e.g., differences in coverage or in the amount of spike-in RNA added.</p>
<p>Figure <a href="#fig:trendplots">19</a> visualizes the quality of the batch-specific trend fits and Figure <a href="#fig:trendplots-fits">20</a> highlights the need for batch-specific estimates of these fits. The analysis of each plate yields estimates of the biological and technical components for each gene, which are averaged across plates to take advantage of information from multiple batches.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>xlim</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span><span class='va'>var_fit</span><span class='op'>$</span><span class='va'>per.block</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span> <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>$</span><span class='va'>mean</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>ylim</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span><span class='va'>var_fit</span><span class='op'>$</span><span class='va'>per.block</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span> <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>$</span><span class='va'>total</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/par.html'>par</a></span><span class='op'>(</span>mfrow <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>2</span>, <span class='fl'>3</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>blocked_stats</span> <span class='op'>&lt;-</span> <span class='va'>var_fit</span><span class='op'>$</span><span class='va'>per.block</span>
<span class='kw'>for</span> <span class='op'>(</span><span class='va'>i</span> <span class='kw'>in</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>blocked_stats</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='va'>current</span> <span class='op'>&lt;-</span> <span class='va'>blocked_stats</span><span class='op'>[[</span><span class='va'>i</span><span class='op'>]</span><span class='op'>]</span>
  <span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span>
    <span class='va'>current</span><span class='op'>$</span><span class='va'>mean</span>,
    <span class='va'>current</span><span class='op'>$</span><span class='va'>total</span>,
    main <span class='op'>=</span> <span class='va'>i</span>,
    pch <span class='op'>=</span> <span class='fl'>16</span>,
    cex <span class='op'>=</span> <span class='fl'>0.5</span>,
    xlab <span class='op'>=</span> <span class='st'>"Mean of log-expression"</span>,
    ylab <span class='op'>=</span> <span class='st'>"Variance of log-expression"</span>,
    xlim <span class='op'>=</span> <span class='va'>xlim</span>,
    ylim <span class='op'>=</span> <span class='va'>ylim</span><span class='op'>)</span>
  <span class='va'>curfit</span> <span class='op'>&lt;-</span> <span class='fu'>metadata</span><span class='op'>(</span><span class='va'>current</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/graphics/points.html'>points</a></span><span class='op'>(</span><span class='va'>curfit</span><span class='op'>$</span><span class='va'>mean</span>, <span class='va'>curfit</span><span class='op'>$</span><span class='va'>var</span>, col <span class='op'>=</span> <span class='st'>"red"</span>, pch <span class='op'>=</span> <span class='fl'>16</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/graphics/curve.html'>curve</a></span><span class='op'>(</span><span class='va'>curfit</span><span class='op'>$</span><span class='fu'>trend</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>, col <span class='op'>=</span> <span class='va'>batch_colours</span><span class='op'>[[</span><span class='va'>i</span><span class='op'>]</span><span class='op'>]</span>, add <span class='op'>=</span> <span class='cn'>TRUE</span>, lwd <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span> 
<span class='op'>}</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:trendplots"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/trendplots-1.png" alt="Variance of normalized log-expression values for each gene in each plate, plotted against the mean log-expression. The coloured line represents the mean-dependent trend fitted to the variances of the spike-in transcripts (red)." width="624" />
<p class="caption">
Figure 19: Variance of normalized log-expression values for each gene in each plate, plotted against the mean log-expression. The coloured line represents the mean-dependent trend fitted to the variances of the spike-in transcripts (red).
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/graphics/par.html'>par</a></span><span class='op'>(</span>mfrow <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>
<span class='co'># NOTE: 'Fake' plot with right dimensions to overlay trends.</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span>
  x <span class='op'>=</span> <span class='va'>current</span><span class='op'>$</span><span class='va'>mean</span>, 
  y <span class='op'>=</span> <span class='va'>current</span><span class='op'>$</span><span class='va'>total</span>, 
  pch <span class='op'>=</span> <span class='fl'>0</span>,
  cex <span class='op'>=</span> <span class='fl'>0.6</span>, 
  xlab <span class='op'>=</span> <span class='st'>"Mean log-expression"</span>, 
  ylab <span class='op'>=</span> <span class='st'>"Variance of log-expression"</span>, 
  main <span class='op'>=</span> <span class='st'>"Trends"</span>, 
  xlim <span class='op'>=</span> <span class='va'>xlim</span>, 
  ylim <span class='op'>=</span> <span class='va'>ylim</span>,
  col <span class='op'>=</span> <span class='st'>"white"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>lapply</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>blocked_stats</span><span class='op'>)</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>b</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='fu'><a href='https://rdrr.io/r/graphics/curve.html'>curve</a></span><span class='op'>(</span>
    <span class='fu'>metadata</span><span class='op'>(</span><span class='va'>var_fit</span><span class='op'>$</span><span class='va'>per.block</span><span class='op'>[[</span><span class='va'>b</span><span class='op'>]</span><span class='op'>]</span><span class='op'>)</span><span class='op'>$</span><span class='fu'>trend</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>, 
    col <span class='op'>=</span> <span class='va'>batch_colours</span><span class='op'>[</span><span class='va'>b</span><span class='op'>]</span>, 
    lwd <span class='op'>=</span> <span class='fl'>1</span>, 
    add <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='op'>}</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/legend.html'>legend</a></span><span class='op'>(</span>
  <span class='st'>"topright"</span>, 
  legend <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>batch</span><span class='op'>)</span>, 
  col <span class='op'>=</span> <span class='va'>batch_colours</span>,
  lwd <span class='op'>=</span> <span class='fl'>2</span>,
  title <span class='op'>=</span> <span class='st'>"Plate number"</span>,
  ncol <span class='op'>=</span> <span class='fl'>2</span>,
  cex <span class='op'>=</span> <span class='fl'>0.8</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:trendplots-fits"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/trendplots-fits-1.png" alt="An overlay of the trend fits from the previous figure, highlighting the need for the batch-specific trend fits. Each line is a the trend line for a particular batch (with colours matching the previous plot)." width="624" />
<p class="caption">
Figure 20: An overlay of the trend fits from the previous figure, highlighting the need for the batch-specific trend fits. Each line is a the trend line for a particular batch (with colours matching the previous plot).
</p>
</div>
</div>
<h2 id="selecting-highly-variable-genes-hvgs">Selecting highly variable genes (HVGs)</h2>
<p>Once we have quantified the per-gene variation, the next step is to select the subset of HVGs to use in downstream analyses. A larger subset will reduce the risk of discarding interesting biological signal by retaining more potentially relevant genes, at the cost of increasing noise from irrelevant genes that might obscure said signal. It is difficult to determine the optimal trade-off for any given application as noise in one context may be useful signal in another. For example, heterogeneity in T cell activation responses is an interesting phenomena but may be irrelevant noise in studies that only care about distinguishing the major immunophenotypes.</p>
<p>We opt to only remove the obviously uninteresting genes with variances below the trend. By doing so, we avoid the need to make any judgement calls regarding what level of variation is interesting enough to retain. This approach represents one extreme of the bias-variance trade-off where bias is minimized at the cost of maximizing noise.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>hvg</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scran/man/getTopHVGs.html'>getTopHVGs</a></span><span class='op'>(</span><span class='va'>var_fit</span>, var.threshold <span class='op'>=</span> <span class='fl'>0</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<h3 id="exclusion-of-gene-sets-from-hvgs">Exclusion of gene sets from HVGs</h3>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>is_mito</span> <span class='op'>&lt;-</span> <span class='va'>hvg</span> <span class='op'>%in%</span> <span class='va'>mito_set</span>
<span class='va'>is_ribo</span> <span class='op'>&lt;-</span> <span class='va'>hvg</span> <span class='op'>%in%</span> <span class='va'>ribo_set</span>
<span class='va'>is_pseudogene</span> <span class='op'>&lt;-</span> <span class='va'>hvg</span> <span class='op'>%in%</span> <span class='va'>pseudogene_set</span>
<span class='va'>is_sex</span> <span class='op'>&lt;-</span> <span class='va'>hvg</span> <span class='op'>%in%</span> <span class='va'>sex_set</span>
</code></pre>
</div>
</details>
</div>
<p>We find that the most highly variable genes in this dataset are somewhat enriched for pseudogenes (n = 4543), with Figure <a href="#fig:hvg-check">21</a> showing around 1/3 of the top-500 HVGs are pseudogenes. Figure <a href="#fig:hvg-check">21</a> also shows the percentage of the top-k HVGs that are ribosomal protein genes<a href="#fn7" class="footnote-ref" id="fnref7" role="doc-noteref"><sup>7</sup></a> (n = 859), genes on the sex chromosomes (n = 873), and mitochondrial genes (n = 27).</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span>
  <span class='fl'>100</span> <span class='op'>*</span> <span class='fu'><a href='https://rdrr.io/r/base/cumsum.html'>cumsum</a></span><span class='op'>(</span><span class='va'>is_mito</span><span class='op'>)</span> <span class='op'>/</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_along</a></span><span class='op'>(</span><span class='va'>hvg</span><span class='op'>)</span>,
  log <span class='op'>=</span> <span class='st'>"x"</span>,
  ylim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>100</span><span class='op'>)</span>,
  ylab <span class='op'>=</span> <span class='st'>"%"</span>,
  xlab <span class='op'>=</span> <span class='st'>"Top-k genes"</span>,
  col <span class='op'>=</span> <span class='st'>"#DF536B"</span>,
  pch <span class='op'>=</span> <span class='fl'>16</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/points.html'>points</a></span><span class='op'>(</span><span class='fl'>100</span> <span class='op'>*</span> <span class='fu'><a href='https://rdrr.io/r/base/cumsum.html'>cumsum</a></span><span class='op'>(</span><span class='va'>is_ribo</span><span class='op'>)</span> <span class='op'>/</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_along</a></span><span class='op'>(</span><span class='va'>hvg</span><span class='op'>)</span>, col <span class='op'>=</span> <span class='st'>"#61D04F"</span>, pch <span class='op'>=</span> <span class='fl'>16</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/points.html'>points</a></span><span class='op'>(</span><span class='fl'>100</span> <span class='op'>*</span> <span class='fu'><a href='https://rdrr.io/r/base/cumsum.html'>cumsum</a></span><span class='op'>(</span><span class='va'>is_pseudogene</span><span class='op'>)</span> <span class='op'>/</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_along</a></span><span class='op'>(</span><span class='va'>hvg</span><span class='op'>)</span>, col <span class='op'>=</span> <span class='st'>"#2297E6"</span>, pch <span class='op'>=</span> <span class='fl'>16</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/points.html'>points</a></span><span class='op'>(</span><span class='fl'>100</span> <span class='op'>*</span> <span class='fu'><a href='https://rdrr.io/r/base/cumsum.html'>cumsum</a></span><span class='op'>(</span><span class='va'>is_sex</span><span class='op'>)</span> <span class='op'>/</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_along</a></span><span class='op'>(</span><span class='va'>hvg</span><span class='op'>)</span>, col <span class='op'>=</span> <span class='st'>"#28E2E5"</span>, pch <span class='op'>=</span> <span class='fl'>16</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/points.html'>points</a></span><span class='op'>(</span>
  <span class='fl'>100</span> <span class='op'>*</span> <span class='fu'><a href='https://rdrr.io/r/base/cumsum.html'>cumsum</a></span><span class='op'>(</span><span class='va'>is_mito</span> <span class='op'>|</span> <span class='va'>is_ribo</span> <span class='op'>|</span> <span class='va'>is_pseudogene</span> <span class='op'>|</span> <span class='va'>is_sex</span><span class='op'>)</span> <span class='op'>/</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_along</a></span><span class='op'>(</span><span class='va'>hvg</span><span class='op'>)</span>, 
  col <span class='op'>=</span> <span class='st'>"black"</span>,
  pch <span class='op'>=</span> <span class='fl'>16</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/legend.html'>legend</a></span><span class='op'>(</span>
  <span class='st'>"topright"</span>,
  pch <span class='op'>=</span> <span class='fl'>16</span>,
  col <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"black"</span>,  <span class='st'>"#2297E6"</span>, <span class='st'>"#61D04F"</span>, <span class='st'>"#28E2E5"</span>, <span class='st'>"#DF536B"</span><span class='op'>)</span>,
  legend <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Union"</span>, <span class='st'>"Pseudogene"</span>, <span class='st'>"Ribosomal protein"</span>, <span class='st'>"Sex"</span>, <span class='st'>"Mitochondrial"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:hvg-check"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/hvg-check-1.png" alt="Percentage of top-K HVGs that are genes of a given class." width="624" />
<p class="caption">
Figure 21: Percentage of top-K HVGs that are genes of a given class.
</p>
</div>
</div>
<p>Generally speaking, pseudogenes, ribosomal protein genes, and mitochondrial genes are of lesser biological relevance, so we tend may exclude them from the HVGs. This means that these genes can no longer <strong>directly</strong> influence some of the subsequent steps in the analysis, including:</p>
<ul>
<li>Dimensionality reduction
<ul>
<li>Principal component analysis (PCA)</li>
<li>Uniform manifold approximation and projection (UMAP)</li>
</ul></li>
<li>Data integration
<ul>
<li>Mutual nearest neighbours (MNN)</li>
</ul></li>
<li>Clustering</li>
</ul>
<p>Although the exclusion of these genes from the HVGs prevents them from directly influencing these analyses, they may still <strong>indirectly</strong> associated with these steps or their outcomes. For example, if there is a set of (non ribosomal protein) genes that are strongly associated with ribosomal protein gene expression, then we may still see a cluster associated with ribosomal protein gene expression.</p>
<p>Finally, and to emphasise, we have only excluded these gene sets from the HVGs, i.e. we have not excluded them entirely from the dataset. In particular, this means that these genes may appear in downstream results (e.g., gene lists resulting from a differential expression analysis between cells in different clusters or with different genotypes).</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>hvg</span> <span class='op'>&lt;-</span> <span class='va'>hvg</span><span class='op'>[</span><span class='op'>!</span><span class='op'>(</span><span class='va'>is_mito</span> <span class='op'>|</span> <span class='va'>is_ribo</span> <span class='op'>|</span> <span class='va'>is_pseudogene</span><span class='op'>)</span><span class='op'>]</span>
</code></pre>
</div>
</details>
</div>
<h2 id="summary-2">Summary</h2>
<p>We are left with 18,248 HVGs by this approach<a href="#fn8" class="footnote-ref" id="fnref8" role="doc-noteref"><sup>8</sup></a>. Figure <a href="#fig:hvgs">22</a> shows the expression of the top-10 HVGs.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>plotExpression</span><span class='op'>(</span>object <span class='op'>=</span> <span class='va'>sce</span>, features <span class='op'>=</span> <span class='va'>hvg</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>10</span><span class='op'>]</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:hvgs"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/hvgs-1.png" alt="Violin plots of normalized log-expression values for the top-10 HVGs. Each point represents the log-expression value in a single cell." width="624" />
<p class="caption">
Figure 22: Violin plots of normalized log-expression values for the top-10 HVGs. Each point represents the log-expression value in a single cell.
</p>
</div>
</div>
<h1 id="dimensionality-reduction">Dimensionality reduction</h1>
<p>Many scRNA-seq analysis procedures involve comparing cells based on their expression values across multiple genes. In these applications, each individual gene represents a dimension of the data. If we had a scRNA-seq data set with two genes, we could make a two-dimensional plot where each axis represents the expression of one gene and each point in the plot represents a cell. Dimensionality reduction extends this idea to data sets with thousands of genes where each cell’s expression profile defines its location in the high-dimensional expression space.</p>
<p>As the name suggests, dimensionality reduction aims to reduce the number of separate dimensions in the data. This is possible because different genes are correlated if they are affected by the same biological process. Thus, we do not need to store separate information for individual genes, but can instead compress multiple features into a single dimension. This reduces computational work in downstream analyses, as calculations only need to be performed for a few dimensions rather than thousands of genes; reduces noise by averaging across multiple genes to obtain a more precise representation of the patterns in the data; and enables effective plotting of the data, for those of us who are not capable of visualizing more than 3 dimensions.</p>
<h2 id="principal-component-analysis">Principal component analysis</h2>
<p>Principal components analysis (PCA) discovers axes in high-dimensional space that capture the largest amount of variation. In PCA, the first axis (or “principal component,” PC) is chosen such that it captures the greatest variance across cells. The next PC is chosen such that it is orthogonal to the first and captures the greatest remaining amount of variation, and so on.</p>
<p>By definition, the top PCs capture the dominant factors of heterogeneity in the data set. Thus, we can perform dimensionality reduction by restricting downstream analyses to the top PCs. This strategy is simple, highly effective and widely used throughout the data sciences.</p>
<p>When applying PCA to scRNA-seq data, our assumption is that biological processes affect multiple genes in a coordinated manner. This means that the earlier PCs are likely to represent biological structure as more variation can be captured by considering the correlated behaviour of many genes. By comparison, random technical or biological noise is expected to affect each gene independently. There is unlikely to be an axis that can capture random variation across many genes, meaning that noise should mostly be concentrated in the later PCs. This motivates the use of the earlier PCs in our downstream analyses, which concentrates the biological signal to simultaneously reduce computational work and remove noise.</p>
<p>We perform the PCA on the log-normalized expression values. PCA is generally robust to random noise but an excess of it may cause the earlier PCs to capture noise instead of biological structure. This effect can be avoided - or at least mitigated - by restricting the PCA to a subset of HVGs, as done in <a href="#feature-selection">Feature selection</a>.</p>
<p>The choice of the number of PCs is a decision that is analogous to the choice of the number of HVGs to use. Using more PCs will avoid discarding biological signal in later PCs, at the cost of retaining more noise.</p>
<p>We use the strategy of retaining all PCs until the percentage of total variation explained reaches some threshold. We derive a suitable value for this threshold by calculating the proportion of variance in the data that is attributed to the biological component. This is done using the the variance modelling results from <a href="#quantifying-per-gene-variation">Quantifying per-gene variation</a>.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>557</span><span class='op'>)</span>
<span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scran/man/denoisePCA.html'>denoisePCA</a></span><span class='op'>(</span><span class='va'>sce</span>, <span class='va'>var_fit</span>, subset.row <span class='op'>=</span> <span class='va'>hvg</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>This retains 50 dimensions, which represents the lower bound on the number of PCs required to retain all biological variation. Any fewer PCs will definitely discard some aspect of biological signal<a href="#fn9" class="footnote-ref" id="fnref9" role="doc-noteref"><sup>9</sup></a>. This approach provides a reasonable choice when we want to retain as much signal as possible while still removing some noise.</p>
<h2 id="dimensionality-reduction-for-visualization">Dimensionality reduction for visualization</h2>
<p>We use the uniform manifold approximation and projection (UMAP) method <span class="citation" data-cites="McInnes2018-hy">(<a href="#ref-McInnes2018-hy" role="doc-biblioref">McInnes, Healy, and Melville 2018</a>)</span> to perform further dimensionality reduction for visualization.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/jlmelville/uwot'>uwot</a></span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>785146</span><span class='op'>)</span>
<span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='fu'>runUMAP</span><span class='op'>(</span><span class='va'>sce</span>, dimred <span class='op'>=</span> <span class='st'>"PCA"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>Figure <a href="#fig:umap">23</a> visualizes the cells using the UMAP co-ordinates, colouring the cells by key experimental factors, and Figures <a href="#fig:umap-tissue-post-hoc-sample-gate">28</a> use facetting to further tease out associations between UMAP co-ordinates and experimental factors.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='fu'>plotUMAP</span><span class='op'>(</span><span class='va'>sce</span>, colour_by <span class='op'>=</span> <span class='st'>"donor"</span>, point_size <span class='op'>=</span> <span class='fl'>0.5</span>, point_alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>donor_colours</span>, name <span class='op'>=</span> <span class='st'>"donor"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>10</span><span class='op'>)</span>
<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='fu'>plotUMAP</span><span class='op'>(</span>
  <span class='va'>sce</span>, 
  colour_by <span class='op'>=</span> <span class='st'>"plate_number"</span>, 
  point_size <span class='op'>=</span> <span class='fl'>0.5</span>, 
  point_alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>plate_number_colours</span>, name <span class='op'>=</span> <span class='st'>"plate_number"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>10</span><span class='op'>)</span>
<span class='va'>p3</span> <span class='op'>&lt;-</span> <span class='fu'>plotUMAP</span><span class='op'>(</span>
  <span class='va'>sce</span>, 
  colour_by <span class='op'>=</span> <span class='st'>"stage"</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span>, 
  point_alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>
    values <span class='op'>=</span> <span class='va'>stage_colours</span>,
    name <span class='op'>=</span> <span class='st'>"stage"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>10</span><span class='op'>)</span>
<span class='va'>p4</span> <span class='op'>&lt;-</span> <span class='fu'>plotUMAP</span><span class='op'>(</span><span class='va'>sce</span>, colour_by <span class='op'>=</span> <span class='st'>"tissue"</span>, point_size <span class='op'>=</span> <span class='fl'>0.5</span>, point_alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>tissue_colours</span>, name <span class='op'>=</span> <span class='st'>"tissue"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>10</span><span class='op'>)</span>
<span class='va'>p5</span> <span class='op'>&lt;-</span> <span class='fu'>plotUMAP</span><span class='op'>(</span>
  <span class='va'>sce</span>,
  colour_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>`log10(sum)` <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log10</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>sum</span><span class='op'>)</span>, check.names <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span>, 
  point_alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>10</span><span class='op'>)</span>

<span class='va'>p1</span> <span class='op'>+</span> <span class='va'>p2</span> <span class='op'>+</span> <span class='va'>p3</span> <span class='op'>+</span> <span class='va'>p4</span> <span class='op'>+</span> <span class='va'>p5</span> <span class='op'>+</span> <span class='fu'><a href='https://patchwork.data-imaginist.com/reference/plot_layout.html'>plot_layout</a></span><span class='op'>(</span>ncol <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:umap"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/umap-1.png" alt="UMAP plot of the dataset. Each point represents a cell and is coloured according to the legend." width="624" />
<p class="caption">
Figure 23: UMAP plot of the dataset. Each point represents a cell and is coloured according to the legend.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>umap_df</span> <span class='op'>&lt;-</span> <span class='fu'>makePerCellDF</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>
<span class='va'>bg</span> <span class='op'>&lt;-</span> <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/select.html'>select</a></span><span class='op'>(</span><span class='va'>umap_df</span>, <span class='op'>-</span><span class='va'>donor</span><span class='op'>)</span>
<span class='fu'><a href='https://ggplot2.tidyverse.org/reference/ggplot.html'>ggplot</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>UMAP.1</span>, y <span class='op'>=</span> <span class='va'>UMAP.2</span><span class='op'>)</span>, data <span class='op'>=</span> <span class='va'>umap_df</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span>data <span class='op'>=</span> <span class='va'>bg</span>, colour <span class='op'>=</span> <span class='fu'>scales</span><span class='fu'>::</span><span class='fu'><a href='https://scales.r-lib.org/reference/alpha.html'>alpha</a></span><span class='op'>(</span><span class='st'>"grey"</span>, <span class='fl'>0.5</span><span class='op'>)</span>, size <span class='op'>=</span> <span class='fl'>0.125</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='va'>donor</span><span class='op'>)</span>, alpha <span class='op'>=</span> <span class='fl'>1</span>, size <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>donor_colours</span>, name <span class='op'>=</span> <span class='st'>"donor"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>donor_colours</span>, name <span class='op'>=</span> <span class='st'>"donor"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>10</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>"UMAP 1"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"UMAP 2"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_wrap.html'>facet_wrap</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>donor</span>, ncol <span class='op'>=</span> <span class='fl'>2</span>, labeller <span class='op'>=</span> <span class='va'>label_both</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guide_legend.html'>guide_legend</a></span><span class='op'>(</span>override.aes <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>2</span>, alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:umap-donor"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/umap-donor-1.png" alt="UMAP plot of the dataset. Each point represents a cell and each panel highlights cells from a particular `donor`." width="624" />
<p class="caption">
Figure 24: UMAP plot of the dataset. Each point represents a cell and each panel highlights cells from a particular <code>donor</code>.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>umap_df</span> <span class='op'>&lt;-</span> <span class='fu'>makePerCellDF</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>
<span class='va'>bg</span> <span class='op'>&lt;-</span> <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/select.html'>select</a></span><span class='op'>(</span><span class='va'>umap_df</span>, <span class='op'>-</span><span class='va'>plate_number</span><span class='op'>)</span>
<span class='fu'><a href='https://ggplot2.tidyverse.org/reference/ggplot.html'>ggplot</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>UMAP.1</span>, y <span class='op'>=</span> <span class='va'>UMAP.2</span><span class='op'>)</span>, data <span class='op'>=</span> <span class='va'>umap_df</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span>data <span class='op'>=</span> <span class='va'>bg</span>, colour <span class='op'>=</span> <span class='fu'>scales</span><span class='fu'>::</span><span class='fu'><a href='https://scales.r-lib.org/reference/alpha.html'>alpha</a></span><span class='op'>(</span><span class='st'>"grey"</span>, <span class='fl'>0.5</span><span class='op'>)</span>, size <span class='op'>=</span> <span class='fl'>0.125</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='va'>plate_number</span><span class='op'>)</span>, alpha <span class='op'>=</span> <span class='fl'>1</span>, size <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>plate_number_colours</span>, name <span class='op'>=</span> <span class='st'>"plate_number"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>plate_number_colours</span>, name <span class='op'>=</span> <span class='st'>"plate_number"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>10</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>"UMAP 1"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"UMAP 2"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_wrap.html'>facet_wrap</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>plate_number</span>, ncol <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guide_legend.html'>guide_legend</a></span><span class='op'>(</span>override.aes <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>2</span>, alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:umap-plate-number"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/umap-plate-number-1.png" alt="UMAP plot of the dataset. Each point represents a cell and each panel highlights cells from a particular `plate_number`." width="624" />
<p class="caption">
Figure 25: UMAP plot of the dataset. Each point represents a cell and each panel highlights cells from a particular <code>plate_number</code>.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>umap_df</span> <span class='op'>&lt;-</span> <span class='fu'>makePerCellDF</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>
<span class='va'>bg</span> <span class='op'>&lt;-</span> <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/select.html'>select</a></span><span class='op'>(</span><span class='va'>umap_df</span>, <span class='op'>-</span><span class='va'>stage</span><span class='op'>)</span>
<span class='fu'><a href='https://ggplot2.tidyverse.org/reference/ggplot.html'>ggplot</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>UMAP.1</span>, y <span class='op'>=</span> <span class='va'>UMAP.2</span><span class='op'>)</span>, data <span class='op'>=</span> <span class='va'>umap_df</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span>data <span class='op'>=</span> <span class='va'>bg</span>, colour <span class='op'>=</span> <span class='fu'>scales</span><span class='fu'>::</span><span class='fu'><a href='https://scales.r-lib.org/reference/alpha.html'>alpha</a></span><span class='op'>(</span><span class='st'>"grey"</span>, <span class='fl'>0.5</span><span class='op'>)</span>, size <span class='op'>=</span> <span class='fl'>0.125</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='va'>stage</span><span class='op'>)</span>, alpha <span class='op'>=</span> <span class='fl'>1</span>, size <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>
    values <span class='op'>=</span> <span class='va'>stage_colours</span>,
    name <span class='op'>=</span> <span class='st'>"stage"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>
    values <span class='op'>=</span> <span class='va'>stage_colours</span>,
    name <span class='op'>=</span> <span class='st'>"stage"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>10</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>"UMAP 1"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"UMAP 2"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_wrap.html'>facet_wrap</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>stage</span>, ncol <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guide_legend.html'>guide_legend</a></span><span class='op'>(</span>override.aes <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>2</span>, alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:umap-post-hoc-sample-gate"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/umap-post-hoc-sample-gate-1.png" alt="UMAP plot of the dataset. Each point represents a cell and each panel highlights cells from a particular `stage`." width="624" />
<p class="caption">
Figure 26: UMAP plot of the dataset. Each point represents a cell and each panel highlights cells from a particular <code>stage</code>.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>umap_df</span> <span class='op'>&lt;-</span> <span class='fu'>makePerCellDF</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>
<span class='va'>bg</span> <span class='op'>&lt;-</span> <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/select.html'>select</a></span><span class='op'>(</span><span class='va'>umap_df</span>, <span class='op'>-</span><span class='va'>tissue</span><span class='op'>)</span>
<span class='fu'><a href='https://ggplot2.tidyverse.org/reference/ggplot.html'>ggplot</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>UMAP.1</span>, y <span class='op'>=</span> <span class='va'>UMAP.2</span><span class='op'>)</span>, data <span class='op'>=</span> <span class='va'>umap_df</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span>data <span class='op'>=</span> <span class='va'>bg</span>, colour <span class='op'>=</span> <span class='fu'>scales</span><span class='fu'>::</span><span class='fu'><a href='https://scales.r-lib.org/reference/alpha.html'>alpha</a></span><span class='op'>(</span><span class='st'>"grey"</span>, <span class='fl'>0.5</span><span class='op'>)</span>, size <span class='op'>=</span> <span class='fl'>0.125</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='va'>tissue</span><span class='op'>)</span>, alpha <span class='op'>=</span> <span class='fl'>1</span>, size <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>tissue_colours</span>, name <span class='op'>=</span> <span class='st'>"tissue"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>tissue_colours</span>, name <span class='op'>=</span> <span class='st'>"tissue"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>10</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>"UMAP 1"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"UMAP 2"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_wrap.html'>facet_wrap</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>tissue</span>, ncol <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guide_legend.html'>guide_legend</a></span><span class='op'>(</span>override.aes <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>2</span>, alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:umap-tissue"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/umap-tissue-1.png" alt="UMAP plot of the dataset. Each point represents a cell and each panel highlights cells from a particular `tissue`." width="624" />
<p class="caption">
Figure 27: UMAP plot of the dataset. Each point represents a cell and each panel highlights cells from a particular <code>tissue</code>.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>umap_df</span> <span class='op'>&lt;-</span> <span class='fu'>makePerCellDF</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>
<span class='va'>bg</span> <span class='op'>&lt;-</span> <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/select.html'>select</a></span><span class='op'>(</span><span class='va'>umap_df</span>, <span class='op'>-</span><span class='va'>tissue</span>, <span class='op'>-</span><span class='va'>stage</span><span class='op'>)</span>
<span class='fu'><a href='https://ggplot2.tidyverse.org/reference/ggplot.html'>ggplot</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>UMAP.1</span>, y <span class='op'>=</span> <span class='va'>UMAP.2</span><span class='op'>)</span>, data <span class='op'>=</span> <span class='va'>umap_df</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span>data <span class='op'>=</span> <span class='va'>bg</span>, colour <span class='op'>=</span> <span class='fu'>scales</span><span class='fu'>::</span><span class='fu'><a href='https://scales.r-lib.org/reference/alpha.html'>alpha</a></span><span class='op'>(</span><span class='st'>"grey"</span>, <span class='fl'>0.5</span><span class='op'>)</span>, size <span class='op'>=</span> <span class='fl'>0.125</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span><span class='op'>(</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='va'>plate_number</span><span class='op'>)</span>, alpha <span class='op'>=</span> <span class='fl'>1</span>, size <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_fill_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>plate_number_colours</span>, name <span class='op'>=</span> <span class='st'>"plate_number"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>scale_colour_manual</a></span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>plate_number_colours</span>, name <span class='op'>=</span> <span class='st'>"plate_number"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span>font_size <span class='op'>=</span> <span class='fl'>10</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>xlab</a></span><span class='op'>(</span><span class='st'>"UMAP 1"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ylab</a></span><span class='op'>(</span><span class='st'>"UMAP 2"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_grid.html'>facet_grid</a></span><span class='op'>(</span><span class='va'>stage</span> <span class='op'>~</span> <span class='va'>tissue</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guides.html'>guides</a></span><span class='op'>(</span>colour <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/guide_legend.html'>guide_legend</a></span><span class='op'>(</span>override.aes <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>2</span>, alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:umap-tissue-post-hoc-sample-gate"></span>
<img src="C094_Pellicci.single-cell.preprocess_files/umap-tissue-post-hoc-sample-gate-1.png" alt="UMAP plot of the dataset. Each point represents a cell and each panel highlights cells from a particular `tissue` and `stage` and further coloured by `plate_number`." width="624" />
<p class="caption">
Figure 28: UMAP plot of the dataset. Each point represents a cell and each panel highlights cells from a particular <code>tissue</code> and <code>stage</code> and further coloured by <code>plate_number</code>.
</p>
</div>
</div>
<h2 id="summary-3">Summary</h2>
<p>Although Figures <a href="#fig:umap">23</a> - <a href="#fig:umap-tissue-post-hoc-sample-gate">28</a> are only preliminary summaries of the data, there are few points worth highlighting:</p>
<ul>
<li>There are no strong <code>donor</code>-associated differences.</li>
<li>There are strong <code>plate_number</code>-associated differences.</li>
<li>There are strong <code>stage</code>-associated differences.</li>
<li>There are strong <code>tissue</code>-associated differences.</li>
<li>There are library size-associated differences, which is itself largely driven by differences in library size between the <code>Cell line</code> and non-<code>Cell line</code> samples.</li>
</ul>
<p>However, it is premature to draw any strong conclusions We will seek to mitigate the differences associated with technical factors (e.g., <code>plate_number</code> and library size) in downstream analyses so that we might better investigate the biological factors (i.e. <code>tissue</code> and <code>stage</code>).</p>
<h1 id="concluding-remarks">Concluding remarks</h1>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span>
  <span class='va'>sce</span>,
  <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"data"</span>, <span class='st'>"SCEs"</span>, <span class='st'>"C094_Pellicci.single-cell.preprocessed.SCE.rds"</span><span class='op'>)</span>,
  compress <span class='op'>=</span> <span class='st'>"xz"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>The processed <em>SingleCellExperiment</em> object is available (see <a href="../data/SCEs/C094_Pellicci.preprocessed.SCE.rds"><code>data/SCEs/C094_Pellicci.preprocessed.SCE.rds</code></a>). This will be used in downstream analyses, e.g., selecting biologically relevant cells.</p>
<h1 class="appendix" id="additional-information">Additional information</h1>
<p>The following are available on request:</p>
<ul>
<li>Full CSV tables of any data presented.</li>
<li>PDF/PNG files of any static plots.</li>
</ul>
<h1 class="appendix" id="session-info">Session info</h1>
<summary>
The analysis and this document were prepared using the following software (click triangle to expand)
</summary>
<details>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>sessioninfo</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/sessioninfo/man/session_info.html'>session_info</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<pre><code>─ Session info ─────────────────────────────────────────────────────
 setting  value                       
 version  R version 4.0.3 (2020-10-10)
 os       CentOS Linux 7 (Core)       
 system   x86_64, linux-gnu           
 ui       X11                         
 language (EN)                        
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 
 tz       Australia/Melbourne         
 date     2021-08-05                  

─ Packages ─────────────────────────────────────────────────────────
 ! package                           * version  date       lib
 P annotate                            1.68.0   2020-10-27 [?]
 P AnnotationDbi                     * 1.52.0   2020-10-27 [?]
 P AnnotationFilter                  * 1.14.0   2020-10-27 [?]
 P AnnotationHub                     * 2.22.0   2020-10-27 [?]
 P askpass                             1.1      2019-01-13 [?]
 P assertthat                          0.2.1    2019-03-21 [?]
 P base64enc                           0.1-3    2015-07-28 [?]
 P beachmat                            2.6.4    2020-12-20 [?]
 P beeswarm                            0.3.1    2021-03-07 [?]
 P Biobase                           * 2.50.0   2020-10-27 [?]
 P BiocFileCache                     * 1.14.0   2020-10-27 [?]
 P BiocGenerics                      * 0.36.0   2020-10-27 [?]
 P BiocManager                         1.30.12  2021-03-28 [?]
 P BiocNeighbors                       1.8.2    2020-12-07 [?]
 P BiocParallel                        1.24.1   2020-11-06 [?]
 P BiocSingular                        1.6.0    2020-10-27 [?]
 P BiocStyle                         * 2.18.1   2020-11-24 [?]
 P BiocVersion                         3.12.0   2020-04-27 [?]
 P biomaRt                             2.46.3   2021-02-09 [?]
 P Biostrings                          2.58.0   2020-10-27 [?]
 P bit                                 4.0.4    2020-08-04 [?]
 P bit64                               4.0.5    2020-08-30 [?]
 P bitops                              1.0-6    2013-08-17 [?]
 P blob                                1.2.1    2020-01-20 [?]
 P bluster                             1.0.0    2020-10-27 [?]
 P bslib                               0.2.4    2021-01-25 [?]
 P cachem                              1.0.4    2021-02-13 [?]
 P cellranger                          1.1.0    2016-07-27 [?]
 P cli                                 2.4.0    2021-04-05 [?]
 P colorspace                          2.0-0    2020-11-11 [?]
 P cowplot                           * 1.1.1    2020-12-30 [?]
 P crayon                              1.4.1    2021-02-08 [?]
 P curl                                4.3      2019-12-02 [?]
 P DBI                                 1.1.1    2021-01-15 [?]
 P dbplyr                            * 2.1.0    2021-02-03 [?]
 P DelayedArray                        0.16.3   2021-03-24 [?]
 P DelayedMatrixStats                  1.12.3   2021-02-03 [?]
 P DESeq2                              1.30.1   2021-02-19 [?]
 P digest                              0.6.27   2020-10-24 [?]
 P distill                             1.2      2021-01-13 [?]
 P downlit                             0.2.1    2020-11-04 [?]
 P dplyr                             * 1.0.5    2021-03-05 [?]
 P dqrng                               0.2.1    2019-05-17 [?]
 P edgeR                             * 3.32.1   2021-01-14 [?]
 P ellipsis                            0.3.1    2020-05-15 [?]
 P ensembldb                         * 2.14.0   2020-10-27 [?]
 P evaluate                            0.14     2019-05-28 [?]
 P fansi                               0.4.2    2021-01-15 [?]
 P farver                              2.1.0    2021-02-28 [?]
 P fastmap                             1.1.0    2021-01-25 [?]
 P FNN                                 1.1.3    2019-02-15 [?]
 P genefilter                          1.72.1   2021-01-21 [?]
 P geneplotter                         1.68.0   2020-10-27 [?]
 P generics                            0.1.0    2020-10-31 [?]
 P GenomeInfoDb                      * 1.26.4   2021-03-10 [?]
 P GenomeInfoDbData                    1.2.4    2020-12-07 [?]
 P GenomicAlignments                   1.26.0   2020-10-27 [?]
 P GenomicFeatures                   * 1.42.3   2021-04-01 [?]
 P GenomicRanges                     * 1.42.0   2020-10-27 [?]
 P ggbeeswarm                          0.6.0    2017-08-07 [?]
 P ggplot2                           * 3.3.3    2020-12-30 [?]
 P Glimma                            * 2.0.0    2020-10-27 [?]
 P glue                                1.4.2    2020-08-27 [?]
 P GO.db                             * 3.12.1   2021-01-29 [?]
 P graph                               1.68.0   2020-10-27 [?]
 P gridExtra                           2.3      2017-09-09 [?]
 P gtable                              0.3.0    2019-03-25 [?]
 P here                              * 1.0.1    2020-12-13 [?]
   highr                               0.9      2021-04-16 [1]
 P hms                                 1.0.0    2021-01-13 [?]
 P Homo.sapiens                      * 1.3.1    2021-02-18 [?]
 P htmltools                           0.5.1.1  2021-01-22 [?]
 P htmlwidgets                         1.5.3    2020-12-10 [?]
 P httpuv                              1.5.5    2021-01-13 [?]
 P httr                                1.4.2    2020-07-20 [?]
 P igraph                              1.2.6    2020-10-06 [?]
 P interactiveDisplayBase              1.28.0   2020-10-27 [?]
 P IRanges                           * 2.24.1   2020-12-12 [?]
 P irlba                               2.3.3    2019-02-05 [?]
 P janitor                           * 2.1.0    2021-01-05 [?]
 P jquerylib                           0.1.3    2020-12-17 [?]
 P jsonlite                            1.7.2    2020-12-09 [?]
   knitr                               1.33     2021-04-24 [1]
 P labeling                            0.4.2    2020-10-20 [?]
 P later                               1.1.0.1  2020-06-05 [?]
   lattice                             0.20-41  2020-04-02 [2]
 P lazyeval                            0.2.2    2019-03-15 [?]
 P lifecycle                           1.0.0    2021-02-15 [?]
 P limma                             * 3.46.0   2020-10-27 [?]
 P locfit                              1.5-9.4  2020-03-25 [?]
 P lubridate                           1.7.10   2021-02-26 [?]
 P magrittr                            2.0.1    2020-11-17 [?]
   Matrix                            * 1.2-18   2019-11-27 [2]
 P MatrixGenerics                    * 1.2.1    2021-01-30 [?]
 P matrixStats                       * 0.58.0   2021-01-29 [?]
 P memoise                             2.0.0    2021-01-26 [?]
   mime                                0.11     2021-06-23 [1]
 P msigdbr                           * 7.2.1    2020-10-02 [?]
 P munsell                             0.5.0    2018-06-12 [?]
 P openssl                             1.4.3    2020-09-18 [?]
 P org.Hs.eg.db                      * 3.12.0   2020-12-08 [?]
 P OrganismDbi                       * 1.32.0   2020-10-27 [?]
 P patchwork                         * 1.1.1    2020-12-17 [?]
 P pheatmap                            1.0.12   2019-01-04 [?]
 P pillar                              1.5.1    2021-03-05 [?]
 P pkgconfig                           2.0.3    2019-09-22 [?]
 P prettyunits                         1.1.1    2020-01-24 [?]
 P progress                            1.2.2    2019-05-16 [?]
 P promises                            1.2.0.1  2021-02-11 [?]
 P ProtGenerics                        1.22.0   2020-10-27 [?]
 P purrr                               0.3.4    2020-04-17 [?]
 P R6                                  2.5.0    2020-10-28 [?]
 P rappdirs                            0.3.3    2021-01-31 [?]
 P RBGL                                1.66.0   2020-10-27 [?]
 P RColorBrewer                        1.1-2    2014-12-07 [?]
 P Rcpp                                1.0.6    2021-01-15 [?]
 P RCurl                               1.98-1.3 2021-03-16 [?]
 P readxl                            * 1.3.1    2019-03-13 [?]
 P repr                                1.1.3    2021-01-21 [?]
   rlang                               0.4.11   2021-04-30 [1]
 P rmarkdown                           2.7      2021-02-19 [?]
 P rprojroot                           2.0.2    2020-11-15 [?]
 P Rsamtools                           2.6.0    2020-10-27 [?]
 P RSpectra                            0.16-0   2019-12-01 [?]
 P RSQLite                             2.2.5    2021-03-27 [?]
 P rstudioapi                          0.13     2020-11-12 [?]
 P rsvd                                1.0.3    2020-02-17 [?]
 P rtracklayer                         1.50.0   2020-10-27 [?]
 P S4Vectors                         * 0.28.1   2020-12-09 [?]
 P sass                                0.3.1    2021-01-24 [?]
 P scales                            * 1.1.1    2020-05-11 [?]
 P scater                            * 1.18.6   2021-02-26 [?]
 P scran                             * 1.18.5   2021-02-04 [?]
 P scuttle                             1.0.4    2020-12-17 [?]
 P sessioninfo                         1.1.1    2018-11-05 [?]
 P shiny                               1.6.0    2021-01-25 [?]
 P SingleCellExperiment              * 1.12.0   2020-10-27 [?]
 P skimr                               2.1.3    2021-03-07 [?]
 P snakecase                           0.11.0   2019-05-25 [?]
 P sparseMatrixStats                   1.2.1    2021-02-02 [?]
 P statmod                             1.4.35   2020-10-19 [?]
   stringi                             1.7.3    2021-07-16 [1]
 P stringr                             1.4.0    2019-02-10 [?]
 P SummarizedExperiment              * 1.20.0   2020-10-27 [?]
   survival                            3.2-7    2020-09-28 [2]
 P tibble                              3.1.0    2021-02-25 [?]
 P tidyr                               1.1.3    2021-03-03 [?]
 P tidyselect                          1.1.0    2020-05-11 [?]
 P TxDb.Hsapiens.UCSC.hg19.knownGene * 3.2.2    2021-02-18 [?]
 P utf8                                1.2.1    2021-03-12 [?]
 P uwot                              * 0.1.10   2020-12-15 [?]
 P vctrs                               0.3.7    2021-03-29 [?]
 P vipor                               0.4.5    2017-03-22 [?]
 P viridis                             0.5.1    2018-03-29 [?]
 P viridisLite                         0.3.0    2018-02-01 [?]
 P withr                               2.4.1    2021-01-26 [?]
   xfun                                0.24     2021-06-15 [1]
 P XML                                 3.99-0.6 2021-03-16 [?]
 P xml2                                1.3.2    2020-04-23 [?]
 P xtable                              1.8-4    2019-04-21 [?]
 P XVector                             0.30.0   2020-10-27 [?]
 P yaml                                2.2.1    2020-02-01 [?]
 P zlibbioc                            1.36.0   2020-10-27 [?]
 source        
 Bioconductor  
 Bioconductor  
 Bioconductor  
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 Bioconductor  
 CRAN (R 4.0.2)
 Bioconductor  
 Bioconductor  
 Bioconductor  
 CRAN (R 4.0.3)
 Bioconductor  
 Bioconductor  
 Bioconductor  
 Bioconductor  
 Bioconductor  
 Bioconductor  
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 Bioconductor  
 CRAN (R 4.0.2)
 CRAN (R 4.0.2)
 CRAN (R 4.0.0)
 CRAN (R 4.0.3)
 CRAN (R 4.0.2)
 CRAN (R 4.0.0)
 CRAN (R 4.0.3)
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 CRAN (R 4.0.2)
 Bioconductor  
 Bioconductor  
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 CRAN (R 4.0.0)
 Bioconductor  
 CRAN (R 4.0.0)
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 CRAN (R 4.0.2)
 CRAN (R 4.0.2)
 CRAN (R 4.0.0)
 Bioconductor  
 Bioconductor  
 CRAN (R 4.0.0)
 Bioconductor  
 Bioconductor  
 Bioconductor  
 Bioconductor  
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 Bioconductor  
 CRAN (R 4.0.0)
 Bioconductor  
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.3)
 CRAN (R 4.0.2)
 Bioconductor  
 CRAN (R 4.0.2)
 CRAN (R 4.0.2)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 Bioconductor  
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.3)
 CRAN (R 4.0.2)
 CRAN (R 4.0.2)
 CRAN (R 4.0.3)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.3)
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.3)
 CRAN (R 4.0.2)
 CRAN (R 4.0.3)
 Bioconductor  
 CRAN (R 4.0.2)
 CRAN (R 4.0.2)
 CRAN (R 4.0.3)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 Bioconductor  
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 CRAN (R 4.0.5)
 CRAN (R 4.0.0)
 CRAN (R 4.0.3)
 CRAN (R 4.0.3)
 CRAN (R 4.0.2)
 CRAN (R 4.0.2)
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.3)
 CRAN (R 4.0.2)
 CRAN (R 4.0.0)
 Bioconductor  
 Bioconductor  
 CRAN (R 4.0.2)
 CRAN (R 4.0.0)
 Bioconductor  
 Bioconductor  
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 Bioconductor  
 CRAN (R 4.0.3)
 CRAN (R 4.0.0)
 Bioconductor  
 CRAN (R 4.0.0)
 CRAN (R 4.0.3)
 CRAN (R 4.0.0)
 Bioconductor  
 CRAN (R 4.0.3)
 CRAN (R 4.0.2)
 CRAN (R 4.0.3)
 CRAN (R 4.0.0)
 Bioconductor  
 CRAN (R 4.0.2)
 CRAN (R 4.0.0)
 CRAN (R 4.0.3)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 CRAN (R 4.0.2)
 CRAN (R 4.0.3)
 CRAN (R 4.0.5)
 CRAN (R 4.0.0)
 CRAN (R 4.0.0)
 Bioconductor  
 CRAN (R 4.0.0)
 Bioconductor  

[1] /stornext/Home/data/allstaff/h/ho.w/C094_Pellicci/renv/library/R-4.0/x86_64-pc-linux-gnu
[2] /stornext/System/data/apps/R/R-4.0.3/lib64/R/library

 P ── Loaded and on-disk path mismatch.</code></pre>
</div>
</details>
<div class="sourceCode" id="cb2"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography">
<div id="ref-anders2010differential" class="csl-entry" role="doc-biblioentry">
Anders, S., and W. Huber. 2010. <span>“Differential Expression Analysis for Sequence Count Data.”</span> <em>Genome Biol.</em> 11 (10): R106. <a href="https://www.ncbi.nlm.nih.gov/pubmed/20979621">https://www.ncbi.nlm.nih.gov/pubmed/20979621</a>.
</div>
<div id="ref-bourgon2010independent" class="csl-entry" role="doc-biblioentry">
Bourgon, R., R. Gentleman, and W. Huber. 2010. <span>“Independent Filtering Increases Detection Power for High-Throughput Experiments.”</span> <em>Proc. Natl. Acad. Sci. U.S.A.</em> 107 (21): 9546–51. <a href="https://www.ncbi.nlm.nih.gov/pubmed/20460310">https://www.ncbi.nlm.nih.gov/pubmed/20460310</a>.
</div>
<div id="ref-love2014moderated" class="csl-entry" role="doc-biblioentry">
Love, M. I., W. Huber, and S. Anders. 2014. <span>“Moderated Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2.”</span> <em>Genome Biol.</em> 15 (12): 550. <a href="https://www.ncbi.nlm.nih.gov/pubmed/25516281">https://www.ncbi.nlm.nih.gov/pubmed/25516281</a>.
</div>
<div id="ref-lun2016pooling" class="csl-entry" role="doc-biblioentry">
Lun, A. T., K. Bach, and J. C. Marioni. 2016. <em>Genome Biol.</em> 17 (April): 75. <a href="https://www.ncbi.nlm.nih.gov/pubmed/27122128">https://www.ncbi.nlm.nih.gov/pubmed/27122128</a>.
</div>
<div id="ref-lun2017assessing" class="csl-entry" role="doc-biblioentry">
Lun, A. T. L., F. J. Calero-Nieto, L. Haim-Vilmovsky, B. Gottgens, and J. C. Marioni. 2017. <span>“Assessing the Reliability of Spike-in Normalization for Analyses of Single-Cell RNA Sequencing Data.”</span> <em>Genome Res.</em> 27 (11): 1795–1806.
</div>
<div id="ref-lun2016step" class="csl-entry" role="doc-biblioentry">
Lun, A. T. L., D. J. McCarthy, and J. C. Marioni. 2016. <span>“A Step-by-Step Workflow for Low-Level Analysis of Single-Cell <span class="nocase">RNA-seq</span> Data.”</span> <em>F1000Res.</em> 5 (August).
</div>
<div id="ref-McInnes2018-hy" class="csl-entry" role="doc-biblioentry">
McInnes, Leland, John Healy, and James Melville. 2018. <span>“UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction,”</span> February. <a href="http://arxiv.org/abs/1802.03426">http://arxiv.org/abs/1802.03426</a>.
</div>
<div id="ref-robinson2010scaling" class="csl-entry" role="doc-biblioentry">
Robinson, M. D., and A. Oshlack. 2010. <span>“A Scaling Normalization Method for Differential Expression Analysis of RNA-Seq Data.”</span> <em>Genome Biol.</em> 11 (3): R25. <a href="https://www.ncbi.nlm.nih.gov/pubmed/20196867">https://www.ncbi.nlm.nih.gov/pubmed/20196867</a>.
</div>
<div id="ref-stegle2015computational" class="csl-entry" role="doc-biblioentry">
Stegle, O., S. A. Teichmann, and J. C. Marioni. 2015. <span>“Computational and Analytical Challenges in Single-Cell Transcriptomics.”</span> <em>Nat. Rev. Genet.</em> 16 (3): 133–45. <a href="https://www.ncbi.nlm.nih.gov/pubmed/25628217">https://www.ncbi.nlm.nih.gov/pubmed/25628217</a>.
</div>
</div>
<section class="footnotes" role="doc-endnotes">
<hr />
<ol>
<li id="fn1" role="doc-endnote"><p>It should be remembered that the gates are set using more cells than are collected for scRNA-seq and so the boundaries may appear more defined in the original FACS files. Furthermore, the data here are shown on a ‘pseudolog’ scale, which may be different to the scale used when viewing the data in FlowJo.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2" role="doc-endnote"><p>Some care is taken to account for missing and duplicate gene symbols; missing symbols are replaced with the Ensembl identifier and duplicated symbols are concatenated with the (unique) Ensembl identifiers.<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3" role="doc-endnote"><p>These marker genes are identified by a Wilcox test comparing the <code>Blood</code> and <code>Thymus</code> single-cells on all plates except <code>LCE511</code><a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4" role="doc-endnote"><p>The number of expressed genes refers to the number of genes which have non-zero counts (ie. they have been identified in the cell at least once)<a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn5" role="doc-endnote"><p>This is consistent with the use of UMI counts rather than read counts, as each transcript molecule can only produce one UMI count but can yield many reads after fragmentation.<a href="#fnref5" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn6" role="doc-endnote"><p>Bearing in mind that the ‘experimental’ cells, which are γδ T cells, have a ‘smaller’ transcriptome than the ‘control’ cells.<a href="#fnref6" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn7" role="doc-endnote"><p>Here, ribosomal protein genes are defined as those gene symbols starting with “Rpl” or “Rps” or otherwise found in the <a href="https://www.gsea-msigdb.org/gsea/msigdb/cards/KEGG_RIBOSOME.html">“KEGG_RIBOSOME” gene set published by MSigDB</a>, namely: FAU, MRPL13, RPL10, RPL10A, RPL10AP1, RPL10AP2, RPL10AP3, RPL10AP5, RPL10AP6, RPL10L, RPL10P1, RPL10P10, RPL10P11, RPL10P12, RPL10P13, RPL10P14, RPL10P15, RPL10P16, RPL10P17, RPL10P18, RPL10P19, RPL10P2, RPL10P3, RPL10P4, RPL10P5, RPL10P6, RPL10P7, RPL10P8, RPL10P9, RPL11, RPL11P3, RPL11P4, RPL11P5, RPL12, RPL12P1, RPL12P10, RPL12P11, RPL12P12, RPL12P13, RPL12P14, RPL12P15, RPL12P17, RPL12P18, RPL12P19, RPL12P2, RPL12P21, RPL12P25, RPL12P26, RPL12P27, RPL12P29, RPL12P31, RPL12P32, RPL12P33, RPL12P34, RPL12P35, RPL12P37, RPL12P38, RPL12P39, RPL12P4, RPL12P40, RPL12P41, RPL12P42, RPL12P44, RPL12P45, RPL12P47, RPL12P48, RPL12P50, RPL12P6, RPL12P7, RPL12P8, RPL13, RPL13A, RPL13AP, RPL13AP19, RPL13AP2, RPL13AP20, RPL13AP23, RPL13AP25, RPL13AP3, RPL13AP5, RPL13AP6, RPL13AP7, RPL13P, RPL13P12, RPL13P2, RPL13P5, RPL13P6, RPL14, RPL14P1, RPL14P3, RPL14P5, RPL15, RPL15P1, RPL15P13, RPL15P14, RPL15P18, RPL15P2, RPL15P20, RPL15P21, RPL15P3, RPL17, RPL17P1, RPL17P10, RPL17P11, RPL17P2, RPL17P20, RPL17P22, RPL17P25, RPL17P26, RPL17P3, RPL17P33, RPL17P34, RPL17P36, RPL17P37, RPL17P41, RPL17P43, RPL17P44, RPL17P49, RPL17P50, RPL17P51, RPL18, RPL18A, RPL18AP14, RPL18AP15, RPL18AP16, RPL18AP17, RPL18AP3, RPL18AP7, RPL18AP8, RPL18P10, RPL18P11, RPL18P13, RPL19, RPL19P1, RPL19P11, RPL19P13, RPL19P14, RPL19P16, RPL19P18, RPL19P20, RPL19P21, RPL21, RPL21P1, RPL21P10, RPL21P103, RPL21P104, RPL21P106, RPL21P107, RPL21P108, RPL21P11, RPL21P110, RPL21P111, RPL21P112, RPL21P116, RPL21P119, RPL21P12, RPL21P120, RPL21P121, RPL21P123, RPL21P126, RPL21P127, RPL21P129, RPL21P13, RPL21P131, RPL21P133, RPL21P134, RPL21P135, RPL21P136, RPL21P16, RPL21P17, RPL21P18, RPL21P2, RPL21P23, RPL21P24, RPL21P28, RPL21P3, RPL21P32, RPL21P33, RPL21P38, RPL21P39, RPL21P4, RPL21P40, RPL21P41, RPL21P42, RPL21P43, RPL21P44, RPL21P5, RPL21P50, RPL21P54, RPL21P65, RPL21P66, RPL21P67, RPL21P7, RPL21P75, RPL21P8, RPL21P88, RPL21P9, RPL21P90, RPL21P91, RPL21P92, RPL21P93, RPL21P99, RPL22, RPL22L1, RPL22P1, RPL22P10, RPL22P11, RPL22P12, RPL22P13, RPL22P14, RPL22P16, RPL22P18, RPL22P19, RPL22P2, RPL22P20, RPL22P22, RPL22P23, RPL22P24, RPL22P3, RPL22P5, RPL22P6, RPL22P7, RPL22P8, RPL23, RPL23A, RPL23AP10, RPL23AP11, RPL23AP12, RPL23AP14, RPL23AP15, RPL23AP16, RPL23AP17, RPL23AP18, RPL23AP19, RPL23AP2, RPL23AP20, RPL23AP24, RPL23AP25, RPL23AP26, RPL23AP27, RPL23AP28, RPL23AP3, RPL23AP30, RPL23AP31, RPL23AP32, RPL23AP34, RPL23AP35, RPL23AP37, RPL23AP38, RPL23AP39, RPL23AP4, RPL23AP41, RPL23AP42, RPL23AP43, RPL23AP44, RPL23AP45, RPL23AP46, RPL23AP47, RPL23AP48, RPL23AP49, RPL23AP52, RPL23AP53, RPL23AP55, RPL23AP57, RPL23AP58, RPL23AP59, RPL23AP6, RPL23AP60, RPL23AP61, RPL23AP64, RPL23AP65, RPL23AP67, RPL23AP69, RPL23AP7, RPL23AP71, RPL23AP73, RPL23AP74, RPL23AP75, RPL23AP77, RPL23AP78, RPL23AP8, RPL23AP82, RPL23AP83, RPL23AP84, RPL23AP85, RPL23AP86, RPL23AP87, RPL23AP88, RPL23AP89, RPL23AP94, RPL23AP95, RPL23AP96, RPL23AP97, RPL23P11, RPL23P2, RPL23P4, RPL23P8, RPL24, RPL24P2, RPL24P4, RPL24P7, RPL24P8, RPL26, RPL26L1, RPL26P19, RPL26P26, RPL26P27, RPL26P28, RPL26P29, RPL26P30, RPL26P31, RPL26P35, RPL26P37, RPL26P6, RPL26P9, RPL27, RPL27A, RPL27AP, RPL27AP5, RPL27AP8, RPL27P12, RPL28, RPL29, RPL29P11, RPL29P12, RPL29P14, RPL29P19, RPL29P2, RPL29P22, RPL29P24, RPL29P25, RPL29P28, RPL29P3, RPL29P30, RPL29P33, RPL29P7, RPL3, RPL30, RPL30P1, RPL30P10, RPL30P11, RPL30P12, RPL30P13, RPL30P14, RPL30P15, RPL30P16, RPL30P2, RPL30P3, RPL30P4, RPL30P5, RPL30P7, RPL31, RPL31P1, RPL31P11, RPL31P12, RPL31P2, RPL31P28, RPL31P33, RPL31P35, RPL31P41, RPL31P44, RPL31P45, RPL31P47, RPL31P49, RPL31P50, RPL31P52, RPL31P57, RPL31P58, RPL31P59, RPL31P61, RPL31P63, RPL31P7, RPL31P9, RPL32, RPL32P1, RPL32P26, RPL32P3, RPL32P31, RPL32P33, RPL32P34, RPL32P9, RPL34, RPL34-AS1, RPL34P1, RPL34P17, RPL34P18, RPL34P19, RPL34P20, RPL34P21, RPL34P22, RPL34P23, RPL34P26, RPL34P27, RPL34P3, RPL34P31, RPL34P33, RPL34P34, RPL34P6, RPL35, RPL35A, RPL35AP, RPL35AP15, RPL35AP18, RPL35AP19, RPL35AP21, RPL35AP23, RPL35AP24, RPL35AP26, RPL35AP28, RPL35AP30, RPL35AP31, RPL35AP32, RPL35AP33, RPL35AP35, RPL35AP4, RPL35P1, RPL35P2, RPL35P3, RPL35P4, RPL35P5, RPL35P6, RPL35P8, RPL35P9, RPL36, RPL36A, RPL36AL, RPL36AP10, RPL36AP13, RPL36AP15, RPL36AP21, RPL36AP26, RPL36AP29, RPL36AP33, RPL36AP36, RPL36AP39, RPL36AP40, RPL36AP41, RPL36AP43, RPL36AP45, RPL36AP48, RPL36AP51, RPL36AP53, RPL36AP55, RPL36AP6, RPL36P16, RPL36P18, RPL36P2, RPL36P4, RPL37, RPL37A, RPL37AP1, RPL37AP8, RPL37P1, RPL37P10, RPL37P2, RPL37P23, RPL37P25, RPL37P3, RPL37P4, RPL37P6, RPL38, RPL38P1, RPL38P3, RPL38P4, RPL39, RPL39L, RPL39P, RPL39P18, RPL39P25, RPL39P26, RPL39P29, RPL39P3, RPL39P31, RPL39P33, RPL39P36, RPL39P38, RPL39P39, RPL39P40, RPL39P5, RPL39P6, RPL3L, RPL3P1, RPL3P10, RPL3P11, RPL3P12, RPL3P13, RPL3P2, RPL3P3, RPL3P4, RPL3P5, RPL3P6, RPL3P7, RPL3P8, RPL3P9, RPL4, RPL41, RPL41P1, RPL41P2, RPL41P5, RPL4P1, RPL4P2, RPL4P3, RPL4P4, RPL4P5, RPL4P6, RPL5, RPL5P1, RPL5P10, RPL5P11, RPL5P12, RPL5P14, RPL5P15, RPL5P16, RPL5P17, RPL5P18, RPL5P22, RPL5P23, RPL5P24, RPL5P25, RPL5P26, RPL5P27, RPL5P28, RPL5P29, RPL5P3, RPL5P30, RPL5P32, RPL5P33, RPL5P34, RPL5P35, RPL5P4, RPL5P5, RPL5P6, RPL5P7, RPL5P8, RPL5P9, RPL6, RPL6P2, RPL6P21, RPL6P24, RPL6P25, RPL6P27, RPL6P29, RPL6P30, RPL7, RPL7A, RPL7AP10, RPL7AP11, RPL7AP14, RPL7AP15, RPL7AP23, RPL7AP26, RPL7AP28, RPL7AP3, RPL7AP30, RPL7AP31, RPL7AP33, RPL7AP34, RPL7AP38, RPL7AP4, RPL7AP50, RPL7AP53, RPL7AP57, RPL7AP58, RPL7AP6, RPL7AP60, RPL7AP61, RPL7AP66, RPL7AP7, RPL7AP71, RPL7AP73, RPL7AP8, RPL7AP9, RPL7L1, RPL7L1P1, RPL7L1P12, RPL7L1P2, RPL7L1P3, RPL7L1P8, RPL7L1P9, RPL7P1, RPL7P10, RPL7P11, RPL7P12, RPL7P13, RPL7P15, RPL7P16, RPL7P17, RPL7P18, RPL7P19, RPL7P2, RPL7P20, RPL7P21, RPL7P22, RPL7P23, RPL7P24, RPL7P25, RPL7P26, RPL7P27, RPL7P28, RPL7P31, RPL7P32, RPL7P33, RPL7P36, RPL7P37, RPL7P38, RPL7P39, RPL7P4, RPL7P41, RPL7P44, RPL7P45, RPL7P46, RPL7P47, RPL7P48, RPL7P49, RPL7P5, RPL7P50, RPL7P51, RPL7P52, RPL7P53, RPL7P56, RPL7P57, RPL7P58, RPL7P59, RPL7P6, RPL7P60, RPL7P61, RPL7P7, RPL7P8, RPL7P9, RPL8, RPL8P2, RPL9, RPL9P18, RPL9P2, RPL9P21, RPL9P28, RPL9P3, RPL9P30, RPL9P31, RPL9P32, RPL9P5, RPL9P7, RPLP0, RPLP0P11, RPLP0P2, RPLP0P6, RPLP0P7, RPLP1, RPLP1P10, RPLP1P13, RPLP1P6, RPLP2, RPLP2P1, RPS10, RPS10P14, RPS10P16, RPS10P18, RPS10P2, RPS10P21, RPS10P27, RPS10P28, RPS10P3, RPS10P5, RPS10P7, RPS11, RPS11P1, RPS11P5, RPS11P7, RPS12, RPS12P16, RPS12P17, RPS12P18, RPS12P2, RPS12P20, RPS12P23, RPS12P24, RPS12P26, RPS12P27, RPS12P5, RPS13, RPS13P2, RPS14, RPS14P4, RPS14P8, RPS15, RPS15A, RPS15AP1, RPS15AP10, RPS15AP11, RPS15AP12, RPS15AP16, RPS15AP17, RPS15AP18, RPS15AP24, RPS15AP25, RPS15AP28, RPS15AP29, RPS15AP3, RPS15AP30, RPS15AP36, RPS15AP38, RPS15AP40, RPS15AP5, RPS15AP6, RPS15AP9, RPS15P5, RPS15P9, RPS16, RPS16P5, RPS16P8, RPS16P9, RPS17, RPS17P1, RPS17P13, RPS17P14, RPS17P15, RPS17P2, RPS17P5, RPS17P8, RPS18, RPS18P12, RPS18P13, RPS18P6, RPS18P9, RPS19, RPS19BP1, RPS19P1, RPS19P3, RPS19P7, RPS2, RPS20, RPS20P1, RPS20P10, RPS20P14, RPS20P2, RPS20P22, RPS20P24, RPS20P32, RPS20P33, RPS20P4, RPS20P5, RPS21, RPS21P1, RPS21P4, RPS23, RPS23P1, RPS23P10, RPS23P2, RPS23P3, RPS23P5, RPS23P6, RPS23P7, RPS23P8, RPS23P9, RPS24, RPS24P13, RPS24P16, RPS24P17, RPS24P18, RPS24P6, RPS24P7, RPS24P8, RPS25, RPS25P9, RPS26, RPS26P11, RPS26P13, RPS26P15, RPS26P2, RPS26P21, RPS26P28, RPS26P3, RPS26P31, RPS26P34, RPS26P35, RPS26P38, RPS26P39, RPS26P4, RPS26P40, RPS26P42, RPS26P43, RPS26P45, RPS26P47, RPS26P52, RPS26P55, RPS26P56, RPS26P6, RPS26P8, RPS27, RPS27A, RPS27AP1, RPS27AP11, RPS27AP12, RPS27AP15, RPS27AP16, RPS27AP17, RPS27AP19, RPS27AP2, RPS27AP3, RPS27AP6, RPS27AP7, RPS27AP9, RPS27L, RPS27P14, RPS27P15, RPS27P21, RPS27P23, RPS27P25, RPS27P26, RPS27P27, RPS27P29, RPS28, RPS28P7, RPS29, RPS29P11, RPS29P12, RPS29P14, RPS29P15, RPS29P16, RPS29P17, RPS29P19, RPS29P20, RPS29P27, RPS29P28, RPS29P29, RPS29P3, RPS29P31, RPS29P32, RPS29P5, RPS29P6, RPS29P7, RPS29P8, RPS29P9, RPS2P1, RPS2P2, RPS2P24, RPS2P28, RPS2P32, RPS2P35, RPS2P36, RPS2P39, RPS2P4, RPS2P41, RPS2P44, RPS2P45, RPS2P46, RPS2P48, RPS2P5, RPS2P55, RPS2P6, RPS2P7, RPS3, RPS3A, RPS3AP12, RPS3AP14, RPS3AP15, RPS3AP22, RPS3AP23, RPS3AP25, RPS3AP26, RPS3AP29, RPS3AP3, RPS3AP30, RPS3AP33, RPS3AP34, RPS3AP35, RPS3AP36, RPS3AP37, RPS3AP38, RPS3AP39, RPS3AP4, RPS3AP40, RPS3AP41, RPS3AP42, RPS3AP43, RPS3AP44, RPS3AP47, RPS3AP49, RPS3AP5, RPS3AP53, RPS3AP54, RPS3AP6, RPS3P1, RPS3P2, RPS3P6, RPS3P7, RPS4X, RPS4XP1, RPS4XP10, RPS4XP11, RPS4XP13, RPS4XP14, RPS4XP15, RPS4XP16, RPS4XP17, RPS4XP18, RPS4XP19, RPS4XP2, RPS4XP20, RPS4XP21, RPS4XP22, RPS4XP23, RPS4XP3, RPS4XP4, RPS4XP5, RPS4XP6, RPS4XP7, RPS4XP8, RPS4XP9, RPS4Y1, RPS4Y2, RPS5, RPS5P2, RPS5P3, RPS5P7, RPS6, RPS6KA1, RPS6KA2, RPS6KA3, RPS6KA4, RPS6KA5, RPS6KA6, RPS6KB1, RPS6KB2, RPS6KB2-AS1, RPS6KC1, RPS6KL1, RPS6P15, RPS6P16, RPS6P2, RPS6P20, RPS6P21, RPS6P22, RPS6P25, RPS6P4, RPS7, RPS7P1, RPS7P10, RPS7P11, RPS7P12, RPS7P13, RPS7P14, RPS7P15, RPS7P3, RPS7P4, RPS7P5, RPS7P6, RPS7P7, RPS7P8, RPS8, RPS8P10, RPS8P3, RPS8P4, RPS9, RPS9P1, RPSA, RPSAP1, RPSAP10, RPSAP11, RPSAP12, RPSAP13, RPSAP14, RPSAP15, RPSAP16, RPSAP17, RPSAP18, RPSAP19, RPSAP2, RPSAP20, RPSAP21, RPSAP22, RPSAP23, RPSAP24, RPSAP27, RPSAP28, RPSAP29, RPSAP3, RPSAP31, RPSAP33, RPSAP36, RPSAP38, RPSAP39, RPSAP4, RPSAP40, RPSAP43, RPSAP44, RPSAP45, RPSAP46, RPSAP47, RPSAP49, RPSAP5, RPSAP51, RPSAP52, RPSAP53, RPSAP54, RPSAP55, RPSAP57, RPSAP58, RPSAP59, RPSAP6, RPSAP61, RPSAP63, RPSAP64, RPSAP66, RPSAP69, RPSAP7, RPSAP73, RPSAP74, RPSAP75, RPSAP76, RPSAP8, RPSAP9, RSL24D1, RSL24D1P11, UBA52<a href="#fnref7" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn8" role="doc-endnote"><p>This is a large number of HVGs but the results don’t qualitatively change by reducing this to, say, the top-1000 HVGs.<a href="#fnref8" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn9" role="doc-endnote"><p>Note that the converse is not true, i.e., there is no guarantee that the retained PCs capture all of the signal, which is only generally possible if no dimensionality reduction is performed at all.<a href="#fnref9" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
<h3 id="references">References</h3>
<div id="references-listing"></div>
</div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
