<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #20794d; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007ba5; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #007ba5; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #007ba5; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #20794d; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

  <!--radix_placeholder_meta_tags-->
  <title>C094: Trajectory analysis of the Pellicci gamma-delta T-cell single-cell data set</title>



  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2021-10-01"/>
  <meta property="article:created" itemprop="dateCreated" content="2021-10-01"/>
  <meta name="article:author" content="Peter Hickey"/>

  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="C094: Trajectory analysis of the Pellicci gamma-delta T-cell single-cell data set"/>
  <meta property="og:type" content="article"/>
  <meta property="og:locale" content="en_US"/>
  <meta property="og:site_name" content="C094"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="C094: Trajectory analysis of the Pellicci gamma-delta T-cell single-cell data set"/>

  <!--/radix_placeholder_meta_tags-->
  
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","author","date","editor_options","bibliography"]}},"value":[{"type":"character","attributes":{},"value":["Trajectory analysis of the Pellicci gamma-delta T-cell single-cell data set"]},{"type":"NULL"},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","url","affiliation","affiliation_url"]}},"value":[{"type":"character","attributes":{},"value":["Peter Hickey"]},{"type":"character","attributes":{},"value":["https://peterhickey.org"]},{"type":"character","attributes":{},"value":["Single Cell Open Research Endeavour (SCORE), WEHI"]},{"type":"character","attributes":{},"value":["https://www.wehi.edu.au/people/rory-bowden/4536/wehi-advanced-genomics-facility"]}]}]},{"type":"character","attributes":{},"value":["2021-10-01"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["chunk_output_type"]}},"value":[{"type":"character","attributes":{},"value":["console"]}]},{"type":"character","attributes":{},"value":["ref.bib"]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  <!--radix_placeholder_navigation_in_header-->
  <meta name="distill:offset" content=""/>

  <script type="application/javascript">

    window.headroom_prevent_pin = false;

    window.document.addEventListener("DOMContentLoaded", function (event) {

      // initialize headroom for banner
      var header = $('header').get(0);
      var headerHeight = header.offsetHeight;
      var headroom = new Headroom(header, {
        tolerance: 5,
        onPin : function() {
          if (window.headroom_prevent_pin) {
            window.headroom_prevent_pin = false;
            headroom.unpin();
          }
        }
      });
      headroom.init();
      if(window.location.hash)
        headroom.unpin();
      $(header).addClass('headroom--transition');

      // offset scroll location for banner on hash change
      // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
      window.addEventListener("hashchange", function(event) {
        window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
      });

      // responsive menu
      $('.distill-site-header').each(function(i, val) {
        var topnav = $(this);
        var toggle = topnav.find('.nav-toggle');
        toggle.on('click', function() {
          topnav.toggleClass('responsive');
        });
      });

      // nav dropdowns
      $('.nav-dropbtn').click(function(e) {
        $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
        $(this).parent().siblings('.nav-dropdown')
           .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $("body").click(function(e){
        $('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $(".nav-dropdown").click(function(e){
        e.stopPropagation();
      });
    });
  </script>

  <style type="text/css">

  /* Theme (user-documented overrideables for nav appearance) */

  .distill-site-nav {
    color: rgba(255, 255, 255, 0.8);
    background-color: #0F2E3D;
    font-size: 15px;
    font-weight: 300;
  }

  .distill-site-nav a {
    color: inherit;
    text-decoration: none;
  }

  .distill-site-nav a:hover {
    color: white;
  }

  @media print {
    .distill-site-nav {
      display: none;
    }
  }

  .distill-site-header {

  }

  .distill-site-footer {

  }


  /* Site Header */

  .distill-site-header {
    width: 100%;
    box-sizing: border-box;
    z-index: 3;
  }

  .distill-site-header .nav-left {
    display: inline-block;
    margin-left: 8px;
  }

  @media screen and (max-width: 768px) {
    .distill-site-header .nav-left {
      margin-left: 0;
    }
  }


  .distill-site-header .nav-right {
    float: right;
    margin-right: 8px;
  }

  .distill-site-header a,
  .distill-site-header .title {
    display: inline-block;
    text-align: center;
    padding: 14px 10px 14px 10px;
  }

  .distill-site-header .title {
    font-size: 18px;
    min-width: 150px;
  }

  .distill-site-header .logo {
    padding: 0;
  }

  .distill-site-header .logo img {
    display: none;
    max-height: 20px;
    width: auto;
    margin-bottom: -4px;
  }

  .distill-site-header .nav-image img {
    max-height: 18px;
    width: auto;
    display: inline-block;
    margin-bottom: -3px;
  }



  @media screen and (min-width: 1000px) {
    .distill-site-header .logo img {
      display: inline-block;
    }
    .distill-site-header .nav-left {
      margin-left: 20px;
    }
    .distill-site-header .nav-right {
      margin-right: 20px;
    }
    .distill-site-header .title {
      padding-left: 12px;
    }
  }


  .distill-site-header .nav-toggle {
    display: none;
  }

  .nav-dropdown {
    display: inline-block;
    position: relative;
  }

  .nav-dropdown .nav-dropbtn {
    border: none;
    outline: none;
    color: rgba(255, 255, 255, 0.8);
    padding: 16px 10px;
    background-color: transparent;
    font-family: inherit;
    font-size: inherit;
    font-weight: inherit;
    margin: 0;
    margin-top: 1px;
    z-index: 2;
  }

  .nav-dropdown-content {
    display: none;
    position: absolute;
    background-color: white;
    min-width: 200px;
    border: 1px solid rgba(0,0,0,0.15);
    border-radius: 4px;
    box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
    z-index: 1;
    margin-top: 2px;
    white-space: nowrap;
    padding-top: 4px;
    padding-bottom: 4px;
  }

  .nav-dropdown-content hr {
    margin-top: 4px;
    margin-bottom: 4px;
    border: none;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .nav-dropdown-active {
    display: block;
  }

  .nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
    color: black;
    padding: 6px 24px;
    text-decoration: none;
    display: block;
    text-align: left;
  }

  .nav-dropdown-content .nav-dropdown-header {
    display: block;
    padding: 5px 24px;
    padding-bottom: 0;
    text-transform: uppercase;
    font-size: 14px;
    color: #999999;
    white-space: nowrap;
  }

  .nav-dropdown:hover .nav-dropbtn {
    color: white;
  }

  .nav-dropdown-content a:hover {
    background-color: #ddd;
    color: black;
  }

  .nav-right .nav-dropdown-content {
    margin-left: -45%;
    right: 0;
  }

  @media screen and (max-width: 768px) {
    .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
    .distill-site-header a.nav-toggle {
      float: right;
      display: block;
    }
    .distill-site-header .title {
      margin-left: 0;
    }
    .distill-site-header .nav-right {
      margin-right: 0;
    }
    .distill-site-header {
      overflow: hidden;
    }
    .nav-right .nav-dropdown-content {
      margin-left: 0;
    }
  }


  @media screen and (max-width: 768px) {
    .distill-site-header.responsive {position: relative; min-height: 500px; }
    .distill-site-header.responsive a.nav-toggle {
      position: absolute;
      right: 0;
      top: 0;
    }
    .distill-site-header.responsive a,
    .distill-site-header.responsive .nav-dropdown {
      display: block;
      text-align: left;
    }
    .distill-site-header.responsive .nav-left,
    .distill-site-header.responsive .nav-right {
      width: 100%;
    }
    .distill-site-header.responsive .nav-dropdown {float: none;}
    .distill-site-header.responsive .nav-dropdown-content {position: relative;}
    .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
      display: block;
      width: 100%;
      text-align: left;
    }
  }

  /* Site Footer */

  .distill-site-footer {
    width: 100%;
    overflow: hidden;
    box-sizing: border-box;
    z-index: 3;
    margin-top: 30px;
    padding-top: 30px;
    padding-bottom: 30px;
    text-align: center;
  }

  /* Headroom */

  d-title {
    padding-top: 6rem;
  }

  @media print {
    d-title {
      padding-top: 4rem;
    }
  }

  .headroom {
    z-index: 1000;
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
  }

  .headroom--transition {
    transition: all .4s ease-in-out;
  }

  .headroom--unpinned {
    top: -100px;
  }

  .headroom--pinned {
    top: 0;
  }

  /* adjust viewport for navbar height */
  /* helps vertically center bootstrap (non-distill) content */
  .min-vh-100 {
    min-height: calc(100vh - 100px) !important;
  }

  </style>

  <script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
  <link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet"/>
  <link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet"/>
  <script src="site_libs/headroom-0.9.4/headroom.min.js"></script>
  <script src="site_libs/autocomplete-0.37.1/autocomplete.min.js"></script>
  <script src="site_libs/fuse-6.4.1/fuse.min.js"></script>

  <script type="application/javascript">

  function getMeta(metaName) {
    var metas = document.getElementsByTagName('meta');
    for (let i = 0; i < metas.length; i++) {
      if (metas[i].getAttribute('name') === metaName) {
        return metas[i].getAttribute('content');
      }
    }
    return '';
  }

  function offsetURL(url) {
    var offset = getMeta('distill:offset');
    return offset ? offset + '/' + url : url;
  }

  function createFuseIndex() {

    // create fuse index
    var options = {
      keys: [
        { name: 'title', weight: 20 },
        { name: 'categories', weight: 15 },
        { name: 'description', weight: 10 },
        { name: 'contents', weight: 5 },
      ],
      ignoreLocation: true,
      threshold: 0
    };
    var fuse = new window.Fuse([], options);

    // fetch the main search.json
    return fetch(offsetURL('search.json'))
      .then(function(response) {
        if (response.status == 200) {
          return response.json().then(function(json) {
            // index main articles
            json.articles.forEach(function(article) {
              fuse.add(article);
            });
            // download collections and index their articles
            return Promise.all(json.collections.map(function(collection) {
              return fetch(offsetURL(collection)).then(function(response) {
                if (response.status === 200) {
                  return response.json().then(function(articles) {
                    articles.forEach(function(article) {
                      fuse.add(article);
                    });
                  })
                } else {
                  return Promise.reject(
                    new Error('Unexpected status from search index request: ' +
                              response.status)
                  );
                }
              });
            })).then(function() {
              return fuse;
            });
          });

        } else {
          return Promise.reject(
            new Error('Unexpected status from search index request: ' +
                        response.status)
          );
        }
      });
  }

  window.document.addEventListener("DOMContentLoaded", function (event) {

    // get search element (bail if we don't have one)
    var searchEl = window.document.getElementById('distill-search');
    if (!searchEl)
      return;

    createFuseIndex()
      .then(function(fuse) {

        // make search box visible
        searchEl.classList.remove('hidden');

        // initialize autocomplete
        var options = {
          autoselect: true,
          hint: false,
          minLength: 2,
        };
        window.autocomplete(searchEl, options, [{
          source: function(query, callback) {
            const searchOptions = {
              isCaseSensitive: false,
              shouldSort: true,
              minMatchCharLength: 2,
              limit: 10,
            };
            var results = fuse.search(query, searchOptions);
            callback(results
              .map(function(result) { return result.item; })
            );
          },
          templates: {
            suggestion: function(suggestion) {
              var img = suggestion.preview && Object.keys(suggestion.preview).length > 0
                ? `<img src="${offsetURL(suggestion.preview)}"</img>`
                : '';
              var html = `
                <div class="search-item">
                  <h3>${suggestion.title}</h3>
                  <div class="search-item-description">
                    ${suggestion.description || ''}
                  </div>
                  <div class="search-item-preview">
                    ${img}
                  </div>
                </div>
              `;
              return html;
            }
          }
        }]).on('autocomplete:selected', function(event, suggestion) {
          window.location.href = offsetURL(suggestion.path);
        });
        // remove inline display style on autocompleter (we want to
        // manage responsive display via css)
        $('.algolia-autocomplete').css("display", "");
      })
      .catch(function(error) {
        console.log(error);
      });

  });

  </script>

  <style type="text/css">

  .nav-search {
    font-size: x-small;
  }

  /* Algolioa Autocomplete */

  .algolia-autocomplete {
    display: inline-block;
    margin-left: 10px;
    vertical-align: sub;
    background-color: white;
    color: black;
    padding: 6px;
    padding-top: 8px;
    padding-bottom: 0;
    border-radius: 6px;
    border: 1px #0F2E3D solid;
    width: 180px;
  }


  @media screen and (max-width: 768px) {
    .distill-site-nav .algolia-autocomplete {
      display: none;
      visibility: hidden;
    }
    .distill-site-nav.responsive .algolia-autocomplete {
      display: inline-block;
      visibility: visible;
    }
    .distill-site-nav.responsive .algolia-autocomplete .aa-dropdown-menu {
      margin-left: 0;
      width: 400px;
      max-height: 400px;
    }
  }

  .algolia-autocomplete .aa-input, .algolia-autocomplete .aa-hint {
    width: 90%;
    outline: none;
    border: none;
  }

  .algolia-autocomplete .aa-hint {
    color: #999;
  }
  .algolia-autocomplete .aa-dropdown-menu {
    width: 550px;
    max-height: 70vh;
    overflow-x: visible;
    overflow-y: scroll;
    padding: 5px;
    margin-top: 3px;
    margin-left: -150px;
    background-color: #fff;
    border-radius: 5px;
    border: 1px solid #999;
    border-top: none;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion {
    cursor: pointer;
    padding: 5px 4px;
    border-bottom: 1px solid #eee;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion:last-of-type {
    border-bottom: none;
    margin-bottom: 2px;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item {
    overflow: hidden;
    font-size: 0.8em;
    line-height: 1.4em;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item h3 {
    font-size: 1rem;
    margin-block-start: 0;
    margin-block-end: 5px;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-description {
    display: inline-block;
    overflow: hidden;
    height: 2.8em;
    width: 80%;
    margin-right: 4%;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview {
    display: inline-block;
    width: 15%;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview img {
    height: 3em;
    width: auto;
    display: none;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview img[src] {
    display: initial;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion.aa-cursor {
    background-color: #eee;
  }
  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion em {
    font-weight: bold;
    font-style: normal;
  }

  </style>


  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: auto;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  d-article details {
    grid-column: text;
    margin-bottom: 0.8em;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure img {
    width: 100%;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }

  /* Citation hover box */

  .tippy-box[data-theme~=light-border] {
    background-color: rgba(250, 250, 250, 0.95);
  }

  .tippy-content > p {
    margin-bottom: 0;
    padding: 2px;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('details, div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme
    $('code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // hoverable references
      $('span.citation[data-cites]').each(function() {
        var refHtml = $('#ref-' + $(this).attr('data-cites')).html();
        window.tippy(this, {
          allowHTML: true,
          content: refHtml,
          maxWidth: 500,
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        });
      });

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace("index.html", "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <!--/radix_placeholder_distill-->
  <script src="site_libs/header-attrs-2.7/header-attrs.js"></script>
  <script src="site_libs/popper-2.6.0/popper.min.js"></script>
  <link href="site_libs/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="site_libs/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="site_libs/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="site_libs/anchor-4.2.2/anchor.min.js"></script>
  <script src="site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <!--/radix_placeholder_site_in_header-->


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"Trajectory analysis of the Pellicci gamma-delta T-cell single-cell data set","authors":[{"author":"Peter Hickey","authorURL":"https://peterhickey.org","affiliation":"Single Cell Open Research Endeavour (SCORE), WEHI","affiliationURL":"https://www.wehi.edu.au/people/rory-bowden/4536/wehi-advanced-genomics-facility","orcidID":""}],"publishedDate":"2021-10-01T00:00:00.000+10:00","citationText":"Hickey, 2021"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a href="index.html" class="title">C094</a>
<a href="index.html">Introduction</a>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Preprocessing
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="C094_Pellicci.single-cell.preprocess.html">Single cell</a>
<a href="C094_Pellicci.mini-bulk.preprocess.html">Mini bulk</a>
</div>
</div>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Selecting biologically relevant cells
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="C094_Pellicci.single-cell.cell_selection.html">Single cell</a>
</div>
</div>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Merging cells
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="C094_Pellicci.single-cell.merge.whole_cell.html">Single cell (whole cell)</a>
<a href="C094_Pellicci.single-cell.merge.S1_S2_only.html">Single cell (S1 and S2 only)</a>
<a href="C094_Pellicci.single-cell.merge.S3_only.html">Single cell (S3 only)</a>
<a href="C094_Pellicci.single-cell.merge.thymus_only.html">Single cell (thymus only)</a>
<a href="C094_Pellicci.single-cell.merge.thymusS1_thymusS2_only.html">Single cell (thymus.S1 and thymus.S2 only)</a>
</div>
</div>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Annotating cells
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="C094_Pellicci.single-cell.annotate.whole_cell.html">Single cell (whole cell)</a>
<a href="C094_Pellicci.single-cell.annotate.S1_S2_only.html">Single cell (S1 and S2 only)</a>
<a href="C094_Pellicci.single-cell.annotate.S3_only.html">Single cell (S3 only)</a>
<a href="C094_Pellicci.single-cell.annotate.thymus_only.html">Single cell (thymus only)</a>
<a href="C094_Pellicci.single-cell.annotate.thymusS1_thymusS2_only.html">Single cell (thymus.S1 and thymus.S2 only)</a>
</div>
</div>
<a href="C094_Pellicci.single-cell.trajectory.html">Trajectory analysis</a>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Mini-bulk differential analyses
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="C094_Pellicci.mini-bulk.multi-sample_comparisons.all_samples.html">All samples</a>
<a href="C094_Pellicci.mini-bulk.multi-sample_comparisons.excluding_blood_1-3.html">Excluding Blood.1 - Blood.3</a>
<a href="C094_Pellicci.mini-bulk.multi-sample_comparisons.excluding_donors_1-3.html">Excluding donors 1-3</a>
</div>
</div>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Supplementary files
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="../output/scPipe/">scPipe reports</a>
<a href="../output/DEGs/">Differentially expressed genes</a>
<a href="../output/glimma-plots/">Glimma plots</a>
</div>
</div>
</div>
<div class="nav-right">
<a href="https://github.com/WEHISCORE/C094_Pellicci">
<i class="fa fa-github" aria-hidden="true"></i>
</a>
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>Trajectory analysis of the Pellicci gamma-delta T-cell single-cell data set</h1>
<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->

</div>

<div class="d-byline">
  Peter Hickey <a href="https://peterhickey.org" class="uri">https://peterhickey.org</a> (Single Cell Open Research Endeavour (SCORE), WEHI)<a href="https://www.wehi.edu.au/people/rory-bowden/4536/wehi-advanced-genomics-facility" class="uri">https://www.wehi.edu.au/people/rory-bowden/4536/wehi-advanced-genomics-facility</a>
  
<br/>2021-10-01
</div>

<div class="d-article">
<div class="d-contents d-contents-float">
<nav class="l-text toc figcaption" id="TOC">
<h3>Contents</h3>
<ul>
<li><a href="#setup">Setup</a></li>
<li><a href="#obtaining-pseudotime-orderings">Obtaining pseudotime orderings</a></li>
<li><a href="#changes-along-a-trajectory">Changes along a trajectory</a></li>
<li><a href="#gene-set-testing">Gene set testing</a>
<ul>
<li><a href="#goana">goana</a></li>
<li><a href="#kegga">kegga</a></li>
<li><a href="#camera">camera</a></li>
<li><a href="#visualisation">Visualisation</a></li>
</ul></li>
<li><a href="#figures-for-paper">Figures for paper</a>
<ul>
<li><a href="#figure-5">Figure 5</a></li>
<li><a href="#deg-heatmaps">DEG heatmaps</a></li>
<li><a href="#mock-figure-sent-to-dan-and-co-2021-09-09">Mock figure sent to Dan and co (2021-09-09)</a></li>
</ul></li>
<li><a href="#additional-information">Additional information</a></li>
<li><a href="#session-info">Session info</a></li>
</ul>
</nav>
</div>
<h1 id="setup">Setup</h1>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://here.r-lib.org/'>here</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>SingleCellExperiment</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://bioconductor.org/packages/scater/'>scater</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://patchwork.data-imaginist.com'>patchwork</a></span><span class='op'>)</span>

<span class='va'>outdir</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"output"</span>, <span class='st'>"trajectory"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/files2.html'>dir.create</a></span><span class='op'>(</span><span class='va'>outdir</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>Need to start from this <em>SCE</em> object in order to have consistency with Williams analyses.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"data/SCEs/C094_Pellicci.single-cell.merged.whole_cell.SCE.rds"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># Some useful gene sets</span>
<span class='va'>mito_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/any.html'>any</a></span><span class='op'>(</span><span class='fu'>rowData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>$</span><span class='va'>ENSEMBL.SEQNAME</span> <span class='op'>==</span> <span class='st'>"MT"</span><span class='op'>)</span><span class='op'>]</span>
<span class='va'>ribo_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/grep.html'>grep</a></span><span class='op'>(</span><span class='st'>"^RP(S|L)"</span>, <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>, value <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='co'># NOTE: A more curated approach for identifying ribosomal protein genes</span>
<span class='co'>#       (https://github.com/Bioconductor/OrchestratingSingleCellAnalysis-base/blob/ae201bf26e3e4fa82d9165d8abf4f4dc4b8e5a68/feature-selection.Rmd#L376-L380)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://igordot.github.io/msigdbr'>msigdbr</a></span><span class='op'>)</span>
<span class='va'>c2_sets</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://igordot.github.io/msigdbr/reference/msigdbr.html'>msigdbr</a></span><span class='op'>(</span>species <span class='op'>=</span> <span class='st'>"Homo sapiens"</span>, category <span class='op'>=</span> <span class='st'>"C2"</span><span class='op'>)</span>
<span class='va'>ribo_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/sets.html'>union</a></span><span class='op'>(</span>
  <span class='va'>ribo_set</span>,
  <span class='va'>c2_sets</span><span class='op'>[</span><span class='va'>c2_sets</span><span class='op'>$</span><span class='va'>gs_name</span> <span class='op'>==</span> <span class='st'>"KEGG_RIBOSOME"</span>, <span class='op'>]</span><span class='op'>$</span><span class='va'>gene_symbol</span><span class='op'>)</span>
<span class='va'>ribo_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/sets.html'>intersect</a></span><span class='op'>(</span><span class='va'>ribo_set</span>, <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>sex_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/any.html'>any</a></span><span class='op'>(</span><span class='fu'>rowData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>$</span><span class='va'>ENSEMBL.SEQNAME</span> <span class='op'>%in%</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"X"</span>, <span class='st'>"Y"</span><span class='op'>)</span><span class='op'>)</span><span class='op'>]</span>
<span class='va'>pseudogene_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span>
  <span class='fu'><a href='https://rdrr.io/r/base/any.html'>any</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/grep.html'>grepl</a></span><span class='op'>(</span><span class='st'>"pseudogene"</span>, <span class='fu'>rowData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>$</span><span class='va'>ENSEMBL.GENEBIOTYPE</span><span class='op'>)</span><span class='op'>)</span><span class='op'>]</span>
<span class='va'>protein_coding_gene_set</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span>
  <span class='fu'><a href='https://rdrr.io/r/base/any.html'>any</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/grep.html'>grepl</a></span><span class='op'>(</span><span class='st'>"protein_coding"</span>, <span class='fu'>rowData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>$</span><span class='va'>ENSEMBL.GENEBIOTYPE</span><span class='op'>)</span><span class='op'>)</span><span class='op'>]</span>
</code></pre>
</div>
</details>
</div>
<h1 id="obtaining-pseudotime-orderings">Obtaining pseudotime orderings</h1>
<p>Analysis using <em><a href="https://bioconductor.org/packages/3.12/slingshot">slingshot</a></em>; see <a href="https://bioconductor.org/books/release/OSCA/trajectory-analysis.html#principal-curves">https://bioconductor.org/books/release/OSCA/trajectory-analysis.html#principal-curves</a> for inspiration.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>slingshot</span><span class='op'>)</span>
<span class='co'># NOTE: Can't directly use `slingshot(sce, reducedDim = "corrected")` because</span>
<span class='co'>#       it gives an error:</span>
<span class='co'>#       "Error in solve.default(s1 + s2) :</span>
<span class='co'>#         system is computationally singular: reciprocal condition</span>
<span class='co'>#         number = 2.91591e-17"</span>
<span class='co'>#       This is similar to previously reported issues</span>
<span class='co'>#       (https://github.com/kstreet13/slingshot/issues/87 and</span>
<span class='co'>#       https://github.com/kstreet13/slingshot/issues/35) for which a hacky</span>
<span class='co'>#       workaround is to drop the last dimension.</span>
<span class='va'>rd</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html'>reducedDim</a></span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"corrected"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html'>reducedDim</a></span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"corrected_1"</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='va'>rd</span><span class='op'>[</span>, <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_len</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>ncol</a></span><span class='op'>(</span><span class='va'>rd</span><span class='op'>)</span> <span class='op'>-</span> <span class='fl'>1</span><span class='op'>)</span><span class='op'>]</span>
<span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/slingshot/man/slingshot.html'>slingshot</a></span><span class='op'>(</span><span class='va'>sce</span>, reducedDim <span class='op'>=</span> <span class='st'>"corrected_1"</span><span class='op'>)</span>

<span class='va'>embedded</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/slingshot/man/embedCurves.html'>embedCurves</a></span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"UMAP_corrected"</span><span class='op'>)</span>
<span class='va'>embedded</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/slingshot/man/slingCurves.html'>slingCurves</a></span><span class='op'>(</span><span class='va'>embedded</span><span class='op'>)</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span> <span class='co'># only 1 path.</span>
<span class='va'>embedded</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span><span class='va'>embedded</span><span class='op'>$</span><span class='va'>s</span><span class='op'>[</span><span class='va'>embedded</span><span class='op'>$</span><span class='va'>ord</span>, <span class='op'>]</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotReducedDim.html'>plotReducedDim</a></span><span class='op'>(</span>
  <span class='va'>sce</span>,
  dimred <span class='op'>=</span> <span class='st'>"UMAP_corrected"</span>,
  colour_by <span class='op'>=</span> <span class='st'>"slingPseudotime_1"</span>,
  point_alpha <span class='op'>=</span> <span class='fl'>1</span>,
  point_size <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>geom_path</span><span class='op'>(</span>data <span class='op'>=</span> <span class='va'>embedded</span>, <span class='fu'>aes</span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>Dim.1</span>, y <span class='op'>=</span> <span class='va'>Dim.2</span><span class='op'>)</span>, size <span class='op'>=</span> <span class='fl'>1.2</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotReducedDim.html'>plotReducedDim</a></span><span class='op'>(</span>
    <span class='va'>sce</span>,
    dimred <span class='op'>=</span> <span class='st'>"UMAP_corrected"</span>,
    colour_by <span class='op'>=</span> <span class='st'>"stage"</span>,
    point_alpha <span class='op'>=</span> <span class='fl'>1</span>,
    point_size <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotReducedDim.html'>plotReducedDim</a></span><span class='op'>(</span>
    <span class='va'>sce</span>,
    dimred <span class='op'>=</span> <span class='st'>"UMAP_corrected"</span>,
    colour_by <span class='op'>=</span> <span class='st'>"tissue"</span>,
    point_alpha <span class='op'>=</span> <span class='fl'>1</span>,
    point_size <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotReducedDim.html'>plotReducedDim</a></span><span class='op'>(</span>
    <span class='va'>sce</span>,
    dimred <span class='op'>=</span> <span class='st'>"UMAP_corrected"</span>,
    colour_by <span class='op'>=</span> <span class='st'>"cluster"</span>,
    point_alpha <span class='op'>=</span> <span class='fl'>1</span>,
    point_size <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://patchwork.data-imaginist.com/reference/plot_layout.html'>plot_layout</a></span><span class='op'>(</span>ncol <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="C094_Pellicci.single-cell.trajectory_files/figure-html5/unnamed-chunk-5-1.png" width="624" /></p>
</div>
<h1 id="changes-along-a-trajectory">Changes along a trajectory</h1>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>keep</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colSums.html'>rowSums</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/SingleCellExperiment/man/assays.html'>counts</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>&gt;</span> <span class='fl'>0</span><span class='op'>)</span> <span class='op'>&gt;=</span> <span class='fl'>10</span>
<span class='va'>sce_pseudo</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>[</span><span class='va'>keep</span>, <span class='op'>]</span>
</code></pre>
</div>
</details>
</div>
<p>We only test genes that are detected in at least 10 cells (<span class="math inline">\(n\)</span> = 14,256).</p>
<p>We use methods from the <em><a href="https://bioconductor.org/packages/3.12/tradeSeq">tradeSeq</a></em> R/Bioconductor package.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://statomics.github.io/tradeSeq/index.html'>tradeSeq</a></span><span class='op'>)</span>

<span class='va'>nonna.pseudo</span> <span class='op'>&lt;-</span> <span class='va'>sce_pseudo</span><span class='op'>$</span><span class='va'>slingPseudotime_1</span>
<span class='va'>nonna.pseudo</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/NA.html'>is.na</a></span><span class='op'>(</span><span class='va'>nonna.pseudo</span><span class='op'>)</span><span class='op'>]</span> <span class='op'>&lt;-</span> <span class='fl'>0</span>
<span class='va'>cell.weights</span> <span class='op'>&lt;-</span> <span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/NA.html'>is.na</a></span><span class='op'>(</span><span class='va'>sce_pseudo</span><span class='op'>$</span><span class='va'>slingPseudotime_1</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/mode.html'>storage.mode</a></span><span class='op'>(</span><span class='va'>cell.weights</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='st'>"numeric"</span>
</code></pre>
</div>
</details>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>666</span><span class='op'>)</span>
<span class='co'># NOTE: Takes ~30 mins and can crash due to memory constraints.</span>
<span class='va'>fit</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/tradeSeq/man/fitGAM.html'>fitGAM</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/pkg/SingleCellExperiment/man/assays.html'>counts</a></span><span class='op'>(</span><span class='va'>sce_pseudo</span><span class='op'>)</span>,
  pseudotime <span class='op'>=</span> <span class='va'>nonna.pseudo</span>,
  cellWeights <span class='op'>=</span> <span class='va'>cell.weights</span>,
  parallel <span class='op'>=</span> <span class='cn'>TRUE</span>,
  BPPARAM <span class='op'>=</span> <span class='fu'>BiocParallel</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html'>MulticoreParam</a></span><span class='op'>(</span><span class='fl'>4</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span><span class='va'>fit</span>, <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>outdir</span>, <span class='st'>"tradeSeq_fit.rds"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>fit</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>outdir</span>, <span class='st'>"tradeSeq_fit.rds"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>We perform a statistical test to check for DE between the starting point and the end stage of the trajectory.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>svet</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/tradeSeq/man/startVsEndTest.html'>startVsEndTest</a></span><span class='op'>(</span><span class='va'>fit</span><span class='op'>)</span>
<span class='va'>svet</span> <span class='op'>&lt;-</span> <span class='va'>svet</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span><span class='va'>svet</span><span class='op'>$</span><span class='va'>pvalue</span><span class='op'>)</span>, <span class='op'>]</span>
<span class='va'>svet</span><span class='op'>$</span><span class='va'>FDR</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/p.adjust.html'>p.adjust</a></span><span class='op'>(</span><span class='va'>svet</span><span class='op'>$</span><span class='va'>pvalue</span>, method <span class='op'>=</span> <span class='st'>"BH"</span><span class='op'>)</span>
<span class='va'>svet</span><span class='op'>$</span><span class='va'>ENTREZID</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span>
  <span class='fu'>rowData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>svet</span><span class='op'>)</span>, <span class='op'>]</span><span class='op'>)</span><span class='op'>$</span><span class='va'>NCBI.ENTREZID</span>,
  <span class='st'>"[["</span>,
  <span class='fl'>1</span><span class='op'>)</span>
<span class='va'>svet</span><span class='op'>$</span><span class='va'>logFC</span> <span class='op'>&lt;-</span> <span class='va'>svet</span><span class='op'>$</span><span class='va'>logFClineage1</span>

<span class='fu'><a href='https://rdrr.io/r/utils/write.table.html'>write.csv</a></span><span class='op'>(</span>
  <span class='va'>svet</span><span class='op'>[</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"logFC"</span>, <span class='st'>"pvalue"</span>, <span class='st'>"FDR"</span><span class='op'>)</span><span class='op'>]</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>outdir</span>, <span class='st'>"tradeSeq_results.DEGs.csv"</span><span class='op'>)</span>,
  row.names <span class='op'>=</span> <span class='cn'>TRUE</span>,
  quote <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>

<span class='va'>pseudotime_degs</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
  up_at_start <span class='op'>=</span>
    <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span>
      <span class='fu'>rowData</span><span class='op'>(</span>
        <span class='va'>sce</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>svet</span><span class='op'>[</span><span class='va'>svet</span><span class='op'>$</span><span class='va'>FDR</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span> <span class='op'>&amp;</span> <span class='va'>svet</span><span class='op'>$</span><span class='va'>logFC</span> <span class='op'>&lt;</span> <span class='fl'>0</span>, <span class='op'>]</span><span class='op'>)</span>, <span class='op'>]</span><span class='op'>)</span><span class='op'>$</span><span class='va'>NCBI.ENTREZID</span>,
      <span class='st'>"[["</span>,
      <span class='fl'>1</span><span class='op'>)</span>,
  up_at_end <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span>
    <span class='fu'>rowData</span><span class='op'>(</span>
      <span class='va'>sce</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>svet</span><span class='op'>[</span><span class='va'>svet</span><span class='op'>$</span><span class='va'>FDR</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span> <span class='op'>&amp;</span> <span class='va'>svet</span><span class='op'>$</span><span class='va'>logFC</span> <span class='op'>&gt;</span> <span class='fl'>0</span>, <span class='op'>]</span><span class='op'>)</span>, <span class='op'>]</span><span class='op'>)</span><span class='op'>$</span><span class='va'>NCBI.ENTREZID</span>,
    <span class='st'>"[["</span>,
    <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>There are 653 and 164 genes that are significantly upregulated (<span class="math inline">\(FDR &lt; 0.05\)</span>) at the start and end of the trajectory, respectively. CSV files of the results are available from <a href="../output/trajectory/"><code>output/trajectory/</code></a>. Figures <a href="#fig:up-at-start">1</a> and <a href="#fig:up-at-end">2</a> highlight some of these genes.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotHeatmap.html'>plotHeatmap</a></span><span class='op'>(</span>
  <span class='va'>sce</span>,
  order_columns_by <span class='op'>=</span> <span class='st'>"slingPseudotime_1"</span>,
  colour_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"stage"</span>, <span class='st'>"tissue"</span><span class='op'>)</span>,
  features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://rdrr.io/r/base/sets.html'>setdiff</a></span><span class='op'>(</span>
      <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>svet</span><span class='op'>[</span><span class='va'>svet</span><span class='op'>$</span><span class='va'>FDR</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span> <span class='op'>&amp;</span> <span class='va'>svet</span><span class='op'>$</span><span class='va'>logFC</span> <span class='op'>&lt;</span> <span class='fl'>0</span>, <span class='op'>]</span><span class='op'>)</span>, 
      <span class='va'>mito_set</span><span class='op'>)</span>,
    <span class='fl'>50</span><span class='op'>)</span>,
  center <span class='op'>=</span> <span class='cn'>TRUE</span>,
  symmetric <span class='op'>=</span> <span class='cn'>TRUE</span>,
  zlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>3</span>, <span class='fl'>3</span><span class='op'>)</span>,
  color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>hcl.colors</a></span><span class='op'>(</span><span class='fl'>101</span>, <span class='st'>"Blue-Red 3"</span><span class='op'>)</span>,
  fontsize <span class='op'>=</span> <span class='fl'>6</span>,
  main <span class='op'>=</span> <span class='st'>"Up at start"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:up-at-start"></span>
<img src="C094_Pellicci.single-cell.trajectory_files/figure-html5/up-at-start-1.png" alt="Selected genes that are upregulated at the start of the inferred trajectory." width="624" />
<p class="caption">
Figure 1: Selected genes that are upregulated at the start of the inferred trajectory.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotHeatmap.html'>plotHeatmap</a></span><span class='op'>(</span>
  <span class='va'>sce</span>,
  order_columns_by <span class='op'>=</span> <span class='st'>"slingPseudotime_1"</span>,
  colour_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"stage"</span>, <span class='st'>"tissue"</span><span class='op'>)</span>,
  features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://rdrr.io/r/base/sets.html'>setdiff</a></span><span class='op'>(</span>
      <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>svet</span><span class='op'>[</span><span class='va'>svet</span><span class='op'>$</span><span class='va'>FDR</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span> <span class='op'>&amp;</span> <span class='va'>svet</span><span class='op'>$</span><span class='va'>logFC</span> <span class='op'>&gt;</span> <span class='fl'>0</span>, <span class='op'>]</span><span class='op'>)</span>, 
      <span class='va'>mito_set</span><span class='op'>)</span>,
    <span class='fl'>50</span><span class='op'>)</span>,
  center <span class='op'>=</span> <span class='cn'>TRUE</span>,
  symmetric <span class='op'>=</span> <span class='cn'>TRUE</span>,
  zlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>3</span>, <span class='fl'>3</span><span class='op'>)</span>,
  color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>hcl.colors</a></span><span class='op'>(</span><span class='fl'>101</span>, <span class='st'>"Blue-Red 3"</span><span class='op'>)</span>,
  fontsize <span class='op'>=</span> <span class='fl'>6</span>,
  main <span class='op'>=</span> <span class='st'>"Up at end"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:up-at-end"></span>
<img src="C094_Pellicci.single-cell.trajectory_files/figure-html5/up-at-end-1.png" alt="Selected genes that are upregulated at the end of the inferred trajectory." width="624" />
<p class="caption">
Figure 2: Selected genes that are upregulated at the end of the inferred trajectory.
</p>
</div>
</div>
<h1 id="gene-set-testing">Gene set testing</h1>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://bioinf.wehi.edu.au/limma'>limma</a></span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>We opt to use the <em><a href="https://bioconductor.org/packages/3.12/tradeSeq">tradeSeq</a></em> results and seek to identify gene sets that are differentially expressed with respect to the trajectory.</p>
<h2 id="goana">goana</h2>
<p>We use the <code>goana()</code> function from the <em><a href="https://bioconductor.org/packages/3.12/limma">limma</a></em> R/Bioconductor package to test for over-representation of gene ontology (GO) terms in each of the <code>up_at_start</code> and <code>up_at_end</code> DEG lists.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>go</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/limma/man/goana.html'>goana</a></span><span class='op'>(</span>de <span class='op'>=</span> <span class='va'>pseudotime_degs</span>, species <span class='op'>=</span> <span class='st'>"Hs"</span><span class='op'>)</span>
<span class='va'>gzout</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/connections.html'>gzfile</a></span><span class='op'>(</span>
  description <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>outdir</span>, <span class='st'>"tradeSeq_results.goana.csv.gz"</span><span class='op'>)</span>,
  open <span class='op'>=</span> <span class='st'>"wb"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/write.table.html'>write.csv</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/pkg/limma/man/topGO.html'>topGO</a></span><span class='op'>(</span><span class='va'>go</span>, number <span class='op'>=</span> <span class='cn'>Inf</span><span class='op'>)</span>,
  <span class='va'>gzout</span>,
  row.names <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/connections.html'>close</a></span><span class='op'>(</span><span class='va'>gzout</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>CSV files of the results are available from <a href="../output/trajectory/"><code>output/trajectory/</code></a>. An example of the results are shown below, here ordered by <code>P.up_at_end</code> to highlight GO terms that are enriched in the <code>up_at_end</code> DEG list.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/limma/man/topGO.html'>topGO</a></span><span class='op'>(</span><span class='va'>go</span>, sort <span class='op'>=</span> <span class='st'>"up_at_end"</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>knitr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/knitr/man/kable.html'>kable</a></span><span class='op'>(</span>
    caption <span class='op'>=</span> <span class='st'>'`goana()` produces a table with a row for each GO term and the following columns: **Term** (GO term); **Ont** (ontology that the GO term belongs to. Possible values are "BP", "CC" and "MF"); **N** (number of genes in the GO term); **up_at_start** (number of `up_at_start`-regulated differentially expressed genes); **up_at_end** (number of `up_at_end`-regulated differentially expressed genes); **P.up_at_start** (p-value for over-representation of GO term in `up_at_start`-regulated genes); **P.up_at_end** (p-value for over-representation of GO term in `up_at_end`-regulated genes)'</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<table style="width:100%;">
<caption><span id="tab:unnamed-chunk-13">Table 1: </span><code>goana()</code> produces a table with a row for each GO term and the following columns: <strong>Term</strong> (GO term); <strong>Ont</strong> (ontology that the GO term belongs to. Possible values are BP, CC and MF); <strong>N</strong> (number of genes in the GO term); <strong>up_at_start</strong> (number of <code>up_at_start</code>-regulated differentially expressed genes); <strong>up_at_end</strong> (number of <code>up_at_end</code>-regulated differentially expressed genes); <strong>P.up_at_start</strong> (p-value for over-representation of GO term in <code>up_at_start</code>-regulated genes); <strong>P.up_at_end</strong> (p-value for over-representation of GO term in <code>up_at_end</code>-regulated genes)</caption>
<colgroup>
<col style="width: 7%" />
<col style="width: 54%" />
<col style="width: 2%" />
<col style="width: 3%" />
<col style="width: 8%" />
<col style="width: 6%" />
<col style="width: 9%" />
<col style="width: 8%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">Term</th>
<th style="text-align: left;">Ont</th>
<th style="text-align: right;">N</th>
<th style="text-align: right;">up_at_start</th>
<th style="text-align: right;">up_at_end</th>
<th style="text-align: right;">P.up_at_start</th>
<th style="text-align: right;">P.up_at_end</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><a href="GO:0002376" class="uri">GO:0002376</a></td>
<td style="text-align: left;">immune system process</td>
<td style="text-align: left;">BP</td>
<td style="text-align: right;">3287</td>
<td style="text-align: right;">77</td>
<td style="text-align: right;">61</td>
<td style="text-align: right;">0.2472586</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;"><a href="GO:0002682" class="uri">GO:0002682</a></td>
<td style="text-align: left;">regulation of immune system process</td>
<td style="text-align: left;">BP</td>
<td style="text-align: right;">1646</td>
<td style="text-align: right;">44</td>
<td style="text-align: right;">42</td>
<td style="text-align: right;">0.0878162</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><a href="GO:0048583" class="uri">GO:0048583</a></td>
<td style="text-align: left;">regulation of response to stimulus</td>
<td style="text-align: left;">BP</td>
<td style="text-align: right;">4403</td>
<td style="text-align: right;">105</td>
<td style="text-align: right;">71</td>
<td style="text-align: right;">0.1483313</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;"><a href="GO:0050896" class="uri">GO:0050896</a></td>
<td style="text-align: left;">response to stimulus</td>
<td style="text-align: left;">BP</td>
<td style="text-align: right;">9398</td>
<td style="text-align: right;">211</td>
<td style="text-align: right;">109</td>
<td style="text-align: right;">0.2641959</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><a href="GO:0050776" class="uri">GO:0050776</a></td>
<td style="text-align: left;">regulation of immune response</td>
<td style="text-align: left;">BP</td>
<td style="text-align: right;">1027</td>
<td style="text-align: right;">24</td>
<td style="text-align: right;">32</td>
<td style="text-align: right;">0.3835778</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;"><a href="GO:0006955" class="uri">GO:0006955</a></td>
<td style="text-align: left;">immune response</td>
<td style="text-align: left;">BP</td>
<td style="text-align: right;">2306</td>
<td style="text-align: right;">43</td>
<td style="text-align: right;">48</td>
<td style="text-align: right;">0.8748847</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><a href="GO:0051716" class="uri">GO:0051716</a></td>
<td style="text-align: left;">cellular response to stimulus</td>
<td style="text-align: left;">BP</td>
<td style="text-align: right;">7748</td>
<td style="text-align: right;">174</td>
<td style="text-align: right;">96</td>
<td style="text-align: right;">0.2971464</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;"><a href="GO:0042110" class="uri">GO:0042110</a></td>
<td style="text-align: left;">T cell activation</td>
<td style="text-align: left;">BP</td>
<td style="text-align: right;">483</td>
<td style="text-align: right;">17</td>
<td style="text-align: right;">22</td>
<td style="text-align: right;">0.0355550</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><a href="GO:0046649" class="uri">GO:0046649</a></td>
<td style="text-align: left;">lymphocyte activation</td>
<td style="text-align: left;">BP</td>
<td style="text-align: right;">765</td>
<td style="text-align: right;">24</td>
<td style="text-align: right;">27</td>
<td style="text-align: right;">0.0462375</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;"><a href="GO:0002684" class="uri">GO:0002684</a></td>
<td style="text-align: left;">positive regulation of immune system process</td>
<td style="text-align: left;">BP</td>
<td style="text-align: right;">1110</td>
<td style="text-align: right;">31</td>
<td style="text-align: right;">32</td>
<td style="text-align: right;">0.0906153</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><a href="GO:0002483" class="uri">GO:0002483</a></td>
<td style="text-align: left;">antigen processing and presentation of endogenous peptide antigen</td>
<td style="text-align: left;">BP</td>
<td style="text-align: right;">17</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;"><a href="GO:0019885" class="uri">GO:0019885</a></td>
<td style="text-align: left;">antigen processing and presentation of endogenous peptide antigen via MHC class I</td>
<td style="text-align: left;">BP</td>
<td style="text-align: right;">17</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><a href="GO:0048584" class="uri">GO:0048584</a></td>
<td style="text-align: left;">positive regulation of response to stimulus</td>
<td style="text-align: left;">BP</td>
<td style="text-align: right;">2465</td>
<td style="text-align: right;">62</td>
<td style="text-align: right;">48</td>
<td style="text-align: right;">0.1202028</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;"><a href="GO:0045321" class="uri">GO:0045321</a></td>
<td style="text-align: left;">leukocyte activation</td>
<td style="text-align: left;">BP</td>
<td style="text-align: right;">1321</td>
<td style="text-align: right;">39</td>
<td style="text-align: right;">34</td>
<td style="text-align: right;">0.0316194</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><a href="GO:0001775" class="uri">GO:0001775</a></td>
<td style="text-align: left;">cell activation</td>
<td style="text-align: left;">BP</td>
<td style="text-align: right;">1482</td>
<td style="text-align: right;">41</td>
<td style="text-align: right;">35</td>
<td style="text-align: right;">0.0650995</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;"><a href="GO:0019883" class="uri">GO:0019883</a></td>
<td style="text-align: left;">antigen processing and presentation of endogenous antigen</td>
<td style="text-align: left;">BP</td>
<td style="text-align: right;">24</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">0.0948227</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><a href="GO:0007165" class="uri">GO:0007165</a></td>
<td style="text-align: left;">signal transduction</td>
<td style="text-align: left;">BP</td>
<td style="text-align: right;">6272</td>
<td style="text-align: right;">130</td>
<td style="text-align: right;">80</td>
<td style="text-align: right;">0.7521839</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;"><a href="GO:0002520" class="uri">GO:0002520</a></td>
<td style="text-align: left;">immune system development</td>
<td style="text-align: left;">BP</td>
<td style="text-align: right;">1024</td>
<td style="text-align: right;">31</td>
<td style="text-align: right;">28</td>
<td style="text-align: right;">0.0391822</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><a href="GO:0046651" class="uri">GO:0046651</a></td>
<td style="text-align: left;">lymphocyte proliferation</td>
<td style="text-align: left;">BP</td>
<td style="text-align: right;">283</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">15</td>
<td style="text-align: right;">0.0902015</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;"><a href="GO:0032943" class="uri">GO:0032943</a></td>
<td style="text-align: left;">mononuclear cell proliferation</td>
<td style="text-align: left;">BP</td>
<td style="text-align: right;">286</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">15</td>
<td style="text-align: right;">0.0950264</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
</div>
<h2 id="kegga">kegga</h2>
<p>We use the <code>kegga()</code> function from the <em><a href="https://bioconductor.org/packages/3.12/limma">limma</a></em> R/Bioconductor package to test for over-representation of KEGG pathways in each of the <code>up_at_start</code> and <code>up_at_end</code> DEG lists.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>kegg</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/limma/man/goana.html'>kegga</a></span><span class='op'>(</span>de <span class='op'>=</span> <span class='va'>pseudotime_degs</span>, species <span class='op'>=</span> <span class='st'>"Hs"</span><span class='op'>)</span>
<span class='va'>gzout</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/connections.html'>gzfile</a></span><span class='op'>(</span>
  description <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>outdir</span>, <span class='st'>"tradeSeq_results.kegga.csv.gz"</span><span class='op'>)</span>,
  open <span class='op'>=</span> <span class='st'>"wb"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/write.table.html'>write.csv</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/pkg/limma/man/topGO.html'>topKEGG</a></span><span class='op'>(</span><span class='va'>kegg</span>, number <span class='op'>=</span> <span class='cn'>Inf</span><span class='op'>)</span>,
  <span class='va'>gzout</span>,
  row.names <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/connections.html'>close</a></span><span class='op'>(</span><span class='va'>gzout</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>CSV files of the results are available from <a href="../output/trajectory/"><code>output/trajectory/</code></a>. An example of the results are shown below, here ordered by <code>P.up_at_end</code> to highlight GO terms that are enriched in the <code>up_at_end</code> DEG list.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/limma/man/topGO.html'>topKEGG</a></span><span class='op'>(</span><span class='va'>kegg</span>, sort <span class='op'>=</span> <span class='st'>"up_at_end"</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>knitr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/knitr/man/kable.html'>kable</a></span><span class='op'>(</span>
    caption <span class='op'>=</span> <span class='st'>'`kegga()` produces a table with a row for each KEGG pathway ID and the following columns: **Pathway** (KEGG pathway); **N** (number of genes in the GO term);; **up_at_start** (number of `up_at_start`-regulated differentially expressed genes); **up_at_end** (number of `up_at_end`-regulated differentially expressed genes); **P.up_at_start** (p-value for over-representation of KEGG pathway in `up_at_start`-regulated genes); **P.up_at_end** (p-value for over-representation of KEGG pathway in `up_at_end`-regulated genes)'</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<table style="width:100%;">
<caption><span id="tab:unnamed-chunk-15">Table 2: </span><code>kegga()</code> produces a table with a row for each KEGG pathway ID and the following columns: <strong>Pathway</strong> (KEGG pathway); <strong>N</strong> (number of genes in the GO term);; <strong>up_at_start</strong> (number of <code>up_at_start</code>-regulated differentially expressed genes); <strong>up_at_end</strong> (number of <code>up_at_end</code>-regulated differentially expressed genes); <strong>P.up_at_start</strong> (p-value for over-representation of KEGG pathway in <code>up_at_start</code>-regulated genes); <strong>P.up_at_end</strong> (p-value for over-representation of KEGG pathway in <code>up_at_end</code>-regulated genes)</caption>
<colgroup>
<col style="width: 12%" />
<col style="width: 42%" />
<col style="width: 3%" />
<col style="width: 10%" />
<col style="width: 8%" />
<col style="width: 12%" />
<col style="width: 10%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">Pathway</th>
<th style="text-align: right;">N</th>
<th style="text-align: right;">up_at_start</th>
<th style="text-align: right;">up_at_end</th>
<th style="text-align: right;">P.up_at_start</th>
<th style="text-align: right;">P.up_at_end</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">path:hsa04612</td>
<td style="text-align: left;">Antigen processing and presentation</td>
<td style="text-align: right;">78</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">0.6381545</td>
<td style="text-align: right;">0.0000000</td>
</tr>
<tr class="even">
<td style="text-align: left;">path:hsa05170</td>
<td style="text-align: left;">Human immunodeficiency virus 1 infection</td>
<td style="text-align: right;">212</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">12</td>
<td style="text-align: right;">0.7000293</td>
<td style="text-align: right;">0.0000001</td>
</tr>
<tr class="odd">
<td style="text-align: left;">path:hsa05169</td>
<td style="text-align: left;">Epstein-Barr virus infection</td>
<td style="text-align: right;">202</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">0.4846459</td>
<td style="text-align: right;">0.0000005</td>
</tr>
<tr class="even">
<td style="text-align: left;">path:hsa05163</td>
<td style="text-align: left;">Human cytomegalovirus infection</td>
<td style="text-align: right;">225</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">0.9507946</td>
<td style="text-align: right;">0.0000014</td>
</tr>
<tr class="odd">
<td style="text-align: left;">path:hsa04514</td>
<td style="text-align: left;">Cell adhesion molecules</td>
<td style="text-align: right;">149</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">0.9206000</td>
<td style="text-align: right;">0.0000025</td>
</tr>
<tr class="even">
<td style="text-align: left;">path:hsa05332</td>
<td style="text-align: left;">Graft-versus-host disease</td>
<td style="text-align: right;">42</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.0000193</td>
</tr>
<tr class="odd">
<td style="text-align: left;">path:hsa05416</td>
<td style="text-align: left;">Viral myocarditis</td>
<td style="text-align: right;">60</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.0001110</td>
</tr>
<tr class="even">
<td style="text-align: left;">path:hsa04145</td>
<td style="text-align: left;">Phagosome</td>
<td style="text-align: right;">152</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">0.2415877</td>
<td style="text-align: right;">0.0002004</td>
</tr>
<tr class="odd">
<td style="text-align: left;">path:hsa05330</td>
<td style="text-align: left;">Allograft rejection</td>
<td style="text-align: right;">38</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.0002277</td>
</tr>
<tr class="even">
<td style="text-align: left;">path:hsa04940</td>
<td style="text-align: left;">Type I diabetes mellitus</td>
<td style="text-align: right;">43</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.0003694</td>
</tr>
<tr class="odd">
<td style="text-align: left;">path:hsa05166</td>
<td style="text-align: left;">Human T-cell leukemia virus 1 infection</td>
<td style="text-align: right;">222</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">8</td>
<td style="text-align: right;">0.8660481</td>
<td style="text-align: right;">0.0003704</td>
</tr>
<tr class="even">
<td style="text-align: left;">path:hsa04650</td>
<td style="text-align: left;">Natural killer cell mediated cytotoxicity</td>
<td style="text-align: right;">131</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">0.9750164</td>
<td style="text-align: right;">0.0006052</td>
</tr>
<tr class="odd">
<td style="text-align: left;">path:hsa05320</td>
<td style="text-align: left;">Autoimmune thyroid disease</td>
<td style="text-align: right;">53</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.0008253</td>
</tr>
<tr class="even">
<td style="text-align: left;">path:hsa05167</td>
<td style="text-align: left;">Kaposi sarcoma-associated herpesvirus infection</td>
<td style="text-align: right;">194</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">0.7874172</td>
<td style="text-align: right;">0.0008754</td>
</tr>
<tr class="odd">
<td style="text-align: left;">path:hsa04218</td>
<td style="text-align: left;">Cellular senescence</td>
<td style="text-align: right;">156</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">0.9324198</td>
<td style="text-align: right;">0.0015047</td>
</tr>
<tr class="even">
<td style="text-align: left;">path:hsa05168</td>
<td style="text-align: left;">Herpes simplex virus 1 infection</td>
<td style="text-align: right;">497</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">0.9366117</td>
<td style="text-align: right;">0.0018180</td>
</tr>
<tr class="odd">
<td style="text-align: left;">path:hsa04668</td>
<td style="text-align: left;">TNF signaling pathway</td>
<td style="text-align: right;">112</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0.9571741</td>
<td style="text-align: right;">0.0019787</td>
</tr>
<tr class="even">
<td style="text-align: left;">path:hsa05140</td>
<td style="text-align: left;">Leishmaniasis</td>
<td style="text-align: right;">77</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.0033036</td>
</tr>
<tr class="odd">
<td style="text-align: left;">path:hsa04144</td>
<td style="text-align: left;">Endocytosis</td>
<td style="text-align: right;">252</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">0.8297867</td>
<td style="text-align: right;">0.0038920</td>
</tr>
<tr class="even">
<td style="text-align: left;">path:hsa04210</td>
<td style="text-align: left;">Apoptosis</td>
<td style="text-align: right;">136</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0.1724547</td>
<td style="text-align: right;">0.0045822</td>
</tr>
</tbody>
</table>
</div>
<h2 id="camera">camera</h2>
<p>We use the <code>cameraPR()</code> function<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> from the <em><a href="https://bioconductor.org/packages/3.12/limma">limma</a></em> R/Bioconductor package to test whether a set of genes is highly ranked relative to other genes in terms of differential expression, accounting for inter-gene correlation. Specifically, we test using gene sets from <a href="https://www.gsea-msigdb.org/gsea/msigdb">MSigDB</a>, namely:</p>
<aside>
We download these gene sets from <a href="http://bioinf.wehi.edu.au/MSigDB/index.html">http://bioinf.wehi.edu.au/MSigDB/index.html</a>
</aside>
<ul>
<li><strong>H</strong>: <strong>hallmark gene sets</strong> are coherently expressed signatures derived by aggregating many MSigDB gene sets to represent well-defined biological states or processes.</li>
<li><strong>C2</strong>: <strong>curated gene sets</strong> from online pathway databases, publications in PubMed, and knowledge of domain experts.</li>
<li><strong>C7</strong>: <strong>immunologic signature gene sets</strong> represent cell states and perturbations within the immune system.</li>
</ul>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># NOTE: Using BiocFileCache to avoid re-downloading these gene sets everytime</span>
<span class='co'>#       the report is rendered.</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>BiocFileCache</span><span class='op'>)</span>
<span class='va'>bfc</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html'>BiocFileCache</a></span><span class='op'>(</span><span class='op'>)</span>
<span class='co'># NOTE: Creating list of gene sets in this slightly convoluted way so that each</span>
<span class='co'>#       gene set name is prepended by its origin (e.g. H, C2, or C7).</span>
<span class='va'>msigdb</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/do.call.html'>do.call</a></span><span class='op'>(</span>
  <span class='va'>c</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
    H <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span>
      <span class='fu'><a href='https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html'>bfcrpath</a></span><span class='op'>(</span>
        <span class='va'>bfc</span>,
        <span class='st'>"http://bioinf.wehi.edu.au/MSigDB/v7.1/Hs.h.all.v7.1.entrez.rds"</span><span class='op'>)</span><span class='op'>)</span>,
    C2 <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span>
      <span class='fu'><a href='https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html'>bfcrpath</a></span><span class='op'>(</span>
        <span class='va'>bfc</span>,
        <span class='st'>"http://bioinf.wehi.edu.au/MSigDB/v7.1/Hs.c2.all.v7.1.entrez.rds"</span><span class='op'>)</span><span class='op'>)</span>,
    C7 <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span>
      <span class='fu'><a href='https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html'>bfcrpath</a></span><span class='op'>(</span>
        <span class='va'>bfc</span>,
        <span class='st'>"http://bioinf.wehi.edu.au/MSigDB/v7.1/Hs.c7.all.v7.1.entrez.rds"</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># NOTE: Using the signed Wald statistic as the statistic, following by</span>
<span class='co'>#       analogy what Gordon recommends</span>
<span class='co'>#       (https://support.bioconductor.org/p/112937/#112938).</span>
<span class='va'>svet</span><span class='op'>$</span><span class='va'>statistic</span> <span class='op'>&lt;-</span> <span class='va'>svet</span><span class='op'>$</span><span class='va'>waldStat</span> <span class='op'>*</span> <span class='fu'><a href='https://rdrr.io/r/base/sign.html'>sign</a></span><span class='op'>(</span><span class='va'>svet</span><span class='op'>$</span><span class='va'>logFClineage1</span><span class='op'>)</span>
<span class='va'>y_idx</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/limma/man/ids2indices.html'>ids2indices</a></span><span class='op'>(</span><span class='va'>msigdb</span>, id <span class='op'>=</span> <span class='va'>svet</span><span class='op'>$</span><span class='va'>ENTREZID</span><span class='op'>)</span>
<span class='va'>cam</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/limma/man/camera.html'>cameraPR</a></span><span class='op'>(</span>
  statistic <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/setNames.html'>setNames</a></span><span class='op'>(</span><span class='va'>svet</span><span class='op'>$</span><span class='va'>statistic</span>, <span class='va'>svet</span><span class='op'>$</span><span class='va'>ENTREZID</span><span class='op'>)</span>,
  index <span class='op'>=</span> <span class='va'>y_idx</span>,
  use.ranks <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>

<span class='va'>gzout</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/connections.html'>gzfile</a></span><span class='op'>(</span>
  description <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>outdir</span>, <span class='st'>"tradeSeq_results.camera.csv.gz"</span><span class='op'>)</span>,
  open <span class='op'>=</span> <span class='st'>"wb"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/write.table.html'>write.csv</a></span><span class='op'>(</span>
  <span class='va'>cam</span>,
  <span class='va'>gzout</span>,
  row.names <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/connections.html'>close</a></span><span class='op'>(</span><span class='va'>gzout</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>CSV files of the results are available from <a href="../output/trajectory/"><code>output/trajectory/</code></a>. An example of the results are shown below, here filtered to highlight those gene sets with <code>Direction = Up</code> meaning <code>up_at_end</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>cam</span><span class='op'>[</span><span class='va'>cam</span><span class='op'>$</span><span class='va'>Direction</span> <span class='op'>==</span> <span class='st'>"Up"</span>, <span class='op'>]</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>knitr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/knitr/man/kable.html'>kable</a></span><span class='op'>(</span>
    caption <span class='op'>=</span> <span class='st'>'`camera()` produces a table with a row for each gene set (prepended by which MSigDB collection it comes from) and the following columns: **NGenes** (number of genes in set); **Direction** (direction of change, `Up` = `up_at_end` and `Down` = `up_at_start`); **PValue** (two-tailed p-value); **FDR** (Benjamini and Hochberg FDR adjusted p-value).'</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<table style="width:100%;">
<caption><span id="tab:unnamed-chunk-17">Table 3: </span><code>camera()</code> produces a table with a row for each gene set (prepended by which MSigDB collection it comes from) and the following columns: <strong>NGenes</strong> (number of genes in set); <strong>Direction</strong> (direction of change, <code>Up</code> = <code>up_at_end</code> and <code>Down</code> = <code>up_at_start</code>); <strong>PValue</strong> (two-tailed p-value); <strong>FDR</strong> (Benjamini and Hochberg FDR adjusted p-value).</caption>
<colgroup>
<col style="width: 75%" />
<col style="width: 6%" />
<col style="width: 8%" />
<col style="width: 6%" />
<col style="width: 3%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">NGenes</th>
<th style="text-align: left;">Direction</th>
<th style="text-align: right;">PValue</th>
<th style="text-align: right;">FDR</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">C2.WUNDER_INFLAMMATORY_RESPONSE_AND_CHOLESTEROL_DN</td>
<td style="text-align: right;">3</td>
<td style="text-align: left;">Up</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">C2.REACTOME_ENDOSOMAL_VACUOLAR_PATHWAY</td>
<td style="text-align: right;">9</td>
<td style="text-align: left;">Up</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">C2.KUROKAWA_LIVER_CANCER_EARLY_RECURRENCE_DN</td>
<td style="text-align: right;">6</td>
<td style="text-align: left;">Up</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">C2.BIOCARTA_MHC_PATHWAY</td>
<td style="text-align: right;">12</td>
<td style="text-align: left;">Up</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">C2.REACTOME_NEF_MEDIATED_DOWNREGULATION_OF_MHC_CLASS_I_COMPLEX_CELL_SURFACE_EXPRESSION</td>
<td style="text-align: right;">8</td>
<td style="text-align: left;">Up</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">C2.BIOCARTA_CTL_PATHWAY</td>
<td style="text-align: right;">12</td>
<td style="text-align: left;">Up</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
</div>
<h2 id="visualisation">Visualisation</h2>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>term</span> <span class='op'>&lt;-</span> <span class='st'>"C2.REACTOME_INTERFERON_ALPHA_BETA_SIGNALING"</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>org.Hs.eg.db</span><span class='op'>)</span>
<span class='va'>key</span> <span class='op'>&lt;-</span> <span class='va'>msigdb</span><span class='op'>[[</span><span class='va'>term</span><span class='op'>]</span><span class='op'>]</span>
<span class='va'>tmp</span> <span class='op'>&lt;-</span> <span class='fu'>AnnotationDbi</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html'>select</a></span><span class='op'>(</span>
  <span class='va'>org.Hs.eg.db</span>,
  key <span class='op'>=</span> <span class='va'>key</span>,
  columns <span class='op'>=</span> <span class='st'>"SYMBOL"</span>,
  keytype <span class='op'>=</span> <span class='st'>"ENTREZID"</span><span class='op'>)</span>
<span class='va'>y_index</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/limma/man/ids2indices.html'>ids2indices</a></span><span class='op'>(</span><span class='va'>msigdb</span><span class='op'>[[</span><span class='va'>term</span><span class='op'>]</span><span class='op'>]</span>, <span class='va'>svet</span><span class='op'>$</span><span class='va'>ENTREZID</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>We can visualise the expression of the genes in any given set using a barcode plot. Such figures are often included in publications.</p>
<p>Figure <a href="#fig:gsea-vis">3</a> gives an example showing the expression of genes in the <a href="https://www.gsea-msigdb.org/gsea/msigdb/cards/REACTOME_INTERFERON_ALPHA_BETA_SIGNALING">C2.REACTOME_INTERFERON_ALPHA_BETA_SIGNALING</a> MSigDB gene set.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/limma/man/barcodeplot.html'>barcodeplot</a></span><span class='op'>(</span>statistics <span class='op'>=</span> <span class='va'>svet</span><span class='op'>$</span><span class='va'>statistic</span>, index <span class='op'>=</span> <span class='va'>y_index</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<div class="figure"><span id="fig:gsea-vis"></span>
<img src="C094_Pellicci.single-cell.trajectory_files/figure-html5/gsea-vis-1.png" alt="Barcode plot of genes in MSigDB set `C2.REACTOME_INTERFERON_ALPHA_BETA_SIGNALING`. Genes are represented by bars and are ranked from left to right by increasing signed Wald statistic from the trajectory DE analysis. This forms the barcode-like pattern. The line (or *worm*) above the barcode shows the relative local enrichment of the vertical bars in each part of the plot. The dotted horizontal line indicates neutral enrichment; the worm above the dotted line shows enrichment while the worm below the dotted line shows depletion." width="624" />
<p class="caption">
Figure 3: Barcode plot of genes in MSigDB set <code>C2.REACTOME_INTERFERON_ALPHA_BETA_SIGNALING</code>. Genes are represented by bars and are ranked from left to right by increasing signed Wald statistic from the trajectory DE analysis. This forms the barcode-like pattern. The line (or <em>worm</em>) above the barcode shows the relative local enrichment of the vertical bars in each part of the plot. The dotted horizontal line indicates neutral enrichment; the worm above the dotted line shows enrichment while the worm below the dotted line shows depletion.
</p>
</div>
</div>
<h1 id="figures-for-paper">Figures for paper</h1>
<p>Note that exact figure numbers may change.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># Some useful colours</span>
<span class='va'>plate_number_colours</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/setNames.html'>setNames</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>plate_number_colours</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>plate_number_colours</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>plate_number_colours</span> <span class='op'>&lt;-</span> <span class='va'>plate_number_colours</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>plate_number</span><span class='op'>)</span><span class='op'>]</span>
<span class='va'>tissue_colours</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/setNames.html'>setNames</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>tissue_colours</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>tissue_colours</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>tissue_colours</span> <span class='op'>&lt;-</span> <span class='va'>tissue_colours</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>tissue</span><span class='op'>)</span><span class='op'>]</span>
<span class='va'>donor_colours</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/setNames.html'>setNames</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>donor_colours</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>donor_colours</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>donor_colours</span> <span class='op'>&lt;-</span> <span class='va'>donor_colours</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>donor</span><span class='op'>)</span><span class='op'>]</span>
<span class='va'>stage_colours</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/setNames.html'>setNames</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>stage_colours</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>stage_colours</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>stage_colours</span> <span class='op'>&lt;-</span> <span class='va'>stage_colours</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>stage</span><span class='op'>)</span><span class='op'>]</span>
<span class='va'>group_colours</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/setNames.html'>setNames</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>group_colours</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>group_colours</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>group_colours</span> <span class='op'>&lt;-</span> <span class='va'>group_colours</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>group</span><span class='op'>)</span><span class='op'>]</span>
<span class='va'>cluster_colours</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/setNames.html'>setNames</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>cluster_colours</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>colours</span><span class='op'>$</span><span class='va'>cluster_colours</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>cluster_colours</span> <span class='op'>&lt;-</span> <span class='va'>cluster_colours</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>cluster</span><span class='op'>)</span><span class='op'>]</span>
</code></pre>
</div>
</details>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>sce</span><span class='op'>$</span><span class='va'>CD4</span> <span class='op'>&lt;-</span> <span class='fu'>assay</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html'>altExp</a></span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"FACS"</span><span class='op'>)</span>, <span class='st'>"pseudolog"</span><span class='op'>)</span><span class='op'>[</span><span class='st'>"V525_50_A_CD4_BV510"</span>, <span class='op'>]</span>
<span class='va'>sce</span><span class='op'>$</span><span class='va'>CD161</span> <span class='op'>&lt;-</span> <span class='fu'>assay</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html'>altExp</a></span><span class='op'>(</span><span class='va'>sce</span>, <span class='st'>"FACS"</span><span class='op'>)</span>, <span class='st'>"pseudolog"</span><span class='op'>)</span><span class='op'>[</span><span class='st'>"B530_30_A_CD161_FITC"</span>, <span class='op'>]</span>
</code></pre>
</div>
</details>
</div>
<h2 id="figure-5">Figure 5</h2>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotReducedDim.html'>plotReducedDim</a></span><span class='op'>(</span>
  <span class='va'>sce</span>,
  <span class='st'>"UMAP_corrected"</span>,
  colour_by <span class='op'>=</span> <span class='st'>"stage"</span>,
  point_alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>scale_colour_manual</span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>stage_colours</span>, name <span class='op'>=</span> <span class='st'>"stage"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>ggtitle</span><span class='op'>(</span><span class='st'>"All cells"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="C094_Pellicci.single-cell.trajectory_files/figure-html5/unnamed-chunk-21-1.png" width="624" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotReducedDim.html'>plotReducedDim</a></span><span class='op'>(</span>
  <span class='va'>sce</span>,
  <span class='st'>"UMAP_corrected"</span>,
  colour_by <span class='op'>=</span> <span class='st'>"tissue"</span>,
  point_alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>scale_colour_manual</span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>tissue_colours</span>, name <span class='op'>=</span> <span class='st'>"tissue"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>ggtitle</span><span class='op'>(</span><span class='st'>"All cells"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="C094_Pellicci.single-cell.trajectory_files/figure-html5/unnamed-chunk-22-1.png" width="624" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotReducedDim.html'>plotReducedDim</a></span><span class='op'>(</span>
  <span class='va'>sce</span>,
  dimred <span class='op'>=</span> <span class='st'>"UMAP_corrected"</span>,
  colour_by <span class='op'>=</span> <span class='st'>"slingPseudotime_1"</span>,
  point_alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>geom_path</span><span class='op'>(</span>data <span class='op'>=</span> <span class='va'>embedded</span>, <span class='fu'>aes</span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>Dim.1</span>, y <span class='op'>=</span> <span class='va'>Dim.2</span><span class='op'>)</span>, size <span class='op'>=</span> <span class='fl'>1.2</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>ggtitle</span><span class='op'>(</span><span class='st'>"All cells"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="C094_Pellicci.single-cell.trajectory_files/figure-html5/unnamed-chunk-23-1.png" width="624" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotHeatmap.html'>plotHeatmap</a></span><span class='op'>(</span>
  <span class='va'>sce</span>,
  order_columns_by <span class='op'>=</span> <span class='st'>"slingPseudotime_1"</span>,
  colour_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"slingPseudotime_1"</span>, <span class='st'>"CD4"</span>, <span class='st'>"CD161"</span>, <span class='st'>"stage"</span>, <span class='st'>"tissue"</span><span class='op'>)</span>,
  features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span>
      <span class='fu'><a href='https://rdrr.io/r/base/sets.html'>setdiff</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>svet</span><span class='op'>[</span><span class='va'>svet</span><span class='op'>$</span><span class='va'>FDR</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span> <span class='op'>&amp;</span> <span class='va'>svet</span><span class='op'>$</span><span class='va'>logFC</span> <span class='op'>&lt;</span> <span class='fl'>0</span>, <span class='op'>]</span><span class='op'>)</span>, <span class='va'>mito_set</span><span class='op'>)</span>, 
      <span class='fl'>25</span><span class='op'>)</span>,
    <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span>
      <span class='fu'><a href='https://rdrr.io/r/base/sets.html'>setdiff</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>svet</span><span class='op'>[</span><span class='va'>svet</span><span class='op'>$</span><span class='va'>FDR</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span> <span class='op'>&amp;</span> <span class='va'>svet</span><span class='op'>$</span><span class='va'>logFC</span> <span class='op'>&gt;</span> <span class='fl'>0</span>, <span class='op'>]</span><span class='op'>)</span>, <span class='va'>mito_set</span><span class='op'>)</span>, 
      <span class='fl'>25</span><span class='op'>)</span><span class='op'>)</span>,
  center <span class='op'>=</span> <span class='cn'>TRUE</span>,
  symmetric <span class='op'>=</span> <span class='cn'>TRUE</span>,
  zlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>3</span>, <span class='fl'>3</span><span class='op'>)</span>,
  color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>hcl.colors</a></span><span class='op'>(</span><span class='fl'>101</span>, <span class='st'>"Blue-Red 3"</span><span class='op'>)</span>,
  fontsize <span class='op'>=</span> <span class='fl'>6</span>,
  main <span class='op'>=</span> <span class='st'>"All cells: Selected pseudotime-associated genes"</span>,
  column_annotation_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
    stage <span class='op'>=</span> <span class='va'>stage_colours</span>,
    tissue <span class='op'>=</span> <span class='va'>tissue_colours</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="C094_Pellicci.single-cell.trajectory_files/figure-html5/unnamed-chunk-24-1.png" width="624" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotReducedDim.html'>plotReducedDim</a></span><span class='op'>(</span>
  <span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>stage</span> <span class='op'>!=</span> <span class='st'>"Unknown"</span><span class='op'>]</span>,
  <span class='st'>"UMAP_corrected"</span>,
  colour_by <span class='op'>=</span> <span class='st'>"stage"</span>,
  point_alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>scale_colour_manual</span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>stage_colours</span>, name <span class='op'>=</span> <span class='st'>"stage"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>ggtitle</span><span class='op'>(</span><span class='st'>"Excluding 'Unknown'-stage cells"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="C094_Pellicci.single-cell.trajectory_files/figure-html5/unnamed-chunk-25-1.png" width="624" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotReducedDim.html'>plotReducedDim</a></span><span class='op'>(</span>
  <span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>stage</span> <span class='op'>!=</span> <span class='st'>"Unknown"</span><span class='op'>]</span>,
  <span class='st'>"UMAP_corrected"</span>,
  colour_by <span class='op'>=</span> <span class='st'>"tissue"</span>,
  point_alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>scale_colour_manual</span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>tissue_colours</span>, name <span class='op'>=</span> <span class='st'>"tissue"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>ggtitle</span><span class='op'>(</span><span class='st'>"Excluding 'Unknown'-stage cells"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="C094_Pellicci.single-cell.trajectory_files/figure-html5/unnamed-chunk-26-1.png" width="624" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotReducedDim.html'>plotReducedDim</a></span><span class='op'>(</span>
  <span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>stage</span> <span class='op'>!=</span> <span class='st'>"Unknown"</span><span class='op'>]</span>,
  dimred <span class='op'>=</span> <span class='st'>"UMAP_corrected"</span>,
  colour_by <span class='op'>=</span> <span class='st'>"slingPseudotime_1"</span>,
  point_alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>geom_path</span><span class='op'>(</span>data <span class='op'>=</span> <span class='va'>embedded</span>, <span class='fu'>aes</span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>Dim.1</span>, y <span class='op'>=</span> <span class='va'>Dim.2</span><span class='op'>)</span>, size <span class='op'>=</span> <span class='fl'>1.2</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>ggtitle</span><span class='op'>(</span><span class='st'>"Excluding 'Unknown'-stage cells"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="C094_Pellicci.single-cell.trajectory_files/figure-html5/unnamed-chunk-27-1.png" width="624" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotHeatmap.html'>plotHeatmap</a></span><span class='op'>(</span>
  <span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>stage</span> <span class='op'>!=</span> <span class='st'>"Unknown"</span><span class='op'>]</span>,
  order_columns_by <span class='op'>=</span> <span class='st'>"slingPseudotime_1"</span>,
  colour_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"slingPseudotime_1"</span>, <span class='st'>"CD4"</span>, <span class='st'>"CD161"</span>, <span class='st'>"stage"</span>, <span class='st'>"tissue"</span><span class='op'>)</span>,
  features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span>
      <span class='fu'><a href='https://rdrr.io/r/base/sets.html'>setdiff</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>svet</span><span class='op'>[</span><span class='va'>svet</span><span class='op'>$</span><span class='va'>FDR</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span> <span class='op'>&amp;</span> <span class='va'>svet</span><span class='op'>$</span><span class='va'>logFC</span> <span class='op'>&lt;</span> <span class='fl'>0</span>, <span class='op'>]</span><span class='op'>)</span>, <span class='va'>mito_set</span><span class='op'>)</span>, 
      <span class='fl'>25</span><span class='op'>)</span>,
    <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span>
      <span class='fu'><a href='https://rdrr.io/r/base/sets.html'>setdiff</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>svet</span><span class='op'>[</span><span class='va'>svet</span><span class='op'>$</span><span class='va'>FDR</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span> <span class='op'>&amp;</span> <span class='va'>svet</span><span class='op'>$</span><span class='va'>logFC</span> <span class='op'>&gt;</span> <span class='fl'>0</span>, <span class='op'>]</span><span class='op'>)</span>, <span class='va'>mito_set</span><span class='op'>)</span>, 
      <span class='fl'>25</span><span class='op'>)</span><span class='op'>)</span>,
  center <span class='op'>=</span> <span class='cn'>TRUE</span>,
  symmetric <span class='op'>=</span> <span class='cn'>TRUE</span>,
  zlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>3</span>, <span class='fl'>3</span><span class='op'>)</span>,
  color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>hcl.colors</a></span><span class='op'>(</span><span class='fl'>101</span>, <span class='st'>"Blue-Red 3"</span><span class='op'>)</span>,
  fontsize <span class='op'>=</span> <span class='fl'>6</span>,
  main <span class='op'>=</span> <span class='st'>"Excluding 'Unknown'-stage cells: Selected pseudotime-associated genes"</span>,
  column_annotation_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
    stage <span class='op'>=</span> <span class='va'>stage_colours</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>stage</span> <span class='op'>!=</span> <span class='st'>"Unknown"</span><span class='op'>]</span><span class='op'>$</span><span class='va'>stage</span><span class='op'>)</span><span class='op'>)</span><span class='op'>]</span>,
    tissue <span class='op'>=</span> <span class='va'>tissue_colours</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="C094_Pellicci.single-cell.trajectory_files/figure-html5/unnamed-chunk-28-1.png" width="624" /></p>
</div>
<h2 id="deg-heatmaps">DEG heatmaps</h2>
<p>Heatmaps sorted by CD4 and CD161 FACS expression showing the DEGs across the gamma delta T cell stages ordered by decreasing either CD4 or KLRB1. This is an example, but, as I explained to Dan, I dont think these are very useful plots.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>tmp</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>group</span> <span class='op'>%in%</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Thymus.S1 (CD4+/CD161-)"</span>, <span class='st'>"Thymus.S2 (CD4-/CD161-)"</span>, <span class='st'>"Thymus.S3 (CD4-/CD161+)"</span>, <span class='st'>"Blood.S3 (CD4-/CD161+)"</span><span class='op'>)</span><span class='op'>]</span>
<span class='va'>tmp</span><span class='op'>$</span><span class='va'>group</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span>
  <span class='va'>tmp</span><span class='op'>$</span><span class='va'>group</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Thymus.S1 (CD4+/CD161-)"</span>, <span class='st'>"Thymus.S2 (CD4-/CD161-)"</span>, <span class='st'>"Thymus.S3 (CD4-/CD161+)"</span>, <span class='st'>"Blood.S3 (CD4-/CD161+)"</span><span class='op'>)</span><span class='op'>)</span>

<span class='va'>x</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/utils/read.table.html'>read.csv</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"output/DEGs/excluding_donors_1-3/Thymus.S3_vs_Thymus.S1.aggregated_tech_reps.DEGs.csv.gz"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotHeatmap.html'>plotHeatmap</a></span><span class='op'>(</span>
  <span class='va'>tmp</span>,
  order_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"group"</span>, <span class='st'>"CD4"</span><span class='op'>)</span>,
  colour_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"CD4"</span>, <span class='st'>"group"</span><span class='op'>)</span>,
  features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>[</span><span class='va'>x</span><span class='op'>$</span><span class='va'>FDR</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span>, <span class='st'>"ENSEMBL.SYMBOL"</span><span class='op'>]</span>, <span class='fl'>50</span><span class='op'>)</span>,
  center <span class='op'>=</span> <span class='cn'>TRUE</span>,
  symmetric <span class='op'>=</span> <span class='cn'>TRUE</span>,
  zlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>3</span>, <span class='fl'>3</span><span class='op'>)</span>,
  color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>hcl.colors</a></span><span class='op'>(</span><span class='fl'>101</span>, <span class='st'>"Blue-Red 3"</span><span class='op'>)</span>,
  fontsize <span class='op'>=</span> <span class='fl'>6</span>,
  main <span class='op'>=</span> <span class='st'>"Mini-bulk DEGs: Thymus.S3 vs. Thymus.S1"</span>,
  column_annotation_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>group <span class='op'>=</span> <span class='va'>group_colours</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>tmp</span><span class='op'>$</span><span class='va'>group</span><span class='op'>)</span><span class='op'>]</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="C094_Pellicci.single-cell.trajectory_files/figure-html5/unnamed-chunk-30-1.png" width="624" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotHeatmap.html'>plotHeatmap</a></span><span class='op'>(</span>
  <span class='va'>tmp</span>,
  order_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"group"</span>, <span class='st'>"CD161"</span><span class='op'>)</span>,
  colour_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"CD161"</span>, <span class='st'>"group"</span><span class='op'>)</span>,
  features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>[</span><span class='va'>x</span><span class='op'>$</span><span class='va'>FDR</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span>, <span class='st'>"ENSEMBL.SYMBOL"</span><span class='op'>]</span>, <span class='fl'>50</span><span class='op'>)</span>,
  center <span class='op'>=</span> <span class='cn'>TRUE</span>,
  symmetric <span class='op'>=</span> <span class='cn'>TRUE</span>,
  zlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>3</span>, <span class='fl'>3</span><span class='op'>)</span>,
  color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>hcl.colors</a></span><span class='op'>(</span><span class='fl'>101</span>, <span class='st'>"Blue-Red 3"</span><span class='op'>)</span>,
  fontsize <span class='op'>=</span> <span class='fl'>6</span>,
  main <span class='op'>=</span> <span class='st'>"Mini-bulk DEGs: Thymus.S3 vs. Thymus.S1"</span>,
  column_annotation_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>group <span class='op'>=</span> <span class='va'>group_colours</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>tmp</span><span class='op'>$</span><span class='va'>group</span><span class='op'>)</span><span class='op'>]</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="C094_Pellicci.single-cell.trajectory_files/figure-html5/unnamed-chunk-31-1.png" width="624" /></p>
</div>
<h2 id="mock-figure-sent-to-dan-and-co-2021-09-09">Mock figure sent to Dan and co (2021-09-09)</h2>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/files2.html'>dir.create</a></span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"output/figures"</span><span class='op'>)</span><span class='op'>)</span>

<span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotReducedDim.html'>plotReducedDim</a></span><span class='op'>(</span>
  <span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>stage</span> <span class='op'>!=</span> <span class='st'>"Unknown"</span><span class='op'>]</span>,
  <span class='st'>"UMAP_corrected"</span>,
  colour_by <span class='op'>=</span> <span class='st'>"stage"</span>,
  point_alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>scale_colour_manual</span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>stage_colours</span>, name <span class='op'>=</span> <span class='st'>"stage"</span><span class='op'>)</span>
<span class='fu'>ggsave</span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"output/figures/Fig5a.pdf"</span><span class='op'>)</span>, <span class='va'>p1</span>, width <span class='op'>=</span> <span class='fl'>6</span>, height <span class='op'>=</span> <span class='fl'>5</span><span class='op'>)</span>

<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotReducedDim.html'>plotReducedDim</a></span><span class='op'>(</span>
  <span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>stage</span> <span class='op'>!=</span> <span class='st'>"Unknown"</span><span class='op'>]</span>,
  <span class='st'>"UMAP_corrected"</span>,
  colour_by <span class='op'>=</span> <span class='st'>"tissue"</span>,
  point_alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>scale_colour_manual</span><span class='op'>(</span>values <span class='op'>=</span> <span class='va'>tissue_colours</span>, name <span class='op'>=</span> <span class='st'>"tissue"</span><span class='op'>)</span>
<span class='fu'>ggsave</span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"output/figures/Fig5b.pdf"</span><span class='op'>)</span>, <span class='va'>p2</span>, width <span class='op'>=</span> <span class='fl'>6</span>, height <span class='op'>=</span> <span class='fl'>5</span><span class='op'>)</span>

<span class='va'>p3</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotReducedDim.html'>plotReducedDim</a></span><span class='op'>(</span>
  <span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>stage</span> <span class='op'>!=</span> <span class='st'>"Unknown"</span><span class='op'>]</span>,
  dimred <span class='op'>=</span> <span class='st'>"UMAP_corrected"</span>,
  colour_by <span class='op'>=</span> <span class='st'>"slingPseudotime_1"</span>,
  point_alpha <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>geom_path</span><span class='op'>(</span>data <span class='op'>=</span> <span class='va'>embedded</span>, <span class='fu'>aes</span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>Dim.1</span>, y <span class='op'>=</span> <span class='va'>Dim.2</span><span class='op'>)</span>, size <span class='op'>=</span> <span class='fl'>1.2</span><span class='op'>)</span>
<span class='fu'>ggsave</span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"output/figures/Fig5c.pdf"</span><span class='op'>)</span>, <span class='va'>p3</span>, width <span class='op'>=</span> <span class='fl'>6</span>, height <span class='op'>=</span> <span class='fl'>5</span><span class='op'>)</span>

<span class='va'>p4</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotHeatmap.html'>plotHeatmap</a></span><span class='op'>(</span>
  <span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>stage</span> <span class='op'>!=</span> <span class='st'>"Unknown"</span><span class='op'>]</span>,
  order_columns_by <span class='op'>=</span> <span class='st'>"slingPseudotime_1"</span>,
  colour_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"slingPseudotime_1"</span>, <span class='st'>"CD4"</span>, <span class='st'>"CD161"</span>, <span class='st'>"stage"</span>, <span class='st'>"tissue"</span><span class='op'>)</span>,
  features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span>
      <span class='fu'><a href='https://rdrr.io/r/base/sets.html'>setdiff</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>svet</span><span class='op'>[</span><span class='va'>svet</span><span class='op'>$</span><span class='va'>FDR</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span> <span class='op'>&amp;</span> <span class='va'>svet</span><span class='op'>$</span><span class='va'>logFC</span> <span class='op'>&lt;</span> <span class='fl'>0</span>, <span class='op'>]</span><span class='op'>)</span>, <span class='va'>mito_set</span><span class='op'>)</span>, 
      <span class='fl'>50</span><span class='op'>)</span>,
    <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span>
      <span class='fu'><a href='https://rdrr.io/r/base/sets.html'>setdiff</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>svet</span><span class='op'>[</span><span class='va'>svet</span><span class='op'>$</span><span class='va'>FDR</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span> <span class='op'>&amp;</span> <span class='va'>svet</span><span class='op'>$</span><span class='va'>logFC</span> <span class='op'>&gt;</span> <span class='fl'>0</span>, <span class='op'>]</span><span class='op'>)</span>, <span class='va'>mito_set</span><span class='op'>)</span>, 
      <span class='fl'>50</span><span class='op'>)</span><span class='op'>)</span>,
  center <span class='op'>=</span> <span class='cn'>TRUE</span>,
  symmetric <span class='op'>=</span> <span class='cn'>TRUE</span>,
  zlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>3</span>, <span class='fl'>3</span><span class='op'>)</span>,
  color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>hcl.colors</a></span><span class='op'>(</span><span class='fl'>101</span>, <span class='st'>"Blue-Red 3"</span><span class='op'>)</span>,
  fontsize <span class='op'>=</span> <span class='fl'>5</span>,
  column_annotation_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
    stage <span class='op'>=</span> <span class='va'>stage_colours</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>stage</span> <span class='op'>!=</span> <span class='st'>"Unknown"</span><span class='op'>]</span><span class='op'>$</span><span class='va'>stage</span><span class='op'>)</span><span class='op'>)</span><span class='op'>]</span>,
    tissue <span class='op'>=</span> <span class='va'>tissue_colours</span><span class='op'>)</span>,
  silent <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='fu'>ggsave</span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"output/figures/Fig5d.pdf"</span><span class='op'>)</span>, <span class='va'>p4</span><span class='op'>$</span><span class='va'>gtable</span>, width <span class='op'>=</span> <span class='fl'>6</span>, height <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span>

<span class='va'>p4a</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/scater/man/plotHeatmap.html'>plotHeatmap</a></span><span class='op'>(</span>
  <span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>stage</span> <span class='op'>!=</span> <span class='st'>"Unknown"</span><span class='op'>]</span>,
  order_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"stage"</span>, <span class='st'>"slingPseudotime_1"</span><span class='op'>)</span>,
  colour_columns_by <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"stage"</span>, <span class='st'>"slingPseudotime_1"</span>, <span class='st'>"CD4"</span>, <span class='st'>"CD161"</span>, <span class='st'>"tissue"</span><span class='op'>)</span>,
  features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span>
      <span class='fu'><a href='https://rdrr.io/r/base/sets.html'>setdiff</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>svet</span><span class='op'>[</span><span class='va'>svet</span><span class='op'>$</span><span class='va'>FDR</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span> <span class='op'>&amp;</span> <span class='va'>svet</span><span class='op'>$</span><span class='va'>logFC</span> <span class='op'>&lt;</span> <span class='fl'>0</span>, <span class='op'>]</span><span class='op'>)</span>, <span class='va'>mito_set</span><span class='op'>)</span>, 
      <span class='fl'>50</span><span class='op'>)</span>,
    <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span>
      <span class='fu'><a href='https://rdrr.io/r/base/sets.html'>setdiff</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>svet</span><span class='op'>[</span><span class='va'>svet</span><span class='op'>$</span><span class='va'>FDR</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span> <span class='op'>&amp;</span> <span class='va'>svet</span><span class='op'>$</span><span class='va'>logFC</span> <span class='op'>&gt;</span> <span class='fl'>0</span>, <span class='op'>]</span><span class='op'>)</span>, <span class='va'>mito_set</span><span class='op'>)</span>, 
      <span class='fl'>50</span><span class='op'>)</span><span class='op'>)</span>,
  center <span class='op'>=</span> <span class='cn'>TRUE</span>,
  symmetric <span class='op'>=</span> <span class='cn'>TRUE</span>,
  zlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>3</span>, <span class='fl'>3</span><span class='op'>)</span>,
  color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>hcl.colors</a></span><span class='op'>(</span><span class='fl'>101</span>, <span class='st'>"Blue-Red 3"</span><span class='op'>)</span>,
  fontsize <span class='op'>=</span> <span class='fl'>5</span>,
  <span class='co'># main = "Excluding 'Unknown'-stage cells: Selected pseudotime-associated genes",</span>
  column_annotation_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
    stage <span class='op'>=</span> <span class='va'>stage_colours</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>[</span>, <span class='va'>sce</span><span class='op'>$</span><span class='va'>stage</span> <span class='op'>!=</span> <span class='st'>"Unknown"</span><span class='op'>]</span><span class='op'>$</span><span class='va'>stage</span><span class='op'>)</span><span class='op'>)</span><span class='op'>]</span>,
    tissue <span class='op'>=</span> <span class='va'>tissue_colours</span><span class='op'>)</span>,
  silent <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='fu'>ggsave</span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"output/figures/Fig5d_alternative.pdf"</span><span class='op'>)</span>, <span class='va'>p4a</span><span class='op'>$</span><span class='va'>gtable</span>, width <span class='op'>=</span> <span class='fl'>6</span>, height <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span>
</code></pre>
</div>
</details>
</div>
<p>Initial proposal:</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>p1</span> <span class='op'>+</span> <span class='va'>p2</span> <span class='op'>+</span> <span class='va'>p3</span> <span class='op'>+</span> <span class='va'>p4</span><span class='op'>$</span><span class='va'>gtable</span> <span class='op'>+</span>
  <span class='fu'><a href='https://patchwork.data-imaginist.com/reference/plot_annotation.html'>plot_annotation</a></span><span class='op'>(</span>
    title <span class='op'>=</span> <span class='st'>"Summary of scRNA-seq data"</span>,
    subtitle <span class='op'>=</span> <span class='st'>"Excluding 'Unknown'-stage cells"</span>,
    tag_levels <span class='op'>=</span> <span class='st'>"a"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="C094_Pellicci.single-cell.trajectory_files/figure-html5/unnamed-chunk-33-1.png" width="624" /></p>
</div>
<p>Alternative:</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>p1</span> <span class='op'>+</span> <span class='va'>p2</span> <span class='op'>+</span> <span class='va'>p3</span> <span class='op'>+</span> <span class='va'>p4a</span><span class='op'>$</span><span class='va'>gtable</span> <span class='op'>+</span>
  <span class='fu'><a href='https://patchwork.data-imaginist.com/reference/plot_annotation.html'>plot_annotation</a></span><span class='op'>(</span>
    title <span class='op'>=</span> <span class='st'>"Summary of scRNA-seq data"</span>,
    subtitle <span class='op'>=</span> <span class='st'>"Excluding 'Unknown'-stage cells"</span>,
    tag_levels <span class='op'>=</span> <span class='st'>"a"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="C094_Pellicci.single-cell.trajectory_files/figure-html5/unnamed-chunk-34-1.png" width="624" /></p>
</div>
<h1 class="appendix" id="additional-information">Additional information</h1>
<p>The following are available on request:</p>
<ul>
<li>Full CSV tables of any data presented.</li>
<li>PDF/PNG files of any static plots.</li>
</ul>
<h1 class="appendix" id="session-info">Session info</h1>
<summary>
The analysis and this document were prepared using the following software (click triangle to expand)
</summary>
<details>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>sessioninfo</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/sessioninfo/man/session_info.html'>session_info</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<pre><code> Session info 
 setting  value                       
 version  R version 4.0.3 (2020-10-10)
 os       CentOS Linux 7 (Core)       
 system   x86_64, linux-gnu           
 ui       X11                         
 language (EN)                        
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 
 tz       Australia/Melbourne         
 date     2021-10-01                  

 Packages 
 ! package              * version  date       lib source        
 P AnnotationDbi        * 1.52.0   2020-10-27 [?] Bioconductor  
 P ape                    5.5      2021-04-25 [?] CRAN (R 4.0.5)
 P assertthat             0.2.1    2019-03-21 [?] CRAN (R 4.0.0)
 P beachmat               2.6.4    2020-12-20 [?] Bioconductor  
 P beeswarm               0.3.1    2021-03-07 [?] CRAN (R 4.0.3)
 P Biobase              * 2.50.0   2020-10-27 [?] Bioconductor  
 P BiocFileCache        * 1.14.0   2020-10-27 [?] Bioconductor  
 P BiocGenerics         * 0.36.0   2020-10-27 [?] Bioconductor  
 P BiocManager            1.30.12  2021-03-28 [?] CRAN (R 4.0.3)
 P BiocNeighbors          1.8.2    2020-12-07 [?] Bioconductor  
 P BiocParallel           1.24.1   2020-11-06 [?] Bioconductor  
 P BiocSingular           1.6.0    2020-10-27 [?] Bioconductor  
 P BiocStyle              2.18.1   2020-11-24 [?] Bioconductor  
 P bit                    4.0.4    2020-08-04 [?] CRAN (R 4.0.0)
 P bit64                  4.0.5    2020-08-30 [?] CRAN (R 4.0.0)
 P bitops                 1.0-6    2013-08-17 [?] CRAN (R 4.0.0)
 P blob                   1.2.1    2020-01-20 [?] CRAN (R 4.0.0)
 P bslib                  0.2.4    2021-01-25 [?] CRAN (R 4.0.3)
 P cachem                 1.0.4    2021-02-13 [?] CRAN (R 4.0.3)
 P cli                    2.4.0    2021-04-05 [?] CRAN (R 4.0.3)
 P cluster                2.1.0    2019-06-19 [3] CRAN (R 4.0.3)
 P colorspace             2.0-0    2020-11-11 [?] CRAN (R 4.0.3)
 P combinat               0.0-8    2012-10-29 [?] CRAN (R 4.0.3)
 P cowplot                1.1.1    2020-12-30 [?] CRAN (R 4.0.3)
 P crayon                 1.4.1    2021-02-08 [?] CRAN (R 4.0.3)
 P curl                   4.3      2019-12-02 [?] CRAN (R 4.0.0)
 P DBI                    1.1.1    2021-01-15 [?] CRAN (R 4.0.3)
 P dbplyr               * 2.1.0    2021-02-03 [?] CRAN (R 4.0.3)
 P DDRTree                0.1.5    2017-04-30 [?] CRAN (R 4.0.3)
 P DelayedArray           0.16.3   2021-03-24 [?] Bioconductor  
 P DelayedMatrixStats     1.12.3   2021-02-03 [?] Bioconductor  
 P densityClust           0.3      2017-10-24 [?] CRAN (R 4.0.3)
 P digest                 0.6.27   2020-10-24 [?] CRAN (R 4.0.2)
 P distill                1.2      2021-01-13 [?] CRAN (R 4.0.3)
 P docopt                 0.7.1    2020-06-24 [?] CRAN (R 4.0.3)
 P downlit                0.2.1    2020-11-04 [?] CRAN (R 4.0.3)
 P dplyr                  1.0.5    2021-03-05 [?] CRAN (R 4.0.3)
 P edgeR                  3.32.1   2021-01-14 [?] Bioconductor  
 P ellipsis               0.3.1    2020-05-15 [?] CRAN (R 4.0.0)
 P evaluate               0.14     2019-05-28 [?] CRAN (R 4.0.0)
 P fansi                  0.4.2    2021-01-15 [?] CRAN (R 4.0.3)
 P farver                 2.1.0    2021-02-28 [?] CRAN (R 4.0.3)
 P fastICA                1.2-2    2019-07-08 [?] CRAN (R 4.0.3)
 P fastmap                1.1.0    2021-01-25 [?] CRAN (R 4.0.3)
 P FNN                    1.1.3    2019-02-15 [?] CRAN (R 4.0.0)
 P generics               0.1.0    2020-10-31 [?] CRAN (R 4.0.3)
 P GenomeInfoDb         * 1.26.4   2021-03-10 [?] Bioconductor  
 P GenomeInfoDbData       1.2.4    2020-10-20 [?] Bioconductor  
 P GenomicRanges        * 1.42.0   2020-10-27 [?] Bioconductor  
 P ggbeeswarm             0.6.0    2017-08-07 [?] CRAN (R 4.0.0)
 P ggplot2              * 3.3.3    2020-12-30 [?] CRAN (R 4.0.3)
 P ggrepel                0.9.1    2021-01-15 [?] CRAN (R 4.0.3)
 P glue                   1.4.2    2020-08-27 [?] CRAN (R 4.0.0)
 P GO.db                  3.12.1   2020-11-04 [?] Bioconductor  
 P gridExtra              2.3      2017-09-09 [?] CRAN (R 4.0.0)
 P gtable                 0.3.0    2019-03-25 [?] CRAN (R 4.0.0)
 P here                 * 1.0.1    2020-12-13 [?] CRAN (R 4.0.3)
 P highr                  0.9      2021-04-16 [?] CRAN (R 4.0.3)
 P HSMMSingleCell         1.10.0   2020-10-29 [?] Bioconductor  
 P htmltools              0.5.1.1  2021-01-22 [?] CRAN (R 4.0.3)
 P httr                   1.4.2    2020-07-20 [?] CRAN (R 4.0.0)
 P igraph                 1.2.6    2020-10-06 [?] CRAN (R 4.0.2)
 P IRanges              * 2.24.1   2020-12-12 [?] Bioconductor  
 P irlba                  2.3.3    2019-02-05 [?] CRAN (R 4.0.0)
 P jquerylib              0.1.3    2020-12-17 [?] CRAN (R 4.0.3)
 P jsonlite               1.7.2    2020-12-09 [?] CRAN (R 4.0.3)
 P knitr                  1.33     2021-04-24 [?] CRAN (R 4.0.3)
 P labeling               0.4.2    2020-10-20 [?] CRAN (R 4.0.0)
 P lattice                0.20-41  2020-04-02 [3] CRAN (R 4.0.3)
 P lifecycle              1.0.0    2021-02-15 [?] CRAN (R 4.0.3)
 P limma                * 3.46.0   2020-10-27 [?] Bioconductor  
 P locfit                 1.5-9.4  2020-03-25 [?] CRAN (R 4.0.0)
 P magrittr               2.0.1    2020-11-17 [?] CRAN (R 4.0.3)
 P Matrix                 1.2-18   2019-11-27 [3] CRAN (R 4.0.3)
 P MatrixGenerics       * 1.2.1    2021-01-30 [?] Bioconductor  
 P matrixStats          * 0.58.0   2021-01-29 [?] CRAN (R 4.0.3)
 P memoise                2.0.0    2021-01-26 [?] CRAN (R 4.0.3)
 P mgcv                   1.8-33   2020-08-27 [3] CRAN (R 4.0.3)
 P monocle                2.18.0   2020-10-27 [?] Bioconductor  
 P msigdbr              * 7.2.1    2020-10-02 [?] CRAN (R 4.0.2)
 P munsell                0.5.0    2018-06-12 [?] CRAN (R 4.0.0)
 P nlme                   3.1-149  2020-08-23 [3] CRAN (R 4.0.3)
 P org.Hs.eg.db         * 3.12.0   2020-10-20 [?] Bioconductor  
 P patchwork            * 1.1.1    2020-12-17 [?] CRAN (R 4.0.3)
 P pbapply                1.4-3    2020-08-18 [?] CRAN (R 4.0.3)
 P pheatmap               1.0.12   2019-01-04 [?] CRAN (R 4.0.0)
 P pillar                 1.5.1    2021-03-05 [?] CRAN (R 4.0.3)
 P pkgconfig              2.0.3    2019-09-22 [?] CRAN (R 4.0.0)
 P plyr                   1.8.6    2020-03-03 [?] CRAN (R 4.0.0)
 P princurve            * 2.1.6    2021-01-18 [?] CRAN (R 4.0.3)
 P purrr                  0.3.4    2020-04-17 [?] CRAN (R 4.0.0)
 P qlcMatrix              0.9.7    2018-04-20 [?] CRAN (R 4.0.3)
 P R6                     2.5.0    2020-10-28 [?] CRAN (R 4.0.2)
 P RANN                   2.6.1    2019-01-08 [?] CRAN (R 4.0.0)
 P rappdirs               0.3.3    2021-01-31 [?] CRAN (R 4.0.3)
 P RColorBrewer           1.1-2    2014-12-07 [?] CRAN (R 4.0.0)
 P Rcpp                   1.0.6    2021-01-15 [?] CRAN (R 4.0.3)
 P RCurl                  1.98-1.3 2021-03-16 [?] CRAN (R 4.0.3)
 P reshape2               1.4.4    2020-04-09 [?] CRAN (R 4.0.0)
 P rlang                  0.4.11   2021-04-30 [?] CRAN (R 4.0.5)
 P rmarkdown              2.7      2021-02-19 [?] CRAN (R 4.0.3)
 P rprojroot              2.0.2    2020-11-15 [?] CRAN (R 4.0.3)
 P RSQLite                2.2.5    2021-03-27 [?] CRAN (R 4.0.3)
 P rsvd                   1.0.3    2020-02-17 [?] CRAN (R 4.0.0)
 P Rtsne                  0.15     2018-11-10 [?] CRAN (R 4.0.0)
 P S4Vectors            * 0.28.1   2020-12-09 [?] Bioconductor  
 P sass                   0.3.1    2021-01-24 [?] CRAN (R 4.0.3)
 P scales                 1.1.1    2020-05-11 [?] CRAN (R 4.0.0)
 P scater               * 1.18.6   2021-02-26 [?] Bioconductor  
 P scuttle                1.0.4    2020-12-17 [?] Bioconductor  
 P sessioninfo            1.1.1    2018-11-05 [?] CRAN (R 4.0.0)
 P SingleCellExperiment * 1.12.0   2020-10-27 [?] Bioconductor  
 P slam                   0.1-48   2020-12-03 [?] CRAN (R 4.0.3)
 P slingshot            * 1.8.0    2020-10-27 [?] Bioconductor  
 P sparseMatrixStats      1.2.1    2021-02-02 [?] Bioconductor  
 P sparsesvd              0.2      2019-07-15 [?] CRAN (R 4.0.3)
 P stringi                1.7.3    2021-07-16 [?] CRAN (R 4.0.3)
 P stringr                1.4.0    2019-02-10 [?] CRAN (R 4.0.0)
 P SummarizedExperiment * 1.20.0   2020-10-27 [?] Bioconductor  
 P tibble                 3.1.0    2021-02-25 [?] CRAN (R 4.0.3)
 P tidyselect             1.1.0    2020-05-11 [?] CRAN (R 4.0.0)
 P tradeSeq             * 1.4.0    2020-10-27 [?] Bioconductor  
 P utf8                   1.2.1    2021-03-12 [?] CRAN (R 4.0.3)
 P vctrs                  0.3.7    2021-03-29 [?] CRAN (R 4.0.3)
 P VGAM                   1.1-5    2021-01-14 [?] CRAN (R 4.0.3)
 P vipor                  0.4.5    2017-03-22 [?] CRAN (R 4.0.0)
 P viridis                0.5.1    2018-03-29 [?] CRAN (R 4.0.0)
 P viridisLite            0.3.0    2018-02-01 [?] CRAN (R 4.0.0)
 P withr                  2.4.1    2021-01-26 [?] CRAN (R 4.0.3)
 P xfun                   0.24     2021-06-15 [?] CRAN (R 4.0.3)
 P XVector                0.30.0   2020-10-27 [?] Bioconductor  
 P yaml                   2.2.1    2020-02-01 [?] CRAN (R 4.0.0)
 P zlibbioc               1.36.0   2020-10-27 [?] Bioconductor  

[1] /stornext/Projects/score/Analyses/C094_Pellicci/renv/library/R-4.0/x86_64-pc-linux-gnu
[2] /tmp/RtmptK80kW/renv-system-library
[3] /stornext/System/data/apps/R/R-4.0.3/lib64/R/library

 P  Loaded and on-disk path mismatch.</code></pre>
</div>
</details>
<div class="sourceCode" id="cb2"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<section class="footnotes" role="doc-endnotes">
<hr />
<ol>
<li id="fn1" role="doc-endnote"><p><code>cameraPR()</code> is a pre-ranked version of <code>camera()</code> where the genes are pre-ranked according to a pre-computed statistic. We cannot use the regular <code>camera()</code> function because the pseudotime DE analysis does not provide the standard limma/edgeR output required for <code>camera()</code>, but <code>cameraPR()</code> is a suitable replacement in this case.<a href="#fnref1" class="footnote-back" role="doc-backlink"></a></p></li>
</ol>
</section>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
<h3 id="references">References</h3>
<div id="references-listing"></div>
</div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
