---
title: "Trajectory analysis of the Pellicci gamma-delta T-cell single-cell data set"
description: |
author:
  - name: Peter Hickey 
    url: https://peterhickey.org
    affiliation: Single Cell Open Research Endeavour (SCORE), WEHI
    affiliation_url: https://www.wehi.edu.au/people/rory-bowden/4536/wehi-advanced-genomics-facility
date: "`r Sys.Date()`"
editor_options: 
  chunk_output_type: console
bibliography: ref.bib
---

# Setup

```{r}
library(here)
library(SingleCellExperiment)
library(scater)
library(patchwork)

outdir <- here("output", "trajectory")
dir.create(outdir)
```

Need to start from this *SCE* object in order to have consistency with William's analyses.

```{r}
sce <- readRDS(
  here("data/SCEs/C094_Pellicci.single-cell.merged.whole_cell.SCE.rds"))
```

```{r}
# Some useful gene sets
mito_set <- rownames(sce)[any(rowData(sce)$ENSEMBL.SEQNAME == "MT")]
ribo_set <- grep("^RP(S|L)", rownames(sce), value = TRUE)
# NOTE: A more curated approach for identifying ribosomal protein genes
#       (https://github.com/Bioconductor/OrchestratingSingleCellAnalysis-base/blob/ae201bf26e3e4fa82d9165d8abf4f4dc4b8e5a68/feature-selection.Rmd#L376-L380)
library(msigdbr)
c2_sets <- msigdbr(species = "Homo sapiens", category = "C2")
ribo_set <- union(
  ribo_set,
  c2_sets[c2_sets$gs_name == "KEGG_RIBOSOME", ]$gene_symbol)
ribo_set <- intersect(ribo_set, rownames(sce))
sex_set <- rownames(sce)[any(rowData(sce)$ENSEMBL.SEQNAME %in% c("X", "Y"))]
pseudogene_set <- rownames(sce)[
  any(grepl("pseudogene", rowData(sce)$ENSEMBL.GENEBIOTYPE))]
protein_coding_gene_set <- rownames(sce)[
  any(grepl("protein_coding", rowData(sce)$ENSEMBL.GENEBIOTYPE))]
```

# Obtaining pseudotime orderings

Analysis using `r BiocStyle::Biocpkg("slingshot")`; see [https://bioconductor.org/books/release/OSCA/trajectory-analysis.html#principal-curves](https://bioconductor.org/books/release/OSCA/trajectory-analysis.html#principal-curves) for inspiration.

```{r, fig.asp = 0.7}
library(slingshot)
# NOTE: Can't directly use `slingshot(sce, reducedDim = "corrected")` because
#       it gives an error:
#       "Error in solve.default(s1 + s2) :
#         system is computationally singular: reciprocal condition
#         number = 2.91591e-17"
#       This is similar to previously reported issues
#       (https://github.com/kstreet13/slingshot/issues/87 and
#       https://github.com/kstreet13/slingshot/issues/35) for which a hacky
#       workaround is to drop the last dimension.
rd <- reducedDim(sce, "corrected")
reducedDim(sce, "corrected_1") <- rd[, seq_len(ncol(rd) - 1)]
sce <- slingshot(sce, reducedDim = "corrected_1")

embedded <- embedCurves(sce, "UMAP_corrected")
embedded <- slingCurves(embedded)[[1]] # only 1 path.
embedded <- data.frame(embedded$s[embedded$ord, ])
```

```{r, fig.asp = 0.7}
plotReducedDim(
  sce,
  dimred = "UMAP_corrected",
  colour_by = "slingPseudotime_1",
  point_alpha = 1,
  point_size = 0.5) +
  geom_path(data = embedded, aes(x = Dim.1, y = Dim.2), size = 1.2) +
  plotReducedDim(
    sce,
    dimred = "UMAP_corrected",
    colour_by = "stage",
    point_alpha = 1,
    point_size = 0.5) +
  plotReducedDim(
    sce,
    dimred = "UMAP_corrected",
    colour_by = "tissue",
    point_alpha = 1,
    point_size = 0.5) +
  plotReducedDim(
    sce,
    dimred = "UMAP_corrected",
    colour_by = "cluster",
    point_alpha = 1,
    point_size = 0.5) +
  plot_layout(ncol = 2)
```

# Changes along a trajectory

We only test genes that are detected in at least 10 cells.

```{r}
sce_pseudo <- sce[rowSums(counts(sce) > 0) >= 10, ]
```

We use methods from the `r BiocStyle::Biocpkg("tradeSeq")` R/Bioconductor package.

```{r}
library(tradeSeq)

nonna.pseudo <- sce_pseudo$slingPseudotime_1
nonna.pseudo[is.na(nonna.pseudo)] <- 0
cell.weights <- !is.na(sce_pseudo$slingPseudotime_1)
storage.mode(cell.weights) <- "numeric"
```

```{r, eval = FALSE}
set.seed(666)
# NOTE: Takes ~30 mins and can crash due to memory constraints.
fit <- fitGAM(
  counts(sce_pseudo),
  pseudotime = nonna.pseudo,
  cellWeights = cell.weights,
  parallel = TRUE,
  BPPARAM = BiocParallel::MulticoreParam(4))
saveRDS(fit, file.path(outdir, "tradeSeq_fit.rds"))
```

```{r}
fit <- readRDS(file.path(outdir, "tradeSeq_fit.rds"))
```

We perform a statistical test to check for DE between the starting point and the end stage of the trajectory.

```{r}
svet <- startVsEndTest(fit)
svet <- svet[order(svet$pvalue), ]
svet$FDR <- p.adjust(svet$pvalue, method = "BH")

write.csv(
  svet[, c("logFC", "p.value", "FDR")],
  file.path(outdir, "tradeSeq_results.DEGs.csv"),
  row.names = TRUE,
  quote = FALSE)

svet$ENTREZID <- sapply(
  rowData(sce[rownames(svet), ])$NCBI.ENTREZID,
  "[[",
  1)
svet$logFC <- svet$logFClineage1

pseudotime_degs <- list(
  up_at_start =
    sapply(
      rowData(
        sce[rownames(svet[svet$FDR < 0.05 & svet$logFC < 0, ]), ])$NCBI.ENTREZID,
      "[[",
      1),
  up_at_end = sapply(
    rowData(
      sce[rownames(svet[svet$FDR < 0.05 & svet$logFC > 0, ]), ])$NCBI.ENTREZID,
    "[[",
    1))
```

There are `r scales::number(length(pseudotime_degs$up_at_start), big.mark = ",")` and `r scales::number(length(pseudotime_degs$up_at_end), big.mark = ",")` genes that are significantly upregulated at the start and end of the trajectory, respectively.
CSV files of the results are available from [`output/trajectory/`](../output/trajectory/).
Figures \@ref(fig:up-at-start) and \@ref(fig:up-at-end) highlight some of these genes.

```{r up-at-start, fig.cap = "Selected genes that are upregulated at the start of the inferred trajectory.", fig.asp = 1}
plotHeatmap(
  sce,
  order_columns_by = "slingPseudotime_1",
  colour_columns_by = c("stage", "tissue"),
  features = head(rownames(svet[svet$FDR < 0.05 & svet$logFC < 0, ]), 25),
  center = TRUE,
  symmetric = TRUE,
  zlim = c(-3, 3),
  color = hcl.colors(101, "Blue-Red 3"),
  fontsize = 6,
  main = "Up at start")
```

```{r up-at-end, fig.cap = "Selected genes that are upregulated at the end of the inferred trajectory.", fig.asp = 1}
plotHeatmap(
  sce,
  order_columns_by = "slingPseudotime_1",
  colour_columns_by = c("stage", "tissue"),
  features = head(rownames(svet[svet$FDR < 0.05 & svet$logFC > 0, ]), 25),
  center = TRUE,
  symmetric = TRUE,
  zlim = c(-3, 3),
  color = hcl.colors(101, "Blue-Red 3"),
  fontsize = 6,
  main = "Up at end")
```

# Gene set testing

```{r}
library(limma)
```

We opt to use the `r BiocStyle::Biocpkg("tradeSeq")` results and seek to identify gene sets that are differentially expressed with respect to the trajectory.

## goana

We use the `goana()` function from the `r BiocStyle::Biocpkg("limma")` R/Bioconductor package to test for over-representation of gene ontology (GO) terms in each of the `up_at_start` and `up_at_end` DEG lists.

```{r}
go <- goana(de = pseudotime_degs, species = "Hs")
gzout <- gzfile(
  description = file.path(outdir, "tradeSeq_results.goana.csv.gz"),
  open = "wb")
write.csv(
  topGO(go, number = Inf),
  gzout,
  row.names = TRUE)
close(gzout)
```

CSV files of the results are available from [`output/trajectory/`](../output/trajectory/).
An example of the results are shown below, here ordered by `P.up_at_end` to highlight GO terms that are enriched in the `up_at_end` DEG list.

```{r}
topGO(go, sort = "up_at_end") %>%
  knitr::kable(
    caption = '`goana()` produces a table with a row for each GO term and the following columns: **Term** (GO term); **Ont** (ontology that the GO term belongs to. Possible values are "BP", "CC" and "MF"); **N** (number of genes in the GO term); **up_at_start** (number of `up_at_start`-regulated differentially expressed genes); **up_at_end** (number of `up_at_end`-regulated differentially expressed genes); **P.up_at_start** (p-value for over-representation of GO term in `up_at_start`-regulated genes); **P.up_at_end** (p-value for over-representation of GO term in `up_at_end`-regulated genes)')
```

## kegga

We use the `kegga()` function from the `r BiocStyle::Biocpkg("limma")` R/Bioconductor package to test for over-representation of KEGG pathways in each of the `up_at_start` and `up_at_end` DEG lists.

```{r}
kegg <- kegga(de = pseudotime_degs, species = "Hs")
gzout <- gzfile(
  description = file.path(outdir, "tradeSeq_results.kegga.csv.gz"),
  open = "wb")
write.csv(
  topKEGG(kegg, number = Inf),
  gzout,
  row.names = TRUE)
close(gzout)
```

CSV files of the results are available from [`output/trajectory/`](../output/trajectory/).
An example of the results are shown below, here ordered by `P.up_at_end` to highlight GO terms that are enriched in the `up_at_end` DEG list.

```{r}
topKEGG(kegg, sort = "up_at_end") %>%
  knitr::kable(
    caption = '`kegga()` produces a table with a row for each KEGG pathway ID and the following columns: **Pathway** (KEGG pathway); **N** (number of genes in the GO term);; **up_at_start** (number of `up_at_start`-regulated differentially expressed genes); **up_at_end** (number of `up_at_end`-regulated differentially expressed genes); **P.up_at_start** (p-value for over-representation of KEGG pathway in `up_at_start`-regulated genes); **P.up_at_end** (p-value for over-representation of KEGG pathway in `up_at_end`-regulated genes)')
```

## camera

We use the `cameraPR()` function^[`cameraPR()` is a "pre-ranked" version of `camera()` where the genes are pre-ranked according to a pre-computed statistic. We cannot use the regular `camera()` function because the pseudotime DE analysis does not provide the standard limma/edgeR output required for `camera()`, but `cameraPR()` is a suitable replacement in this case.] from the `r BiocStyle::Biocpkg("limma")` R/Bioconductor package to test whether a set of genes is highly ranked relative to other genes in terms of differential expression, accounting for inter-gene correlation.
Specifically, we test using gene sets from [MSigDB](https://www.gsea-msigdb.org/gsea/msigdb), namely:

<aside>
We download these gene sets from [http://bioinf.wehi.edu.au/MSigDB/index.html](http://bioinf.wehi.edu.au/MSigDB/index.html)
</aside>

- **H**: **hallmark gene sets** are coherently expressed signatures derived by aggregating many MSigDB gene sets to represent well-defined biological states or processes.
- **C2**: **curated gene sets** from online pathway databases, publications in PubMed, and knowledge of domain experts.
- **C7**: **immunologic signature gene sets** represent cell states and perturbations within the immune system.

```{r}
# NOTE: Using BiocFileCache to avoid re-downloading these gene sets everytime
#       the report is rendered.
library(BiocFileCache)
bfc <- BiocFileCache()
# NOTE: Creating list of gene sets in this slightly convoluted way so that each
#       gene set name is prepended by its origin (e.g. H, C2, or C7).
msigdb <- do.call(
  c,
  list(
    H = readRDS(
      bfcrpath(
        bfc,
        "http://bioinf.wehi.edu.au/MSigDB/v7.1/Hs.h.all.v7.1.entrez.rds")),
    C2 = readRDS(
      bfcrpath(
        bfc,
        "http://bioinf.wehi.edu.au/MSigDB/v7.1/Hs.c2.all.v7.1.entrez.rds")),
    C7 = readRDS(
      bfcrpath(
        bfc,
        "http://bioinf.wehi.edu.au/MSigDB/v7.1/Hs.c7.all.v7.1.entrez.rds"))))

# NOTE: Using the signed Wald statistic as the statistic, following by
#       analogy what Gordon recommends
#       (https://support.bioconductor.org/p/112937/#112938).
svet$statistic <- svet$waldStat * sign(svet$logFClineage1)
y_idx <- ids2indices(msigdb, id = svet$ENTREZID)
cam <- cameraPR(
  statistic = setNames(svet$statistic, svet$ENTREZID),
  index = y_idx,
  use.ranks = FALSE)

gzout <- gzfile(
  description = file.path(outdir, "tradeSeq_results.camera.csv.gz"),
  open = "wb")
write.csv(
  cam,
  gzout,
  row.names = TRUE)
close(gzout)
```

CSV files of the results are available from [`output/trajectory/`](../output/trajectory/).
An example of the results are shown below, here filtered to highlight those gene sets with `Direction = Up` meaning `up_at_end`.

```{r}
head(cam[cam$Direction == "Up", ]) %>%
  knitr::kable(
    caption = '`camera()` produces a table with a row for each gene set (prepended by which MSigDB collection it comes from) and the following columns: **NGenes** (number of genes in set); **Direction** (direction of change, `Up` = `up_at_end` and `Down` = `up_at_start`); **PValue** (two-tailed p-value); **FDR** (Benjamini and Hochberg FDR adjusted p-value).')
```

## Visualisation

```{r}
term <- "C2.REACTOME_INTERFERON_ALPHA_BETA_SIGNALING"
library(org.Hs.eg.db)
key <- msigdb[[term]]
tmp <- AnnotationDbi::select(
  org.Hs.eg.db,
  key = key,
  columns = "SYMBOL",
  keytype = "ENTREZID")
y_index <- ids2indices(msigdb[[term]], svet$ENTREZID)
```

We can visualise the expression of the genes in any given set using a 'barcode plot'.
Such figures are often included in publications.

Figure \@ref(fig:gsea-vis) gives an example showing the expression of genes in the [`r term`](https://www.gsea-msigdb.org/gsea/msigdb/cards/REACTOME_INTERFERON_ALPHA_BETA_SIGNALING) MSigDB gene set.

```{r gsea-vis, fig.asp = 1 / 2, fig.cap = "Barcode plot of genes in MSigDB set `C2.REACTOME_INTERFERON_ALPHA_BETA_SIGNALING`. Genes are represented by bars and are ranked from left to right by increasing signed Wald statistic from the trajectory DE analysis. This forms the barcode-like pattern. The line (or *worm*) above the barcode shows the relative local enrichment of the vertical bars in each part of the plot. The dotted horizontal line indicates neutral enrichment; the worm above the dotted line shows enrichment while the worm below the dotted line shows depletion."}
barcodeplot(statistics = svet$statistic, index = y_index[[1]])
```

# Figures for paper

Note that exact figure numbers may change.

```{r}
# Some useful colours
plate_number_colours <- setNames(
  unique(sce$colours$plate_number_colours),
  unique(names(sce$colours$plate_number_colours)))
plate_number_colours <- plate_number_colours[levels(sce$plate_number)]
tissue_colours <- setNames(
  unique(sce$colours$tissue_colours),
  unique(names(sce$colours$tissue_colours)))
tissue_colours <- tissue_colours[levels(sce$tissue)]
donor_colours <- setNames(
  unique(sce$colours$donor_colours),
  unique(names(sce$colours$donor_colours)))
donor_colours <- donor_colours[levels(sce$donor)]
stage_colours <- setNames(
  unique(sce$colours$stage_colours),
  unique(names(sce$colours$stage_colours)))
stage_colours <- stage_colours[levels(sce$stage)]
group_colours <- setNames(
  unique(sce$colours$group_colours),
  unique(names(sce$colours$group_colours)))
group_colours <- group_colours[levels(sce$group)]
cluster_colours <- setNames(
  unique(sce$colours$cluster_colours),
  unique(names(sce$colours$cluster_colours)))
cluster_colours <- cluster_colours[levels(sce$cluster)]
```

```{r}
sce$CD4 <- assay(altExp(sce, "FACS"), "pseudolog")["V525_50_A_CD4_BV510", ]
sce$CD161 <- assay(altExp(sce, "FACS"), "pseudolog")["B530_30_A_CD161_FITC", ]
```

## Figure 5

```{r, fig.asp = 0.8}
plotReducedDim(
  sce,
  "UMAP_corrected",
  colour_by = "stage",
  point_alpha = 1) +
  scale_colour_manual(values = stage_colours, name = "stage") +
  ggtitle("All cells")
```

```{r, fig.asp = 0.8}
plotReducedDim(
  sce,
  "UMAP_corrected",
  colour_by = "tissue",
  point_alpha = 1) +
  scale_colour_manual(values = tissue_colours, name = "tissue") +
  ggtitle("All cells")
```

```{r, fig.asp = 0.8}
plotReducedDim(
  sce,
  dimred = "UMAP_corrected",
  colour_by = "slingPseudotime_1",
  point_alpha = 1) +
  geom_path(data = embedded, aes(x = Dim.1, y = Dim.2), size = 1.2) +
  ggtitle("All cells")
```

```{r, fig.asp = 1}
plotHeatmap(
  sce,
  order_columns_by = "slingPseudotime_1",
  colour_columns_by = c("slingPseudotime_1", "CD4", "CD161", "stage", "tissue"),
  features = c(
    head(
      setdiff(rownames(svet[svet$FDR < 0.05 & svet$logFC < 0, ]), mito_set), 
      25),
    head(
      setdiff(rownames(svet[svet$FDR < 0.05 & svet$logFC > 0, ]), mito_set), 
      25)),
  center = TRUE,
  symmetric = TRUE,
  zlim = c(-3, 3),
  color = hcl.colors(101, "Blue-Red 3"),
  fontsize = 6,
  main = "All cells: Selected pseudotime-associated genes",
  column_annotation_colors = list(
    stage = stage_colours,
    tissue = tissue_colours))
```

```{r, fig.asp = 0.8}
plotReducedDim(
  sce[, sce$stage != "Unknown"],
  "UMAP_corrected",
  colour_by = "stage",
  point_alpha = 1) +
  scale_colour_manual(values = stage_colours, name = "stage") +
  ggtitle("Excluding 'Unknown'-stage cells")
```

```{r, fig.asp = 0.8}
plotReducedDim(
  sce[, sce$stage != "Unknown"],
  "UMAP_corrected",
  colour_by = "tissue",
  point_alpha = 1) +
  scale_colour_manual(values = tissue_colours, name = "tissue") +
  ggtitle("Excluding 'Unknown'-stage cells")
```

```{r, fig.asp = 0.8}

plotReducedDim(
  sce[, sce$stage != "Unknown"],
  dimred = "UMAP_corrected",
  colour_by = "slingPseudotime_1",
  point_alpha = 1) +
  geom_path(data = embedded, aes(x = Dim.1, y = Dim.2), size = 1.2) +
  ggtitle("Excluding 'Unknown'-stage cells")
```

```{r, fig.asp = 1}
plotHeatmap(
  sce[, sce$stage != "Unknown"],
  order_columns_by = "slingPseudotime_1",
  colour_columns_by = c("slingPseudotime_1", "CD4", "CD161", "stage", "tissue"),
  features = c(
    head(
      setdiff(rownames(svet[svet$FDR < 0.05 & svet$logFC < 0, ]), mito_set), 
      25),
    head(
      setdiff(rownames(svet[svet$FDR < 0.05 & svet$logFC > 0, ]), mito_set), 
      25)),
  center = TRUE,
  symmetric = TRUE,
  zlim = c(-3, 3),
  color = hcl.colors(101, "Blue-Red 3"),
  fontsize = 6,
  main = "Excluding 'Unknown'-stage cells: Selected pseudotime-associated genes",
  column_annotation_colors = list(
    stage = stage_colours[levels(factor(sce[, sce$stage != "Unknown"]$stage))],
    tissue = tissue_colours))
```

## DEG heatmaps

Heatmaps sorted by CD4 and CD161 FACS expression "showing the DEGs across the gamma delta T cell stages ordered by decreasing either CD4 or KLRB1."
This is an example, but, as I explained to Dan, I don't think these are very useful plots.

```{r}
tmp <- sce[, sce$group %in% c("Thymus.S1 (CD4+/CD161-)", "Thymus.S2 (CD4-/CD161-)", "Thymus.S3 (CD4-/CD161+)", "Blood.S3 (CD4-/CD161+)")]
tmp$group <- factor(
  tmp$group,
  c("Thymus.S1 (CD4+/CD161-)", "Thymus.S2 (CD4-/CD161-)", "Thymus.S3 (CD4-/CD161+)", "Blood.S3 (CD4-/CD161+)"))

x <- read.csv(
  here("output/DEGs/excluding_donors_1-3/Thymus.S3_vs_Thymus.S1.aggregated_tech_reps.DEGs.csv.gz"))
```

```{r, fig.asp = 1}
plotHeatmap(
  tmp,
  order_columns_by = c("group", "CD4"),
  colour_columns_by = c("CD4", "group"),
  features = head(x[x$FDR < 0.05, "ENSEMBL.SYMBOL"], 50),
  center = TRUE,
  symmetric = TRUE,
  zlim = c(-3, 3),
  color = hcl.colors(101, "Blue-Red 3"),
  fontsize = 6,
  main = "Mini-bulk DEGs: Thymus.S3 vs. Thymus.S1",
  column_annotation_colors = list(group = group_colours[levels(tmp$group)]))
```

```{r, fig.asp = 1}
plotHeatmap(
  tmp,
  order_columns_by = c("group", "CD161"),
  colour_columns_by = c("CD161", "group"),
  features = head(x[x$FDR < 0.05, "ENSEMBL.SYMBOL"], 50),
  center = TRUE,
  symmetric = TRUE,
  zlim = c(-3, 3),
  color = hcl.colors(101, "Blue-Red 3"),
  fontsize = 6,
  main = "Mini-bulk DEGs: Thymus.S3 vs. Thymus.S1",
  column_annotation_colors = list(group = group_colours[levels(tmp$group)]))
```

## Mock figure sent to Dan and co (2021-09-09)

```{r}
dir.create(here("output/figures"))

p1 <- plotReducedDim(
  sce[, sce$stage != "Unknown"],
  "UMAP_corrected",
  colour_by = "stage",
  point_alpha = 1) +
  scale_colour_manual(values = stage_colours, name = "stage")
ggsave(here("output/figures/Fig5a.pdf"), p1, width = 6, height = 5)

p2 <- plotReducedDim(
  sce[, sce$stage != "Unknown"],
  "UMAP_corrected",
  colour_by = "tissue",
  point_alpha = 1) +
  scale_colour_manual(values = tissue_colours, name = "tissue")
ggsave(here("output/figures/Fig5b.pdf"), p2, width = 6, height = 5)

p3 <- plotReducedDim(
  sce[, sce$stage != "Unknown"],
  dimred = "UMAP_corrected",
  colour_by = "slingPseudotime_1",
  point_alpha = 1) +
  geom_path(data = embedded, aes(x = Dim.1, y = Dim.2), size = 1.2)
ggsave(here("output/figures/Fig5c.pdf"), p3, width = 6, height = 5)

p4 <- plotHeatmap(
  sce[, sce$stage != "Unknown"],
  order_columns_by = "slingPseudotime_1",
  colour_columns_by = c("slingPseudotime_1", "CD4", "CD161", "stage", "tissue"),
  features = c(
    head(
      setdiff(rownames(svet[svet$FDR < 0.05 & svet$logFC < 0, ]), mito_set), 
      50),
    head(
      setdiff(rownames(svet[svet$FDR < 0.05 & svet$logFC > 0, ]), mito_set), 
      50)),
  center = TRUE,
  symmetric = TRUE,
  zlim = c(-3, 3),
  color = hcl.colors(101, "Blue-Red 3"),
  fontsize = 5,
  column_annotation_colors = list(
    stage = stage_colours[levels(factor(sce[, sce$stage != "Unknown"]$stage))],
    tissue = tissue_colours),
  silent = TRUE)
ggsave(here("output/figures/Fig5d.pdf"), p4$gtable, width = 6, height = 8)

p4a <- plotHeatmap(
  sce[, sce$stage != "Unknown"],
  order_columns_by = c("stage", "slingPseudotime_1"),
  colour_columns_by = c("stage", "slingPseudotime_1", "CD4", "CD161", "tissue"),
  features = c(
    head(
      setdiff(rownames(svet[svet$FDR < 0.05 & svet$logFC < 0, ]), mito_set), 
      50),
    head(
      setdiff(rownames(svet[svet$FDR < 0.05 & svet$logFC > 0, ]), mito_set), 
      50)),
  center = TRUE,
  symmetric = TRUE,
  zlim = c(-3, 3),
  color = hcl.colors(101, "Blue-Red 3"),
  fontsize = 5,
  # main = "Excluding 'Unknown'-stage cells: Selected pseudotime-associated genes",
  column_annotation_colors = list(
    stage = stage_colours[levels(factor(sce[, sce$stage != "Unknown"]$stage))],
    tissue = tissue_colours),
  silent = TRUE)
ggsave(here("output/figures/Fig5d_alternative.pdf"), p4a$gtable, width = 6, height = 8)
```

Initial proposal:

```{r, fig.asp = 1}
p1 + p2 + p3 + p4$gtable +
  plot_annotation(
    title = "Summary of scRNA-seq data",
    subtitle = "Excluding 'Unknown'-stage cells",
    tag_levels = "a")
```

Alternative:

```{r, fig.asp = 1}
p1 + p2 + p3 + p4a$gtable +
  plot_annotation(
    title = "Summary of scRNA-seq data",
    subtitle = "Excluding 'Unknown'-stage cells",
    tag_levels = "a")
```

# Additional information {.appendix}

The following are available on request:

- Full CSV tables of any data presented.
- PDF/PNG files of any static plots.

# Session info {.appendix}

<summary>The analysis and this document were prepared using the following software (click triangle to expand)</summary>
<details>

```{r}
sessioninfo::session_info()
```

</details>
